<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Basic class | Tutorial for R microeco package (v1.7.0)</title>
  <meta name="description" content="The tutorial for R microeco (v1.7.0), file2meco (v0.7.1), meconetcomp (v0.5.0) and mecoturn packages" />
  <meta name="generator" content="bookdown 0.38 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Basic class | Tutorial for R microeco package (v1.7.0)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The tutorial for R microeco (v1.7.0), file2meco (v0.7.1), meconetcomp (v0.5.0) and mecoturn packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Basic class | Tutorial for R microeco package (v1.7.0)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco (v1.7.0), file2meco (v0.7.1), meconetcomp (v0.5.0) and mecoturn packages" />
  



<meta name="date" content="2024-05-15" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="intro.html"/>
<link rel="next" href="composition-based-class.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#rtools"><i class="fa fa-check"></i><b>2.4</b> RTools</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.5</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#description"><i class="fa fa-check"></i><b>2.5.1</b> Description</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#cran-packages"><i class="fa fa-check"></i><b>2.5.2</b> CRAN packages</a></li>
<li class="chapter" data-level="2.5.3" data-path="intro.html"><a href="intro.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor packages</a></li>
<li class="chapter" data-level="2.5.4" data-path="intro.html"><a href="intro.html#github-packages"><i class="fa fa-check"></i><b>2.5.4</b> Github packages</a></li>
<li class="chapter" data-level="2.5.5" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.5.5</b> Gephi</a></li>
<li class="chapter" data-level="2.5.6" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.5.6</b> Tax4Fun</a></li>
<li class="chapter" data-level="2.5.7" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.5.7</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#plot"><i class="fa fa-check"></i><b>2.6</b> Plot</a></li>
<li class="chapter" data-level="2.7" data-path="intro.html"><a href="intro.html#rstudio"><i class="fa fa-check"></i><b>2.7</b> Rstudio</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#prepare-the-example-data"><i class="fa fa-check"></i><b>3.1.1</b> Prepare the example data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#how-to-read-your-files-to-microtable-object"><i class="fa fa-check"></i><b>3.1.2</b> How to read your files to microtable object?</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#functions-in-microtable-class"><i class="fa fa-check"></i><b>3.1.3</b> Functions in microtable class</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#merge-taxa-or-samples"><i class="fa fa-check"></i><b>3.1.4</b> merge taxa or samples</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-class.html"><a href="basic-class.html#subset-of-samples"><i class="fa fa-check"></i><b>3.1.5</b> subset of samples</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-class.html"><a href="basic-class.html#subset-of-taxa"><i class="fa fa-check"></i><b>3.1.6</b> subset of taxa</a></li>
<li class="chapter" data-level="3.1.7" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.7</b> Other examples</a></li>
<li class="chapter" data-level="3.1.8" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.8</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-2"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-4"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#more-options"><i class="fa fa-check"></i><b>6.1.2</b> More options</a></li>
<li class="chapter" data-level="6.1.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.3</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#network-comparison"><i class="fa fa-check"></i><b>6.2.2</b> Network comparison</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.3</b> Key points</a></li>
<li class="chapter" data-level="6.2.4" data-path="model-based-class.html"><a href="model-based-class.html#other-functions"><i class="fa fa-check"></i><b>6.2.4</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies</a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples</a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points</a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-7"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME</a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2</a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann"><i class="fa fa-check"></i><b>8.3</b> HUMAnN</a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn</a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2braken"><i class="fa fa-check"></i><b>8.5</b> Kraken2/Braken</a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdbpcycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB/PCycDB</a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq</a></li>
<li class="chapter" data-level="8.8" data-path="file2meco-package.html"><a href="file2meco-package.html#picrust2"><i class="fa fa-check"></i><b>8.8</b> PICRUSt2</a></li>
<li class="chapter" data-level="8.9" data-path="file2meco-package.html"><a href="file2meco-package.html#viromescan"><i class="fa fa-check"></i><b>8.9</b> ViromeScan</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset</a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data</a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> <code>clone</code> function</a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> <code>save</code> function</a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#github-issues"><i class="fa fa-check"></i><b>10.3</b> Github Issues</a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.4</b> group order</a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.5</b> add layers to plot</a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#colors-for-more-groups"><i class="fa fa-check"></i><b>10.6</b> colors for more groups</a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.7</b> change object</a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html"><i class="fa fa-check"></i><b>11</b> meconetcomp package</a>
<ul>
<li class="chapter" data-level="11.1" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-modularity-for-all-networks"><i class="fa fa-check"></i><b>11.1</b> Network modularity for all networks</a></li>
<li class="chapter" data-level="11.2" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-topological-attributes-for-all-networks"><i class="fa fa-check"></i><b>11.2</b> Network topological attributes for all networks</a></li>
<li class="chapter" data-level="11.3" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#node-and-edge-properties-extraction-for-all-networks"><i class="fa fa-check"></i><b>11.3</b> Node and edge properties extraction for all networks</a></li>
<li class="chapter" data-level="11.4" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-nodes-across-networks"><i class="fa fa-check"></i><b>11.4</b> Compare nodes across networks</a></li>
<li class="chapter" data-level="11.5" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-edges-across-networks"><i class="fa fa-check"></i><b>11.5</b> Compare edges across networks</a></li>
<li class="chapter" data-level="11.6" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#extract-overlapped-edges-of-networks-to-a-new-network"><i class="fa fa-check"></i><b>11.6</b> Extract overlapped edges of networks to a new network</a></li>
<li class="chapter" data-level="11.7" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-phylogenetic-distances-of-paired-nodes-in-edges"><i class="fa fa-check"></i><b>11.7</b> Compare phylogenetic distances of paired nodes in edges</a></li>
<li class="chapter" data-level="11.8" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-node-sources-of-edges-across-networks"><i class="fa fa-check"></i><b>11.8</b> Compare node sources of edges across networks</a></li>
<li class="chapter" data-level="11.9" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-topological-properties-of-sub-networks"><i class="fa fa-check"></i><b>11.9</b> Compare topological properties of sub-networks</a></li>
<li class="chapter" data-level="11.10" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#robustness-of-network"><i class="fa fa-check"></i><b>11.10</b> Robustness of network</a></li>
<li class="chapter" data-level="11.11" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#vulnerability-of-nodes"><i class="fa fa-check"></i><b>11.11</b> Vulnerability of nodes</a></li>
<li class="chapter" data-level="11.12" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#cohesion"><i class="fa fa-check"></i><b>11.12</b> Cohesion</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>12</b> mecodev package</a>
<ul>
<li class="chapter" data-level="12.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>12.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="12.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>12.2</b> trans_ts</a></li>
<li class="chapter" data-level="12.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>12.3</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mecoturn-package.html"><a href="mecoturn-package.html"><i class="fa fa-check"></i><b>13</b> mecoturn package</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mecoturn-package.html"><a href="mecoturn-package.html#betaturn-class"><i class="fa fa-check"></i><b>13.1</b> betaturn class</a></li>
<li class="chapter" data-level="13.2" data-path="mecoturn-package.html"><a href="mecoturn-package.html#taxaturn-class"><i class="fa fa-check"></i><b>13.2</b> taxaturn class</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>14</b> Other examples</a>
<ul>
<li class="chapter" data-level="14.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>14.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="14.2" data-path="other-examples-1.html"><a href="other-examples-1.html#tidy_taxonomy-function"><i class="fa fa-check"></i><b>14.2</b> tidy_taxonomy function</a></li>
<li class="chapter" data-level="14.3" data-path="other-examples-1.html"><a href="other-examples-1.html#show-the-abundance-of-unknown-taxa"><i class="fa fa-check"></i><b>14.3</b> Show the abundance of unknown taxa</a></li>
<li class="chapter" data-level="14.4" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>14.4</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="14.5" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>14.5</b> The use of phylogenetic tree</a></li>
<li class="chapter" data-level="14.6" data-path="other-examples-1.html"><a href="other-examples-1.html#trans_norm"><i class="fa fa-check"></i><b>14.6</b> trans_norm</a></li>
<li class="chapter" data-level="14.7" data-path="other-examples-1.html"><a href="other-examples-1.html#mantel-test-correlation-heatmap"><i class="fa fa-check"></i><b>14.7</b> Mantel test + Correlation heatmap</a></li>
<li class="chapter" data-level="14.8" data-path="other-examples-1.html"><a href="other-examples-1.html#pcoa-boxplot"><i class="fa fa-check"></i><b>14.8</b> PCoA + boxplot</a></li>
<li class="chapter" data-level="14.9" data-path="other-examples-1.html"><a href="other-examples-1.html#faprotax-differential-test"><i class="fa fa-check"></i><b>14.9</b> FAPROTAX + differential test</a></li>
<li class="chapter" data-level="14.10" data-path="other-examples-1.html"><a href="other-examples-1.html#rda-results-output"><i class="fa fa-check"></i><b>14.10</b> RDA results output</a></li>
<li class="chapter" data-level="14.11" data-path="other-examples-1.html"><a href="other-examples-1.html#betanribetanti-for-each-group"><i class="fa fa-check"></i><b>14.11</b> betaNRI/betaNTI for each group</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v1.7.0)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="basic-class" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Basic class<a href="basic-class.html#basic-class" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The microtable class is the basic class.
All the other classes depend on the microtable class.</p>
<p><img src="Images/microtable_framework.png" width="8000px" style="display: block; margin: auto;" /></p>
<p>The objects inside the rectangle with full line represent functions.
The red rectangle means it is extremely important function.
The dashed line denotes the key objects (input or output of functions) that deserve more attention.</p>
<div id="microtable-class" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> microtable class<a href="basic-class.html#microtable-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　Many tools can be used for the bioinformatic analysis of amplicon sequencing data, such as QIIME <span class="citation">(<a href="#ref-Caporaso_QIIME_2010" role="doc-biblioref">Caporaso et al. 2010</a>)</span>, QIIME2 <span class="citation">(<a href="#ref-Bolyen_Reproducible_2019" role="doc-biblioref">Bolyen et al. 2019</a>)</span>,
usearch (<a href="https://www.drive5.com/usearch/" class="uri">https://www.drive5.com/usearch/</a>), mothur <span class="citation">(<a href="#ref-Schloss_Introducing_2009" role="doc-biblioref">Schloss et al. 2009</a>)</span>,
SILVAngs (<a href="https://ngs.arb-silva.de/silvangs/" class="uri">https://ngs.arb-silva.de/silvangs/</a>), LotuS2 <span class="citation">(<a href="#ref-Ozkurt_LotuS2_2022" role="doc-biblioref">Ozkurt et al. 2022</a>)</span>,
and RDP (<a href="http://rdp.cme.msu.edu/" class="uri">http://rdp.cme.msu.edu/</a>).
Although the formats of result files may vary across tools, the main contents can be generally classified into the following parts:
(1) OTU/ASV table, i.e. the feature-sample abundance table;
(2) taxonomic assignment table;
(3) representative sequences;
(4) phylogenetic tree;
(5) metadata.
It is generally useful to create a detailed sample metadata table to store all the sample information (including the environmental data).</p>
<p>　The microtable class is the basic class and designed to store the basic data for all the downstream analysis in the microeco package.
At least, the OTU table (i.e. feature-sample abundance table) should be provided to create microtable object.
Thus, the microtable class can determine that the sample information table is missing and create a default sample table according to
sample names in otu_table.
To make the file input more convenient,
we also build another R package file2meco (<a href="https://github.com/ChiLiubio/file2meco" class="uri">https://github.com/ChiLiubio/file2meco</a>) to read the output files of some tools into microtable object.
Currently, those tools/softwares include not only commonly-used QIIME <span class="citation">(<a href="#ref-Caporaso_QIIME_2010" role="doc-biblioref">Caporaso et al. 2010</a>)</span> and QIIME2<span class="citation">(<a href="#ref-Bolyen_Reproducible_2019" role="doc-biblioref">Bolyen et al. 2019</a>)</span>,
but also several metagenomic tools, such as HUMAnN <span class="citation">(<a href="#ref-Franzosa_Species_2018" role="doc-biblioref">Franzosa et al. 2018</a>)</span> and kraken2 <span class="citation">(<a href="#ref-Wood_Improved_2019" role="doc-biblioref">Wood, Lu, and Langmead 2019</a>)</span>.
In this tutorial, the data inside the package was employed to show some basic operations.</p>
<div id="prepare-the-example-data" class="section level3 hasAnchor" number="3.1.1">
<h3><span class="header-section-number">3.1.1</span> Prepare the example data<a href="basic-class.html#prepare-the-example-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>　The example data inside the microeco package is used to show the main part of the tutorial.
This dataset arose from 16S rRNA gene Miseq sequencing results of wetland soils in China published by An et al. <span class="citation">(<a href="#ref-An_Soil_2019" role="doc-biblioref">An et al. 2019</a>)</span>,
who surveyed soil prokaryotic communities in Chinese inland wetlands (IW),
coastal wetland (CW) and Tibet plateau wetlands (TW) using amplicon sequencing.
These wetlands include both saline and non-saline samples (classified for the tutorial).
The sample information table has 4 columns: “SampleID”, “Group”, “Type” and “Saline”.
The column “SampleID” is same with the rownames.
The column “Group” represents the IW, CW and TW.
The column “Type” means the sampling region: northeastern region (NE), northwest region (NW), North China area (NC),
middle-lower reaches of the Yangtze River (YML), southern coastal area (SC), upper reaches of the Yangtze River (YU), Qinghai-Tibet Plateau (QTP).
The column “Saline” denotes the saline soils and non-saline soils.
In this dataset, the environmental factor table is separated from the sample information table.
It is also recommended to put all the environmental data into sample information table.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="basic-class.html#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb10-2"><a href="basic-class.html#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># load the example data; 16S rRNA gene amplicon sequencing dataset</span></span>
<span id="cb10-3"><a href="basic-class.html#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># metadata table; data.frame</span></span>
<span id="cb10-4"><a href="basic-class.html#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(sample_info_16S)</span>
<span id="cb10-5"><a href="basic-class.html#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># feature table; data.frame</span></span>
<span id="cb10-6"><a href="basic-class.html#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(otu_table_16S)</span>
<span id="cb10-7"><a href="basic-class.html#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># taxonomic assignment table; data.frame</span></span>
<span id="cb10-8"><a href="basic-class.html#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(taxonomy_table_16S)</span>
<span id="cb10-9"><a href="basic-class.html#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co"># phylogenetic tree; not necessary; use for the phylogenetic analysis</span></span>
<span id="cb10-10"><a href="basic-class.html#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Newick format; use read.tree function of ape package to read a tree</span></span>
<span id="cb10-11"><a href="basic-class.html#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(phylo_tree_16S)</span>
<span id="cb10-12"><a href="basic-class.html#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co"># load the environmental data table if it is not in sample table</span></span>
<span id="cb10-13"><a href="basic-class.html#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(env_data_16S)</span>
<span id="cb10-14"><a href="basic-class.html#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="co"># use pipe operator in magrittr package</span></span>
<span id="cb10-15"><a href="basic-class.html#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb10-16"><a href="basic-class.html#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="co"># fix the random number generation to make the results repeatable</span></span>
<span id="cb10-17"><a href="basic-class.html#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb10-18"><a href="basic-class.html#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="co"># make the plotting background same with the tutorial</span></span>
<span id="cb10-19"><a href="basic-class.html#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb10-20"><a href="basic-class.html#cb10-20" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>())</span></code></pre></div>
<p>Make sure that the data types of sample_table, otu_table and tax_table are all <code>data.frame</code> format as the following part shows.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="basic-class.html#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(otu_table_16S)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="basic-class.html#cb13-1" aria-hidden="true" tabindex="-1"></a>otu_table_16S[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<table style="width:57%;">
<colgroup>
<col width="20%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="8%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">S1</th>
<th align="center">S2</th>
<th align="center">S3</th>
<th align="center">S4</th>
<th align="center">S5</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>OTU_4272</strong></td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_236</strong></td>
<td align="center">1</td>
<td align="center">4</td>
<td align="center">0</td>
<td align="center">2</td>
<td align="center">35</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_399</strong></td>
<td align="center">9</td>
<td align="center">2</td>
<td align="center">2</td>
<td align="center">4</td>
<td align="center">4</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_1556</strong></td>
<td align="center">5</td>
<td align="center">18</td>
<td align="center">7</td>
<td align="center">3</td>
<td align="center">2</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_32</strong></td>
<td align="center">83</td>
<td align="center">9</td>
<td align="center">19</td>
<td align="center">8</td>
<td align="center">102</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="basic-class.html#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(taxonomy_table_16S)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="basic-class.html#cb16-1" aria-hidden="true" tabindex="-1"></a>taxonomy_table_16S[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code></pre></div>
<table style="width:100%;">
<colgroup>
<col width="19%" />
<col width="18%" />
<col width="30%" />
<col width="31%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">Kingdom</th>
<th align="center">Phylum</th>
<th align="center">Class</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>OTU_4272</strong></td>
<td align="center">k__Bacteria</td>
<td align="center">p__Firmicutes</td>
<td align="center">c__Bacilli</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_236</strong></td>
<td align="center">k__Bacteria</td>
<td align="center">p__Chloroflexi</td>
<td align="center">c__</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_399</strong></td>
<td align="center">k__Bacteria</td>
<td align="center">p__Proteobacteria</td>
<td align="center">c__Betaproteobacteria</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_1556</strong></td>
<td align="center">k__Bacteria</td>
<td align="center">p__Acidobacteria</td>
<td align="center">c__Acidobacteria</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_32</strong></td>
<td align="center">k__Archaea</td>
<td align="center">p__Miscellaneous
Crenarchaeotic Group</td>
<td align="center">c__</td>
</tr>
</tbody>
</table>
<p>Generally, users’ taxonomic table has some messy information, such as NA, unidentified and unknown.
These information can potentially influence the following taxonomic abundance calculation and other taxonomy-based analysis.
So it is usually necessary to clean this data using the <code>tidy_taxonomy</code> function.
Another very important result of this operation is to <strong>unify the taxonomic prefix</strong> automatically,
e.g., converting D_1__ to p__ for Phylum level or adding p__ to Phylum directly if no prefix is found.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="basic-class.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make the taxonomic information unified, very important</span></span>
<span id="cb17-2"><a href="basic-class.html#cb17-2" aria-hidden="true" tabindex="-1"></a>taxonomy_table_16S <span class="sc">%&lt;&gt;%</span> tidy_taxonomy</span></code></pre></div>
<p>The rownames of sample_table in microtable object (i.e. sample names) are used for selecting samples/groups in all the related operations in the package.
Using pure number as sample names is <strong>not recommended</strong> in case of unknown disorder or man-made mistake.
<strong>Before creating microtable object, make sure that the rownames of sample information table are sample names</strong>.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="basic-class.html#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(sample_info_16S)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="basic-class.html#cb20-1" aria-hidden="true" tabindex="-1"></a>sample_info_16S[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<table style="width:74%;">
<colgroup>
<col width="12%" />
<col width="15%" />
<col width="11%" />
<col width="9%" />
<col width="25%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">SampleID</th>
<th align="center">Group</th>
<th align="center">Type</th>
<th align="center">Saline</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">S1</td>
<td align="center">IW</td>
<td align="center">NE</td>
<td align="center">Non-saline soil</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">S2</td>
<td align="center">IW</td>
<td align="center">NE</td>
<td align="center">Non-saline soil</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">S3</td>
<td align="center">IW</td>
<td align="center">NE</td>
<td align="center">Non-saline soil</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">S4</td>
<td align="center">IW</td>
<td align="center">NE</td>
<td align="center">Non-saline soil</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">S5</td>
<td align="center">IW</td>
<td align="center">NE</td>
<td align="center">Non-saline soil</td>
</tr>
</tbody>
</table>
<p>In this example, the environmental data is stored in the env_data_16S alone.
The user can also directly integrate those data into the sample information table.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="basic-class.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(env_data_16S)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<table style="width:100%;">
<colgroup>
<col width="12%" />
<col width="15%" />
<col width="16%" />
<col width="15%" />
<col width="19%" />
<col width="21%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">Latitude</th>
<th align="center">Longitude</th>
<th align="center">Altitude</th>
<th align="center">Temperature</th>
<th align="center">Precipitation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">52.96</td>
<td align="center">122.6</td>
<td align="center">432</td>
<td align="center">-4.2</td>
<td align="center">445</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">52.95</td>
<td align="center">122.6</td>
<td align="center">445</td>
<td align="center">-4.3</td>
<td align="center">449</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">52.95</td>
<td align="center">122.6</td>
<td align="center">430</td>
<td align="center">-4.3</td>
<td align="center">449</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">52.95</td>
<td align="center">122.6</td>
<td align="center">430</td>
<td align="center">-4.3</td>
<td align="center">449</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">52.95</td>
<td align="center">122.6</td>
<td align="center">429</td>
<td align="center">-4.3</td>
<td align="center">449</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="basic-class.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(phylo_tree_16S)</span></code></pre></div>
<pre><code>## [1] &quot;phylo&quot;</code></pre>
<p>Then, we create an object of microtable class.
This operation is very similar with the package phyloseq<span class="citation">(<a href="#ref-Mcmurdie_phyloseq_2013" role="doc-biblioref">Mcmurdie and Holmes 2013</a>)</span>, but in microeco it is more brief.
The otu_table in the microtable class must be the feature-sample format: rownames - OTU/ASV/pathway/other names; colnames - sample names.
<strong>The colnames in otu_table must have overlap with rownames of sample_table</strong>.
Otherwise, the following check can filter all the samples of otu_table because of no same sample names between otu_table and sample_table.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="basic-class.html#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># In R6 class, &#39;$new&#39; is the original method used to create a new object of class</span></span>
<span id="cb25-2"><a href="basic-class.html#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># If you only provide abundance table, the class can help you create a sample info table</span></span>
<span id="cb25-3"><a href="basic-class.html#cb25-3" aria-hidden="true" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">otu_table =</span> otu_table_16S)</span></code></pre></div>
<pre><code>## No sample_table provided, automatically use colnames in otu_table to create one ...</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="basic-class.html#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(dataset)</span></code></pre></div>
<pre><code>## [1] &quot;microtable&quot; &quot;R6&quot;</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="basic-class.html#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generally add the metadata</span></span>
<span id="cb29-2"><a href="basic-class.html#cb29-2" aria-hidden="true" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">otu_table =</span> otu_table_16S, <span class="at">sample_table =</span> sample_info_16S)</span>
<span id="cb29-3"><a href="basic-class.html#cb29-3" aria-hidden="true" tabindex="-1"></a>dataset</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 4 columns
## otu_table have 13628 rows and 90 columns</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="basic-class.html#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let&#39;s create a microtable object with more information</span></span>
<span id="cb31-2"><a href="basic-class.html#cb31-2" aria-hidden="true" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">sample_table =</span> sample_info_16S, <span class="at">otu_table =</span> otu_table_16S, <span class="at">tax_table =</span> taxonomy_table_16S, <span class="at">phylo_tree =</span> phylo_tree_16S)</span>
<span id="cb31-3"><a href="basic-class.html#cb31-3" aria-hidden="true" tabindex="-1"></a>dataset</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 4 columns
## otu_table have 13628 rows and 90 columns
## tax_table have 13628 rows and 7 columns
## phylo_tree have 14096 tips</code></pre>
</div>
<div id="how-to-read-your-files-to-microtable-object" class="section level3 hasAnchor" number="3.1.2">
<h3><span class="header-section-number">3.1.2</span> How to read your files to microtable object?<a href="basic-class.html#how-to-read-your-files-to-microtable-object" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The above-mentioned example data are directly loaded from microeco package.
So the question is <strong>how to read your data to create a microtable object?</strong><br />
There are two ways:</p>
<p>▲ 1. <strong>Use file2meco package</strong><br />
R package file2meco (<a href="https://chiliubio.github.io/microeco_tutorial/file2meco-package.html" class="uri">https://chiliubio.github.io/microeco_tutorial/file2meco-package.html</a>) is designed to directly read the output files of some famous tools into microtable object.
Currently, it supports QIIME <span class="citation">(<a href="#ref-Caporaso_QIIME_2010" role="doc-biblioref">Caporaso et al. 2010</a>)</span>, QIIME2<span class="citation">(<a href="#ref-Bolyen_Reproducible_2019" role="doc-biblioref">Bolyen et al. 2019</a>)</span>,
HUMAnN <span class="citation">(<a href="#ref-Franzosa_Species_2018" role="doc-biblioref">Franzosa et al. 2018</a>)</span>, MetaPhlAn <span class="citation">(<a href="#ref-Truong_MeTApHLaN2_2015" role="doc-biblioref">Truong et al. 2015</a>)</span>, kraken2 <span class="citation">(<a href="#ref-Wood_Improved_2019" role="doc-biblioref">Wood, Lu, and Langmead 2019</a>)</span>, phyloseq <span class="citation">(<a href="#ref-Mcmurdie_phyloseq_2013" role="doc-biblioref">Mcmurdie and Holmes 2013</a>)</span>, etc.
Please read the tutorial of file2meco package for more detailed information (<a href="https://chiliubio.github.io/microeco_tutorial/file2meco-package.html" class="uri">https://chiliubio.github.io/microeco_tutorial/file2meco-package.html</a>).</p>
<p>▲ 2. <strong>Other cases</strong><br />
To transform customized files to microtable object,
there should be two steps:<br />
<strong>I) read files to R</strong><br />
The required format of microtable$new parameters, <strong>otu_table</strong>, <strong>sample_table</strong> and <strong>tax_table</strong>, are all the data.frame, which is the most frequently-used data format in R.
So no matter what the format the files are, they should be first read into R with some functions, such as <code>read.table</code> and <code>read.csv</code>.
If the user want to perform phylogenetic analysis, please also read your phylogenetic tree using <code>read.tree</code> function of ape package and
provide the tree to the <strong>phylo_tree</strong> parameter of microtable$new function like the above example.<br />
<strong>II) create the microtable object</strong><br />
Then the user can create the microtable object like the operation in the last section.
Please also see the help document of the microtable class for detailed descriptions using the following help command.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="basic-class.html#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># search the class name, not the function name</span></span>
<span id="cb33-2"><a href="basic-class.html#cb33-2" aria-hidden="true" tabindex="-1"></a>?microtable</span>
<span id="cb33-3"><a href="basic-class.html#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co"># then see microtable$new()</span></span></code></pre></div>
</div>
<div id="functions-in-microtable-class" class="section level3 hasAnchor" number="3.1.3">
<h3><span class="header-section-number">3.1.3</span> Functions in microtable class<a href="basic-class.html#functions-in-microtable-class" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Then, we remove OTUs which are not assigned in the Kingdom “k__Archaea” or “k__Bacteria”.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="basic-class.html#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use R subset function to filter taxa in tax_table</span></span>
<span id="cb34-2"><a href="basic-class.html#cb34-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>tax_table <span class="sc">%&lt;&gt;%</span> base<span class="sc">::</span><span class="fu">subset</span>(Kingdom <span class="sc">==</span> <span class="st">&quot;k__Archaea&quot;</span> <span class="sc">|</span> Kingdom <span class="sc">==</span> <span class="st">&quot;k__Bacteria&quot;</span>)</span>
<span id="cb34-3"><a href="basic-class.html#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="co"># another way with grepl function</span></span>
<span id="cb34-4"><a href="basic-class.html#cb34-4" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>tax_table <span class="sc">%&lt;&gt;%</span> .[<span class="fu">grepl</span>(<span class="st">&quot;Bacteria|Archaea&quot;</span>, .<span class="sc">$</span>Kingdom), ]</span>
<span id="cb34-5"><a href="basic-class.html#cb34-5" aria-hidden="true" tabindex="-1"></a>dataset</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 4 columns
## otu_table have 13628 rows and 90 columns
## tax_table have 13330 rows and 7 columns
## phylo_tree have 14096 tips</code></pre>
<p>We also remove OTUs with the taxonomic assignments “mitochondria” or “chloroplast”.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="basic-class.html#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This will remove the lines containing the taxa word regardless of taxonomic ranks and ignoring word case in the tax_table.</span></span>
<span id="cb36-2"><a href="basic-class.html#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co"># So if you want to filter some taxa not considerd pollutions, please use subset like the previous operation to filter tax_table.</span></span>
<span id="cb36-3"><a href="basic-class.html#cb36-3" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">filter_pollution</span>(<span class="at">taxa =</span> <span class="fu">c</span>(<span class="st">&quot;mitochondria&quot;</span>, <span class="st">&quot;chloroplast&quot;</span>))</span></code></pre></div>
<pre><code>## Total 34 features are removed from tax_table ...</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="basic-class.html#cb38-1" aria-hidden="true" tabindex="-1"></a>dataset</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 4 columns
## otu_table have 13628 rows and 90 columns
## tax_table have 13296 rows and 7 columns
## phylo_tree have 14096 tips</code></pre>
<p>To make the OTU and sample information consistent across all files in the dataset object, we use function <code>tidy_dataset</code> to trim the dataset.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="basic-class.html#cb40-1" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb40-2"><a href="basic-class.html#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(dataset)</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 4 columns
## otu_table have 13296 rows and 90 columns
## tax_table have 13296 rows and 7 columns
## phylo_tree have 13296 tips</code></pre>
<p>Then let’s use sample_sums() to check the sequence numbers in each sample.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="basic-class.html#cb42-1" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">sample_sums</span>() <span class="sc">%&gt;%</span> range</span></code></pre></div>
<pre><code>## [1] 10316 37087</code></pre>
<p>Sometimes, in order to reduce the effects of sequencing depth on the diversity measurements,
it is optional to perform the resampling to make the sequence number equal for each sample.
The function <code>rarefy_samples</code> can invoke the function <code>tidy_dataset</code> automatically before and after the rarefying.
In v0.19.0, <code>method = 'SRS'</code> is available to perfom normalization by scaling with ranked subsampling <span class="citation">(<a href="#ref-Beule_normalization_2020" role="doc-biblioref">Beule and Karlovsky 2020</a>)</span>.
The default method is <code>method = 'rarefying'</code>.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="basic-class.html#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># As an example, use 10000 sequences in each sample</span></span>
<span id="cb44-2"><a href="basic-class.html#cb44-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">rarefy_samples</span>(<span class="at">sample.size =</span> <span class="dv">10000</span>)</span></code></pre></div>
<pre><code>## 530 features are removed because they are no longer present in any sample after random subsampling ...</code></pre>
<pre><code>## 530 taxa with 0 abundance are removed from the otu_table ...</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="basic-class.html#cb47-1" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">sample_sums</span>() <span class="sc">%&gt;%</span> range</span></code></pre></div>
<pre><code>## [1] 10000 10000</code></pre>
<p>For v0.17.0, the function <code>save_table</code> can be performed to save all the basic data in microtable object to local files,
including feature abundance, metadata, taxonomic table, phylogenetic tree and representative sequences.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="basic-class.html#cb49-1" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">save_table</span>(<span class="at">dirpath =</span> <span class="st">&quot;basic_files&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;,&quot;</span>)</span></code></pre></div>
<p>Then, let’s calculate the taxa abundance at each taxonomic rank using <code>cal_abund()</code>.
This function <strong>generate a list called taxa_abund stored in the microtable object</strong>.
This list contain several data frame of the abundance information at each taxonomic rank.
It’s worth noting that the <code>cal_abund()</code> function can be used to <strong>solve more complicated cases with special parameters</strong>,
such as supporting both the relative and absolute abundance calculation and selecting the partial ‘taxonomic’ columns.
Those have been shown in file2meco package part (<a href="https://chiliubio.github.io/microeco_tutorial/file2meco-package.html#humann-metagenomic-results" class="uri">https://chiliubio.github.io/microeco_tutorial/file2meco-package.html#humann-metagenomic-results</a>) with complex metagenomic dataset.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="basic-class.html#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use default parameters</span></span>
<span id="cb50-2"><a href="basic-class.html#cb50-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">cal_abund</span>()</span></code></pre></div>
<pre><code>## The result is stored in object$taxa_abund ...</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="basic-class.html#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return dataset$taxa_abund</span></span>
<span id="cb52-2"><a href="basic-class.html#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(dataset<span class="sc">$</span>taxa_abund)</span></code></pre></div>
<pre><code>## [1] &quot;list&quot;</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="basic-class.html#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show part of the relative abundance at Phylum level</span></span>
<span id="cb54-2"><a href="basic-class.html#cb54-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>taxa_abund<span class="sc">$</span>Phylum[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<table style="width:100%;">
<colgroup>
<col width="44%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">S1</th>
<th align="center">S2</th>
<th align="center">S3</th>
<th align="center">S4</th>
<th align="center">S5</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">**k__Bacteria|p__Proteobacteria**</td>
<td align="center">0.2008</td>
<td align="center">0.1996</td>
<td align="center">0.2151</td>
<td align="center">0.261</td>
<td align="center">0.1663</td>
</tr>
<tr class="even">
<td align="center">**k__Bacteria|p__Chloroflexi**</td>
<td align="center">0.1215</td>
<td align="center">0.1937</td>
<td align="center">0.1588</td>
<td align="center">0.1471</td>
<td align="center">0.3098</td>
</tr>
<tr class="odd">
<td align="center">**k__Bacteria|p__Bacteroidetes**</td>
<td align="center">0.1816</td>
<td align="center">0.0359</td>
<td align="center">0.0267</td>
<td align="center">0.0215</td>
<td align="center">0.0266</td>
</tr>
<tr class="even">
<td align="center">**k__Bacteria|p__Acidobacteria**</td>
<td align="center">0.1215</td>
<td align="center">0.2467</td>
<td align="center">0.2532</td>
<td align="center">0.262</td>
<td align="center">0.2482</td>
</tr>
<tr class="odd">
<td align="center">**k__Bacteria|p__Actinobacteria**</td>
<td align="center">0.1182</td>
<td align="center">0.0861</td>
<td align="center">0.0875</td>
<td align="center">0.0954</td>
<td align="center">0.0824</td>
</tr>
</tbody>
</table>
<p>The function <code>save_abund()</code> can be used to save the taxa abundance file to a local place easily.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="basic-class.html#cb55-1" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">save_abund</span>(<span class="at">dirpath =</span> <span class="st">&quot;taxa_abund&quot;</span>)</span></code></pre></div>
<p>All the abundance tables can also be merged into one to save from v0.15.0.
This type of file format can be opened directly by other software, such as STAMP.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="basic-class.html#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tab-delimited, i.e. mpa format</span></span>
<span id="cb56-2"><a href="basic-class.html#cb56-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">save_abund</span>(<span class="at">merge_all =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span>
<span id="cb56-3"><a href="basic-class.html#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co"># remove those unclassified</span></span>
<span id="cb56-4"><a href="basic-class.html#cb56-4" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">save_abund</span>(<span class="at">merge_all =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">rm_un =</span> <span class="cn">TRUE</span>, <span class="at">rm_pattern =</span> <span class="st">&quot;__$|Sedis$&quot;</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Then, let’s calculate the alpha diversity.
The result is also stored in the object microtable automatically.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="basic-class.html#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to add Faith&#39;s phylogenetic diversity, use PD = TRUE, this will be a little slow</span></span>
<span id="cb57-2"><a href="basic-class.html#cb57-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">cal_alphadiv</span>(<span class="at">PD =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## The result is stored in object$alpha_diversity ...</code></pre>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="basic-class.html#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return dataset$alpha_diversity</span></span>
<span id="cb59-2"><a href="basic-class.html#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(dataset<span class="sc">$</span>alpha_diversity)</span></code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="basic-class.html#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save dataset$alpha_diversity to a directory</span></span>
<span id="cb61-2"><a href="basic-class.html#cb61-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">save_alphadiv</span>(<span class="at">dirpath =</span> <span class="st">&quot;alpha_diversity&quot;</span>)</span></code></pre></div>
<p>Let’s go on to beta diversity with function <code>cal_betadiv()</code>.
If method parameter is not provided, the function automatically calculates Bray-curtis, Jaccard, weighted Unifrac and unweighted unifrac matrixes <span class="citation">(<a href="#ref-Lozupone_UniFrac_2005" role="doc-biblioref">Lozupone and Knight 2005</a>)</span>.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="basic-class.html#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># unifrac = FALSE means do not calculate unifrac metric</span></span>
<span id="cb62-2"><a href="basic-class.html#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require GUniFrac package installed</span></span>
<span id="cb62-3"><a href="basic-class.html#cb62-3" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">cal_betadiv</span>(<span class="at">unifrac =</span> <span class="cn">TRUE</span>)</span>
<span id="cb62-4"><a href="basic-class.html#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="co"># return dataset$beta_diversity</span></span>
<span id="cb62-5"><a href="basic-class.html#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(dataset<span class="sc">$</span>beta_diversity)</span>
<span id="cb62-6"><a href="basic-class.html#cb62-6" aria-hidden="true" tabindex="-1"></a><span class="co"># save dataset$beta_diversity to a directory</span></span>
<span id="cb62-7"><a href="basic-class.html#cb62-7" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">save_betadiv</span>(<span class="at">dirpath =</span> <span class="st">&quot;beta_diversity&quot;</span>)</span></code></pre></div>
</div>
<div id="merge-taxa-or-samples" class="section level3 hasAnchor" number="3.1.4">
<h3><span class="header-section-number">3.1.4</span> merge taxa or samples<a href="basic-class.html#merge-taxa-or-samples" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Merging taxa according to a specific taxonomic rank level of tax_table can generate a new microtable object.
In the new microtable object, each feature in otu_table represents one taxon at the output level.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="basic-class.html#cb63-1" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_taxa</span>(<span class="at">taxa =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb63-2"><a href="basic-class.html#cb63-2" aria-hidden="true" tabindex="-1"></a>test</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 4 columns
## otu_table have 1245 rows and 90 columns
## tax_table have 1245 rows and 6 columns</code></pre>
<p>Similarly, merging samples according to a specific group of sample_table can also generate a new microtable object.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="basic-class.html#cb65-1" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb65-2"><a href="basic-class.html#cb65-2" aria-hidden="true" tabindex="-1"></a>test</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 3 rows and 1 columns
## otu_table have 12766 rows and 3 columns
## tax_table have 12766 rows and 7 columns
## phylo_tree have 12766 tips</code></pre>
</div>
<div id="subset-of-samples" class="section level3 hasAnchor" number="3.1.5">
<h3><span class="header-section-number">3.1.5</span> subset of samples<a href="basic-class.html#subset-of-samples" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We donnot provide a special function to filter samples in microtable object, as we think it is redundant.
<strong>We recommend manipulating the sample_table in microtable object directly.</strong>
For example, if you want to extract samples of ‘CW’ group, please do like this:</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="basic-class.html#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remember first clone the whole dataset</span></span>
<span id="cb67-2"><a href="basic-class.html#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="co"># see https://chiliubio.github.io/microeco_tutorial/notes.html#clone-function</span></span>
<span id="cb67-3"><a href="basic-class.html#cb67-3" aria-hidden="true" tabindex="-1"></a>group_CW <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb67-4"><a href="basic-class.html#cb67-4" aria-hidden="true" tabindex="-1"></a><span class="co"># select &#39;CW&#39;</span></span>
<span id="cb67-5"><a href="basic-class.html#cb67-5" aria-hidden="true" tabindex="-1"></a>group_CW<span class="sc">$</span>sample_table <span class="ot">&lt;-</span> <span class="fu">subset</span>(group_CW<span class="sc">$</span>sample_table, Group <span class="sc">==</span> <span class="st">&quot;CW&quot;</span>)</span>
<span id="cb67-6"><a href="basic-class.html#cb67-6" aria-hidden="true" tabindex="-1"></a><span class="co"># or: group_CW$sample_table &lt;- subset(group_CW$sample_table, grepl(&quot;CW&quot;, Group))</span></span>
<span id="cb67-7"><a href="basic-class.html#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="co"># use tidy_dataset to trim all the basic files</span></span>
<span id="cb67-8"><a href="basic-class.html#cb67-8" aria-hidden="true" tabindex="-1"></a>group_CW<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb67-9"><a href="basic-class.html#cb67-9" aria-hidden="true" tabindex="-1"></a>group_CW</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 30 rows and 4 columns
## otu_table have 9727 rows and 30 columns
## tax_table have 9727 rows and 7 columns
## phylo_tree have 9727 tips
## Taxa abundance: calculated for Kingdom,Phylum,Class,Order,Family,Genus,Species 
## Alpha diversity: calculated for Observed,Chao1,se.chao1,ACE,se.ACE,Shannon,Simpson,InvSimpson,Fisher,Pielou,Coverage 
## Beta diversity: calculated for bray,jaccard</code></pre>
</div>
<div id="subset-of-taxa" class="section level3 hasAnchor" number="3.1.6">
<h3><span class="header-section-number">3.1.6</span> subset of taxa<a href="basic-class.html#subset-of-taxa" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Similar with above operation, <strong>subset of features can be achieved by manipulating the tax_table in microtable object directly.</strong></p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="basic-class.html#cb69-1" aria-hidden="true" tabindex="-1"></a>proteo <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb69-2"><a href="basic-class.html#cb69-2" aria-hidden="true" tabindex="-1"></a>proteo<span class="sc">$</span>tax_table <span class="ot">&lt;-</span> <span class="fu">subset</span>(proteo<span class="sc">$</span>tax_table, Phylum <span class="sc">==</span> <span class="st">&quot;p__Proteobacteria&quot;</span>)</span>
<span id="cb69-3"><a href="basic-class.html#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="co"># or: proteo$tax_table &lt;- subset(proteo$tax_table, grepl(&quot;Proteobacteria&quot;, Phylum))</span></span>
<span id="cb69-4"><a href="basic-class.html#cb69-4" aria-hidden="true" tabindex="-1"></a>proteo<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb69-5"><a href="basic-class.html#cb69-5" aria-hidden="true" tabindex="-1"></a>proteo</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 4 columns
## otu_table have 3052 rows and 90 columns
## tax_table have 3052 rows and 7 columns
## phylo_tree have 3052 tips
## Taxa abundance: calculated for Kingdom,Phylum,Class,Order,Family,Genus,Species 
## Alpha diversity: calculated for Observed,Chao1,se.chao1,ACE,se.ACE,Shannon,Simpson,InvSimpson,Fisher,Pielou,Coverage 
## Beta diversity: calculated for bray,jaccard</code></pre>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="basic-class.html#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># proteo is a new microtable object with all OTUs coming from phylum Proteobacteria</span></span>
<span id="cb71-2"><a href="basic-class.html#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="co"># beta diversity dissimilarity for Proteobacteria</span></span>
<span id="cb71-3"><a href="basic-class.html#cb71-3" aria-hidden="true" tabindex="-1"></a>proteo<span class="sc">$</span><span class="fu">cal_betadiv</span>()</span></code></pre></div>
<pre><code>## The result is stored in object$beta_diversity ...</code></pre>
</div>
<div id="other-examples" class="section level3 hasAnchor" number="3.1.7">
<h3><span class="header-section-number">3.1.7</span> Other examples<a href="basic-class.html#other-examples" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The function <code>add_rownames2taxonomy</code> can add the rownames of tax_table as the last column of tax_table directly.
This operation is very useful in some analysis, e.g. biomarker finding at OTU/ASV level with the relative abundance.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="basic-class.html#cb73-1" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb73-2"><a href="basic-class.html#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(test<span class="sc">$</span>tax_table)</span>
<span id="cb73-3"><a href="basic-class.html#cb73-3" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">add_rownames2taxonomy</span>(<span class="at">use_name =</span> <span class="st">&quot;OTU&quot;</span>)</span>
<span id="cb73-4"><a href="basic-class.html#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(test<span class="sc">$</span>tax_table)</span></code></pre></div>
<p>The <code>filter_taxa</code> function can be applied to filter the features with low abundance or occurrence frequency when needed.
For other operations on the features, please directly manipulate the otu_table of your microtable object.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="basic-class.html#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># It is better to have a backup before filtering features</span></span>
<span id="cb74-2"><a href="basic-class.html#cb74-2" aria-hidden="true" tabindex="-1"></a>dataset_filter <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb74-3"><a href="basic-class.html#cb74-3" aria-hidden="true" tabindex="-1"></a>dataset_filter</span>
<span id="cb74-4"><a href="basic-class.html#cb74-4" aria-hidden="true" tabindex="-1"></a><span class="co"># In this example, mean relative abundance threshold 0.0001</span></span>
<span id="cb74-5"><a href="basic-class.html#cb74-5" aria-hidden="true" tabindex="-1"></a><span class="co"># occurrence frequency 0.1; 10% samples have the target features</span></span>
<span id="cb74-6"><a href="basic-class.html#cb74-6" aria-hidden="true" tabindex="-1"></a>dataset_filter<span class="sc">$</span><span class="fu">filter_taxa</span>(<span class="at">rel_abund =</span> <span class="fl">0.0001</span>, <span class="at">freq =</span> <span class="fl">0.1</span>)</span>
<span id="cb74-7"><a href="basic-class.html#cb74-7" aria-hidden="true" tabindex="-1"></a>dataset_filter</span></code></pre></div>
<p>In microtable$new, if auto_tidy = TRUE, the function can automatically use <code>tidy_dataset</code> to make all files uniform.
Then, all other functions in microtable will also do this. But if the user changes the file in microtable object,
the class can not recognize this modification, the user should use <code>tidy_dataset</code> function to manually trim the microtable object.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="basic-class.html#cb75-1" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">sample_table =</span> sample_info_16S[<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, ], <span class="at">otu_table =</span> otu_table_16S, <span class="at">auto_tidy =</span> <span class="cn">FALSE</span>)</span>
<span id="cb75-2"><a href="basic-class.html#cb75-2" aria-hidden="true" tabindex="-1"></a>test</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 40 rows and 4 columns
## otu_table have 13628 rows and 90 columns</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="basic-class.html#cb77-1" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">sample_table =</span> sample_info_16S[<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>, ], <span class="at">otu_table =</span> otu_table_16S, <span class="at">auto_tidy =</span> <span class="cn">TRUE</span>)</span>
<span id="cb77-2"><a href="basic-class.html#cb77-2" aria-hidden="true" tabindex="-1"></a>test1</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 40 rows and 4 columns
## otu_table have 12747 rows and 40 columns</code></pre>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="basic-class.html#cb79-1" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span>sample_table <span class="sc">%&lt;&gt;%</span> .[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, ]</span>
<span id="cb79-2"><a href="basic-class.html#cb79-2" aria-hidden="true" tabindex="-1"></a>test1</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 10 rows and 4 columns
## otu_table have 12747 rows and 40 columns</code></pre>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="basic-class.html#cb81-1" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb81-2"><a href="basic-class.html#cb81-2" aria-hidden="true" tabindex="-1"></a>test1</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 10 rows and 4 columns
## otu_table have 8864 rows and 10 columns</code></pre>
<p>The phylogenetic tree can be read with <code>read.tree</code> function in ape package.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="basic-class.html#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use the example data rep_phylo.tre in file2meco package https://chiliubio.github.io/microeco_tutorial/file2meco-package.html#qiime</span></span>
<span id="cb83-2"><a href="basic-class.html#cb83-2" aria-hidden="true" tabindex="-1"></a>phylo_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;rep_phylo.tre&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb83-3"><a href="basic-class.html#cb83-3" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> ape<span class="sc">::</span><span class="fu">read.tree</span>(phylo_file_path)</span></code></pre></div>
<p>Other functions and examples are listed here.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="basic-class.html#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clone a complete dataset named test</span></span>
<span id="cb84-2"><a href="basic-class.html#cb84-2" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb84-3"><a href="basic-class.html#cb84-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-4"><a href="basic-class.html#cb84-4" aria-hidden="true" tabindex="-1"></a><span class="co"># rename features in all the files of microtable object</span></span>
<span id="cb84-5"><a href="basic-class.html#cb84-5" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">rename_taxa</span>(<span class="at">newname_prefix =</span> <span class="st">&quot;new_name_&quot;</span>)</span>
<span id="cb84-6"><a href="basic-class.html#cb84-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(test<span class="sc">$</span>otu_table)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb84-7"><a href="basic-class.html#cb84-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(test<span class="sc">$</span>tax_table)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb84-8"><a href="basic-class.html#cb84-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-9"><a href="basic-class.html#cb84-9" aria-hidden="true" tabindex="-1"></a><span class="co"># sum the abundance for each taxon</span></span>
<span id="cb84-10"><a href="basic-class.html#cb84-10" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">taxa_sums</span>()</span>
<span id="cb84-11"><a href="basic-class.html#cb84-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-12"><a href="basic-class.html#cb84-12" aria-hidden="true" tabindex="-1"></a><span class="co"># output sample names of microtable object</span></span>
<span id="cb84-13"><a href="basic-class.html#cb84-13" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">sample_names</span>()[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb84-14"><a href="basic-class.html#cb84-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-15"><a href="basic-class.html#cb84-15" aria-hidden="true" tabindex="-1"></a><span class="co"># output taxa names of microtable object</span></span>
<span id="cb84-16"><a href="basic-class.html#cb84-16" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">taxa_names</span>()[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
</div>
<div id="key-points" class="section level3 hasAnchor" number="3.1.8">
<h3><span class="header-section-number">3.1.8</span> Key points<a href="basic-class.html#key-points" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>sample_table: rownames of sample_table must be sample names used</li>
<li>otu_table: rownames must be feature names; colnames must be sample names</li>
<li><code>microtable</code> class: creating microtable object requires at least one file input (otu_table)</li>
<li><code>tidy_taxonomy()</code>: necessary to make taxonomic table have unified format</li>
<li><code>tidy_dataset()</code>: necessary to trim files in microtable object</li>
<li><code>add_rownames2taxonomy()</code>: add the rownames of tax_table as the last column of tax_table</li>
<li><code>cal_abund()</code>: powerful and flexible to cope with complex cases in tax_table, see the parameters</li>
<li>taxa_abund: taxa_abund is a list stored in microtable object and have several data frame</li>
<li>beta_diversity: beta_diversity is a list stored in microtable object and have several distance matrix</li>
</ul>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-An_Soil_2019" class="csl-entry">
An, Jiaxing, Chi Liu, Qian Wang, Minjie Yao, Junpeng Rui, Shiheng Zhang, and Xiangzhen Li. 2019. <span>“Soil Bacterial Community Structure in Chinese Wetlands.”</span> Journal Article. <em>Geoderma</em> 337: 290–99.
</div>
<div id="ref-Beule_normalization_2020" class="csl-entry">
Beule, Lukas, and Petr Karlovsky. 2020. <span>“<span class="nocase">Improved normalization of species count data in ecology by scaling with ranked subsampling (SRS): application to microbial communities</span>.”</span> Journal Article. <em>PeerJ</em> 8: e9593. <a href="https://doi.org/10.7717/peerj.9593">https://doi.org/10.7717/peerj.9593</a>.
</div>
<div id="ref-Bolyen_Reproducible_2019" class="csl-entry">
Bolyen, E., J. R. Rideout, M. R. Dillon, N. A. Bokulich, C. C. Abnet, G. A. Al-Ghalith, H. Alexander, et al. 2019. <span>“Reproducible, Interactive, Scalable and Extensible Microbiome Data Science Using QIIME 2.”</span> Journal Article. <em>Nat Biotechnol</em> 37 (8): 852–57. <a href="https://doi.org/10.1038/s41587-019-0209-9">https://doi.org/10.1038/s41587-019-0209-9</a>.
</div>
<div id="ref-Caporaso_QIIME_2010" class="csl-entry">
Caporaso, J. Gregory, Justin Kuczynski, Jesse Stombaugh, Kyle Bittinger, Frederic D. Bushman, Elizabeth K. Costello, Noah Fierer, et al. 2010. <span>“QIIME Allows Analysis of High-Throughput Community Sequencing Data.”</span> Journal Article. <em>Nature Methods</em> 7 (5): 335–36. https://doi.org/<a href="http://www.nature.com/nmeth/journal/v7/n5/suppinfo/nmeth.f.303_S1.html">http://www.nature.com/nmeth/journal/v7/n5/suppinfo/nmeth.f.303_S1.html</a>.
</div>
<div id="ref-Franzosa_Species_2018" class="csl-entry">
Franzosa, Eric A., Lauren J. McIver, Gholamali Rahnavard, Luke R. Thompson, Melanie Schirmer, George Weingart, Karen Schwarzberg Lipson, et al. 2018. <span>“Species-Level Functional Profiling of Metagenomes and Metatranscriptomes.”</span> Journal Article. <em>Nature Methods</em> 15 (11): 962–68. <a href="https://doi.org/10.1038/s41592-018-0176-y">https://doi.org/10.1038/s41592-018-0176-y</a>.
</div>
<div id="ref-Lozupone_UniFrac_2005" class="csl-entry">
Lozupone, C., and R. Knight. 2005. <span>“<span class="nocase">UniFrac: a new phylogenetic method for comparing microbial communities</span>.”</span> Journal Article. <em>Applied and Environmental Microbiology</em> 71 (12): 8228–35. <a href="https://doi.org/10.1128/AEM.71.12.8228-8235.2005">https://doi.org/10.1128/AEM.71.12.8228-8235.2005</a>.
</div>
<div id="ref-Mcmurdie_phyloseq_2013" class="csl-entry">
Mcmurdie, Paul J, and Susan Holmes. 2013. <span>“Phyloseq: An r Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.”</span> Journal Article. <em>Plos One</em> 8 (4): e61217.
</div>
<div id="ref-Ozkurt_LotuS2_2022" class="csl-entry">
Ozkurt, E., J. Fritscher, N. Soranzo, D. Y. K. Ng, R. P. Davey, M. Bahram, and F. Hildebrand. 2022. <span>“LotuS2: An Ultrafast and Highly Accurate Tool for Amplicon Sequencing Analysis.”</span> Journal Article. <em>Microbiome</em> 10 (1): 176. <a href="https://doi.org/10.1186/s40168-022-01365-1">https://doi.org/10.1186/s40168-022-01365-1</a>.
</div>
<div id="ref-Schloss_Introducing_2009" class="csl-entry">
Schloss, P. D., S. L. Westcott, T. Ryabin, J. R. Hall, M. Hartmann, E. B. Hollister, R. A. Lesniewski, et al. 2009. <span>“Introducing Mothur: Open-Source, Platform-Independent, Community-Supported Software for Describing and Comparing Microbial Communities.”</span> Journal Article. <em>Applied and Environmental Microbiology</em> 75 (23): 7537–41. <a href="https://doi.org/10.1128/aem.01541-09">https://doi.org/10.1128/aem.01541-09</a>.
</div>
<div id="ref-Truong_MeTApHLaN2_2015" class="csl-entry">
Truong, Duy Tin, Eric A Franzosa, Timothy L Tickle, Matthias Scholz, George Weingart, Edoardo Pasolli, Adrian Tett, Curtis Huttenhower, and Nicola Segata. 2015. <span>“MetaPhlAn2 for Enhanced Metagenomic Taxonomic Profiling.”</span> Journal Article. <em>Nature Methods</em> 12 (10): 902–3.
</div>
<div id="ref-Wood_Improved_2019" class="csl-entry">
Wood, D. E., J. Lu, and B. Langmead. 2019. <span>“Improved Metagenomic Analysis with Kraken 2.”</span> Journal Article. <em>Genome Biol</em> 20 (1): 257. <a href="https://doi.org/10.1186/s13059-019-1891-0">https://doi.org/10.1186/s13059-019-1891-0</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="intro.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="composition-based-class.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/02-Basic_class.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
