<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Composition-based class | Tutorial for R microeco package (v0.17.0)</title>
  <meta name="description" content="The tutorial for R microeco, file2meco, meconetcomp and mecoturn packages" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Composition-based class | Tutorial for R microeco package (v0.17.0)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecoturn packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Composition-based class | Tutorial for R microeco package (v0.17.0)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecoturn packages" />
  

<meta name="author" content="Chi Liu, Felipe R. P. Mansoldo, Umer Zeeshan Ijaz, Chenhao Li, Yang Cao, Jarrod J. Scott, Yaoming Cui, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2023-05-05" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="basic-class.html"/>
<link rel="next" href="diversity-based-class.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#rtools"><i class="fa fa-check"></i><b>2.4</b> RTools</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.5</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#description"><i class="fa fa-check"></i><b>2.5.1</b> Description</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#cran-packages"><i class="fa fa-check"></i><b>2.5.2</b> CRAN packages</a></li>
<li class="chapter" data-level="2.5.3" data-path="intro.html"><a href="intro.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor packages</a></li>
<li class="chapter" data-level="2.5.4" data-path="intro.html"><a href="intro.html#github-packages"><i class="fa fa-check"></i><b>2.5.4</b> Github packages</a></li>
<li class="chapter" data-level="2.5.5" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.5.5</b> Gephi</a></li>
<li class="chapter" data-level="2.5.6" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.5.6</b> Tax4Fun</a></li>
<li class="chapter" data-level="2.5.7" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.5.7</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#plot"><i class="fa fa-check"></i><b>2.6</b> Plot</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#prepare-the-example-data"><i class="fa fa-check"></i><b>3.1.1</b> Prepare the example data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#how-to-read-your-files-to-microtable-object"><i class="fa fa-check"></i><b>3.1.2</b> How to read your files to microtable object?</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#functions-in-microtable-class"><i class="fa fa-check"></i><b>3.1.3</b> Functions in microtable class</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#merge-taxa-or-samples"><i class="fa fa-check"></i><b>3.1.4</b> merge taxa or samples</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-class.html"><a href="basic-class.html#subset-of-samples"><i class="fa fa-check"></i><b>3.1.5</b> subset of samples</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-class.html"><a href="basic-class.html#subset-of-taxa"><i class="fa fa-check"></i><b>3.1.6</b> subset of taxa</a></li>
<li class="chapter" data-level="3.1.7" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.7</b> Other examples</a></li>
<li class="chapter" data-level="3.1.8" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.8</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-2"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-4"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#network-comparison"><i class="fa fa-check"></i><b>6.2.2</b> Network comparison</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.3</b> Key points</a></li>
<li class="chapter" data-level="6.2.4" data-path="model-based-class.html"><a href="model-based-class.html#other-functions"><i class="fa fa-check"></i><b>6.2.4</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies</a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples</a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points</a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-7"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME</a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2</a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann"><i class="fa fa-check"></i><b>8.3</b> HUMAnN</a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn</a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2"><i class="fa fa-check"></i><b>8.5</b> Kraken2</a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB</a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq</a></li>
<li class="chapter" data-level="8.8" data-path="file2meco-package.html"><a href="file2meco-package.html#picrust2"><i class="fa fa-check"></i><b>8.8</b> PICRUSt2</a></li>
<li class="chapter" data-level="8.9" data-path="file2meco-package.html"><a href="file2meco-package.html#viromescan"><i class="fa fa-check"></i><b>8.9</b> ViromeScan</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset</a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data</a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> <code>clone</code> function</a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> <code>save</code> function</a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#github-issues"><i class="fa fa-check"></i><b>10.3</b> Github Issues</a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.4</b> change object</a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.5</b> group order</a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.6</b> add layers to plot</a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#colors-for-more-groups"><i class="fa fa-check"></i><b>10.7</b> colors for more groups</a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>11</b> mecodev package</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>11.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="11.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_norm"><i class="fa fa-check"></i><b>11.2</b> trans_norm</a></li>
<li class="chapter" data-level="11.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>11.3</b> trans_ts</a></li>
<li class="chapter" data-level="11.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>11.4</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html"><i class="fa fa-check"></i><b>12</b> meconetcomp package</a>
<ul>
<li class="chapter" data-level="12.1" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-modularity-for-all-networks"><i class="fa fa-check"></i><b>12.1</b> Network modularity for all networks</a></li>
<li class="chapter" data-level="12.2" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-topological-attributes-for-all-networks"><i class="fa fa-check"></i><b>12.2</b> Network topological attributes for all networks</a></li>
<li class="chapter" data-level="12.3" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#node-and-edge-properties-extraction-for-all-networks"><i class="fa fa-check"></i><b>12.3</b> Node and edge properties extraction for all networks</a></li>
<li class="chapter" data-level="12.4" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-nodes-across-networks"><i class="fa fa-check"></i><b>12.4</b> Compare nodes across networks</a></li>
<li class="chapter" data-level="12.5" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-edges-across-networks"><i class="fa fa-check"></i><b>12.5</b> Compare edges across networks</a></li>
<li class="chapter" data-level="12.6" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#extract-overlapped-edges-of-networks-to-a-new-network"><i class="fa fa-check"></i><b>12.6</b> Extract overlapped edges of networks to a new network</a></li>
<li class="chapter" data-level="12.7" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-phylogenetic-distances-of-paired-nodes-in-edges"><i class="fa fa-check"></i><b>12.7</b> Compare phylogenetic distances of paired nodes in edges</a></li>
<li class="chapter" data-level="12.8" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-node-sources-of-edges-across-networks"><i class="fa fa-check"></i><b>12.8</b> Compare node sources of edges across networks</a></li>
<li class="chapter" data-level="12.9" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-topological-properties-of-sub-networks"><i class="fa fa-check"></i><b>12.9</b> Compare topological properties of sub-networks</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mecoturn-package.html"><a href="mecoturn-package.html"><i class="fa fa-check"></i><b>13</b> mecoturn package</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mecoturn-package.html"><a href="mecoturn-package.html#betaturn-class"><i class="fa fa-check"></i><b>13.1</b> betaturn class</a></li>
<li class="chapter" data-level="13.2" data-path="mecoturn-package.html"><a href="mecoturn-package.html#taxaturn-class"><i class="fa fa-check"></i><b>13.2</b> taxaturn class</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>14</b> Other examples</a>
<ul>
<li class="chapter" data-level="14.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>14.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="14.2" data-path="other-examples-1.html"><a href="other-examples-1.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>14.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="14.3" data-path="other-examples-1.html"><a href="other-examples-1.html#show-the-abundance-of-unknown-taxa"><i class="fa fa-check"></i><b>14.3</b> Show the abundance of unknown taxa</a></li>
<li class="chapter" data-level="14.4" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>14.4</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="14.5" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>14.5</b> The use of phylogenetic tree</a></li>
<li class="chapter" data-level="14.6" data-path="other-examples-1.html"><a href="other-examples-1.html#mantel-test-correlation-heatmap"><i class="fa fa-check"></i><b>14.6</b> Mantel test + Correlation heatmap</a></li>
<li class="chapter" data-level="14.7" data-path="other-examples-1.html"><a href="other-examples-1.html#pcoa-boxplot"><i class="fa fa-check"></i><b>14.7</b> PCoA + boxplot</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v0.17.0)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="composition-based-class" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">Chapter 4</span> Composition-based class<a href="composition-based-class.html#composition-based-class" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The trans_abund class and trans_venn class are organised into the section ‘Composition-based class’,
since they are mainly used to show the composition information of communities.</p>
<div id="trans_abund-class" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> trans_abund class<a href="composition-based-class.html#trans_abund-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　The trans_abund class has several functions to visualize taxonomic abundance based on the ggplot2 package.</p>
<div id="example" class="section level3 hasAnchor" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Example<a href="composition-based-class.html#example" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We first show the bar plot example.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="composition-based-class.html#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_abund object</span></span>
<span id="cb85-2"><a href="composition-based-class.html#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="co"># select top 8 abundant Phyla.</span></span>
<span id="cb85-3"><a href="composition-based-class.html#cb85-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>)</span></code></pre></div>
<pre><code>## The transformed abundance data is stored in object$data_abund ...</code></pre>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="composition-based-class.html#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># t1 object now include the transformed abundance data t1$abund_data and other elements for the following plotting</span></span></code></pre></div>
<p>As the sample number is large, we do not show the sample names in x axis and add the facet to show abundance according to groups.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="composition-based-class.html#cb88-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">others_color =</span> <span class="st">&quot;grey70&quot;</span>, <span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">xtext_keep =</span> <span class="cn">FALSE</span>, <span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span>
<span id="cb88-2"><a href="composition-based-class.html#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return a ggplot2 object</span></span></code></pre></div>
<p><img src="Images/trans_abund_barplot1.png" width="750px" style="display: block; margin: auto;" /></p>
<p>Two or more facets are supported with the facet parameter from v0.14.0 by providing a vector with multiple elements.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="composition-based-class.html#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require package ggh4x, first run install.packages(&quot;ggh4x&quot;) if not installed</span></span>
<span id="cb89-2"><a href="composition-based-class.html#cb89-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">others_color =</span> <span class="st">&quot;grey70&quot;</span>, <span class="at">facet =</span> <span class="fu">c</span>(<span class="st">&quot;Group&quot;</span>, <span class="st">&quot;Type&quot;</span>), <span class="at">xtext_keep =</span> <span class="cn">FALSE</span>, <span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>, <span class="at">barwidth =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Images/trans_abund_barplot_facet2.png" width="750px" style="display: block; margin: auto;" /></p>
<p>The default operation can filter all the unclassified taxa (i.e. p__ or g__ in tax_table that has been processed by <code>tidy_taxonomy</code> function),
as those unknown taxa are generally meaningless.
However sometimes, these unknown taxa may be meaningful for users.
For example, if one want to isolate some unknown species, it is valuable to check the abundance of those unknown taxa.
At this time, please see this topic (<a href="https://github.com/ChiLiubio/microeco/issues/165" class="uri">https://github.com/ChiLiubio/microeco/issues/165</a>) to resolve the issue that how to show unknown taxa with hierarchical taxonomy classification.
The alluvial plot is also implemented in the plot_bar function with use_alluvium parameter.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="composition-based-class.html#cb90-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>)</span>
<span id="cb90-2"><a href="composition-based-class.html#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require ggalluvial package</span></span>
<span id="cb90-3"><a href="composition-based-class.html#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use_alluvium = TRUE make the alluvial plot, clustering =TRUE can be used to reorder the samples by clustering</span></span>
<span id="cb90-4"><a href="composition-based-class.html#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="co"># bar_type = &quot;notfull&quot; can discard &#39;others&#39;; select another color palette</span></span>
<span id="cb90-5"><a href="composition-based-class.html#cb90-5" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">bar_type =</span> <span class="st">&quot;notfull&quot;</span>, <span class="at">use_alluvium =</span> <span class="cn">TRUE</span>, <span class="at">clustering =</span> <span class="cn">TRUE</span>, <span class="at">xtext_angle =</span> <span class="dv">30</span>, <span class="at">xtext_size =</span> <span class="dv">3</span>, <span class="at">color_values =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Set2&quot;</span>))</span></code></pre></div>
<p><img src="Images/trans_abund_bar_allu.png" width="720" style="display: block; margin: auto;" /></p>
<p>The bar plot can also be performed with group mean values.
Note that, from v0.16.0, the parameter <code>group_morestats = TRUE</code> can be used to add more summary statistics in the return <code>data_abund</code> when <code>groupmean</code> parameter is provided.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="composition-based-class.html#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The groupmean parameter can be used to obtain the group-mean barplot.</span></span>
<span id="cb91-2"><a href="composition-based-class.html#cb91-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb91-3"><a href="composition-based-class.html#cb91-3" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">others_color =</span> <span class="st">&quot;grey70&quot;</span>, <span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span>
<span id="cb91-4"><a href="composition-based-class.html#cb91-4" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>))</span></code></pre></div>
<p><img src="Images/trans_abund_barplot_groupmean.png" width="400px" style="display: block; margin: auto;" /></p>
<p>The box plot is an excellent way to intuitionally show abundance distribution across groups.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="composition-based-class.html#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show 15 taxa at Class level</span></span>
<span id="cb92-2"><a href="composition-based-class.html#cb92-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Class&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">15</span>)</span>
<span id="cb92-3"><a href="composition-based-class.html#cb92-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_box</span>(<span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">xtext_angle =</span> <span class="dv">30</span>)</span></code></pre></div>
<p><img src="Images/trans_abund_boxplot.png" width="700px" style="display: block; margin: auto;" /></p>
<p>Then we show the heatmap with the high abundant genera.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="composition-based-class.html#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show 40 taxa at Genus level</span></span>
<span id="cb93-2"><a href="composition-based-class.html#cb93-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">40</span>)</span>
<span id="cb93-3"><a href="composition-based-class.html#cb93-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_heatmap</span>(<span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">xtext_keep =</span> <span class="cn">FALSE</span>, <span class="at">withmargin =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/trans_abund_heatmap.png" width="750px" style="display: block; margin: auto;" /></p>
<p>Line chart is very useful to show the abundance change of taxa along time, space or other gradients.</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="composition-based-class.html#cb94-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">5</span>)</span>
<span id="cb94-2"><a href="composition-based-class.html#cb94-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_line</span>()</span>
<span id="cb94-3"><a href="composition-based-class.html#cb94-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">5</span>, <span class="at">groupmean =</span> <span class="st">&quot;Type&quot;</span>)</span>
<span id="cb94-4"><a href="composition-based-class.html#cb94-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_line</span>(<span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="fl">0.3</span>), <span class="at">xtext_angle =</span> <span class="dv">0</span>)</span></code></pre></div>
<p><img src="Images/trans_abund_line.png" width="750px" style="display: block; margin: auto;" /></p>
<p>Then, we show the pie chart with the group mean values.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="composition-based-class.html#cb95-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">6</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb95-2"><a href="composition-based-class.html#cb95-2" aria-hidden="true" tabindex="-1"></a><span class="co"># all pie chart in one row</span></span>
<span id="cb95-3"><a href="composition-based-class.html#cb95-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_pie</span>(<span class="at">facet_nrow =</span> <span class="dv">1</span>)</span>
<span id="cb95-4"><a href="composition-based-class.html#cb95-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_pie</span>(<span class="at">facet_nrow =</span> <span class="dv">1</span>, <span class="at">add_label =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/trans_abund_pie.png" width="600px" style="display: block; margin: auto;" /></p>
<p>The donut and radar charts are implemented from v0.17.0.
Please install the dependent packages according to the steps (<a href="https://chiliubio.github.io/microeco_tutorial/intro.html#dependence" class="uri">https://chiliubio.github.io/microeco_tutorial/intro.html#dependence</a>).</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="composition-based-class.html#cb96-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb96-2"><a href="composition-based-class.html#cb96-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_donut</span>(<span class="at">label =</span> <span class="cn">FALSE</span>)</span>
<span id="cb96-3"><a href="composition-based-class.html#cb96-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_donut</span>(<span class="at">label =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/trans_abund_donut.png" width="650px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="composition-based-class.html#cb97-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb97-2"><a href="composition-based-class.html#cb97-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_radar</span>(<span class="at">values.radar =</span> <span class="fu">c</span>(<span class="st">&quot;0%&quot;</span>, <span class="st">&quot;25%&quot;</span>, <span class="st">&quot;50%&quot;</span>), <span class="at">grid.min =</span> <span class="dv">0</span>, <span class="at">grid.mid =</span> <span class="fl">0.25</span>, <span class="at">grid.max =</span> <span class="fl">0.5</span>)</span>
<span id="cb97-3"><a href="composition-based-class.html#cb97-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>, <span class="at">groupmean =</span> <span class="st">&quot;Type&quot;</span>)</span>
<span id="cb97-4"><a href="composition-based-class.html#cb97-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_radar</span>(<span class="at">values.radar =</span> <span class="fu">c</span>(<span class="st">&quot;0%&quot;</span>, <span class="st">&quot;25%&quot;</span>, <span class="st">&quot;50%&quot;</span>), <span class="at">grid.min =</span> <span class="dv">0</span>, <span class="at">grid.mid =</span> <span class="fl">0.25</span>, <span class="at">grid.max =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="Images/trans_abund_radar.png" width="700px" style="display: block; margin: auto;" /></p>
<p>The ternary plot can be used for the case with three samples/groups.</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="composition-based-class.html#cb98-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb98-2"><a href="composition-based-class.html#cb98-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_tern</span>()</span></code></pre></div>
<p><img src="Images/trans_abund_ternary.png" width="600px" style="display: block; margin: auto;" /></p>
<p>When the hierarchical abundance data of two levels is needed to be shown in bar plot, the nested legend can be used.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="composition-based-class.html#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require ggnested package; see https://chiliubio.github.io/microeco_tutorial/intro.html#dependence</span></span>
<span id="cb99-2"><a href="composition-based-class.html#cb99-2" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Class&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">high_level =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">prefix =</span> <span class="st">&quot;</span><span class="sc">\\</span><span class="st">|&quot;</span>)</span>
<span id="cb99-3"><a href="composition-based-class.html#cb99-3" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">ggnested =</span> <span class="cn">TRUE</span>, <span class="at">facet =</span> <span class="fu">c</span>(<span class="st">&quot;Group&quot;</span>, <span class="st">&quot;Type&quot;</span>), <span class="at">xtext_angle =</span> <span class="dv">30</span>)</span>
<span id="cb99-4"><a href="composition-based-class.html#cb99-4" aria-hidden="true" tabindex="-1"></a><span class="co"># fixed number in each phylum</span></span>
<span id="cb99-5"><a href="composition-based-class.html#cb99-5" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Class&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">30</span>, <span class="at">show =</span> <span class="dv">0</span>, <span class="at">high_level =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">high_level_fix_nsub =</span> <span class="dv">4</span>)</span>
<span id="cb99-6"><a href="composition-based-class.html#cb99-6" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">ggnested =</span> <span class="cn">TRUE</span>, <span class="at">xtext_angle =</span> <span class="dv">30</span>, <span class="at">facet =</span> <span class="fu">c</span>(<span class="st">&quot;Group&quot;</span>, <span class="st">&quot;Type&quot;</span>))</span>
<span id="cb99-7"><a href="composition-based-class.html#cb99-7" aria-hidden="true" tabindex="-1"></a><span class="co"># sum others in each phylum</span></span>
<span id="cb99-8"><a href="composition-based-class.html#cb99-8" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Class&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">20</span>, <span class="at">show =</span> <span class="dv">0</span>, <span class="at">high_level =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">high_level_fix_nsub =</span> <span class="dv">3</span>, <span class="at">prefix =</span> <span class="st">&quot;</span><span class="sc">\\</span><span class="st">|&quot;</span>)</span>
<span id="cb99-9"><a href="composition-based-class.html#cb99-9" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">ggnested =</span> <span class="cn">TRUE</span>, <span class="at">high_level_add_other =</span> <span class="cn">TRUE</span>, <span class="at">xtext_angle =</span> <span class="dv">30</span>, <span class="at">facet =</span> <span class="fu">c</span>(<span class="st">&quot;Group&quot;</span>, <span class="st">&quot;Type&quot;</span>))</span></code></pre></div>
<p><img src="Images/trans_abund_barplot_ggnested.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<div id="key-points-1" class="section level3 hasAnchor" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Key points<a href="composition-based-class.html#key-points-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>trans_abund$new: creating trans_abund object can invoke taxa_abund in microtable for transformation</li>
<li>color_values parameter: color_values parameter in each function is used for colors selection</li>
<li>input_taxaname parameter: input_taxaname parameter in trans_abund$new can be used to select interested customized taxa instead of abundance-based selection</li>
<li>use_percentage parameter: use_percentage parameter in trans_abund$new - whether show the abundance percentage</li>
</ul>
</div>
</div>
<div id="trans_venn-class" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> trans_venn class<a href="composition-based-class.html#trans_venn-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The trans_venn class is developed for venn analysis, i.e. shared and unique taxa across samples/groups.</p>
<div id="example-1" class="section level3 hasAnchor" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Example<a href="composition-based-class.html#example-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This part can be performed using samples or groups at OTU/ASV level or higher taxonomic level.
To analyze the unique and shared OTUs of groups,
we first merge samples according to the “Group” column of sample_table.</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="composition-based-class.html#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co"># merge samples as one community for each group</span></span>
<span id="cb100-2"><a href="composition-based-class.html#cb100-2" aria-hidden="true" tabindex="-1"></a>dataset1 <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb100-3"><a href="composition-based-class.html#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="co"># dataset1 is a new microtable object</span></span>
<span id="cb100-4"><a href="composition-based-class.html#cb100-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_venn object</span></span>
<span id="cb100-5"><a href="composition-based-class.html#cb100-5" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1, <span class="at">ratio =</span> <span class="cn">NULL</span>)</span>
<span id="cb100-6"><a href="composition-based-class.html#cb100-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_venn</span>()</span></code></pre></div>
<p><img src="Images/trans_venn_0.png" width="500px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="composition-based-class.html#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create venn plot with more information</span></span>
<span id="cb101-2"><a href="composition-based-class.html#cb101-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1, <span class="at">ratio =</span> <span class="st">&quot;seqratio&quot;</span>)</span>
<span id="cb101-3"><a href="composition-based-class.html#cb101-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_venn</span>()</span>
<span id="cb101-4"><a href="composition-based-class.html#cb101-4" aria-hidden="true" tabindex="-1"></a><span class="co"># The integer is OTU number</span></span>
<span id="cb101-5"><a href="composition-based-class.html#cb101-5" aria-hidden="true" tabindex="-1"></a><span class="co"># The percentage data is the sequence number/total sequence number</span></span></code></pre></div>
<p><img src="Images/trans_venn_1.png" width="500px" style="display: block; margin: auto;" /></p>
<p>When the groups are too many to show with venn plot, using petal plot is better.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="composition-based-class.html#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use &quot;Type&quot; column in sample_table</span></span>
<span id="cb102-2"><a href="composition-based-class.html#cb102-2" aria-hidden="true" tabindex="-1"></a>dataset1 <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Type&quot;</span>)</span>
<span id="cb102-3"><a href="composition-based-class.html#cb102-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1)</span>
<span id="cb102-4"><a href="composition-based-class.html#cb102-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_venn</span>(<span class="at">petal_plot =</span> <span class="cn">TRUE</span>, <span class="at">petal_center_size =</span> <span class="dv">50</span>, <span class="at">petal_r =</span> <span class="fl">1.5</span>, <span class="at">petal_a =</span> <span class="dv">3</span>, <span class="at">petal_move_xy =</span> <span class="fl">3.8</span>, <span class="at">petal_color_center =</span> <span class="st">&quot;#BEBADA&quot;</span>)</span></code></pre></div>
<p><img src="Images/trans_venn_2.png" width="500px" style="display: block; margin: auto;" /></p>
<p>Another way to plot the results is to use <code>plot_bar</code> function, which is especially useful for a large number of samples/groups.
This way is generally called UpSet plot. Please see the help document for more parameters to adjust the plot.</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="composition-based-class.html#cb103-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Type&quot;</span>)</span>
<span id="cb103-2"><a href="composition-based-class.html#cb103-2" aria-hidden="true" tabindex="-1"></a>tmp</span>
<span id="cb103-3"><a href="composition-based-class.html#cb103-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> tmp)</span>
<span id="cb103-4"><a href="composition-based-class.html#cb103-4" aria-hidden="true" tabindex="-1"></a><span class="co"># only show some sets with large intersection numbers</span></span>
<span id="cb103-5"><a href="composition-based-class.html#cb103-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>data_summary <span class="sc">%&lt;&gt;%</span> .[.[, <span class="dv">1</span>] <span class="sc">&gt;</span> <span class="dv">20</span>, ]</span>
<span id="cb103-6"><a href="composition-based-class.html#cb103-6" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">left_plot =</span> <span class="cn">TRUE</span>, <span class="at">bottom_height =</span> <span class="fl">0.5</span>, <span class="at">left_width =</span> <span class="fl">0.15</span>, <span class="at">up_bar_fill =</span> <span class="st">&quot;grey50&quot;</span>, <span class="at">left_bar_fill =</span> <span class="st">&quot;grey50&quot;</span>, <span class="at">bottom_point_color =</span> <span class="st">&quot;black&quot;</span>)</span>
<span id="cb103-7"><a href="composition-based-class.html#cb103-7" aria-hidden="true" tabindex="-1"></a>g1</span>
<span id="cb103-8"><a href="composition-based-class.html#cb103-8" aria-hidden="true" tabindex="-1"></a><span class="co"># g1 is aplot class and can be saved with ggplot2::ggsave, aplot::ggsave or cowplot::save_plot function</span></span>
<span id="cb103-9"><a href="composition-based-class.html#cb103-9" aria-hidden="true" tabindex="-1"></a><span class="co"># as g1 is comprised of several sub-plots, please adjust the details for each sub-plot</span></span>
<span id="cb103-10"><a href="composition-based-class.html#cb103-10" aria-hidden="true" tabindex="-1"></a>g1[[<span class="dv">1</span>]]</span>
<span id="cb103-11"><a href="composition-based-class.html#cb103-11" aria-hidden="true" tabindex="-1"></a>g1[[<span class="dv">2</span>]]</span></code></pre></div>
<p><img src="Images/trans_venn_3.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Generally, after getting the intersection results, we do not know who those shared or unique taxa are.
The composition of the unique or shared species may account for the different and similar parts of ecological characteristics across groups<span class="citation">(<a href="#ref-Mendes_Deciphering_2011" role="doc-biblioref">Mendes et al. 2011</a>)</span>.
So, it is interesting to further analyze the composition of unique and shared species.
For this goal, we first transform the results of venn plot to the traditional feature-sample table, that is, another object of microtable class.</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="composition-based-class.html#cb104-1" aria-hidden="true" tabindex="-1"></a>dataset1 <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb104-2"><a href="composition-based-class.html#cb104-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1)</span></code></pre></div>
<pre><code>## The details of each venn part is stored in object$data_details ...</code></pre>
<pre><code>## The venn summary table used for plot is stored in object$data_summary ...</code></pre>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="composition-based-class.html#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># transform venn results to the sample-species table, here do not consider abundance, only use presence/absence.</span></span>
<span id="cb107-2"><a href="composition-based-class.html#cb107-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">trans_comm</span>(<span class="at">use_frequency =</span> <span class="cn">TRUE</span>)</span>
<span id="cb107-3"><a href="composition-based-class.html#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t2 is a new microtable class, each part is considered a sample</span></span>
<span id="cb107-4"><a href="composition-based-class.html#cb107-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(t2)</span></code></pre></div>
<pre><code>## [1] &quot;microtable&quot; &quot;R6&quot;</code></pre>
<p>We use bar plot to show the composition at the Genus level.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="composition-based-class.html#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate taxa abundance, that is, the frequency</span></span>
<span id="cb109-2"><a href="composition-based-class.html#cb109-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb109-3"><a href="composition-based-class.html#cb109-3" aria-hidden="true" tabindex="-1"></a><span class="co"># transform and plot</span></span>
<span id="cb109-4"><a href="composition-based-class.html#cb109-4" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> t2, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>)</span>
<span id="cb109-5"><a href="composition-based-class.html#cb109-5" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">bar_type =</span> <span class="st">&quot;part&quot;</span>, <span class="at">legend_text_italic =</span> T, <span class="at">xtext_angle =</span> <span class="dv">30</span>, <span class="at">color_values =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Set2&quot;</span>),</span>
<span id="cb109-6"><a href="composition-based-class.html#cb109-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">order_x =</span> <span class="fu">c</span>(<span class="st">&quot;IW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;TW&quot;</span>, <span class="st">&quot;IW&amp;CW&quot;</span>, <span class="st">&quot;IW&amp;TW&quot;</span>, <span class="st">&quot;CW&amp;TW&quot;</span>, <span class="st">&quot;IW&amp;CW&amp;TW&quot;</span>)) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Frequency (%)&quot;</span>)</span></code></pre></div>
<p><img src="Images/trans_venn_bar.png" width="650px" style="display: block; margin: auto;" /></p>
<p>We also try to use pie chart to show the compositions at the Phylum level.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="composition-based-class.html#cb110-1" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> t2, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>)</span>
<span id="cb110-2"><a href="composition-based-class.html#cb110-2" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span>data_abund<span class="sc">$</span>Sample <span class="sc">%&lt;&gt;%</span> <span class="fu">factor</span>(., <span class="at">levels =</span> <span class="fu">unique</span>(.))</span>
<span id="cb110-3"><a href="composition-based-class.html#cb110-3" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">plot_pie</span>(<span class="at">facet_nrow =</span> <span class="dv">3</span>, <span class="at">color_values =</span> <span class="fu">c</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Dark2&quot;</span>), <span class="st">&quot;grey50&quot;</span>))</span></code></pre></div>
<p><img src="Images/trans_venn_pie.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Other examples:</p>
<p>To reorder samples in the plots, please manipulate the sample_table in the object to adjust the orders.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="composition-based-class.html#cb111-1" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Type&quot;</span>)</span>
<span id="cb111-2"><a href="composition-based-class.html#cb111-2" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>sample_table <span class="sc">%&lt;&gt;%</span> .[<span class="fu">c</span>(<span class="st">&quot;YML&quot;</span>, <span class="st">&quot;NE&quot;</span>, <span class="st">&quot;NW&quot;</span>, <span class="st">&quot;NC&quot;</span>, <span class="st">&quot;QTP&quot;</span>, <span class="st">&quot;SC&quot;</span>), , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb111-3"><a href="composition-based-class.html#cb111-3" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb111-4"><a href="composition-based-class.html#cb111-4" aria-hidden="true" tabindex="-1"></a><span class="co"># The columns of otu_table can also be reordered according to the sample_table after running tidy_dataset function</span></span>
<span id="cb111-5"><a href="composition-based-class.html#cb111-5" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(test)</span>
<span id="cb111-6"><a href="composition-based-class.html#cb111-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">sort_samples =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>The parameter <code>sort_samples = TRUE</code> in <code>plot_bar</code> function can be applied to sort samples in the y axis according to the number of features.
The left bar plot can be removed when the parameter <code>left_plot = FALSE</code>.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="composition-based-class.html#cb112-1" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Type&quot;</span>)</span>
<span id="cb112-2"><a href="composition-based-class.html#cb112-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(test)</span>
<span id="cb112-3"><a href="composition-based-class.html#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="co"># remove left bar in the UpSet plot</span></span>
<span id="cb112-4"><a href="composition-based-class.html#cb112-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">left_plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb112-5"><a href="composition-based-class.html#cb112-5" aria-hidden="true" tabindex="-1"></a><span class="co"># sort samples in the axis according to the number of features</span></span>
<span id="cb112-6"><a href="composition-based-class.html#cb112-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">sort_samples =</span> <span class="cn">TRUE</span>)</span>
<span id="cb112-7"><a href="composition-based-class.html#cb112-7" aria-hidden="true" tabindex="-1"></a><span class="co"># original orders in test$sample_table</span></span>
<span id="cb112-8"><a href="composition-based-class.html#cb112-8" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">sort_samples =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
</div>
<div id="key-points-2" class="section level3 hasAnchor" number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Key points<a href="composition-based-class.html#key-points-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>ratio parameter: ratio parameter in trans_abund$new control whether and what content appear below the taxa number in venn plot</li>
<li>return data: using trans_venn$new() return data_details and data_summary stored in trans_venn object for further ploting</li>
</ul>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Mendes_Deciphering_2011" class="csl-entry">
Mendes, R, M Kruijt, Bruijn I De, E Dekkers, der Voort M Van, J. H. Schneider, Y. M. Piceno, T. Z. Desantis, G. L. Andersen, and P. A. Bakker. 2011. <span>“Deciphering the Rhizosphere Microbiome for Disease-Suppressive Bacteria.”</span> Journal Article. <em>Science</em> 332 (6033): 1097–1100.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="basic-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="diversity-based-class.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/03-Composition-based_class.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
