<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Composition-based class | Tutorial for R microeco package (v0.7.0)</title>
  <meta name="description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Composition-based class | Tutorial for R microeco package (v0.7.0)" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Composition-based class | Tutorial for R microeco package (v0.7.0)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  

<meta name="author" content="Chi Liu, Felipe R. P. Mansoldo, Umer Zeeshan Ijaz, Chenhao Li, Yang Cao, Jarrod J. Scott, Yaoming Cui, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2022-03-19" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="basic-class.html"/>
<link rel="next" href="diversity-based-class.html"/>
<script src="libs/header-attrs-2.12/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background<span></span></a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction<span></span></a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.1</b> R6 Class<span></span></a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.2</b> Help<span></span></a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.3</b> Dependence<span></span></a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="intro.html"><a href="intro.html#important-packages"><i class="fa fa-check"></i><b>2.3.1</b> Important packages<span></span></a></li>
<li class="chapter" data-level="2.3.2" data-path="intro.html"><a href="intro.html#ggtree"><i class="fa fa-check"></i><b>2.3.2</b> ggtree<span></span></a></li>
<li class="chapter" data-level="2.3.3" data-path="intro.html"><a href="intro.html#spieceasi"><i class="fa fa-check"></i><b>2.3.3</b> SpiecEasi<span></span></a></li>
<li class="chapter" data-level="2.3.4" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.3.4</b> Gephi<span></span></a></li>
<li class="chapter" data-level="2.3.5" data-path="intro.html"><a href="intro.html#wgcna"><i class="fa fa-check"></i><b>2.3.5</b> WGCNA<span></span></a></li>
<li class="chapter" data-level="2.3.6" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.3.6</b> Tax4Fun<span></span></a></li>
<li class="chapter" data-level="2.3.7" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.3.7</b> Tax4Fun2<span></span></a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#plotting"><i class="fa fa-check"></i><b>2.4</b> Plotting<span></span></a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class<span></span></a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class<span></span></a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#example"><i class="fa fa-check"></i><b>3.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.2</b> Other examples<span></span></a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.3</b> Key points<span></span></a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#other-functions"><i class="fa fa-check"></i><b>3.1.4</b> Other functions<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class<span></span></a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class<span></span></a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class<span></span></a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-2"><i class="fa fa-check"></i><b>4.2.1</b> Example<span></span></a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class<span></span></a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class<span></span></a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class<span></span></a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-4"><i class="fa fa-check"></i><b>5.2.1</b> Example<span></span></a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class<span></span></a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class<span></span></a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.2</b> Key points<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class<span></span></a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.2.1</b> Example<span></span></a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.2</b> Key points<span></span></a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#other-functions-1"><i class="fa fa-check"></i><b>6.2.3</b> Other functions<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class<span></span></a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-7"><i class="fa fa-check"></i><b>6.3.1</b> Example<span></span></a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points<span></span></a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class<span></span></a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class<span></span></a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-8"><i class="fa fa-check"></i><b>7.1.2</b> Key points<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class<span></span></a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-9"><i class="fa fa-check"></i><b>7.2.1</b> Example<span></span></a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.2.2</b> Key points<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package<span></span></a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME<span></span></a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2<span></span></a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann-metagenomic-results"><i class="fa fa-check"></i><b>8.3</b> HUMAnN metagenomic results<span></span></a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn<span></span></a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2"><i class="fa fa-check"></i><b>8.5</b> Kraken2<span></span></a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB<span></span></a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq<span></span></a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset<span></span></a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data<span></span></a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data<span></span></a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes<span></span></a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> clone() function<span></span></a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> save() function<span></span></a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#subset-of-samples"><i class="fa fa-check"></i><b>10.3</b> subset of samples<span></span></a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.4</b> change object<span></span></a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.5</b> group order<span></span></a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.6</b> add layers to plot<span></span></a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#colors-for-many-groups"><i class="fa fa-check"></i><b>10.7</b> colors for many groups<span></span></a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input<span></span></a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>11</b> mecodev package<span></span></a>
<ul>
<li class="chapter" data-level="11.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>11.1</b> trans_rarefy<span></span></a></li>
<li class="chapter" data-level="11.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_convert"><i class="fa fa-check"></i><b>11.2</b> trans_convert<span></span></a></li>
<li class="chapter" data-level="11.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_netchord"><i class="fa fa-check"></i><b>11.3</b> trans_netchord<span></span></a></li>
<li class="chapter" data-level="11.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>11.4</b> trans_ts<span></span></a></li>
<li class="chapter" data-level="11.5" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>11.5</b> trans_gamma<span></span></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>12</b> Other examples<span></span></a>
<ul>
<li class="chapter" data-level="12.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>12.1</b> Custom taxa order in bar plot<span></span></a></li>
<li class="chapter" data-level="12.2" data-path="other-examples-1.html"><a href="other-examples-1.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>12.2</b> The importance of tidy_taxonomy function<span></span></a></li>
<li class="chapter" data-level="12.3" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>12.3</b> Question of prefix in the taxa<span></span></a></li>
<li class="chapter" data-level="12.4" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>12.4</b> The use of phylogenetic tree<span></span></a></li>
</ul></li>
<li><a href="references.html#references">References<span></span></a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v0.7.0)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="composition-based-class" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">Chapter 4</span> Composition-based class<a href="composition-based-class.html#composition-based-class" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The trans_abund class and trans_venn class are grouped into ‘Composition-based class,’
since they are mainly used to show the composition information of communities.</p>
<div id="trans_abund-class" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> trans_abund class<a href="composition-based-class.html#trans_abund-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　The trans_abund class is used to transform taxonomic abundance data to plots with the ggplot2 package.
As a premise, the <strong>taxa_abund list in the object of microtable class must be first calculated</strong>.</p>
<div id="example-1" class="section level3 hasAnchor" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Example<a href="composition-based-class.html#example-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We first show the bar plot example.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="composition-based-class.html#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_abund object</span></span>
<span id="cb70-2"><a href="composition-based-class.html#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="co"># use 10 Phyla with the highest abundance in the dataset.</span></span>
<span id="cb70-3"><a href="composition-based-class.html#cb70-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>)</span>
<span id="cb70-4"><a href="composition-based-class.html#cb70-4" aria-hidden="true" tabindex="-1"></a><span class="co"># t1 object now include the transformed abundance data t1$abund_data and other elements for the following plotting</span></span></code></pre></div>
<p>As the sample number is large, we do not show the sample names in x axis and add the facet to show abundance according to groups.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="composition-based-class.html#cb71-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">others_color =</span> <span class="st">&quot;grey70&quot;</span>, <span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">xtext_keep =</span> <span class="cn">FALSE</span>, <span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span>
<span id="cb71-2"><a href="composition-based-class.html#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return a ggplot2 object</span></span></code></pre></div>
<p><img src="Images/plot_bar.png" width="750px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="composition-based-class.html#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># two facets example</span></span>
<span id="cb72-2"><a href="composition-based-class.html#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require package ggh4x, please run install.packages(&quot;ggh4x&quot;) if not installed</span></span>
<span id="cb72-3"><a href="composition-based-class.html#cb72-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">others_color =</span> <span class="st">&quot;grey70&quot;</span>, <span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">facet2 =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">xtext_keep =</span> <span class="cn">FALSE</span>, <span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>, <span class="at">barwidth =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Images/plot_bar_facet2.png" width="750px" style="display: block; margin: auto;" /></p>
<p>The alluvial plot is also implemented in the plot_bar function with use_alluvium parameter.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="composition-based-class.html#cb73-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>)</span>
<span id="cb73-2"><a href="composition-based-class.html#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require ggalluvial package</span></span>
<span id="cb73-3"><a href="composition-based-class.html#cb73-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use_alluvium = TRUE make the alluvial plot, clustering =TRUE can be used to reorder the samples by clustering</span></span>
<span id="cb73-4"><a href="composition-based-class.html#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="co"># select another color palette</span></span>
<span id="cb73-5"><a href="composition-based-class.html#cb73-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">use_alluvium =</span> <span class="cn">TRUE</span>, <span class="at">clustering =</span> <span class="cn">TRUE</span>, <span class="at">xtext_type_hor =</span> <span class="cn">FALSE</span>, <span class="at">xtext_size =</span> <span class="dv">6</span>, <span class="at">use_colors =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Set2&quot;</span>))</span></code></pre></div>
<p><img src="Images/plot_bar_allu.png" width="1440" style="display: block; margin: auto;" /></p>
<p>The bar plot can also be performed with group mean values.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="composition-based-class.html#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The groupmean parameter can be used to obtain the group-mean barplot.</span></span>
<span id="cb74-2"><a href="composition-based-class.html#cb74-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb74-3"><a href="composition-based-class.html#cb74-3" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">others_color =</span> <span class="st">&quot;grey70&quot;</span>, <span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span>
<span id="cb74-4"><a href="composition-based-class.html#cb74-4" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>))</span></code></pre></div>
<p><img src="Images/plot_bar_mean.png" width="400px" style="display: block; margin: auto;" /></p>
<p>The box plot is an excellent way to intuitionally show data distribution across groups.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="composition-based-class.html#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show 15 taxa at Class level</span></span>
<span id="cb75-2"><a href="composition-based-class.html#cb75-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Class&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">15</span>)</span>
<span id="cb75-3"><a href="composition-based-class.html#cb75-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_box</span>(<span class="at">group =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_box.png" width="700px" style="display: block; margin: auto;" /></p>
<p>Then we show the heatmap with the high abundant genera.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="composition-based-class.html#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show 40 taxa at Genus level</span></span>
<span id="cb76-2"><a href="composition-based-class.html#cb76-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">40</span>)</span>
<span id="cb76-3"><a href="composition-based-class.html#cb76-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_heatmap</span>(<span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">xtext_keep =</span> <span class="cn">FALSE</span>, <span class="at">withmargin =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/plot_heatmap.png" width="750px" style="display: block; margin: auto;" /></p>
<p>Then, we show the pie chart with the group mean values.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="composition-based-class.html#cb77-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">6</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb77-2"><a href="composition-based-class.html#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="co"># all pie chart in one row</span></span>
<span id="cb77-3"><a href="composition-based-class.html#cb77-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_pie</span>(<span class="at">facet_nrow =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Images/plot_pie.png" width="600px" style="display: block; margin: auto;" /></p>
</div>
<div id="key-points-1" class="section level3 hasAnchor" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Key points<a href="composition-based-class.html#key-points-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>trans_abund$new: creating trans_abund object can invoke taxa_abund in microtable for transformation</li>
<li>use_colors parameter: use_colors parameter in each function is used for colors selection</li>
<li>input_taxaname parameter: input_taxaname parameter in trans_abund$new can be used to select interested customized taxa instead of abundance-based selection</li>
<li>use_percentage parameter: use_percentage parameter in trans_abund$new - whether show the abundance percentage</li>
</ul>
</div>
</div>
<div id="trans_venn-class" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> trans_venn class<a href="composition-based-class.html#trans_venn-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The trans_venn class is used for venn analysis, i.e. shared and unique taxa.</p>
<div id="example-2" class="section level3 hasAnchor" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Example<a href="composition-based-class.html#example-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This part can be performed using samples or groups at OTU/ASV level or higher taxonomic level.
To analyze the unique and shared OTUs of groups,
we first merge samples according to the “Group” column of sample_table.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="composition-based-class.html#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># merge samples as one community for each group</span></span>
<span id="cb78-2"><a href="composition-based-class.html#cb78-2" aria-hidden="true" tabindex="-1"></a>dataset1 <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb78-3"><a href="composition-based-class.html#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="co"># dataset1 is a new microtable object</span></span>
<span id="cb78-4"><a href="composition-based-class.html#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_venn object</span></span>
<span id="cb78-5"><a href="composition-based-class.html#cb78-5" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1, <span class="at">ratio =</span> <span class="cn">NULL</span>)</span>
<span id="cb78-6"><a href="composition-based-class.html#cb78-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_venn</span>()</span></code></pre></div>
<p><img src="Images/trans_venn_0.png" width="500px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="composition-based-class.html#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create venn plot with more information</span></span>
<span id="cb79-2"><a href="composition-based-class.html#cb79-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1, <span class="at">ratio =</span> <span class="st">&quot;seqratio&quot;</span>)</span>
<span id="cb79-3"><a href="composition-based-class.html#cb79-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_venn</span>()</span>
<span id="cb79-4"><a href="composition-based-class.html#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="co"># The integer is OTU number</span></span>
<span id="cb79-5"><a href="composition-based-class.html#cb79-5" aria-hidden="true" tabindex="-1"></a><span class="co"># The percentage data is the sequence number/total sequence number</span></span></code></pre></div>
<p><img src="Images/trans_venn_1.png" width="500px" style="display: block; margin: auto;" /></p>
<p>When the groups are too many to show with venn plot, using petal plot is better.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="composition-based-class.html#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use &quot;Type&quot; column in sample_table</span></span>
<span id="cb80-2"><a href="composition-based-class.html#cb80-2" aria-hidden="true" tabindex="-1"></a>dataset1 <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Type&quot;</span>)</span>
<span id="cb80-3"><a href="composition-based-class.html#cb80-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1)</span>
<span id="cb80-4"><a href="composition-based-class.html#cb80-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_venn</span>(<span class="at">petal_plot =</span> <span class="cn">TRUE</span>, <span class="at">petal_center_size =</span> <span class="dv">50</span>, <span class="at">petal_r =</span> <span class="fl">1.5</span>, <span class="at">petal_a =</span> <span class="dv">3</span>, <span class="at">petal_move_xy =</span> <span class="fl">3.8</span>, <span class="at">petal_color_center =</span> <span class="st">&quot;#BEBADA&quot;</span>)</span></code></pre></div>
<p><img src="Images/trans_venn_2.png" width="500px" style="display: block; margin: auto;" /></p>
<p>Generally, after getting the venn plot, we do not know what those shared or unique taxa are.
The composition of the unique or shared species may account for the different and similar parts of ecological characteristics across groups<span class="citation">(<a href="#ref-Mendes_Deciphering_2011" role="doc-biblioref">Mendes et al. 2011</a>)</span>.
So, it is interesting to further analyze the composition of unique and shared species.
For this goal, we first transform the results of venn plot to the traditional species-sample table, that is, another object of microtable class.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="composition-based-class.html#cb81-1" aria-hidden="true" tabindex="-1"></a>dataset1 <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb81-2"><a href="composition-based-class.html#cb81-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1)</span></code></pre></div>
<pre><code>## The result is stored in object$venn_table and object$venn_count_abund ...</code></pre>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="composition-based-class.html#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co"># transform venn results to the sample-species table, here do not consider abundance, only use presence/absence.</span></span>
<span id="cb83-2"><a href="composition-based-class.html#cb83-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">trans_venn_com</span>(<span class="at">use_frequency =</span> <span class="cn">TRUE</span>)</span>
<span id="cb83-3"><a href="composition-based-class.html#cb83-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t2 is a new microtable class, each part is considered a sample</span></span>
<span id="cb83-4"><a href="composition-based-class.html#cb83-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(t2)</span></code></pre></div>
<pre><code>## [1] &quot;microtable&quot; &quot;R6&quot;</code></pre>
<p>We use bar plot to show the composition at the Genus level.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="composition-based-class.html#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate taxa abundance, that is, the frequency</span></span>
<span id="cb85-2"><a href="composition-based-class.html#cb85-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb85-3"><a href="composition-based-class.html#cb85-3" aria-hidden="true" tabindex="-1"></a><span class="co"># transform and plot</span></span>
<span id="cb85-4"><a href="composition-based-class.html#cb85-4" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> t2, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>)</span>
<span id="cb85-5"><a href="composition-based-class.html#cb85-5" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">bar_type =</span> <span class="st">&quot;part&quot;</span>, <span class="at">legend_text_italic =</span> T, <span class="at">ylab_title =</span> <span class="st">&quot;Frequency (%)&quot;</span>, <span class="at">xtext_type_hor =</span> <span class="cn">FALSE</span>, <span class="at">use_colors =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Set2&quot;</span>),</span>
<span id="cb85-6"><a href="composition-based-class.html#cb85-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">order_x =</span> <span class="fu">c</span>(<span class="st">&quot;IW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;TW&quot;</span>, <span class="st">&quot;IW-CW&quot;</span>, <span class="st">&quot;IW-TW&quot;</span>, <span class="st">&quot;CW-TW&quot;</span>, <span class="st">&quot;IW-CW-TW&quot;</span>))</span></code></pre></div>
<p><img src="Images/trans_venn_bar.png" width="650px" style="display: block; margin: auto;" /></p>
<p>We also try to use pie chart to show the compositions at the Phylum level.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="composition-based-class.html#cb86-1" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> t2, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>)</span>
<span id="cb86-2"><a href="composition-based-class.html#cb86-2" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">plot_pie</span>(<span class="at">facet_nrow =</span> <span class="dv">3</span>, <span class="at">use_colors =</span> <span class="fu">rev</span>(<span class="fu">c</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Dark2&quot;</span>), <span class="st">&quot;grey50&quot;</span>)))</span></code></pre></div>
<p><img src="Images/trans_venn_pie.png" width="800px" style="display: block; margin: auto;" /></p>
</div>
<div id="key-points-2" class="section level3 hasAnchor" number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Key points<a href="composition-based-class.html#key-points-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>ratio parameter: ratio parameter in trans_abund$new control whether and what content appear under the taxa number in venn plot</li>
<li>returned data: using trans_venn$new() return venn_table and venn_count_abund stored in trans_venn object for further ploting</li>
</ul>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Mendes_Deciphering_2011" class="csl-entry">
Mendes, R, M Kruijt, Bruijn I De, E Dekkers, der Voort M Van, J. H. Schneider, Y. M. Piceno, T. Z. Desantis, G. L. Andersen, and P. A. Bakker. 2011. <span>“Deciphering the Rhizosphere Microbiome for Disease-Suppressive Bacteria.”</span> Journal Article. <em>Science</em> 332 (6033): 1097–1100.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="basic-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="diversity-based-class.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/03-Composition-based_class.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
