<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Diversity-based class | Tutorial for R microeco package (v0.14.1)</title>
  <meta name="description" content="The tutorial for R microeco, file2meco, meconetcomp and mecodev packages" />
  <meta name="generator" content="bookdown 0.30 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Diversity-based class | Tutorial for R microeco package (v0.14.1)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecodev packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Diversity-based class | Tutorial for R microeco package (v0.14.1)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecodev packages" />
  

<meta name="author" content="Chi Liu, Felipe R. P. Mansoldo, Umer Zeeshan Ijaz, Chenhao Li, Yang Cao, Jarrod J. Scott, Yaoming Cui, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2023-03-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="composition-based-class.html"/>
<link rel="next" href="model-based-class.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#rtools"><i class="fa fa-check"></i><b>2.4</b> RTools</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.5</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#important-packages"><i class="fa fa-check"></i><b>2.5.1</b> Important packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#cran-packages"><i class="fa fa-check"></i><b>2.5.2</b> CRAN packages</a></li>
<li class="chapter" data-level="2.5.3" data-path="intro.html"><a href="intro.html#ggtree"><i class="fa fa-check"></i><b>2.5.3</b> ggtree</a></li>
<li class="chapter" data-level="2.5.4" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.5.4</b> Gephi</a></li>
<li class="chapter" data-level="2.5.5" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.5.5</b> Tax4Fun</a></li>
<li class="chapter" data-level="2.5.6" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.5.6</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#plotting"><i class="fa fa-check"></i><b>2.6</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#prepare-the-example-data"><i class="fa fa-check"></i><b>3.1.1</b> Prepare the example data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#how-to-read-your-files-to-microtable-object"><i class="fa fa-check"></i><b>3.1.2</b> How to read your files to microtable object?</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#functions-in-microtable-class"><i class="fa fa-check"></i><b>3.1.3</b> Functions in microtable class</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#subset-of-samples"><i class="fa fa-check"></i><b>3.1.4</b> subset of samples</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-class.html"><a href="basic-class.html#filter-features"><i class="fa fa-check"></i><b>3.1.5</b> filter features</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.6</b> Other examples</a></li>
<li class="chapter" data-level="3.1.7" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.7</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-2"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-4"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#network-comparison"><i class="fa fa-check"></i><b>6.2.2</b> Network comparison</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.3</b> Key points</a></li>
<li class="chapter" data-level="6.2.4" data-path="model-based-class.html"><a href="model-based-class.html#other-functions"><i class="fa fa-check"></i><b>6.2.4</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies</a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples</a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points</a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-7"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME</a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2</a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann"><i class="fa fa-check"></i><b>8.3</b> HUMAnN</a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn</a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2"><i class="fa fa-check"></i><b>8.5</b> Kraken2</a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB</a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq</a></li>
<li class="chapter" data-level="8.8" data-path="file2meco-package.html"><a href="file2meco-package.html#picrust2"><i class="fa fa-check"></i><b>8.8</b> PICRUSt2</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset</a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data</a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> <code>clone</code> function</a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> <code>save</code> function</a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#github-issues"><i class="fa fa-check"></i><b>10.3</b> Github Issues</a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.4</b> change object</a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.5</b> group order</a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.6</b> add layers to plot</a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#colors-for-many-groups"><i class="fa fa-check"></i><b>10.7</b> colors for many groups</a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>11</b> mecodev package</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>11.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="11.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_norm"><i class="fa fa-check"></i><b>11.2</b> trans_norm</a></li>
<li class="chapter" data-level="11.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>11.3</b> trans_ts</a></li>
<li class="chapter" data-level="11.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>11.4</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html"><i class="fa fa-check"></i><b>12</b> meconetcomp package</a>
<ul>
<li class="chapter" data-level="12.1" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-modularity-for-all-networks"><i class="fa fa-check"></i><b>12.1</b> Network modularity for all networks</a></li>
<li class="chapter" data-level="12.2" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-topological-attributes-for-all-networks"><i class="fa fa-check"></i><b>12.2</b> Network topological attributes for all networks</a></li>
<li class="chapter" data-level="12.3" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#node-and-edge-properties-extraction-for-all-networks"><i class="fa fa-check"></i><b>12.3</b> Node and edge properties extraction for all networks</a></li>
<li class="chapter" data-level="12.4" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-nodes-across-networks"><i class="fa fa-check"></i><b>12.4</b> Compare nodes across networks</a></li>
<li class="chapter" data-level="12.5" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-edges-across-networks"><i class="fa fa-check"></i><b>12.5</b> Compare edges across networks</a></li>
<li class="chapter" data-level="12.6" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#extract-overlapped-edges-of-networks-to-a-new-network"><i class="fa fa-check"></i><b>12.6</b> Extract overlapped edges of networks to a new network</a></li>
<li class="chapter" data-level="12.7" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-phylogenetic-distances-of-paired-nodes-in-edges"><i class="fa fa-check"></i><b>12.7</b> Compare phylogenetic distances of paired nodes in edges</a></li>
<li class="chapter" data-level="12.8" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-node-sources-of-edges-across-networks"><i class="fa fa-check"></i><b>12.8</b> Compare node sources of edges across networks</a></li>
<li class="chapter" data-level="12.9" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-topological-properties-of-sub-networks"><i class="fa fa-check"></i><b>12.9</b> Compare topological properties of sub-networks</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mecoturn-package.html"><a href="mecoturn-package.html"><i class="fa fa-check"></i><b>13</b> mecoturn package</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mecoturn-package.html"><a href="mecoturn-package.html#betaturn-class"><i class="fa fa-check"></i><b>13.1</b> betaturn class</a></li>
<li class="chapter" data-level="13.2" data-path="mecoturn-package.html"><a href="mecoturn-package.html#taxaturn-class"><i class="fa fa-check"></i><b>13.2</b> taxaturn class</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>14</b> Other examples</a>
<ul>
<li class="chapter" data-level="14.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>14.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="14.2" data-path="other-examples-1.html"><a href="other-examples-1.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>14.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="14.3" data-path="other-examples-1.html"><a href="other-examples-1.html#show-the-abundance-of-unknown-taxa"><i class="fa fa-check"></i><b>14.3</b> Show the abundance of unknown taxa</a></li>
<li class="chapter" data-level="14.4" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>14.4</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="14.5" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>14.5</b> The use of phylogenetic tree</a></li>
<li class="chapter" data-level="14.6" data-path="other-examples-1.html"><a href="other-examples-1.html#mantel-test-correlation-heatmap"><i class="fa fa-check"></i><b>14.6</b> Mantel test + Correlation heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v0.14.1)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="diversity-based-class" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> Diversity-based class<a href="diversity-based-class.html#diversity-based-class" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Diversity is one of the core topics in community ecology.
It refers to alpha diversity, beta diversity and gamma diversity.</p>
<div id="trans_alpha-class" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> trans_alpha class<a href="diversity-based-class.html#trans_alpha-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　Alpha diversity can be transformed and plotted using trans_alpha class.
Creating the object of trans_alpha class can invoke the alpha_diversity data stored in the microtable object.</p>
<div id="example-2" class="section level3 hasAnchor" number="5.1.1">
<h3><span class="header-section-number">5.1.1</span> Example<a href="diversity-based-class.html#example-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Creating trans_alpha object have two return data.frame with prefix ‘data_’: data_alpha and data_stat.
The data_alpha is used for the following differential test and plotting.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="diversity-based-class.html#cb99-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_alpha<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb99-2"><a href="diversity-based-class.html#cb99-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$data_stat</span></span>
<span id="cb99-3"><a href="diversity-based-class.html#cb99-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>data_stat[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<pre><code>## The transformed diversity data is stored in object$data_alpha ...</code></pre>
<pre><code>## The group statistics are stored in object$data_stat ...</code></pre>
<table style="width:75%;">
<colgroup>
<col width="11%" />
<col width="15%" />
<col width="6%" />
<col width="12%" />
<col width="13%" />
<col width="15%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Group</th>
<th align="center">Measure</th>
<th align="center">N</th>
<th align="center">Mean</th>
<th align="center">SD</th>
<th align="center">SE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">CW</td>
<td align="center">Observed</td>
<td align="center">30</td>
<td align="center">1843</td>
<td align="center">220.6</td>
<td align="center">40.27</td>
</tr>
<tr class="even">
<td align="center">CW</td>
<td align="center">Chao1</td>
<td align="center">30</td>
<td align="center">2553</td>
<td align="center">338.1</td>
<td align="center">61.73</td>
</tr>
<tr class="odd">
<td align="center">CW</td>
<td align="center">ACE</td>
<td align="center">30</td>
<td align="center">2716</td>
<td align="center">367</td>
<td align="center">67.01</td>
</tr>
<tr class="even">
<td align="center">CW</td>
<td align="center">Shannon</td>
<td align="center">30</td>
<td align="center">6.308</td>
<td align="center">0.5355</td>
<td align="center">0.09777</td>
</tr>
<tr class="odd">
<td align="center">CW</td>
<td align="center">Simpson</td>
<td align="center">30</td>
<td align="center">0.9897</td>
<td align="center">0.01305</td>
<td align="center">0.002382</td>
</tr>
</tbody>
</table>
<p>Then, we test the differences among groups using Kruskal-Wallis Rank Sum Test (overall test when groups &gt; 2), Wilcoxon Rank Sum Tests (for paired groups),
Dunn’s Kruskal-Wallis Multiple Comparisons (for paired groups when groups &gt; 2) and anova with multiple comparisons.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="diversity-based-class.html#cb102-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;KW&quot;</span>)</span>
<span id="cb102-2"><a href="diversity-based-class.html#cb102-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_diff</span></span>
<span id="cb102-3"><a href="diversity-based-class.html#cb102-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_diff[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<pre><code>## The result is stored in object$res_diff ...</code></pre>
<table style="width:96%;">
<colgroup>
<col width="20%" />
<col width="15%" />
<col width="11%" />
<col width="13%" />
<col width="13%" />
<col width="20%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Comparison</th>
<th align="center">Measure</th>
<th align="center">Group</th>
<th align="center">P.unadj</th>
<th align="center">P.adj</th>
<th align="center">Significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">IW - CW - TW</td>
<td align="center">Observed</td>
<td align="center">IW</td>
<td align="center">0.155</td>
<td align="center">0.2713</td>
<td align="center">ns</td>
</tr>
<tr class="even">
<td align="center">IW - CW - TW</td>
<td align="center">Chao1</td>
<td align="center">IW</td>
<td align="center">0.01696</td>
<td align="center">0.05936</td>
<td align="center">ns</td>
</tr>
<tr class="odd">
<td align="center">IW - CW - TW</td>
<td align="center">ACE</td>
<td align="center">IW</td>
<td align="center">0.01333</td>
<td align="center">0.05936</td>
<td align="center">ns</td>
</tr>
<tr class="even">
<td align="center">IW - CW - TW</td>
<td align="center">Shannon</td>
<td align="center">IW</td>
<td align="center">0.5319</td>
<td align="center">0.7446</td>
<td align="center">ns</td>
</tr>
<tr class="odd">
<td align="center">IW - CW - TW</td>
<td align="center">Simpson</td>
<td align="center">CW</td>
<td align="center">0.8083</td>
<td align="center">0.8083</td>
<td align="center">ns</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="diversity-based-class.html#cb104-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;KW_dunn&quot;</span>)</span>
<span id="cb104-2"><a href="diversity-based-class.html#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_diff</span></span>
<span id="cb104-3"><a href="diversity-based-class.html#cb104-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_diff[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<pre><code>## The result is stored in object$res_diff ...</code></pre>
<table>
<colgroup>
<col width="14%" />
<col width="10%" />
<col width="16%" />
<col width="11%" />
<col width="14%" />
<col width="12%" />
<col width="19%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Measure</th>
<th align="center">Group</th>
<th align="center">Comparison</th>
<th align="center">Z</th>
<th align="center">P.unadj</th>
<th align="center">P.adj</th>
<th align="center">Significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Observed</td>
<td align="center">IW</td>
<td align="center">CW - IW</td>
<td align="center">-1.93</td>
<td align="center">0.05364</td>
<td align="center">0.1609</td>
<td align="center">ns</td>
</tr>
<tr class="even">
<td align="center">Observed</td>
<td align="center">TW</td>
<td align="center">CW - TW</td>
<td align="center">-1.02</td>
<td align="center">0.3075</td>
<td align="center">0.615</td>
<td align="center">ns</td>
</tr>
<tr class="odd">
<td align="center">Observed</td>
<td align="center">IW</td>
<td align="center">IW - TW</td>
<td align="center">0.9093</td>
<td align="center">0.3632</td>
<td align="center">0.3632</td>
<td align="center">ns</td>
</tr>
<tr class="even">
<td align="center">Chao1</td>
<td align="center">IW</td>
<td align="center">CW - IW</td>
<td align="center">-2.841</td>
<td align="center">0.004491</td>
<td align="center">0.01347</td>
<td align="center">*</td>
</tr>
<tr class="odd">
<td align="center">Chao1</td>
<td align="center">TW</td>
<td align="center">CW - TW</td>
<td align="center">-1.176</td>
<td align="center">0.2395</td>
<td align="center">0.2395</td>
<td align="center">ns</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="diversity-based-class.html#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co"># more options</span></span>
<span id="cb106-2"><a href="diversity-based-class.html#cb106-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>)</span>
<span id="cb106-3"><a href="diversity-based-class.html#cb106-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;t.test&quot;</span>)</span></code></pre></div>
<p>Then, let’s try to use anova.</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="diversity-based-class.html#cb107-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;anova&quot;</span>)</span>
<span id="cb107-2"><a href="diversity-based-class.html#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_diff</span></span>
<span id="cb107-3"><a href="diversity-based-class.html#cb107-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_diff</span></code></pre></div>
<pre><code>## Registered S3 methods overwritten by &#39;klaR&#39;:
##   method      from 
##   predict.rda vegan
##   print.rda   vegan
##   plot.rda    vegan</code></pre>
<pre><code>## The result is stored in object$res_diff ...</code></pre>
<table>
<colgroup>
<col width="10%" />
<col width="14%" />
<col width="10%" />
<col width="8%" />
<col width="13%" />
<col width="13%" />
<col width="17%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Group</th>
<th align="center">Observed</th>
<th align="center">Chao1</th>
<th align="center">ACE</th>
<th align="center">Shannon</th>
<th align="center">Simpson</th>
<th align="center">InvSimpson</th>
<th align="center">Fisher</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">IW</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
</tr>
<tr class="even">
<td align="center">TW</td>
<td align="center">a</td>
<td align="center">ab</td>
<td align="center">b</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
</tr>
<tr class="odd">
<td align="center">CW</td>
<td align="center">a</td>
<td align="center">b</td>
<td align="center">b</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
</tr>
</tbody>
</table>
<p>The multi-factor analysis of variance is also supported.
It is notable that the result res_diff will be a list instead of a data.frame when formula is provided.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="diversity-based-class.html#cb110-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_alpha<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb110-2"><a href="diversity-based-class.html#cb110-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;anova&quot;</span>, <span class="at">formula =</span> <span class="st">&quot;Group+Type&quot;</span>)</span>
<span id="cb110-3"><a href="diversity-based-class.html#cb110-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(t1<span class="sc">$</span>res_diff)</span>
<span id="cb110-4"><a href="diversity-based-class.html#cb110-4" aria-hidden="true" tabindex="-1"></a><span class="co"># see the help document for the usage of formula</span></span></code></pre></div>
<p>The plot_alpha function add the significance label by searching the results in <strong>object$res_diff</strong> instead of calculating the significance again.
Now, let us plot the mean and se of alpha diversity for each group, and add the anova result.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="diversity-based-class.html#cb111-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;anova&quot;</span>)</span>
<span id="cb111-2"><a href="diversity-based-class.html#cb111-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_alpha</span>(<span class="at">measure =</span> <span class="st">&quot;Chao1&quot;</span>)</span>
<span id="cb111-3"><a href="diversity-based-class.html#cb111-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_alpha</span>(<span class="at">measure =</span> <span class="st">&quot;Chao1&quot;</span>, <span class="at">order_x_mean =</span> <span class="cn">TRUE</span>, <span class="at">add_sig_text_size =</span> <span class="dv">6</span>)</span></code></pre></div>
<p><img src="Images/plot_alpha_letter.png" width="600px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="diversity-based-class.html#cb112-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>)</span>
<span id="cb112-2"><a href="diversity-based-class.html#cb112-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_alpha</span>(<span class="at">measure =</span> <span class="st">&quot;Chao1&quot;</span>, <span class="at">shape =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_alpha_wilcox.png" width="600px" style="display: block; margin: auto;" /></p>
<p>Let’s try to remove the ns in the label by operating the object$res_diff file.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="diversity-based-class.html#cb113-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_diff <span class="sc">%&lt;&gt;%</span> base<span class="sc">::</span><span class="fu">subset</span>(Significance <span class="sc">!=</span> <span class="st">&quot;ns&quot;</span>)</span>
<span id="cb113-2"><a href="diversity-based-class.html#cb113-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_alpha</span>(<span class="at">measure =</span> <span class="st">&quot;Chao1&quot;</span>, <span class="at">xtext_size =</span> <span class="dv">15</span>)</span></code></pre></div>
<p><img src="Images/plot_alpha_wilcox_nons.png" width="600px" style="display: block; margin: auto;" /></p>
<p>From the v0.12.0, the trans_alpha class supports the differential test of groups within each group by using the by_group parameter.</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="diversity-based-class.html#cb114-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_alpha<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">by_group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb114-2"><a href="diversity-based-class.html#cb114-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>)</span>
<span id="cb114-3"><a href="diversity-based-class.html#cb114-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_alpha</span>(<span class="at">measure =</span> <span class="st">&quot;Shannon&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_alpha_wilcox_bygroup.png" width="600px" style="display: block; margin: auto;" /></p>
</div>
<div id="key-points-3" class="section level3 hasAnchor" number="5.1.2">
<h3><span class="header-section-number">5.1.2</span> Key points<a href="diversity-based-class.html#key-points-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>trans_alpha$new: creating trans_alpha object can invoke alpha_diversity in microtable for transformation</li>
<li>cal_diff: formula parameter can be used for multi-factor analysis of variance</li>
<li>plot_alpha: the significance label comes from the object$res_diff</li>
</ul>
</div>
</div>
<div id="trans_beta-class" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> trans_beta class<a href="diversity-based-class.html#trans_beta-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　The trans_beta class is developed for the beta diversity analysis, i.e. the dissimilarities among samples.
Beta diversity can be defined at different forms<span class="citation">(<a href="#ref-Tuomisto_diversity_2010" role="doc-biblioref">Tuomisto 2010</a>)</span> and can be explored with different ways<span class="citation">(<a href="#ref-Anderson_Navigating_2011" role="doc-biblioref">Anderson et al. 2011</a>)</span>,
So we encapsulate some commonly-used approaches in microbial ecology<span class="citation">(<a href="#ref-Ramette_Multivariate_2007" role="doc-biblioref">Ramette 2007</a>)</span>.
Note that the parts related with “Environmental interpretation” are placed into the trans_env class.
When needed, the beta_diversity list in microtable object will be invoked for transformation and ploting using trans_beta class.
The analysis referred to the beta diversity in this class mainly include ordination, group distance, clustering and manova.</p>
<div id="example-3" class="section level3 hasAnchor" number="5.2.1">
<h3><span class="header-section-number">5.2.1</span> Example<a href="diversity-based-class.html#example-3" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We first show the ordination using PCoA.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="diversity-based-class.html#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we first create an trans_beta object</span></span>
<span id="cb115-2"><a href="diversity-based-class.html#cb115-2" aria-hidden="true" tabindex="-1"></a><span class="co"># measure parameter can invoke the distance matrix in dataset$beta_diversity</span></span>
<span id="cb115-3"><a href="diversity-based-class.html#cb115-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;bray&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="diversity-based-class.html#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use PCoA as an example, PCA or NMDS is also available</span></span>
<span id="cb116-2"><a href="diversity-based-class.html#cb116-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">ordination =</span> <span class="st">&quot;PCoA&quot;</span>)</span>
<span id="cb116-3"><a href="diversity-based-class.html#cb116-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_ordination is the ordination result list</span></span>
<span id="cb116-4"><a href="diversity-based-class.html#cb116-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(t1<span class="sc">$</span>res_ordination)</span>
<span id="cb116-5"><a href="diversity-based-class.html#cb116-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the PCoA result with confidence ellipse</span></span>
<span id="cb116-6"><a href="diversity-based-class.html#cb116-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;ellipse&quot;</span>))</span></code></pre></div>
<p><img src="Images/plot_ordination.png" width="650px" style="display: block; margin: auto;" /></p>
<p>Try other interesting options in the plotting.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="diversity-based-class.html#cb117-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">plot_type =</span> <span class="st">&quot;point&quot;</span>)</span>
<span id="cb117-2"><a href="diversity-based-class.html#cb117-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">point_size =</span> <span class="dv">5</span>, <span class="at">point_alpha =</span> .<span class="dv">2</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;ellipse&quot;</span>), <span class="at">ellipse_chull_fill =</span> <span class="cn">FALSE</span>)</span>
<span id="cb117-3"><a href="diversity-based-class.html#cb117-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;centroid&quot;</span>))</span>
<span id="cb117-4"><a href="diversity-based-class.html#cb117-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;ellipse&quot;</span>, <span class="st">&quot;centroid&quot;</span>))</span>
<span id="cb117-5"><a href="diversity-based-class.html#cb117-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;chull&quot;</span>))</span>
<span id="cb117-6"><a href="diversity-based-class.html#cb117-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;chull&quot;</span>, <span class="st">&quot;centroid&quot;</span>))</span>
<span id="cb117-7"><a href="diversity-based-class.html#cb117-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;chull&quot;</span>, <span class="st">&quot;centroid&quot;</span>))</span>
<span id="cb117-8"><a href="diversity-based-class.html#cb117-8" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;chull&quot;</span>, <span class="st">&quot;centroid&quot;</span>), <span class="at">add_sample_label =</span> <span class="st">&quot;SampleID&quot;</span>)</span>
<span id="cb117-9"><a href="diversity-based-class.html#cb117-9" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="st">&quot;centroid&quot;</span>)</span>
<span id="cb117-10"><a href="diversity-based-class.html#cb117-10" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="st">&quot;centroid&quot;</span>, <span class="at">centroid_segment_alpha =</span> <span class="fl">0.9</span>, <span class="at">centroid_segment_size =</span> <span class="dv">1</span>, <span class="at">centroid_segment_linetype =</span> <span class="dv">1</span>)</span>
<span id="cb117-11"><a href="diversity-based-class.html#cb117-11" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;centroid&quot;</span>), <span class="at">plot_color =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">centroid_segment_linetype =</span> <span class="dv">1</span>)</span>
<span id="cb117-12"><a href="diversity-based-class.html#cb117-12" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Saline&quot;</span>, <span class="at">point_size =</span> <span class="dv">5</span>, <span class="at">point_alpha =</span> .<span class="dv">2</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;chull&quot;</span>), <span class="at">ellipse_chull_fill =</span> <span class="cn">FALSE</span>, <span class="at">ellipse_chull_alpha =</span> <span class="fl">0.1</span>)</span>
<span id="cb117-13"><a href="diversity-based-class.html#cb117-13" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">panel.grid =</span> <span class="fu">element_blank</span>()) <span class="sc">+</span> <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span></code></pre></div>
<p>Then we plot and compare the group distances.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="diversity-based-class.html#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate and plot sample distances within groups</span></span>
<span id="cb118-2"><a href="diversity-based-class.html#cb118-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_group_distance</span>(<span class="at">within_group =</span> <span class="cn">TRUE</span>)</span>
<span id="cb118-3"><a href="diversity-based-class.html#cb118-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_group_distance</span></span>
<span id="cb118-4"><a href="diversity-based-class.html#cb118-4" aria-hidden="true" tabindex="-1"></a><span class="co"># perform Wilcoxon Rank Sum and Signed Rank Tests</span></span>
<span id="cb118-5"><a href="diversity-based-class.html#cb118-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_group_distance_diff</span>(<span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>)</span>
<span id="cb118-6"><a href="diversity-based-class.html#cb118-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">boxplot_add =</span> <span class="st">&quot;mean&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_group_distance_within.png" width="500px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="diversity-based-class.html#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate and plot sample distances between groups</span></span>
<span id="cb119-2"><a href="diversity-based-class.html#cb119-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_group_distance</span>(<span class="at">within_group =</span> <span class="cn">FALSE</span>)</span>
<span id="cb119-3"><a href="diversity-based-class.html#cb119-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_group_distance_diff</span>(<span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>)</span>
<span id="cb119-4"><a href="diversity-based-class.html#cb119-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">boxplot_add =</span> <span class="st">&quot;mean&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_group_distance_between.png" width="500px" style="display: block; margin: auto;" /></p>
<p>Clustering plot is also a frequently used method.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="diversity-based-class.html#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use replace_name to set the label name, group parameter used to set the color</span></span>
<span id="cb120-2"><a href="diversity-based-class.html#cb120-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_clustering</span>(<span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">replace_name =</span> <span class="fu">c</span>(<span class="st">&quot;Saline&quot;</span>, <span class="st">&quot;Type&quot;</span>))</span></code></pre></div>
<p><img src="Images/plot_clustering.png" width="550px" style="display: block; margin: auto;" /></p>
<p>perMANOVA<span class="citation">(<a href="#ref-Anderson_Austral_2001" role="doc-biblioref">Anderson 2001</a>)</span> is often used in the differential test of distances among groups.</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="diversity-based-class.html#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="co"># manova for all groups when manova_all = TRUE</span></span>
<span id="cb121-2"><a href="diversity-based-class.html#cb121-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_manova</span>(<span class="at">manova_all =</span> <span class="cn">TRUE</span>)</span>
<span id="cb121-3"><a href="diversity-based-class.html#cb121-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_manova</span></code></pre></div>
<pre><code>## The result is stored in object$res_manova ...</code></pre>
<table style="width:79%;">
<caption>Permutation test for adonis under reduced model</caption>
<colgroup>
<col width="20%" />
<col width="6%" />
<col width="15%" />
<col width="12%" />
<col width="11%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">Df</th>
<th align="center">SumOfSqs</th>
<th align="center">R2</th>
<th align="center">F</th>
<th align="center">Pr(&gt;F)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Group</strong></td>
<td align="center">2</td>
<td align="center">6.121</td>
<td align="center">0.1955</td>
<td align="center">10.57</td>
<td align="center">0.001</td>
</tr>
<tr class="even">
<td align="center"><strong>Residual</strong></td>
<td align="center">87</td>
<td align="center">25.18</td>
<td align="center">0.8045</td>
<td align="center">NA</td>
<td align="center">NA</td>
</tr>
<tr class="odd">
<td align="center"><strong>Total</strong></td>
<td align="center">89</td>
<td align="center">31.3</td>
<td align="center">1</td>
<td align="center">NA</td>
<td align="center">NA</td>
</tr>
</tbody>
</table>
<p>The parameter manova_all = FALSE can be used to calculate significance for each paired group.</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="diversity-based-class.html#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="co"># manova for each paired groups</span></span>
<span id="cb123-2"><a href="diversity-based-class.html#cb123-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_manova</span>(<span class="at">manova_all =</span> <span class="cn">FALSE</span>)</span>
<span id="cb123-3"><a href="diversity-based-class.html#cb123-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_manova</span></code></pre></div>
<pre><code>## The result is stored in object$res_manova ...</code></pre>
<table>
<colgroup>
<col width="14%" />
<col width="13%" />
<col width="10%" />
<col width="11%" />
<col width="13%" />
<col width="17%" />
<col width="19%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Groups</th>
<th align="center">measure</th>
<th align="center">F</th>
<th align="center">R2</th>
<th align="center">p.value</th>
<th align="center">p.adjusted</th>
<th align="center">Significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">IW vs CW</td>
<td align="center">bray</td>
<td align="center">11.01</td>
<td align="center">0.1595</td>
<td align="center">0.001</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="even">
<td align="center">IW vs TW</td>
<td align="center">bray</td>
<td align="center">9.992</td>
<td align="center">0.147</td>
<td align="center">0.001</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="odd">
<td align="center">CW vs TW</td>
<td align="center">bray</td>
<td align="center">10.69</td>
<td align="center">0.1556</td>
<td align="center">0.001</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
</tbody>
</table>
<p>The parameter manova_set has higher priority than manova_all. If manova_set is provided, manova_all parameter will be disabled.</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="diversity-based-class.html#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="co"># manova for specified group set: such as &quot;Group + Type&quot;</span></span>
<span id="cb125-2"><a href="diversity-based-class.html#cb125-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_manova</span>(<span class="at">manova_set =</span> <span class="st">&quot;Group + Type&quot;</span>)</span>
<span id="cb125-3"><a href="diversity-based-class.html#cb125-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_manova</span></code></pre></div>
<pre><code>## The result is stored in object$res_manova ...</code></pre>
<table style="width:79%;">
<caption>Permutation test for adonis under reduced model</caption>
<colgroup>
<col width="20%" />
<col width="6%" />
<col width="15%" />
<col width="12%" />
<col width="11%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">Df</th>
<th align="center">SumOfSqs</th>
<th align="center">R2</th>
<th align="center">F</th>
<th align="center">Pr(&gt;F)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Group</strong></td>
<td align="center">2</td>
<td align="center">6.121</td>
<td align="center">0.1955</td>
<td align="center">12.01</td>
<td align="center">0.001</td>
</tr>
<tr class="even">
<td align="center"><strong>Type</strong></td>
<td align="center">3</td>
<td align="center">3.783</td>
<td align="center">0.1208</td>
<td align="center">4.949</td>
<td align="center">0.001</td>
</tr>
<tr class="odd">
<td align="center"><strong>Residual</strong></td>
<td align="center">84</td>
<td align="center">21.4</td>
<td align="center">0.6836</td>
<td align="center">NA</td>
<td align="center">NA</td>
</tr>
<tr class="even">
<td align="center"><strong>Total</strong></td>
<td align="center">89</td>
<td align="center">31.3</td>
<td align="center">1</td>
<td align="center">NA</td>
<td align="center">NA</td>
</tr>
</tbody>
</table>
<p>PERMDISP<span class="citation">(<a href="#ref-Anderson_Navigating_2011" role="doc-biblioref">Anderson et al. 2011</a>)</span> is also implemented to check multivariate homogeneity of groups dispersions (variances).</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="diversity-based-class.html#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for the whole comparison and for each paired groups</span></span>
<span id="cb127-2"><a href="diversity-based-class.html#cb127-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_betadisper</span>()</span></code></pre></div>
<pre><code>## The result is stored in object$res_betadisper ...</code></pre>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="diversity-based-class.html#cb129-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_betadisper</span></code></pre></div>
<pre><code>## 
## Permutation test for homogeneity of multivariate dispersions
## Permutation: free
## Number of permutations: 999
## 
## Response: Distances
##           Df  Sum Sq   Mean Sq      F N.Perm Pr(&gt;F)  
## Groups     2 0.04131 0.0206545 4.1682    999  0.021 *
## Residuals 87 0.43110 0.0049552                       
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Pairwise comparisons:
## (Observed p-value below diagonal, permuted p-value above diagonal)
##           CW        IW    TW
## CW           0.4690000 0.063
## IW 0.4621193           0.005
## TW 0.0566190 0.0050319</code></pre>
<p>For the explanation of statistical methods in microbial ecology, please read the references <span class="citation">(<a href="#ref-Ramette_Multivariate_2007" role="doc-biblioref">Ramette 2007</a>; <a href="#ref-Buttigieg_guide_2014" role="doc-biblioref">Buttigieg and Ramette 2014</a>)</span>.</p>
</div>
<div id="key-points-4" class="section level3 hasAnchor" number="5.2.2">
<h3><span class="header-section-number">5.2.2</span> Key points<a href="diversity-based-class.html#key-points-4" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>trans_beta$new: creating trans_beta object with measure parameter can invoke beta_diversity in microtable object for transformation</li>
<li>cal_ordination(): PCoA, PCA and NMDS approaches are all available</li>
<li>cal_manova(): cal_manova function can be used for paired comparisons, overall test and multi-factor test</li>
<li>plot_group_distance(): hide_ns and pair_compare_filter_select parameters can be used to control the significance labels shown in the plot</li>
</ul>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Anderson_Austral_2001" class="csl-entry">
Anderson, M. J. 2001. <span>“A New Method for Non-Parametric Multivariate Analysis of Variance.”</span> Journal Article. <em>Austral Ecology</em>, no. 26: 32–46.
</div>
<div id="ref-Anderson_Navigating_2011" class="csl-entry">
Anderson, M. J., T. O. Crist, J. M. Chase, M. Vellend, B. D. Inouye, A. L. Freestone, N. J. Sanders, et al. 2011. <span>“Navigating the Multiple Meanings of Beta Diversity: A Roadmap for the Practicing Ecologist.”</span> Journal Article. <em>Ecology Letters</em> 14 (1): 19–28. <a href="https://doi.org/10.1111/j.1461-0248.2010.01552.x">https://doi.org/10.1111/j.1461-0248.2010.01552.x</a>.
</div>
<div id="ref-Buttigieg_guide_2014" class="csl-entry">
Buttigieg, Pier Luigi, and Alban Ramette. 2014. <span>“A Guide to Statistical Analysis in Microbial Ecology: A Community-Focused, Living Review of Multivariate Data Analyses.”</span> Journal Article. <em>FEMS Microbiology Ecology</em> 90 (3): 543–50. <a href="https://doi.org/10.1111/1574-6941.12437">https://doi.org/10.1111/1574-6941.12437</a>.
</div>
<div id="ref-Ramette_Multivariate_2007" class="csl-entry">
Ramette, A. 2007. <span>“Multivariate Analyses in Microbial Ecology.”</span> Journal Article. <em>FEMS Microbiol Ecol</em> 62 (2): 142–60. <a href="https://doi.org/10.1111/j.1574-6941.2007.00375.x">https://doi.org/10.1111/j.1574-6941.2007.00375.x</a>.
</div>
<div id="ref-Tuomisto_diversity_2010" class="csl-entry">
Tuomisto, Hanna. 2010. <span>“A Diversity of Beta Diversities: Straightening up a Concept Gone Awry. Part 1. Defining Beta Diversity as a Function of Alpha and Gamma Diversity.”</span> Journal Article. <em>Ecography</em> 33 (1): 2–22.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="composition-based-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="model-based-class.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/04-Diversity-based_class.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
