<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Explainable class | Tutorial for R microeco package (v0.6.5)</title>
  <meta name="description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Explainable class | Tutorial for R microeco package (v0.6.5)" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Explainable class | Tutorial for R microeco package (v0.6.5)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  

<meta name="author" content="Chi Liu, Felipe R. P. Mansoldo, Umer Zeeshan Ijaz, Chenhao Li, Yang Cao, Jarrod J. Scott, Yaoming Cui, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2022-01-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="model-based-class.html"/>
<link rel="next" href="other-dataset.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.1</b> R6 Class</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.2</b> Help</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.3</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="intro.html"><a href="intro.html#important-packages"><i class="fa fa-check"></i><b>2.3.1</b> Important packages</a></li>
<li class="chapter" data-level="2.3.2" data-path="intro.html"><a href="intro.html#ggtree"><i class="fa fa-check"></i><b>2.3.2</b> ggtree</a></li>
<li class="chapter" data-level="2.3.3" data-path="intro.html"><a href="intro.html#spieceasi"><i class="fa fa-check"></i><b>2.3.3</b> SpiecEasi</a></li>
<li class="chapter" data-level="2.3.4" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.3.4</b> Gephi</a></li>
<li class="chapter" data-level="2.3.5" data-path="intro.html"><a href="intro.html#wgcna"><i class="fa fa-check"></i><b>2.3.5</b> WGCNA</a></li>
<li class="chapter" data-level="2.3.6" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.3.6</b> Tax4Fun</a></li>
<li class="chapter" data-level="2.3.7" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.3.7</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#plotting"><i class="fa fa-check"></i><b>2.4</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#example"><i class="fa fa-check"></i><b>3.1.1</b> Example</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.2</b> Key points</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#other-functions"><i class="fa fa-check"></i><b>3.1.3</b> Other functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-2"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-4"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.2</b> Key points</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#other-functions-1"><i class="fa fa-check"></i><b>6.2.3</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-7"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-8"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-9"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>8</b> Other dataset</a>
<ul>
<li class="chapter" data-level="8.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>8.1</b> Fungi data</a></li>
<li class="chapter" data-level="8.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>8.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="8.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>8.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>9</b> Notes</a>
<ul>
<li class="chapter" data-level="9.1" data-path="notes.html"><a href="notes.html#clone"><i class="fa fa-check"></i><b>9.1</b> clone</a></li>
<li class="chapter" data-level="9.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>9.2</b> save() function</a></li>
<li class="chapter" data-level="9.3" data-path="notes.html"><a href="notes.html#subset-of-samples"><i class="fa fa-check"></i><b>9.3</b> subset of samples</a></li>
<li class="chapter" data-level="9.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>9.4</b> change object</a></li>
<li class="chapter" data-level="9.5" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>9.5</b> group order</a></li>
<li class="chapter" data-level="9.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>9.6</b> add layers to plot</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>10</b> mecodev package</a>
<ul>
<li class="chapter" data-level="10.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>10.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="10.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_convert"><i class="fa fa-check"></i><b>10.2</b> trans_convert</a></li>
<li class="chapter" data-level="10.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_netchord"><i class="fa fa-check"></i><b>10.3</b> trans_netchord</a></li>
<li class="chapter" data-level="10.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>10.4</b> trans_ts</a></li>
<li class="chapter" data-level="10.5" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>10.5</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="other-examples.html"><a href="other-examples.html"><i class="fa fa-check"></i><b>11</b> Other examples</a>
<ul>
<li class="chapter" data-level="11.1" data-path="other-examples.html"><a href="other-examples.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>11.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="11.2" data-path="other-examples.html"><a href="other-examples.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>11.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="11.3" data-path="other-examples.html"><a href="other-examples.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>11.3</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="11.4" data-path="other-examples.html"><a href="other-examples.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>11.4</b> The use of phylogenetic tree</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v0.6.5)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="explainable-class" class="section level1" number="7">
<h1><span class="header-section-number">Chapter 7</span> Explainable class</h1>
<p>We group trans_env and trans_func classes into ‘Explainable class,’
as the environmental measurements and functional data are generally explainable factors.</p>
<div id="trans_env-class" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> trans_env class</h2>
<p>　The environmental variables are explainable indices in analyzing microbial community structure and assembly mechanisms.</p>
<div id="example-8" class="section level3" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> Example</h3>
<p>We first show the RDA analysis (db-RDA and RDA).</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="explainable-class.html#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add_data is used to add the environmental data</span></span>
<span id="cb138-2"><a href="explainable-class.html#cb138-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span></code></pre></div>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="explainable-class.html#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use bray-curtis distance to do dbrda</span></span>
<span id="cb139-2"><a href="explainable-class.html#cb139-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_rda</span>(<span class="at">use_dbrda =</span> <span class="cn">TRUE</span>, <span class="at">use_measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb139-3"><a href="explainable-class.html#cb139-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rda is the result list stored in the object</span></span>
<span id="cb139-4"><a href="explainable-class.html#cb139-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">trans_rda</span>(<span class="at">adjust_arrow_length =</span> <span class="cn">TRUE</span>, <span class="at">max_perc_env =</span> <span class="dv">10</span>)</span>
<span id="cb139-5"><a href="explainable-class.html#cb139-5" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rda_trans is the transformed result for plotting</span></span>
<span id="cb139-6"><a href="explainable-class.html#cb139-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_rda</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_rda_dbrda.png" width="650px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="explainable-class.html#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Genus</span></span>
<span id="cb140-2"><a href="explainable-class.html#cb140-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_rda</span>(<span class="at">use_dbrda =</span> <span class="cn">FALSE</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb140-3"><a href="explainable-class.html#cb140-3" aria-hidden="true" tabindex="-1"></a><span class="co"># As the main results of RDA are related with the projection and angles between different arrows,</span></span>
<span id="cb140-4"><a href="explainable-class.html#cb140-4" aria-hidden="true" tabindex="-1"></a><span class="co"># we adjust the length of the arrow to show them clearly using several parameters.</span></span>
<span id="cb140-5"><a href="explainable-class.html#cb140-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">trans_rda</span>(<span class="at">show_taxa =</span> <span class="dv">10</span>, <span class="at">adjust_arrow_length =</span> <span class="cn">TRUE</span>, <span class="at">max_perc_env =</span> <span class="dv">1500</span>, <span class="at">max_perc_tax =</span> <span class="dv">3000</span>, <span class="at">min_perc_env =</span> <span class="dv">200</span>, <span class="at">min_perc_tax =</span> <span class="dv">300</span>)</span>
<span id="cb140-6"><a href="explainable-class.html#cb140-6" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rda_trans is the transformed result for plotting</span></span>
<span id="cb140-7"><a href="explainable-class.html#cb140-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_rda</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_rda_genus.png" width="650px" style="display: block; margin: auto;" /></p>
<p>Mantel test can be used to check whether there is significant correlations between environmental variables and distance matrix.</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="explainable-class.html#cb141-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_mantel</span>(<span class="at">use_measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb141-2"><a href="explainable-class.html#cb141-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_mantel</span></span>
<span id="cb141-3"><a href="explainable-class.html#cb141-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_mantel</span></code></pre></div>
<pre><code>## The result is stored in object$res_mantel or object$res_mantel_partial ...</code></pre>
<table style="width:88%;">
<colgroup>
<col width="22%" />
<col width="18%" />
<col width="15%" />
<col width="11%" />
<col width="20%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">variable_name</th>
<th align="center">cor_method</th>
<th align="center">corr_res</th>
<th align="center">p_res</th>
<th align="center">significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Temperature</td>
<td align="center">pearson</td>
<td align="center">0.452</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="even">
<td align="center">Precipitation</td>
<td align="center">pearson</td>
<td align="center">0.2791</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="odd">
<td align="center">TOC</td>
<td align="center">pearson</td>
<td align="center">0.13</td>
<td align="center">0.003</td>
<td align="center">**</td>
</tr>
<tr class="even">
<td align="center">NH4</td>
<td align="center">pearson</td>
<td align="center">-0.05539</td>
<td align="center">0.926</td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="center">NO3</td>
<td align="center">pearson</td>
<td align="center">0.06758</td>
<td align="center">0.05</td>
<td align="center">*</td>
</tr>
<tr class="even">
<td align="center">pH</td>
<td align="center">pearson</td>
<td align="center">0.4085</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="odd">
<td align="center">Conductivity</td>
<td align="center">pearson</td>
<td align="center">0.2643</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="even">
<td align="center">TN</td>
<td align="center">pearson</td>
<td align="center">0.1321</td>
<td align="center">0.002</td>
<td align="center">**</td>
</tr>
</tbody>
</table>
<p>The correlations between environmental variables and taxa are important in analyzing and inferring the factors affecting community structure.
Let’s first perform a correlation heatmap using Genus level data.</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="explainable-class.html#cb143-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb143-2"><a href="explainable-class.html#cb143-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">use_data =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;fdr&quot;</span>)</span></code></pre></div>
<pre><code>## The correlation result is stored in object$res_cor ...</code></pre>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="explainable-class.html#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_cor</span></span></code></pre></div>
<p>Then, we can plot the correlation results using plot_cor function.</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="explainable-class.html#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># default ggplot2 method with clustering</span></span>
<span id="cb146-2"><a href="explainable-class.html#cb146-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p>There are too many genera.
We can use the filter_feature parameter in plot_cor to filter some taxa that do not have any significance &lt; 0.001.</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="explainable-class.html#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter genera that do not have at least one ***</span></span>
<span id="cb147-2"><a href="explainable-class.html#cb147-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">filter_feature =</span> <span class="fu">c</span>(<span class="st">&quot;&quot;</span>, <span class="st">&quot;*&quot;</span>, <span class="st">&quot;**&quot;</span>))</span></code></pre></div>
<p>Sometimes, if we want to conduct the correlation analysis between the environmental factors and some important taxa which are detected in the biomarker analysis,
we can use other_taxa parameter in cal_cor function.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="explainable-class.html#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first create trans_diff object as a demonstration</span></span>
<span id="cb148-2"><a href="explainable-class.html#cb148-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;rf&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">rf_taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb148-3"><a href="explainable-class.html#cb148-3" aria-hidden="true" tabindex="-1"></a><span class="co"># then create trans_env object</span></span>
<span id="cb148-4"><a href="explainable-class.html#cb148-4" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb148-5"><a href="explainable-class.html#cb148-5" aria-hidden="true" tabindex="-1"></a><span class="co"># use other_taxa to select taxa you need</span></span>
<span id="cb148-6"><a href="explainable-class.html#cb148-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">use_data =</span> <span class="st">&quot;other&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;fdr&quot;</span>, <span class="at">other_taxa =</span> t2<span class="sc">$</span>res_rf<span class="sc">$</span>Taxa[<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>])</span>
<span id="cb148-7"><a href="explainable-class.html#cb148-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/plot_corr_ggplot.png" width="700px" style="display: block; margin: auto;" /></p>
<p>The pheatmap method is also available.
Note that, besides the color_vector parameter,
color_palette can also be used to control color palette with customized colors.</p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="explainable-class.html#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clustering heatmap; require pheatmap package</span></span>
<span id="cb149-2"><a href="explainable-class.html#cb149-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Let&#39;s take another color pallete</span></span>
<span id="cb149-3"><a href="explainable-class.html#cb149-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">pheatmap =</span> <span class="cn">TRUE</span>, <span class="at">color_palette =</span> <span class="fu">rev</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">9</span>, <span class="at">name =</span> <span class="st">&quot;RdYlBu&quot;</span>)))</span></code></pre></div>
<p><img src="Images/plot_corr_pheatmap.png" width="700px" style="display: block; margin: auto;" /></p>
<p>Sometimes, if it is necessary to study the correlations between environmental variables and taxa for different groups,
by_group parameter can be used for this goal.</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="explainable-class.html#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate correlations for different groups using parameter by_group</span></span>
<span id="cb150-2"><a href="explainable-class.html#cb150-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">by_group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">use_data =</span> <span class="st">&quot;other&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;fdr&quot;</span>, <span class="at">other_taxa =</span> t2<span class="sc">$</span>res_rf<span class="sc">$</span>Taxa[<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>])</span>
<span id="cb150-3"><a href="explainable-class.html#cb150-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_cor</span></span>
<span id="cb150-4"><a href="explainable-class.html#cb150-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/plot_corr_ggplot_groups.png" width="700px" style="display: block; margin: auto;" /></p>
<p>If you are concerned with the relationship between environmental factors and alpha diversity, you can also use this function.</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="explainable-class.html#cb151-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb151-2"><a href="explainable-class.html#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="co"># use add_abund_table parameter to add the extra data table</span></span>
<span id="cb151-3"><a href="explainable-class.html#cb151-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">add_abund_table =</span> dataset<span class="sc">$</span>alpha_diversity)</span>
<span id="cb151-4"><a href="explainable-class.html#cb151-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">pheatmap =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/plot_corr_alpha_diversity.png" width="600px" style="display: block; margin: auto;" /></p>
<p>The function plot_scatterfit() in trans_env class is designed for the scatter plot, adding the fitted line and statistics of correlation or regression.</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="explainable-class.html#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use pH and bray-curtis distance</span></span>
<span id="cb152-2"><a href="explainable-class.html#cb152-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_scatterfit</span>(</span>
<span id="cb152-3"><a href="explainable-class.html#cb152-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;pH&quot;</span>, </span>
<span id="cb152-4"><a href="explainable-class.html#cb152-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> dataset<span class="sc">$</span>beta_diversity<span class="sc">$</span>bray[<span class="fu">rownames</span>(t1<span class="sc">$</span>env_data), <span class="fu">rownames</span>(t1<span class="sc">$</span>env_data)], </span>
<span id="cb152-5"><a href="explainable-class.html#cb152-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">use_cor =</span> <span class="cn">TRUE</span>,</span>
<span id="cb152-6"><a href="explainable-class.html#cb152-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> .<span class="dv">1</span>,</span>
<span id="cb152-7"><a href="explainable-class.html#cb152-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">x_axis_title =</span> <span class="st">&quot;Euclidean distance of pH&quot;</span>, </span>
<span id="cb152-8"><a href="explainable-class.html#cb152-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">y_axis_title =</span> <span class="st">&quot;Bray-Curtis distance&quot;</span>, </span>
<span id="cb152-9"><a href="explainable-class.html#cb152-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">text_x_pos =</span> <span class="dv">4</span>, </span>
<span id="cb152-10"><a href="explainable-class.html#cb152-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">text_y_pos =</span> .<span class="dv">4</span>)</span></code></pre></div>
<p><img src="Images/plot_scatterfit_pHbray.png" width="550px" style="display: block; margin: auto;" /></p>
</div>
<div id="key-points-8" class="section level3" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> Key points</h3>
<ul>
<li>add_abund_table parameter in cal_cor: other customized data can be also provided for the correlation analysis.</li>
<li>use_cor parameter in plot_scatterfit: both the correlation and regression are available in this function.</li>
<li>cal_mantel(): partial_mantel = TRUE can be used for partial mantel test.</li>
</ul>
</div>
</div>
<div id="trans_func-class" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> trans_func class</h2>
<p>　Ecological researchers are usually interested in the the funtional profiles of microbial communities,
because functional or metabolic data is powerful to explain the structure and dynamics of microbial communities and to infer the underlying mechanisms.
As metagenomic sequencing is complicated and expensive, using amplicon sequencing data to predict functional profiles is an alternative choice.
Several software are often used for this goal, such as PICRUSt<span class="citation">(<a href="#ref-Langille_Predictive_2013" role="doc-biblioref">Langille et al. 2013</a>)</span>, Tax4Fun<span class="citation">(<a href="#ref-Aßhauer_Tax4Fun_2015" role="doc-biblioref">Aßhauer et al. 2015</a>)</span> and FAPROTAX<span class="citation">(<a href="#ref-Louca_High_2016" role="doc-biblioref">Stilianos Louca et al. 2016</a>; <a href="#ref-Louca_Decoupling_2016" role="doc-biblioref">S. Louca, Parfrey, and Doebeli 2016</a>)</span>.
These tools are great to be used for the prediction of functional profiles based on the prokaryotic communities from sequencing results.
In addition, it is also important to obtain the functions for each taxa or OTU, not just the whole profile of communities.
But it is hard to know exact functions of each OTU.
FAPROTAX database is a collection of the traits and functions of prokaryotes based on the known research results published in books and literatures.
We match the taxonomic information of prokaryotes against this database to predict the traits of prokaryotes on biogeochemical roles.
The NJC19 database<span class="citation">(<a href="#ref-Lim_Large_2020" role="doc-biblioref">Lim et al. 2020</a>)</span> is also available for animal-associated prokaryotic data, such as human gut microbiota.
We also implement the FUNGuild <span class="citation">(<a href="#ref-Nguyen_FUNGuild_2016" role="doc-biblioref">Nguyen et al. 2016</a>)</span> and FungalTraits <span class="citation">(<a href="#ref-Polme_FungalTraits_2020" role="doc-biblioref">Põlme et al. 2020</a>)</span> databases to predict the fungal traits.</p>
<div id="example-9" class="section level3" number="7.2.1">
<h3><span class="header-section-number">7.2.1</span> Example</h3>
<p>We first identify/predict traits of taxa with the prokaryotic example data.</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="explainable-class.html#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create object of trans_func</span></span>
<span id="cb153-2"><a href="explainable-class.html#cb153-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(dataset)</span>
<span id="cb153-3"><a href="explainable-class.html#cb153-3" aria-hidden="true" tabindex="-1"></a><span class="co"># mapping the taxonomy to the database</span></span>
<span id="cb153-4"><a href="explainable-class.html#cb153-4" aria-hidden="true" tabindex="-1"></a><span class="co"># this can recognize prokaryotes or fungi automatically if the names of taxonomic levels are standard.</span></span>
<span id="cb153-5"><a href="explainable-class.html#cb153-5" aria-hidden="true" tabindex="-1"></a><span class="co"># default database for prokaryotes is FAPROTAX database</span></span>
<span id="cb153-6"><a href="explainable-class.html#cb153-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func</span>(<span class="at">prok_database =</span> <span class="st">&quot;FAPROTAX&quot;</span>)</span></code></pre></div>
<pre><code>## Please also cite the original FAPROTAX paper: Louca et al. (2016).</code></pre>
<pre><code>## Decoupling function and taxonomy in the global ocean microbiome. Science, 353(6305), 1272.</code></pre>
<pre><code>## The functional binary table is stored in object$res_spe_func ...</code></pre>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="explainable-class.html#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return t2$res_spe_func, 1 represent trait exists, 0 represent no or cannot confirmed.</span></span></code></pre></div>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="explainable-class.html#cb158-1" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>res_spe_func[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code></pre></div>
<table style="width:85%;">
<colgroup>
<col width="20%" />
<col width="22%" />
<col width="41%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">methanotrophy</th>
<th align="center">acetoclastic_methanogenesis</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>OTU_4272</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_236</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_399</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_1556</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_32</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>The percentages of the OTUs having the same trait can reflect the functional redundancy of this function in the community or the module in the network.</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="explainable-class.html#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the percentages of OTUs for each trait in each module of network</span></span>
<span id="cb159-2"><a href="explainable-class.html#cb159-2" aria-hidden="true" tabindex="-1"></a><span class="co"># use_community = TRUE; for community</span></span>
<span id="cb159-3"><a href="explainable-class.html#cb159-3" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">use_community =</span> <span class="cn">TRUE</span>)</span>
<span id="cb159-4"><a href="explainable-class.html#cb159-4" aria-hidden="true" tabindex="-1"></a><span class="co"># return t2$res_spe_func_perc for each sample</span></span>
<span id="cb159-5"><a href="explainable-class.html#cb159-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Then, let&#39;s explore network modules; use_community = FALSE</span></span>
<span id="cb159-6"><a href="explainable-class.html#cb159-6" aria-hidden="true" tabindex="-1"></a><span class="co"># node_type_table is generated in the trans_network part</span></span>
<span id="cb159-7"><a href="explainable-class.html#cb159-7" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">use_community =</span> <span class="cn">FALSE</span>, <span class="at">node_type_table =</span> network_node_type)</span>
<span id="cb159-8"><a href="explainable-class.html#cb159-8" aria-hidden="true" tabindex="-1"></a><span class="co"># we only plot some important traits, so we use the default group list to filter and show the traits.</span></span>
<span id="cb159-9"><a href="explainable-class.html#cb159-9" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_spe_func_perc</span>(<span class="at">select_samples =</span> <span class="fu">paste0</span>(<span class="st">&quot;M&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>))</span>
<span id="cb159-10"><a href="explainable-class.html#cb159-10" aria-hidden="true" tabindex="-1"></a><span class="co"># M represents module, ordered by the nodes number from high to low</span></span></code></pre></div>
<p><img src="Images/plot_func_perc.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="explainable-class.html#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to change the group list, reset the list t2$func_group_list</span></span>
<span id="cb160-2"><a href="explainable-class.html#cb160-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>func_group_list</span>
<span id="cb160-3"><a href="explainable-class.html#cb160-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use show_prok_func to see the detailed information of prokaryotic traits</span></span>
<span id="cb160-4"><a href="explainable-class.html#cb160-4" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">show_prok_func</span>(<span class="st">&quot;methanotrophy&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="explainable-class.html#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the percentages for communities</span></span>
<span id="cb161-2"><a href="explainable-class.html#cb161-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">use_community =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## The result table is stored in object$res_spe_func_perc ...</code></pre>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="explainable-class.html#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># t2$res_spe_func_perc[1:5, 1:2]</span></span></code></pre></div>
<table style="width:76%;">
<colgroup>
<col width="12%" />
<col width="22%" />
<col width="41%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">methanotrophy</th>
<th align="center">acetoclastic_methanogenesis</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">0.39</td>
<td align="center">0.04</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">0.27</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">0.48</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">0.48</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">0.56</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="explainable-class.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="co"># then we try to correlate the res_spe_func_perc of communities to environmental variables</span></span>
<span id="cb164-2"><a href="explainable-class.html#cb164-2" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb164-3"><a href="explainable-class.html#cb164-3" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">add_abund_table =</span> t2<span class="sc">$</span>res_spe_func_perc, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb164-4"><a href="explainable-class.html#cb164-4" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">pheatmap =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/plot_func_perc_corr.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Tax4Fun requires a strict input file format associated with the taxonomic information.
To analyze the trimmed or changed OTU data in R with Tax4Fun, we provide a link to the Tax4Fun functional prediction.
Please see <a href="https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun" class="uri">https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun</a> for installing Tax4Fun package and download SILVA123 ref data.</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="explainable-class.html#cb165-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(dataset)</span>
<span id="cb165-2"><a href="explainable-class.html#cb165-2" aria-hidden="true" tabindex="-1"></a><span class="co"># see https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun for the installation</span></span>
<span id="cb165-3"><a href="explainable-class.html#cb165-3" aria-hidden="true" tabindex="-1"></a><span class="co"># and provide the file path of SILVA123</span></span>
<span id="cb165-4"><a href="explainable-class.html#cb165-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_tax4fun</span>(<span class="at">folderReferenceData =</span> <span class="st">&quot;./SILVA123&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in write.table(otu_file, file = output, append = TRUE, quote = FALSE, :
## appending column names to file</code></pre>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="explainable-class.html#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return two files: t1$tax4fun_KO: KO file; t1$tax4fun_path: pathway file.</span></span>
<span id="cb167-2"><a href="explainable-class.html#cb167-2" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$tax4fun_KO$Tax4FunProfile[1:5, 1:2]</span></span></code></pre></div>
<table>
<colgroup>
<col width="12%" />
<col width="43%" />
<col width="43%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">K00001; alcohol dehydrogenase
[EC:1.1.1.1]</th>
<th align="center">K00002; alcohol dehydrogenase
(NADP+) [EC:1.1.1.2]</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">0.0004823</td>
<td align="center">5.942e-06</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">0.0005266</td>
<td align="center">4.017e-06</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">0.0005054</td>
<td align="center">6.168e-06</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">0.0005109</td>
<td align="center">5.888e-06</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">0.0005083</td>
<td align="center">5.547e-06</td>
</tr>
</tbody>
</table>
<p>Now, we use pathway file to analyze the abundance of pathway.</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="explainable-class.html#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="co"># must transpose to taxa row, sample column</span></span>
<span id="cb168-2"><a href="explainable-class.html#cb168-2" aria-hidden="true" tabindex="-1"></a>pathway_file <span class="ot">&lt;-</span> t1<span class="sc">$</span>tax4fun_path<span class="sc">$</span>Tax4FunProfile <span class="sc">%&gt;%</span> t <span class="sc">%&gt;%</span> as.data.frame</span>
<span id="cb168-3"><a href="explainable-class.html#cb168-3" aria-hidden="true" tabindex="-1"></a><span class="co"># filter rownames, only keep ko+number</span></span>
<span id="cb168-4"><a href="explainable-class.html#cb168-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pathway_file) <span class="sc">%&lt;&gt;%</span> <span class="fu">gsub</span>(<span class="st">&quot;(^.*);</span><span class="sc">\\</span><span class="st">s.*&quot;</span>, <span class="st">&quot;</span><span class="sc">\\</span><span class="st">1&quot;</span>, .)</span>
<span id="cb168-5"><a href="explainable-class.html#cb168-5" aria-hidden="true" tabindex="-1"></a><span class="co"># load the pathway hierarchical metadata</span></span>
<span id="cb168-6"><a href="explainable-class.html#cb168-6" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(ko_map)</span>
<span id="cb168-7"><a href="explainable-class.html#cb168-7" aria-hidden="true" tabindex="-1"></a><span class="co"># further create a microtable object, familiar?</span></span>
<span id="cb168-8"><a href="explainable-class.html#cb168-8" aria-hidden="true" tabindex="-1"></a>func1 <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">otu_table =</span> pathway_file, <span class="at">tax_table =</span> ko_map, <span class="at">sample_table =</span> t1<span class="sc">$</span>sample_table)</span>
<span id="cb168-9"><a href="explainable-class.html#cb168-9" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(func1)</span></code></pre></div>
<pre><code>## microtable class:
## sample_table have 90 rows and 4 columns
## otu_table have 284 rows and 90 columns
## tax_table have 341 rows and 3 columns</code></pre>
<p>Now, we need to trim data and calculate abundance.</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="explainable-class.html#cb170-1" aria-hidden="true" tabindex="-1"></a>func1<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb170-2"><a href="explainable-class.html#cb170-2" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate abundance automatically at three levels: level_1, level_2, level_3</span></span>
<span id="cb170-3"><a href="explainable-class.html#cb170-3" aria-hidden="true" tabindex="-1"></a>func1<span class="sc">$</span><span class="fu">cal_abund</span>()</span></code></pre></div>
<pre><code>## The result is stored in object$taxa_abund ...</code></pre>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="explainable-class.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(func1)</span></code></pre></div>
<pre><code>## microtable class:
## sample_table have 90 rows and 4 columns
## otu_table have 284 rows and 90 columns
## tax_table have 284 rows and 3 columns
## Taxa abundance: calculated for level_1,level_2,level_3</code></pre>
<p>Then, we can plot the abundance.</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="explainable-class.html#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="co"># bar plot at level_1</span></span>
<span id="cb174-2"><a href="explainable-class.html#cb174-2" aria-hidden="true" tabindex="-1"></a>func2 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(func1, <span class="at">taxrank =</span> <span class="st">&quot;level_1&quot;</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb174-3"><a href="explainable-class.html#cb174-3" aria-hidden="true" tabindex="-1"></a>func2<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/plot_bar_tax4fun1.png" width="600px" style="display: block; margin: auto;" /></p>
<p>We can also do something else. For example, we can use lefse to test the differences of the abundances and find the important enriched pathways across groups.</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="explainable-class.html#cb175-1" aria-hidden="true" tabindex="-1"></a>func2 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> func1, <span class="at">method =</span> <span class="st">&quot;lefse&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">lefse_subgroup =</span> <span class="cn">NULL</span>)</span>
<span id="cb175-2"><a href="explainable-class.html#cb175-2" aria-hidden="true" tabindex="-1"></a>func2<span class="sc">$</span><span class="fu">plot_lefse_bar</span>(<span class="at">LDA_score =</span> <span class="dv">3</span>, <span class="at">width =</span> <span class="fl">0.8</span>)</span></code></pre></div>
<p><img src="Images/plot_lefse_bar_tax4fun.png" width="600px" style="display: block; margin: auto;" /></p>
<p>Tax4Fun2 <span class="citation">(<a href="#ref-Wemheuer_Tax4Fun2_2020" role="doc-biblioref">Wemheuer et al. 2020</a>)</span> is another R package for the prediction of functional profiles of prokaryotic communities from 16S rRNA gene sequences.
It also provides two indexes for the evaluation of functional gene redundancies.
If you want to use Tax4Fun2 method, you must add the representative fasta file to the microtable object.
Please see <a href="https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun2" class="uri">https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun2</a> for downloading ncbi-blast and Ref99NR/Ref100NR.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="explainable-class.html#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a microtable object with the fasta file</span></span>
<span id="cb176-2"><a href="explainable-class.html#cb176-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(sample_info_16S)</span>
<span id="cb176-3"><a href="explainable-class.html#cb176-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(otu_table_16S)</span>
<span id="cb176-4"><a href="explainable-class.html#cb176-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(taxonomy_table_16S)</span>
<span id="cb176-5"><a href="explainable-class.html#cb176-5" aria-hidden="true" tabindex="-1"></a><span class="co"># representative sequences are necessary for this analysis</span></span>
<span id="cb176-6"><a href="explainable-class.html#cb176-6" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(rep_fasta_16S)</span>
<span id="cb176-7"><a href="explainable-class.html#cb176-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-8"><a href="explainable-class.html#cb176-8" aria-hidden="true" tabindex="-1"></a>use_dataset <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">sample_table =</span> sample_info_16S, <span class="at">otu_table =</span> otu_table_16S, <span class="at">tax_table =</span> taxonomy_table_16S, <span class="at">rep_fasta =</span> rep_fasta_16S)</span>
<span id="cb176-9"><a href="explainable-class.html#cb176-9" aria-hidden="true" tabindex="-1"></a>use_dataset<span class="sc">$</span><span class="fu">filter_pollution</span>(<span class="at">taxa =</span> <span class="fu">c</span>(<span class="st">&quot;mitochondria&quot;</span>, <span class="st">&quot;chloroplast&quot;</span>))</span>
<span id="cb176-10"><a href="explainable-class.html#cb176-10" aria-hidden="true" tabindex="-1"></a>use_dataset<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb176-11"><a href="explainable-class.html#cb176-11" aria-hidden="true" tabindex="-1"></a>use_dataset</span>
<span id="cb176-12"><a href="explainable-class.html#cb176-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-13"><a href="explainable-class.html#cb176-13" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(use_dataset)</span>
<span id="cb176-14"><a href="explainable-class.html#cb176-14" aria-hidden="true" tabindex="-1"></a><span class="co"># create a directory for result and log files</span></span>
<span id="cb176-15"><a href="explainable-class.html#cb176-15" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;test_prediction&quot;</span>)</span>
<span id="cb176-16"><a href="explainable-class.html#cb176-16" aria-hidden="true" tabindex="-1"></a><span class="co"># see https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun2 for installation</span></span>
<span id="cb176-17"><a href="explainable-class.html#cb176-17" aria-hidden="true" tabindex="-1"></a><span class="co"># ignore blast_tool_path parameter if blast tools have been in path</span></span>
<span id="cb176-18"><a href="explainable-class.html#cb176-18" aria-hidden="true" tabindex="-1"></a><span class="co"># the function can search whether blast tool directory is in the path, if not, automatically use provided blast_tool_path parameter</span></span>
<span id="cb176-19"><a href="explainable-class.html#cb176-19" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_tax4fun2</span>(<span class="at">blast_tool_path =</span> <span class="st">&quot;ncbi-blast-2.5.0+/bin&quot;</span>, <span class="at">path_to_reference_data =</span> <span class="st">&quot;Tax4Fun2_ReferenceData_v2&quot;</span>,</span>
<span id="cb176-20"><a href="explainable-class.html#cb176-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">database_mode =</span> <span class="st">&quot;Ref99NR&quot;</span>, <span class="at">path_to_temp_folder =</span> <span class="st">&quot;test_prediction&quot;</span>)</span>
<span id="cb176-21"><a href="explainable-class.html#cb176-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb176-22"><a href="explainable-class.html#cb176-22" aria-hidden="true" tabindex="-1"></a><span class="co"># functional gene redundancies</span></span>
<span id="cb176-23"><a href="explainable-class.html#cb176-23" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_tax4fun2_FRI</span>()</span></code></pre></div>
</div>
<div id="key-points-9" class="section level3" number="7.2.2">
<h3><span class="header-section-number">7.2.2</span> Key points</h3>
<ul>
<li>blast_tool_path parameter in cal_tax4fun2: if the blast tool has been in ‘environment variable’ of computer, it is ok to use blast_tool_path = NULL</li>
<li>blast version: tax4fun2 require NCBI blast tool. However, some errors can come from the latest versions (<a href="https://www.biostars.org/p/413294/" class="uri">https://www.biostars.org/p/413294/</a>). An easy solution is to use previous version (such as 2.5.0).</li>
</ul>

</div>
</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Aßhauer_Tax4Fun_2015" class="csl-entry">
Aßhauer, Kathrin P, Bernd Wemheuer, Rolf Daniel, and Peter Meinicke. 2015. <span>“Tax4Fun: Predicting Functional Profiles from Metagenomic 16s rRNA Data.”</span> Journal Article. <em>Bioinformatics</em> 31 (17): 2882–84.
</div>
<div id="ref-Langille_Predictive_2013" class="csl-entry">
Langille, M. G., J. Zaneveld, J. G. Caporaso, D. McDonald, D. Knights, J. A. Reyes, J. C. Clemente, et al. 2013. <span>“Predictive Functional Profiling of Microbial Communities Using 16s rRNA Marker Gene Sequences.”</span> Journal Article. <em>Nature Biotechnology</em> 31 (9): 814–21. <a href="https://doi.org/10.1038/nbt.2676">https://doi.org/10.1038/nbt.2676</a>.
</div>
<div id="ref-Lim_Large_2020" class="csl-entry">
Lim, Roktaek, Josephine Jill T. Cabatbat, Thomas L. P. Martin, Haneul Kim, Seunghyeon Kim, Jaeyun Sung, Cheol-Min Ghim, and Pan-Jun Kim. 2020. <span>“Large-Scale Metabolic Interaction Network of the Mouse and Human Gut Microbiota.”</span> Journal Article. <em>Scientific Data</em> 7 (1). <a href="https://doi.org/10.1038/s41597-020-0516-5">https://doi.org/10.1038/s41597-020-0516-5</a>.
</div>
<div id="ref-Louca_Decoupling_2016" class="csl-entry">
Louca, S, L. W. Parfrey, and M Doebeli. 2016. <span>“Decoupling Function and Taxonomy in the Global Ocean Microbiome.”</span> Journal Article. <em>Science</em> 353 (6305): 1272.
</div>
<div id="ref-Louca_High_2016" class="csl-entry">
Louca, Stilianos, Saulo MS Jacques, Aliny PF Pires, Juliana S Leal, Diane S Srivastava, Laura Wegener Parfrey, Vinicius F Farjalla, and Michael Doebeli. 2016. <span>“High Taxonomic Variability Despite Stable Functional Structure Across Microbial Communities.”</span> Journal Article. <em>Nature Ecology &amp; Evolution</em> 1: 0015.
</div>
<div id="ref-Nguyen_FUNGuild_2016" class="csl-entry">
Nguyen, Nhu H., Zewei Song, Scott T. Bates, Sara Branco, Leho Tedersoo, Jon Menke, Jonathan S. Schilling, and Peter G. Kennedy. 2016. <span>“FUNGuild: An Open Annotation Tool for Parsing Fungal Community Datasets by Ecological Guild.”</span> Journal Article. <em>Fungal Ecology</em> 20 (1): 241–48.
</div>
<div id="ref-Polme_FungalTraits_2020" class="csl-entry">
Põlme, Sergei, Kessy Abarenkov, R. Henrik Nilsson, Björn D. Lindahl, Karina Engelbrecht Clemmensen, Havard Kauserud, Nhu Nguyen, et al. 2020. <span>“FungalTraits: A User-Friendly Traits Database of Fungi and Fungus-Like Stramenopiles.”</span> Journal Article. <em>Fungal Diversity</em> 105 (1): 1–16. <a href="https://doi.org/10.1007/s13225-020-00466-2">https://doi.org/10.1007/s13225-020-00466-2</a>.
</div>
<div id="ref-Wemheuer_Tax4Fun2_2020" class="csl-entry">
Wemheuer, Franziska, Jessica A. Taylor, Rolf Daniel, Emma Johnston, Peter Meinicke, Torsten Thomas, and Bernd Wemheuer. 2020. <span>“Tax4Fun2: Prediction of Habitat-Specific Functional Profiles and Functional Redundancy Based on 16s rRNA Gene Sequences.”</span> Journal Article. <em>Environmental Microbiome</em> 15 (1). <a href="https://doi.org/10.1186/s40793-020-00358-7">https://doi.org/10.1186/s40793-020-00358-7</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="model-based-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="other-dataset.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/06-Explainable_class.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
