<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Explainable class | Tutorial for R microeco package (v0.15.0)</title>
  <meta name="description" content="The tutorial for R microeco, file2meco, meconetcomp and mecodev packages" />
  <meta name="generator" content="bookdown 0.30 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Explainable class | Tutorial for R microeco package (v0.15.0)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecodev packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Explainable class | Tutorial for R microeco package (v0.15.0)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecodev packages" />
  

<meta name="author" content="Chi Liu, Felipe R. P. Mansoldo, Umer Zeeshan Ijaz, Chenhao Li, Yang Cao, Jarrod J. Scott, Yaoming Cui, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2023-03-31" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="model-based-class.html"/>
<link rel="next" href="file2meco-package.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#rtools"><i class="fa fa-check"></i><b>2.4</b> RTools</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.5</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#important-packages"><i class="fa fa-check"></i><b>2.5.1</b> Important packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#cran-packages"><i class="fa fa-check"></i><b>2.5.2</b> CRAN packages</a></li>
<li class="chapter" data-level="2.5.3" data-path="intro.html"><a href="intro.html#ggtree"><i class="fa fa-check"></i><b>2.5.3</b> ggtree</a></li>
<li class="chapter" data-level="2.5.4" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.5.4</b> Gephi</a></li>
<li class="chapter" data-level="2.5.5" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.5.5</b> Tax4Fun</a></li>
<li class="chapter" data-level="2.5.6" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.5.6</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#plot"><i class="fa fa-check"></i><b>2.6</b> Plot</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#prepare-the-example-data"><i class="fa fa-check"></i><b>3.1.1</b> Prepare the example data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#how-to-read-your-files-to-microtable-object"><i class="fa fa-check"></i><b>3.1.2</b> How to read your files to microtable object?</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#functions-in-microtable-class"><i class="fa fa-check"></i><b>3.1.3</b> Functions in microtable class</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#merge-taxa-or-samples"><i class="fa fa-check"></i><b>3.1.4</b> merge taxa or samples</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-class.html"><a href="basic-class.html#subset-of-samples"><i class="fa fa-check"></i><b>3.1.5</b> subset of samples</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-class.html"><a href="basic-class.html#subset-of-taxa"><i class="fa fa-check"></i><b>3.1.6</b> subset of taxa</a></li>
<li class="chapter" data-level="3.1.7" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.7</b> Other examples</a></li>
<li class="chapter" data-level="3.1.8" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.8</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-2"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-4"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#network-comparison"><i class="fa fa-check"></i><b>6.2.2</b> Network comparison</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.3</b> Key points</a></li>
<li class="chapter" data-level="6.2.4" data-path="model-based-class.html"><a href="model-based-class.html#other-functions"><i class="fa fa-check"></i><b>6.2.4</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies</a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples</a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points</a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-7"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME</a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2</a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann"><i class="fa fa-check"></i><b>8.3</b> HUMAnN</a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn</a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2"><i class="fa fa-check"></i><b>8.5</b> Kraken2</a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB</a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq</a></li>
<li class="chapter" data-level="8.8" data-path="file2meco-package.html"><a href="file2meco-package.html#picrust2"><i class="fa fa-check"></i><b>8.8</b> PICRUSt2</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset</a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data</a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> <code>clone</code> function</a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> <code>save</code> function</a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#github-issues"><i class="fa fa-check"></i><b>10.3</b> Github Issues</a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.4</b> change object</a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.5</b> group order</a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.6</b> add layers to plot</a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#colors-for-many-groups"><i class="fa fa-check"></i><b>10.7</b> colors for many groups</a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>11</b> mecodev package</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>11.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="11.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_norm"><i class="fa fa-check"></i><b>11.2</b> trans_norm</a></li>
<li class="chapter" data-level="11.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>11.3</b> trans_ts</a></li>
<li class="chapter" data-level="11.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>11.4</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html"><i class="fa fa-check"></i><b>12</b> meconetcomp package</a>
<ul>
<li class="chapter" data-level="12.1" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-modularity-for-all-networks"><i class="fa fa-check"></i><b>12.1</b> Network modularity for all networks</a></li>
<li class="chapter" data-level="12.2" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-topological-attributes-for-all-networks"><i class="fa fa-check"></i><b>12.2</b> Network topological attributes for all networks</a></li>
<li class="chapter" data-level="12.3" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#node-and-edge-properties-extraction-for-all-networks"><i class="fa fa-check"></i><b>12.3</b> Node and edge properties extraction for all networks</a></li>
<li class="chapter" data-level="12.4" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-nodes-across-networks"><i class="fa fa-check"></i><b>12.4</b> Compare nodes across networks</a></li>
<li class="chapter" data-level="12.5" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-edges-across-networks"><i class="fa fa-check"></i><b>12.5</b> Compare edges across networks</a></li>
<li class="chapter" data-level="12.6" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#extract-overlapped-edges-of-networks-to-a-new-network"><i class="fa fa-check"></i><b>12.6</b> Extract overlapped edges of networks to a new network</a></li>
<li class="chapter" data-level="12.7" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-phylogenetic-distances-of-paired-nodes-in-edges"><i class="fa fa-check"></i><b>12.7</b> Compare phylogenetic distances of paired nodes in edges</a></li>
<li class="chapter" data-level="12.8" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-node-sources-of-edges-across-networks"><i class="fa fa-check"></i><b>12.8</b> Compare node sources of edges across networks</a></li>
<li class="chapter" data-level="12.9" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-topological-properties-of-sub-networks"><i class="fa fa-check"></i><b>12.9</b> Compare topological properties of sub-networks</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mecoturn-package.html"><a href="mecoturn-package.html"><i class="fa fa-check"></i><b>13</b> mecoturn package</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mecoturn-package.html"><a href="mecoturn-package.html#betaturn-class"><i class="fa fa-check"></i><b>13.1</b> betaturn class</a></li>
<li class="chapter" data-level="13.2" data-path="mecoturn-package.html"><a href="mecoturn-package.html#taxaturn-class"><i class="fa fa-check"></i><b>13.2</b> taxaturn class</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>14</b> Other examples</a>
<ul>
<li class="chapter" data-level="14.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>14.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="14.2" data-path="other-examples-1.html"><a href="other-examples-1.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>14.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="14.3" data-path="other-examples-1.html"><a href="other-examples-1.html#show-the-abundance-of-unknown-taxa"><i class="fa fa-check"></i><b>14.3</b> Show the abundance of unknown taxa</a></li>
<li class="chapter" data-level="14.4" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>14.4</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="14.5" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>14.5</b> The use of phylogenetic tree</a></li>
<li class="chapter" data-level="14.6" data-path="other-examples-1.html"><a href="other-examples-1.html#mantel-test-correlation-heatmap"><i class="fa fa-check"></i><b>14.6</b> Mantel test + Correlation heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v0.15.0)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="explainable-class" class="section level1 hasAnchor" number="7">
<h1><span class="header-section-number">Chapter 7</span> Explainable class<a href="explainable-class.html#explainable-class" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>We group trans_env and trans_func classes into ‘Explainable class’,
as environmental factors and microbial functions can be generally applied to explain microbial community structure and assembly.</p>
<div id="trans_env-class" class="section level2 hasAnchor" number="7.1">
<h2><span class="header-section-number">7.1</span> trans_env class<a href="explainable-class.html#trans_env-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There may be some NA (missing value) in the user’s env data.
If so, please add <code>complete_na = TRUE</code> for interpolation when creating the trans_env object.</p>
<div id="example-7" class="section level3 hasAnchor" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> Example<a href="explainable-class.html#example-7" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Creating trans_env object has at least two ways.
The following is using additional environmental data which is not in the microtable object.</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="explainable-class.html#cb204-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add_data is used to add the environmental data</span></span>
<span id="cb204-2"><a href="explainable-class.html#cb204-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span></code></pre></div>
<pre><code>## Env data is stored in object$data_env ...</code></pre>
<p>Maybe a more general way is to directly use the data from sample_table of your microtable object.
To show this operation, we first merge additional table into sample_table to generate a new microtable object.</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="explainable-class.html#cb206-1" aria-hidden="true" tabindex="-1"></a>new_test <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb206-2"><a href="explainable-class.html#cb206-2" aria-hidden="true" tabindex="-1"></a>new_test<span class="sc">$</span>sample_table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(new_test<span class="sc">$</span>sample_table, env_data_16S[<span class="fu">rownames</span>(new_test<span class="sc">$</span>sample_table), ])</span>
<span id="cb206-3"><a href="explainable-class.html#cb206-3" aria-hidden="true" tabindex="-1"></a><span class="co"># now new_test$sample_table has the whole data</span></span>
<span id="cb206-4"><a href="explainable-class.html#cb206-4" aria-hidden="true" tabindex="-1"></a>new_test</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 15 columns
## otu_table have 12766 rows and 90 columns
## tax_table have 12766 rows and 7 columns
## phylo_tree have 12766 tips
## Taxa abundance: calculated for Kingdom,Phylum,Class,Order,Family,Genus,Species 
## Alpha diversity: calculated for Observed,Chao1,se.chao1,ACE,se.ACE,Shannon,Simpson,InvSimpson,Fisher,Coverage 
## Beta diversity: calculated for bray,jaccard</code></pre>
<p>Now let’s use env_cols to select the required columns from sample_table in the microtable object.</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="explainable-class.html#cb208-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> new_test, <span class="at">env_cols =</span> <span class="dv">8</span><span class="sc">:</span><span class="dv">15</span>)</span></code></pre></div>
<pre><code>## Env data is stored in object$data_env ...</code></pre>
<p>Generally, it is beneficial to the understanding on environmental variables in order to better use more methods.
So, we first show the cal_diff and cal_autocor functions.
The cal_diff function is used to test the significance of variables across groups like we have shown in trans_alpha and trans_diff class parts.</p>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="explainable-class.html#cb210-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Wilcoxon Rank Sum and Signed Rank Tests as an example</span></span>
<span id="cb210-2"><a href="explainable-class.html#cb210-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>)</span>
<span id="cb210-3"><a href="explainable-class.html#cb210-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_diff[, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">7</span>)]</span></code></pre></div>
<pre><code>## The result is stored in object$res_diff ...</code></pre>
<table style="width:89%;">
<colgroup>
<col width="18%" />
<col width="22%" />
<col width="11%" />
<col width="16%" />
<col width="20%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Comparison</th>
<th align="center">Measure</th>
<th align="center">Group</th>
<th align="center">P.adj</th>
<th align="center">Significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">IW - CW</td>
<td align="center">Temperature</td>
<td align="center">CW</td>
<td align="center">3.283e-10</td>
<td align="center">***</td>
</tr>
<tr class="even">
<td align="center">IW - TW</td>
<td align="center">Temperature</td>
<td align="center">TW</td>
<td align="center">0.0001958</td>
<td align="center">***</td>
</tr>
<tr class="odd">
<td align="center">CW - TW</td>
<td align="center">Temperature</td>
<td align="center">CW</td>
<td align="center">3.283e-10</td>
<td align="center">***</td>
</tr>
<tr class="even">
<td align="center">IW - CW</td>
<td align="center">Precipitation</td>
<td align="center">CW</td>
<td align="center">5.239e-09</td>
<td align="center">***</td>
</tr>
<tr class="odd">
<td align="center">IW - TW</td>
<td align="center">Precipitation</td>
<td align="center">IW</td>
<td align="center">0.07758</td>
<td align="center">ns</td>
</tr>
<tr class="even">
<td align="center">CW - TW</td>
<td align="center">Precipitation</td>
<td align="center">CW</td>
<td align="center">4.002e-07</td>
<td align="center">***</td>
</tr>
<tr class="odd">
<td align="center">IW - CW</td>
<td align="center">TOC</td>
<td align="center">IW</td>
<td align="center">0.4687</td>
<td align="center">ns</td>
</tr>
</tbody>
</table>
<p>Let’s perform the anova and show the letters in the box plot. We use list to store all the plots for each factor and plot them together.</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="explainable-class.html#cb212-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;anova&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb212-2"><a href="explainable-class.html#cb212-2" aria-hidden="true" tabindex="-1"></a><span class="co"># place all the plots into a list</span></span>
<span id="cb212-3"><a href="explainable-class.html#cb212-3" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb212-4"><a href="explainable-class.html#cb212-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">colnames</span>(t1<span class="sc">$</span>data_env)){</span>
<span id="cb212-5"><a href="explainable-class.html#cb212-5" aria-hidden="true" tabindex="-1"></a>    tmp[[i]] <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_diff</span>(<span class="at">measure =</span> i, <span class="at">add_sig_text_size =</span> <span class="dv">5</span>, <span class="at">xtext_size =</span> <span class="dv">12</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.margin =</span> <span class="fu">unit</span>(<span class="fu">c</span>(<span class="fl">0.1</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="st">&quot;cm&quot;</span>))</span>
<span id="cb212-6"><a href="explainable-class.html#cb212-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb212-7"><a href="explainable-class.html#cb212-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gridExtra<span class="sc">::</span><span class="fu">arrangeGrob</span>(<span class="at">grobs =</span> tmp, <span class="at">ncol =</span> <span class="dv">3</span>))</span></code></pre></div>
<p><img src="Images/plot_env_diff_all.png" width="750px" style="display: block; margin: auto;" /></p>
<p>From the v0.12.0, the trans_env class supports the differential test of groups within each group by using the by_group parameter in cal_diff function.</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="explainable-class.html#cb213-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">group =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">by_group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">method =</span> <span class="st">&quot;anova&quot;</span>)</span>
<span id="cb213-2"><a href="explainable-class.html#cb213-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff</span>(<span class="at">measure =</span> <span class="st">&quot;pH&quot;</span>, <span class="at">add_sig_text_size =</span> <span class="dv">5</span>)</span></code></pre></div>
<p><img src="Images/plot_env_diff_bygroup.png" width="600px" style="display: block; margin: auto;" /></p>
<p>Then we show the autocorrelations among variables.</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="explainable-class.html#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require GGally package installed</span></span>
<span id="cb214-2"><a href="explainable-class.html#cb214-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_autocor</span>()</span></code></pre></div>
<p><img src="Images/trans_env_autocor1.png" width="800px" style="display: block; margin: auto;" /></p>
<p>For different groups, please use group parameter to show the distributions of variables and the autocorrelations across groups.</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="explainable-class.html#cb215-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_autocor</span>(<span class="at">group =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/trans_env_autocor_group.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Then let’s do the RDA analysis (db-RDA and RDA).</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="explainable-class.html#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use bray-curtis distance to do dbrda</span></span>
<span id="cb216-2"><a href="explainable-class.html#cb216-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">method =</span> <span class="st">&quot;dbRDA&quot;</span>, <span class="at">use_measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb216-3"><a href="explainable-class.html#cb216-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rda is the result list stored in the object</span></span>
<span id="cb216-4"><a href="explainable-class.html#cb216-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">trans_ordination</span>(<span class="at">adjust_arrow_length =</span> <span class="cn">TRUE</span>, <span class="at">max_perc_env =</span> <span class="fl">1.5</span>)</span>
<span id="cb216-5"><a href="explainable-class.html#cb216-5" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rda_trans is the transformed result for plotting</span></span>
<span id="cb216-6"><a href="explainable-class.html#cb216-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_rda_dbrda.png" width="650px" style="display: block; margin: auto;" /></p>
<p>From v0.14.0, the function <code>cal_ordination_anova</code> is implemented to check the significance of the ordination model instead of the encapsulation in <code>cal_ordination</code>.
Furthermore, the function <code>cal_ordination_envfit</code> can be used to get the contribution of each variables to the model.</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="explainable-class.html#cb217-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination_anova</span>()</span>
<span id="cb217-2"><a href="explainable-class.html#cb217-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination_envfit</span>()</span></code></pre></div>
<p>Then, let’s try to do RDA at the Genus level.</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="explainable-class.html#cb218-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Genus</span></span>
<span id="cb218-2"><a href="explainable-class.html#cb218-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">method =</span> <span class="st">&quot;RDA&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb218-3"><a href="explainable-class.html#cb218-3" aria-hidden="true" tabindex="-1"></a><span class="co"># As the main results of RDA are related with the projection and angles between different arrows,</span></span>
<span id="cb218-4"><a href="explainable-class.html#cb218-4" aria-hidden="true" tabindex="-1"></a><span class="co"># we adjust the length of the arrow to show them clearly using several parameters.</span></span>
<span id="cb218-5"><a href="explainable-class.html#cb218-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">trans_ordination</span>(<span class="at">show_taxa =</span> <span class="dv">10</span>, <span class="at">adjust_arrow_length =</span> <span class="cn">TRUE</span>, <span class="at">max_perc_env =</span> <span class="fl">1.5</span>, <span class="at">max_perc_tax =</span> <span class="fl">1.5</span>, <span class="at">min_perc_env =</span> <span class="fl">0.2</span>, <span class="at">min_perc_tax =</span> <span class="fl">0.2</span>)</span>
<span id="cb218-6"><a href="explainable-class.html#cb218-6" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rda_trans is the transformed result for plot</span></span>
<span id="cb218-7"><a href="explainable-class.html#cb218-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_rda_genus.png" width="650px" style="display: block; margin: auto;" /></p>
<p>For more visualization styles, run the following examples.</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="explainable-class.html#cb219-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb219-2"><a href="explainable-class.html#cb219-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;ellipse&quot;</span>))</span>
<span id="cb219-3"><a href="explainable-class.html#cb219-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;centroid&quot;</span>))</span>
<span id="cb219-4"><a href="explainable-class.html#cb219-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;chull&quot;</span>))</span>
<span id="cb219-5"><a href="explainable-class.html#cb219-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;ellipse&quot;</span>, <span class="st">&quot;centroid&quot;</span>))</span>
<span id="cb219-6"><a href="explainable-class.html#cb219-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;chull&quot;</span>, <span class="st">&quot;centroid&quot;</span>), <span class="at">add_sample_label =</span> <span class="st">&quot;SampleID&quot;</span>)</span>
<span id="cb219-7"><a href="explainable-class.html#cb219-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="st">&quot;centroid&quot;</span>, <span class="at">centroid_segment_alpha =</span> <span class="fl">0.9</span>, <span class="at">centroid_segment_size =</span> <span class="dv">1</span>, <span class="at">centroid_segment_linetype =</span> <span class="dv">1</span>)</span>
<span id="cb219-8"><a href="explainable-class.html#cb219-8" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Type&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;centroid&quot;</span>), <span class="at">centroid_segment_linetype =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>Mantel test can be used to check whether there is significant correlations between environmental variables and distance matrix.</p>
<div class="sourceCode" id="cb220"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb220-1"><a href="explainable-class.html#cb220-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_mantel</span>(<span class="at">use_measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb220-2"><a href="explainable-class.html#cb220-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_mantel</span></span>
<span id="cb220-3"><a href="explainable-class.html#cb220-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_mantel</span></code></pre></div>
<pre><code>## The result is stored in object$res_mantel ...</code></pre>
<table>
<caption>Table continues below</caption>
<colgroup>
<col width="13%" />
<col width="25%" />
<col width="32%" />
<col width="12%" />
<col width="16%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">by_group</th>
<th align="center">Correlation method</th>
<th align="center">Correlation coefficient</th>
<th align="center">p.value</th>
<th align="center">p.adjusted</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">All</td>
<td align="center">pearson</td>
<td align="center">0.452</td>
<td align="center">0.001</td>
<td align="center">0.002</td>
</tr>
<tr class="even">
<td align="center">All</td>
<td align="center">pearson</td>
<td align="center">0.2791</td>
<td align="center">0.001</td>
<td align="center">0.002</td>
</tr>
<tr class="odd">
<td align="center">All</td>
<td align="center">pearson</td>
<td align="center">0.13</td>
<td align="center">0.003</td>
<td align="center">0.004</td>
</tr>
<tr class="even">
<td align="center">All</td>
<td align="center">pearson</td>
<td align="center">-0.05539</td>
<td align="center">0.926</td>
<td align="center">0.926</td>
</tr>
<tr class="odd">
<td align="center">All</td>
<td align="center">pearson</td>
<td align="center">0.06758</td>
<td align="center">0.05</td>
<td align="center">0.05714</td>
</tr>
<tr class="even">
<td align="center">All</td>
<td align="center">pearson</td>
<td align="center">0.4085</td>
<td align="center">0.001</td>
<td align="center">0.002</td>
</tr>
<tr class="odd">
<td align="center">All</td>
<td align="center">pearson</td>
<td align="center">0.2643</td>
<td align="center">0.001</td>
<td align="center">0.002</td>
</tr>
<tr class="even">
<td align="center">All</td>
<td align="center">pearson</td>
<td align="center">0.1321</td>
<td align="center">0.002</td>
<td align="center">0.0032</td>
</tr>
</tbody>
</table>
<table style="width:21%;">
<colgroup>
<col width="20%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">**</td>
</tr>
<tr class="even">
<td align="center">**</td>
</tr>
<tr class="odd">
<td align="center">**</td>
</tr>
<tr class="even">
<td align="center">**</td>
</tr>
<tr class="odd">
<td align="center">**</td>
</tr>
<tr class="even">
<td align="center">**</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="explainable-class.html#cb222-1" aria-hidden="true" tabindex="-1"></a><span class="co"># mantel test for different groups</span></span>
<span id="cb222-2"><a href="explainable-class.html#cb222-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_mantel</span>(<span class="at">by_group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">use_measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb222-3"><a href="explainable-class.html#cb222-3" aria-hidden="true" tabindex="-1"></a><span class="co"># partial mantel test</span></span>
<span id="cb222-4"><a href="explainable-class.html#cb222-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_mantel</span>(<span class="at">partial_mantel =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>For the combination of mantel test and correlation heatmap,
please see another example (<a href="https://chiliubio.github.io/microeco_tutorial/other-examples-1.html#mantel-test-correlation-heatmap" class="uri">https://chiliubio.github.io/microeco_tutorial/other-examples-1.html#mantel-test-correlation-heatmap</a>).</p>
<p>The correlations between environmental variables and taxa are important in analyzing and inferring the factors affecting community structure.
Let’s first perform a correlation heatmap using relative abundance data at Genus level with the <code>cal_cor</code> function.
The parameter <code>p_adjust_type</code> can control the p value adjustment type.
The default <code>p_adjust_type = "Env"</code> means p adjustment is performed for each environmental variable separately.
If the user needs to adjust p values for all the results together, please use <code>p_adjust_type = "Type"</code>.</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="explainable-class.html#cb223-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span></code></pre></div>
<pre><code>## Env data is stored in object$data_env ...</code></pre>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="explainable-class.html#cb225-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">use_data =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;fdr&quot;</span>, <span class="at">p_adjust_type =</span> <span class="st">&quot;Env&quot;</span>)</span></code></pre></div>
<pre><code>## The correlation result is stored in object$res_cor ...</code></pre>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="explainable-class.html#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_cor</span></span></code></pre></div>
<p>Then, we can plot the correlation results using plot_cor function.</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="explainable-class.html#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="co"># default ggplot2 method with clustering</span></span>
<span id="cb228-2"><a href="explainable-class.html#cb228-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p>There are too many genera.
We can use the filter_feature parameter in plot_cor to filter some taxa that do not have any significance &lt; 0.001.</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="explainable-class.html#cb229-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter genera that donot have at least one ***</span></span>
<span id="cb229-2"><a href="explainable-class.html#cb229-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">filter_feature =</span> <span class="fu">c</span>(<span class="st">&quot;&quot;</span>, <span class="st">&quot;*&quot;</span>, <span class="st">&quot;**&quot;</span>))</span></code></pre></div>
<p>Sometimes, if the user wants to do the correlation analysis between the environmental factors and some important taxa detected in the biomarker analysis,
please use <strong>other_taxa</strong> parameter in cal_cor function.</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="explainable-class.html#cb230-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first create trans_diff object as a demonstration</span></span>
<span id="cb230-2"><a href="explainable-class.html#cb230-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;rf&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">rf_taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb230-3"><a href="explainable-class.html#cb230-3" aria-hidden="true" tabindex="-1"></a><span class="co"># then create trans_env object</span></span>
<span id="cb230-4"><a href="explainable-class.html#cb230-4" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb230-5"><a href="explainable-class.html#cb230-5" aria-hidden="true" tabindex="-1"></a><span class="co"># use other_taxa to select taxa you need</span></span>
<span id="cb230-6"><a href="explainable-class.html#cb230-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">use_data =</span> <span class="st">&quot;other&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;fdr&quot;</span>, <span class="at">other_taxa =</span> t2<span class="sc">$</span>res_diff<span class="sc">$</span>Taxa[<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>])</span>
<span id="cb230-7"><a href="explainable-class.html#cb230-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/plot_corr_ggplot.png" width="700px" style="display: block; margin: auto;" /></p>
<p>The pheatmap method is also available.
Note that, besides the <strong>color_vector parameter</strong>,
<strong>color_palette</strong> can also be used to control color palette with customized colors.</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="explainable-class.html#cb231-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clustering heatmap; require pheatmap package</span></span>
<span id="cb231-2"><a href="explainable-class.html#cb231-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Let&#39;s take another color pallete</span></span>
<span id="cb231-3"><a href="explainable-class.html#cb231-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">pheatmap =</span> <span class="cn">TRUE</span>, <span class="at">color_palette =</span> <span class="fu">rev</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">9</span>, <span class="at">name =</span> <span class="st">&quot;RdYlBu&quot;</span>)))</span></code></pre></div>
<p><img src="Images/plot_corr_pheatmap.png" width="700px" style="display: block; margin: auto;" /></p>
<p>Sometimes, if it is necessary to study the correlations between environmental variables and taxa for different groups,
<strong>by_group parameter</strong> can be used for this goal.</p>
<div class="sourceCode" id="cb232"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb232-1"><a href="explainable-class.html#cb232-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate correlations for different groups using parameter by_group</span></span>
<span id="cb232-2"><a href="explainable-class.html#cb232-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">by_group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">use_data =</span> <span class="st">&quot;other&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;fdr&quot;</span>, <span class="at">other_taxa =</span> t2<span class="sc">$</span>res_diff<span class="sc">$</span>Taxa[<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>])</span>
<span id="cb232-3"><a href="explainable-class.html#cb232-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_cor</span></span>
<span id="cb232-4"><a href="explainable-class.html#cb232-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/plot_corr_ggplot_groups.png" width="700px" style="display: block; margin: auto;" /></p>
<p>If the user is concerned with the relationship between environmental factors and alpha diversity,
please use <strong>add_abund_table parameter</strong> in the cal_cor function.</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="explainable-class.html#cb233-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb233-2"><a href="explainable-class.html#cb233-2" aria-hidden="true" tabindex="-1"></a><span class="co"># use add_abund_table parameter to add the extra data table</span></span>
<span id="cb233-3"><a href="explainable-class.html#cb233-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">add_abund_table =</span> dataset<span class="sc">$</span>alpha_diversity)</span>
<span id="cb233-4"><a href="explainable-class.html#cb233-4" aria-hidden="true" tabindex="-1"></a><span class="co"># try to use ggplot2 with clustering plot</span></span>
<span id="cb233-5"><a href="explainable-class.html#cb233-5" aria-hidden="true" tabindex="-1"></a><span class="co"># require ggtree and aplot packages to be installed (https://chiliubio.github.io/microeco_tutorial/intro.html#dependence)</span></span>
<span id="cb233-6"><a href="explainable-class.html#cb233-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">cluster_ggplot =</span> <span class="st">&quot;row&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_corr_alpha_diversity.png" width="600px" style="display: block; margin: auto;" /></p>
<p>The function plot_scatterfit() in trans_env class is designed for the scatter plot, adding the fitted line and statistics of correlation or regression.</p>
<div class="sourceCode" id="cb234"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb234-1"><a href="explainable-class.html#cb234-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use pH and bray-curtis distance</span></span>
<span id="cb234-2"><a href="explainable-class.html#cb234-2" aria-hidden="true" tabindex="-1"></a><span class="co"># add correlation statistics</span></span>
<span id="cb234-3"><a href="explainable-class.html#cb234-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_scatterfit</span>(</span>
<span id="cb234-4"><a href="explainable-class.html#cb234-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">&quot;pH&quot;</span>, </span>
<span id="cb234-5"><a href="explainable-class.html#cb234-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> dataset<span class="sc">$</span>beta_diversity<span class="sc">$</span>bray[<span class="fu">rownames</span>(t1<span class="sc">$</span>data_env), <span class="fu">rownames</span>(t1<span class="sc">$</span>data_env)], </span>
<span id="cb234-6"><a href="explainable-class.html#cb234-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">&quot;cor&quot;</span>,</span>
<span id="cb234-7"><a href="explainable-class.html#cb234-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">point_alpha =</span> <span class="fl">0.1</span>, <span class="at">label.x.npc =</span> <span class="st">&quot;center&quot;</span>, <span class="at">label.y.npc =</span> <span class="st">&quot;bottom&quot;</span>, </span>
<span id="cb234-8"><a href="explainable-class.html#cb234-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">x_axis_title =</span> <span class="st">&quot;Euclidean distance of pH&quot;</span>, </span>
<span id="cb234-9"><a href="explainable-class.html#cb234-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">y_axis_title =</span> <span class="st">&quot;Bray-Curtis distance&quot;</span></span>
<span id="cb234-10"><a href="explainable-class.html#cb234-10" aria-hidden="true" tabindex="-1"></a>    )</span></code></pre></div>
<p><img src="Images/plot_scatterfit_cor.png" width="550px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="explainable-class.html#cb235-1" aria-hidden="true" tabindex="-1"></a><span class="co"># regression with type = &quot;lm&quot;, use group parameter for different groups</span></span>
<span id="cb235-2"><a href="explainable-class.html#cb235-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_scatterfit</span>(</span>
<span id="cb235-3"><a href="explainable-class.html#cb235-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> dataset<span class="sc">$</span>beta_diversity<span class="sc">$</span>bray[<span class="fu">rownames</span>(t1<span class="sc">$</span>data_env), <span class="fu">rownames</span>(t1<span class="sc">$</span>data_env)],</span>
<span id="cb235-4"><a href="explainable-class.html#cb235-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">&quot;pH&quot;</span>,</span>
<span id="cb235-5"><a href="explainable-class.html#cb235-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">&quot;lm&quot;</span>, </span>
<span id="cb235-6"><a href="explainable-class.html#cb235-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, </span>
<span id="cb235-7"><a href="explainable-class.html#cb235-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;CW&quot;</span>, <span class="st">&quot;TW&quot;</span>, <span class="st">&quot;IW&quot;</span>),</span>
<span id="cb235-8"><a href="explainable-class.html#cb235-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">point_size =</span> <span class="dv">3</span>, <span class="at">point_alpha =</span> <span class="fl">0.3</span>, <span class="at">line_se =</span> <span class="cn">FALSE</span>, <span class="at">line_size =</span> <span class="fl">1.5</span>, <span class="at">shape_values =</span> <span class="fu">c</span>(<span class="dv">16</span>, <span class="dv">17</span>, <span class="dv">7</span>),</span>
<span id="cb235-9"><a href="explainable-class.html#cb235-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">y_axis_title =</span> <span class="st">&quot;Euclidean distance of pH&quot;</span>, <span class="at">x_axis_title =</span> <span class="st">&quot;Bray-Curtis distance&quot;</span></span>
<span id="cb235-10"><a href="explainable-class.html#cb235-10" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="Images/plot_scatterfit_lmgroup.png" width="550px" style="display: block; margin: auto;" /></p>
<p>Other examples.</p>
<div class="sourceCode" id="cb236"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb236-1"><a href="explainable-class.html#cb236-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> new_test, <span class="at">env_cols =</span> <span class="dv">8</span><span class="sc">:</span><span class="dv">15</span>)</span>
<span id="cb236-2"><a href="explainable-class.html#cb236-2" aria-hidden="true" tabindex="-1"></a><span class="co"># with forward selection in RDA</span></span>
<span id="cb236-3"><a href="explainable-class.html#cb236-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">method =</span> <span class="st">&quot;dbRDA&quot;</span>, <span class="at">feature_sel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb236-4"><a href="explainable-class.html#cb236-4" aria-hidden="true" tabindex="-1"></a><span class="co"># CCA, canonical correspondence analysis</span></span>
<span id="cb236-5"><a href="explainable-class.html#cb236-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">method =</span> <span class="st">&quot;CCA&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb236-6"><a href="explainable-class.html#cb236-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">trans_ordination</span>(<span class="at">adjust_arrow_length =</span> <span class="cn">TRUE</span>)</span>
<span id="cb236-7"><a href="explainable-class.html#cb236-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;ellipse&quot;</span>))</span>
<span id="cb236-8"><a href="explainable-class.html#cb236-8" aria-hidden="true" tabindex="-1"></a><span class="co"># correlation analysis without p adjustment</span></span>
<span id="cb236-9"><a href="explainable-class.html#cb236-9" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">use_data =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;none&quot;</span>, <span class="at">use_taxa_num =</span> <span class="dv">30</span>)</span>
<span id="cb236-10"><a href="explainable-class.html#cb236-10" aria-hidden="true" tabindex="-1"></a><span class="co"># correlation heatmap with clustering based on the ggplot2 and aplot packages</span></span>
<span id="cb236-11"><a href="explainable-class.html#cb236-11" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">cluster_ggplot =</span> <span class="st">&quot;both&quot;</span>)</span>
<span id="cb236-12"><a href="explainable-class.html#cb236-12" aria-hidden="true" tabindex="-1"></a>g1</span>
<span id="cb236-13"><a href="explainable-class.html#cb236-13" aria-hidden="true" tabindex="-1"></a><span class="co"># clustering heatmap with ggplot2 depends on aplot package</span></span>
<span id="cb236-14"><a href="explainable-class.html#cb236-14" aria-hidden="true" tabindex="-1"></a><span class="co"># to change the detail in the plot, please manipulate each element of g1</span></span>
<span id="cb236-15"><a href="explainable-class.html#cb236-15" aria-hidden="true" tabindex="-1"></a>g1[[<span class="dv">1</span>]]</span>
<span id="cb236-16"><a href="explainable-class.html#cb236-16" aria-hidden="true" tabindex="-1"></a><span class="co"># standardize x axis text format</span></span>
<span id="cb236-17"><a href="explainable-class.html#cb236-17" aria-hidden="true" tabindex="-1"></a>g1[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> g1[[<span class="dv">1</span>]] <span class="sc">+</span> <span class="fu">scale_x_discrete</span>(<span class="at">labels =</span> <span class="fu">c</span>(<span class="at">NH4 =</span> <span class="fu">expression</span>(NH[<span class="dv">4</span>]<span class="sc">^</span><span class="st">&#39;+&#39;</span><span class="sc">-</span>N), <span class="at">NO3 =</span> <span class="fu">expression</span>(NO[<span class="dv">3</span>]<span class="sc">^</span><span class="st">&#39;-&#39;</span><span class="sc">-</span>N)))</span>
<span id="cb236-18"><a href="explainable-class.html#cb236-18" aria-hidden="true" tabindex="-1"></a>g1[[<span class="dv">1</span>]]</span>
<span id="cb236-19"><a href="explainable-class.html#cb236-19" aria-hidden="true" tabindex="-1"></a>g1</span>
<span id="cb236-20"><a href="explainable-class.html#cb236-20" aria-hidden="true" tabindex="-1"></a>ggplot2<span class="sc">::</span><span class="fu">ggsave</span>(<span class="st">&quot;test.pdf&quot;</span>, g1, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">6</span>)</span>
<span id="cb236-21"><a href="explainable-class.html#cb236-21" aria-hidden="true" tabindex="-1"></a><span class="co"># For regression, lm_equation = FALSE can be applied to not display the equation.</span></span>
<span id="cb236-22"><a href="explainable-class.html#cb236-22" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_scatterfit</span>(<span class="at">x =</span> <span class="dv">1</span>, <span class="at">y =</span> <span class="dv">2</span>, <span class="at">type =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">lm_equation =</span> <span class="cn">TRUE</span>)</span>
<span id="cb236-23"><a href="explainable-class.html#cb236-23" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_scatterfit</span>(<span class="at">x =</span> <span class="dv">1</span>, <span class="at">y =</span> <span class="dv">2</span>, <span class="at">type =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">lm_equation =</span> <span class="cn">FALSE</span>)</span>
<span id="cb236-24"><a href="explainable-class.html#cb236-24" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_scatterfit</span>(<span class="at">x =</span> <span class="dv">1</span>, <span class="at">y =</span> <span class="dv">2</span>, <span class="at">type =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">point_alpha =</span> .<span class="dv">3</span>, <span class="at">line_se =</span> <span class="cn">FALSE</span>)</span>
<span id="cb236-25"><a href="explainable-class.html#cb236-25" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_scatterfit</span>(<span class="at">x =</span> <span class="dv">1</span>, <span class="at">y =</span> <span class="dv">2</span>, <span class="at">type =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">line_se_color =</span> <span class="st">&quot;grey90&quot;</span>, <span class="at">label_sep =</span> <span class="st">&quot;,&quot;</span>, <span class="at">label.x.npc =</span> <span class="st">&quot;center&quot;</span>, <span class="at">label.y.npc =</span> <span class="st">&quot;bottom&quot;</span>)</span>
<span id="cb236-26"><a href="explainable-class.html#cb236-26" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_scatterfit</span>(<span class="at">x =</span> <span class="dv">1</span>, <span class="at">y =</span> <span class="dv">2</span>, <span class="at">line_se =</span> <span class="cn">FALSE</span>, <span class="at">pvalue_trim =</span> <span class="dv">3</span>, <span class="at">cor_coef_trim =</span> <span class="dv">3</span>)</span>
<span id="cb236-27"><a href="explainable-class.html#cb236-27" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_scatterfit</span>(<span class="at">x =</span> <span class="st">&quot;pH&quot;</span>, <span class="at">y =</span> <span class="st">&quot;TOC&quot;</span>, <span class="at">type =</span> <span class="st">&quot;lm&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">line_se =</span> <span class="cn">FALSE</span>, <span class="at">label.x.npc =</span> <span class="st">&quot;center&quot;</span>,</span>
<span id="cb236-28"><a href="explainable-class.html#cb236-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">shape_values =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">x_axis_title =</span> <span class="st">&quot;pH&quot;</span>, <span class="at">y_axis_title =</span> <span class="st">&quot;TOC&quot;</span>)</span></code></pre></div>
</div>
<div id="key-points-9" class="section level3 hasAnchor" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> Key points<a href="explainable-class.html#key-points-9" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>complete_na parameter in trans_env$new: used to fill the NA (missing value) of the environmental data based on the mice package.</li>
<li>env_cols parameter in trans_env$new: select the variables from sample_table of your microtable object.</li>
<li>add_abund_table parameter in cal_cor: other customized data can be also provided for the correlation analysis.</li>
<li>use_cor parameter in plot_scatterfit: both the correlation and regression are available in this function.</li>
<li>cal_mantel(): partial_mantel = TRUE can be used for partial mantel test.</li>
<li>plot_ordination(): use plot_type parameter to select point types and env_nudge_x and taxa_nudge_x (also _y) to adjust the text positions.</li>
</ul>
</div>
</div>
<div id="trans_func-class" class="section level2 hasAnchor" number="7.2">
<h2><span class="header-section-number">7.2</span> trans_func class<a href="explainable-class.html#trans_func-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　Ecological researchers are usually interested in the the funtional profiles of microbial communities,
because functional or metabolic data is powerful to explain the structure and dynamics of microbial communities.
As metagenomic sequencing is complicated and expensive, using amplicon sequencing data to predict functional profiles is an alternative choice.
Several software are often used for this goal, such as PICRUSt <span class="citation">(<a href="#ref-Langille_Predictive_2013" role="doc-biblioref">Langille et al. 2013</a>)</span>, Tax4Fun <span class="citation">(<a href="#ref-Aßhauer_Tax4Fun_2015" role="doc-biblioref">Aßhauer et al. 2015</a>)</span> and FAPROTAX <span class="citation">(<a href="#ref-Louca_High_2016" role="doc-biblioref">Stilianos Louca et al. 2016</a>; <a href="#ref-Louca_Decoupling_2016" role="doc-biblioref">S. Louca, Parfrey, and Doebeli 2016</a>)</span>.
These tools are great to be used for the prediction of functional profiles based on the prokaryotic communities from sequencing results.
In addition, it is also important to obtain the traits or functions for each taxa, not just the whole profile of communities.
FAPROTAX database is a collection of the traits and functions of prokaryotes based on the known research results published in books and literatures.
We match the taxonomic information of prokaryotes against this database to predict the traits of prokaryotes on biogeochemical roles.
The NJC19 database <span class="citation">(<a href="#ref-Lim_Large_2020" role="doc-biblioref">Lim et al. 2020</a>)</span> is also available for animal-associated prokaryotic data, such as human gut microbiota.
We also implement the FUNGuild <span class="citation">(<a href="#ref-Nguyen_FUNGuild_2016" role="doc-biblioref">Nguyen et al. 2016</a>)</span> and FungalTraits <span class="citation">(<a href="#ref-Polme_FungalTraits_2020" role="doc-biblioref">Põlme et al. 2020</a>)</span> databases to predict the fungal traits.
The idea identifying prokaryotic traits and functional redundancy was initially inspired by our another study <span class="citation">(<a href="#ref-Liu_Microbial_2022" role="doc-biblioref">Liu et al. 2022</a>)</span>.</p>
<div id="example-8" class="section level3 hasAnchor" number="7.2.1">
<h3><span class="header-section-number">7.2.1</span> Example<a href="explainable-class.html#example-8" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We first identify/predict traits of taxa with the prokaryotic example data.</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="explainable-class.html#cb237-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create object of trans_func</span></span>
<span id="cb237-2"><a href="explainable-class.html#cb237-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(dataset)</span>
<span id="cb237-3"><a href="explainable-class.html#cb237-3" aria-hidden="true" tabindex="-1"></a><span class="co"># mapping the taxonomy to the database</span></span>
<span id="cb237-4"><a href="explainable-class.html#cb237-4" aria-hidden="true" tabindex="-1"></a><span class="co"># this can recognize prokaryotes or fungi automatically if the names of taxonomic levels are standard.</span></span>
<span id="cb237-5"><a href="explainable-class.html#cb237-5" aria-hidden="true" tabindex="-1"></a><span class="co"># for fungi example, see https://chiliubio.github.io/microeco_tutorial/other-dataset.html#fungi-data</span></span>
<span id="cb237-6"><a href="explainable-class.html#cb237-6" aria-hidden="true" tabindex="-1"></a><span class="co"># default database for prokaryotes is FAPROTAX database</span></span>
<span id="cb237-7"><a href="explainable-class.html#cb237-7" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func</span>(<span class="at">prok_database =</span> <span class="st">&quot;FAPROTAX&quot;</span>)</span></code></pre></div>
<pre><code>## FAPROTAX v1.2.6. Please also cite the original FAPROTAX paper: Louca et al. (2016).</code></pre>
<pre><code>## Decoupling function and taxonomy in the global ocean microbiome. Science, 353(6305), 1272.</code></pre>
<pre><code>## The functional binary table is stored in object$res_spe_func ...</code></pre>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="explainable-class.html#cb241-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return t2$res_spe_func, 1 represent trait exists, 0 represent no or cannot confirmed.</span></span></code></pre></div>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="explainable-class.html#cb242-1" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>res_spe_func[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code></pre></div>
<table style="width:85%;">
<colgroup>
<col width="20%" />
<col width="22%" />
<col width="41%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">methanotrophy</th>
<th align="center">acetoclastic_methanogenesis</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>OTU_4272</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_236</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_399</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_1556</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_32</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>The percentages of the OTUs having the same trait can reflect the functional redundancy of this function in the community.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="explainable-class.html#cb243-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the percentages for communities</span></span>
<span id="cb243-2"><a href="explainable-class.html#cb243-2" aria-hidden="true" tabindex="-1"></a><span class="co"># here do not consider the abundance</span></span>
<span id="cb243-3"><a href="explainable-class.html#cb243-3" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">abundance_weighted =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>## The result table is stored in object$res_spe_func_perc ...</code></pre>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="explainable-class.html#cb245-1" aria-hidden="true" tabindex="-1"></a><span class="co"># t2$res_spe_func_perc[1:5, 1:2]</span></span></code></pre></div>
<table style="width:76%;">
<colgroup>
<col width="12%" />
<col width="22%" />
<col width="41%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">methanotrophy</th>
<th align="center">acetoclastic_methanogenesis</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">0.39</td>
<td align="center">0.04</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">0.27</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">0.48</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">0.48</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">0.56</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>Then we also take an example to show the percentages of the OTUs for each trait in network modules.</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="explainable-class.html#cb246-1" aria-hidden="true" tabindex="-1"></a><span class="co"># construct a network for the example</span></span>
<span id="cb246-2"><a href="explainable-class.html#cb246-2" aria-hidden="true" tabindex="-1"></a>network <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="st">&quot;base&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb246-3"><a href="explainable-class.html#cb246-3" aria-hidden="true" tabindex="-1"></a>network<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">p_thres =</span> <span class="fl">0.01</span>, <span class="at">COR_cut =</span> <span class="fl">0.7</span>)</span>
<span id="cb246-4"><a href="explainable-class.html#cb246-4" aria-hidden="true" tabindex="-1"></a>network<span class="sc">$</span><span class="fu">cal_module</span>()</span>
<span id="cb246-5"><a href="explainable-class.html#cb246-5" aria-hidden="true" tabindex="-1"></a><span class="co"># convert module info to microtable object</span></span>
<span id="cb246-6"><a href="explainable-class.html#cb246-6" aria-hidden="true" tabindex="-1"></a>meco_module <span class="ot">&lt;-</span> network<span class="sc">$</span><span class="fu">trans_comm</span>(<span class="at">use_col =</span> <span class="st">&quot;module&quot;</span>)</span>
<span id="cb246-7"><a href="explainable-class.html#cb246-7" aria-hidden="true" tabindex="-1"></a>meco_module_func <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(meco_module)</span>
<span id="cb246-8"><a href="explainable-class.html#cb246-8" aria-hidden="true" tabindex="-1"></a>meco_module_func<span class="sc">$</span><span class="fu">cal_spe_func</span>(<span class="at">prok_database =</span> <span class="st">&quot;FAPROTAX&quot;</span>)</span>
<span id="cb246-9"><a href="explainable-class.html#cb246-9" aria-hidden="true" tabindex="-1"></a>meco_module_func<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">abundance_weighted =</span> <span class="cn">FALSE</span>)</span>
<span id="cb246-10"><a href="explainable-class.html#cb246-10" aria-hidden="true" tabindex="-1"></a>meco_module_func<span class="sc">$</span><span class="fu">plot_spe_func_perc</span>(<span class="at">order_x =</span> <span class="fu">paste0</span>(<span class="st">&quot;M&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>))</span></code></pre></div>
<p><img src="Images/plot_func_perc_module.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="explainable-class.html#cb247-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to change the group list, reset the list t2$func_group_list</span></span>
<span id="cb247-2"><a href="explainable-class.html#cb247-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>func_group_list</span>
<span id="cb247-3"><a href="explainable-class.html#cb247-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use show_prok_func to see the detailed information of prokaryotic traits</span></span>
<span id="cb247-4"><a href="explainable-class.html#cb247-4" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">show_prok_func</span>(<span class="st">&quot;methanotrophy&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="explainable-class.html#cb248-1" aria-hidden="true" tabindex="-1"></a><span class="co"># then we try to correlate the res_spe_func_perc of communities to environmental variables</span></span>
<span id="cb248-2"><a href="explainable-class.html#cb248-2" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb248-3"><a href="explainable-class.html#cb248-3" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">add_abund_table =</span> t2<span class="sc">$</span>res_spe_func_perc, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb248-4"><a href="explainable-class.html#cb248-4" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">pheatmap =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/plot_func_perc_corr.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Tax4Fun <span class="citation">(<a href="#ref-Aßhauer_Tax4Fun_2015" role="doc-biblioref">Aßhauer et al. 2015</a>)</span> requires a strict input file format associated with the taxonomic information.
To analyze the trimmed or changed OTU data in R with Tax4Fun, we provide a link to the Tax4Fun functional prediction.
Please check out the dependence part <a href="https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun" class="uri">https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun</a> for installing Tax4Fun package and download SILVA123 ref data.</p>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="explainable-class.html#cb249-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(dataset)</span>
<span id="cb249-2"><a href="explainable-class.html#cb249-2" aria-hidden="true" tabindex="-1"></a><span class="co"># https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun for the installation description</span></span>
<span id="cb249-3"><a href="explainable-class.html#cb249-3" aria-hidden="true" tabindex="-1"></a><span class="co"># and provide the file path of SILVA123</span></span>
<span id="cb249-4"><a href="explainable-class.html#cb249-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_tax4fun</span>(<span class="at">folderReferenceData =</span> <span class="st">&quot;./SILVA123&quot;</span>)</span>
<span id="cb249-5"><a href="explainable-class.html#cb249-5" aria-hidden="true" tabindex="-1"></a><span class="co"># return two files: t1$tax4fun_KO: KO file; t1$tax4fun_path: pathway file.</span></span>
<span id="cb249-6"><a href="explainable-class.html#cb249-6" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$tax4fun_KO$Tax4FunProfile[1:5, 1:2]</span></span></code></pre></div>
<table>
<colgroup>
<col width="12%" />
<col width="43%" />
<col width="43%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">K00001; alcohol dehydrogenase
[EC:1.1.1.1]</th>
<th align="center">K00002; alcohol dehydrogenase
(NADP+) [EC:1.1.1.2]</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">0.0004823</td>
<td align="center">5.942e-06</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">0.0005266</td>
<td align="center">4.017e-06</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">0.0005054</td>
<td align="center">6.168e-06</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">0.0005109</td>
<td align="center">5.888e-06</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">0.0005083</td>
<td align="center">5.547e-06</td>
</tr>
</tbody>
</table>
<p>We further analyze the abundance of predicted metabolic pathways.</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="explainable-class.html#cb250-1" aria-hidden="true" tabindex="-1"></a><span class="co"># must transpose to taxa row, sample column</span></span>
<span id="cb250-2"><a href="explainable-class.html#cb250-2" aria-hidden="true" tabindex="-1"></a>pathway_file <span class="ot">&lt;-</span> t1<span class="sc">$</span>tax4fun_path<span class="sc">$</span>Tax4FunProfile <span class="sc">%&gt;%</span> t <span class="sc">%&gt;%</span> as.data.frame</span>
<span id="cb250-3"><a href="explainable-class.html#cb250-3" aria-hidden="true" tabindex="-1"></a><span class="co"># filter rownames, only keep ko+number</span></span>
<span id="cb250-4"><a href="explainable-class.html#cb250-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pathway_file) <span class="sc">%&lt;&gt;%</span> <span class="fu">gsub</span>(<span class="st">&quot;(^.*);</span><span class="sc">\\</span><span class="st">s.*&quot;</span>, <span class="st">&quot;</span><span class="sc">\\</span><span class="st">1&quot;</span>, .)</span>
<span id="cb250-5"><a href="explainable-class.html#cb250-5" aria-hidden="true" tabindex="-1"></a><span class="co"># load the pathway hierarchical metadata</span></span>
<span id="cb250-6"><a href="explainable-class.html#cb250-6" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Tax4Fun2_KEGG)</span>
<span id="cb250-7"><a href="explainable-class.html#cb250-7" aria-hidden="true" tabindex="-1"></a><span class="co"># further create a microtable object, familiar?</span></span>
<span id="cb250-8"><a href="explainable-class.html#cb250-8" aria-hidden="true" tabindex="-1"></a>func1 <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">otu_table =</span> pathway_file, <span class="at">tax_table =</span> Tax4Fun2_KEGG<span class="sc">$</span>ptw_desc, <span class="at">sample_table =</span> t1<span class="sc">$</span>sample_table)</span>
<span id="cb250-9"><a href="explainable-class.html#cb250-9" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(func1)</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 4 columns
## otu_table have 284 rows and 90 columns
## tax_table have 444 rows and 3 columns</code></pre>
<p>Now, we need to trim data and calculate abundance.</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="explainable-class.html#cb252-1" aria-hidden="true" tabindex="-1"></a>func1<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb252-2"><a href="explainable-class.html#cb252-2" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate abundance automatically at three levels: Level.1, Level.2, Level.3</span></span>
<span id="cb252-3"><a href="explainable-class.html#cb252-3" aria-hidden="true" tabindex="-1"></a>func1<span class="sc">$</span><span class="fu">cal_abund</span>()</span></code></pre></div>
<pre><code>## The result is stored in object$taxa_abund ...</code></pre>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="explainable-class.html#cb254-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(func1)</span></code></pre></div>
<pre><code>## microtable-class object:
## sample_table have 90 rows and 4 columns
## otu_table have 282 rows and 90 columns
## tax_table have 282 rows and 3 columns
## Taxa abundance: calculated for Level.1,Level.2,Level.3</code></pre>
<p>Then, we can plot the abundance.</p>
<div class="sourceCode" id="cb256"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb256-1"><a href="explainable-class.html#cb256-1" aria-hidden="true" tabindex="-1"></a><span class="co"># bar plot at Level.1</span></span>
<span id="cb256-2"><a href="explainable-class.html#cb256-2" aria-hidden="true" tabindex="-1"></a>func2 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(func1, <span class="at">taxrank =</span> <span class="st">&quot;Level.1&quot;</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb256-3"><a href="explainable-class.html#cb256-3" aria-hidden="true" tabindex="-1"></a>func2<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/plot_bar_tax4fun1.png" width="600px" style="display: block; margin: auto;" /></p>
<p>We can also do something else. For example, we can use lefse to test the differences of the abundances and find the important enriched pathways across groups.</p>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="explainable-class.html#cb257-1" aria-hidden="true" tabindex="-1"></a>func2 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> func1, <span class="at">method =</span> <span class="st">&quot;lefse&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">lefse_subgroup =</span> <span class="cn">NULL</span>)</span>
<span id="cb257-2"><a href="explainable-class.html#cb257-2" aria-hidden="true" tabindex="-1"></a>func2<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">threshold =</span> <span class="dv">3</span>, <span class="at">width =</span> <span class="fl">0.8</span>)</span></code></pre></div>
<p><img src="Images/plot_lefse_bar_tax4fun.png" width="600px" style="display: block; margin: auto;" /></p>
<p>Tax4Fun2 <span class="citation">(<a href="#ref-Wemheuer_Tax4Fun2_2020" role="doc-biblioref">Wemheuer et al. 2020</a>)</span> is another R package for the prediction of functional profiles of prokaryotic communities from 16S rRNA gene sequences.
It also provides two indexes for the evaluation of functional gene redundancies.
If the user want to use Tax4Fun2 method, the representative fasta file is necessary to be added in the microtable object.
Please check out <a href="https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun2" class="uri">https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun2</a> to see
how to read fasta file with <code>read.fasta</code> of seqinr package or <code>readDNAStringSet</code> of Biostrings package.
Please also see <a href="https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun2" class="uri">https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun2</a> for downloading ncbi-blast and Ref99NR/Ref100NR.
For windows system, ncbi-blast-2.5.0+ is recommended since other versions can not operate well.</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="explainable-class.html#cb258-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first delete the dataset created before</span></span>
<span id="cb258-2"><a href="explainable-class.html#cb258-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(dataset)</span>
<span id="cb258-3"><a href="explainable-class.html#cb258-3" aria-hidden="true" tabindex="-1"></a><span class="co"># load the example dataset from microeco package as there is the rep_fasta object in it</span></span>
<span id="cb258-4"><a href="explainable-class.html#cb258-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataset)</span>
<span id="cb258-5"><a href="explainable-class.html#cb258-5" aria-hidden="true" tabindex="-1"></a>dataset</span>
<span id="cb258-6"><a href="explainable-class.html#cb258-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb258-7"><a href="explainable-class.html#cb258-7" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(dataset)</span>
<span id="cb258-8"><a href="explainable-class.html#cb258-8" aria-hidden="true" tabindex="-1"></a><span class="co"># create a directory for result and log files</span></span>
<span id="cb258-9"><a href="explainable-class.html#cb258-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;test_prediction&quot;</span>)</span>
<span id="cb258-10"><a href="explainable-class.html#cb258-10" aria-hidden="true" tabindex="-1"></a><span class="co"># https://chiliubio.github.io/microeco_tutorial/intro.html#tax4fun2 for installation</span></span>
<span id="cb258-11"><a href="explainable-class.html#cb258-11" aria-hidden="true" tabindex="-1"></a><span class="co"># ignore blast_tool_path parameter if blast tools have been in path</span></span>
<span id="cb258-12"><a href="explainable-class.html#cb258-12" aria-hidden="true" tabindex="-1"></a><span class="co"># the function can search whether blast tool directory is in the path, if not, automatically use provided blast_tool_path parameter</span></span>
<span id="cb258-13"><a href="explainable-class.html#cb258-13" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_tax4fun2</span>(<span class="at">blast_tool_path =</span> <span class="st">&quot;ncbi-blast-2.5.0+/bin&quot;</span>, <span class="at">path_to_reference_data =</span> <span class="st">&quot;Tax4Fun2_ReferenceData_v2&quot;</span>,</span>
<span id="cb258-14"><a href="explainable-class.html#cb258-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">database_mode =</span> <span class="st">&quot;Ref99NR&quot;</span>, <span class="at">path_to_temp_folder =</span> <span class="st">&quot;test_prediction&quot;</span>)</span>
<span id="cb258-15"><a href="explainable-class.html#cb258-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb258-16"><a href="explainable-class.html#cb258-16" aria-hidden="true" tabindex="-1"></a><span class="co"># prepare feature table and metadata</span></span>
<span id="cb258-17"><a href="explainable-class.html#cb258-17" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Tax4Fun2_KEGG)</span>
<span id="cb258-18"><a href="explainable-class.html#cb258-18" aria-hidden="true" tabindex="-1"></a><span class="co"># create a microtable object for pathways</span></span>
<span id="cb258-19"><a href="explainable-class.html#cb258-19" aria-hidden="true" tabindex="-1"></a>func2 <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">otu_table =</span> t1<span class="sc">$</span>res_tax4fun2_pathway, <span class="at">tax_table =</span> Tax4Fun2_KEGG<span class="sc">$</span>ptw_desc, <span class="at">sample_table =</span> dataset<span class="sc">$</span>sample_table)</span>
<span id="cb258-20"><a href="explainable-class.html#cb258-20" aria-hidden="true" tabindex="-1"></a>func2<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb258-21"><a href="explainable-class.html#cb258-21" aria-hidden="true" tabindex="-1"></a>func2<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb258-22"><a href="explainable-class.html#cb258-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb258-23"><a href="explainable-class.html#cb258-23" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate functional redundancies</span></span>
<span id="cb258-24"><a href="explainable-class.html#cb258-24" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_tax4fun2_FRI</span>()</span></code></pre></div>
</div>
<div id="key-points-10" class="section level3 hasAnchor" number="7.2.2">
<h3><span class="header-section-number">7.2.2</span> Key points<a href="explainable-class.html#key-points-10" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>blast_tool_path parameter in cal_tax4fun2: if the blast tool has been in ‘environment variable’ of computer, it is ok to use blast_tool_path = NULL</li>
<li>blast version: tax4fun2 require NCBI blast tool. However, some errors often come from the latest versions (<a href="https://www.biostars.org/p/413294/" class="uri">https://www.biostars.org/p/413294/</a>). An easy solution is to use previous version (such as v2.5.0).</li>
</ul>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Aßhauer_Tax4Fun_2015" class="csl-entry">
Aßhauer, Kathrin P, Bernd Wemheuer, Rolf Daniel, and Peter Meinicke. 2015. <span>“Tax4Fun: Predicting Functional Profiles from Metagenomic 16S rRNA Data.”</span> Journal Article. <em>Bioinformatics</em> 31 (17): 2882–84.
</div>
<div id="ref-Langille_Predictive_2013" class="csl-entry">
Langille, M. G., J. Zaneveld, J. G. Caporaso, D. McDonald, D. Knights, J. A. Reyes, J. C. Clemente, et al. 2013. <span>“Predictive Functional Profiling of Microbial Communities Using 16S rRNA Marker Gene Sequences.”</span> Journal Article. <em>Nature Biotechnology</em> 31 (9): 814–21. <a href="https://doi.org/10.1038/nbt.2676">https://doi.org/10.1038/nbt.2676</a>.
</div>
<div id="ref-Lim_Large_2020" class="csl-entry">
Lim, Roktaek, Josephine Jill T. Cabatbat, Thomas L. P. Martin, Haneul Kim, Seunghyeon Kim, Jaeyun Sung, Cheol-Min Ghim, and Pan-Jun Kim. 2020. <span>“Large-Scale Metabolic Interaction Network of the Mouse and Human Gut Microbiota.”</span> Journal Article. <em>Scientific Data</em> 7 (1). <a href="https://doi.org/10.1038/s41597-020-0516-5">https://doi.org/10.1038/s41597-020-0516-5</a>.
</div>
<div id="ref-Liu_Microbial_2022" class="csl-entry">
Liu, Chi, Xiangzhen Li, Felipe R. P. Mansoldo, Jiaxing An, Yongping Kou, Xiao Zhang, Junming Wang, Jianxiong Zeng, Alane B. Vermelho, and Minjie Yao. 2022. <span>“Microbial Habitat Specificity Largely Affects Microbial Co-Occurrence Patterns and Functional Profiles in Wetland Soils.”</span> Journal Article. <em>Geoderma</em> 418: 115866. <a href="https://doi.org/10.1016/j.geoderma.2022.115866">https://doi.org/10.1016/j.geoderma.2022.115866</a>.
</div>
<div id="ref-Louca_Decoupling_2016" class="csl-entry">
Louca, S, L. W. Parfrey, and M Doebeli. 2016. <span>“Decoupling Function and Taxonomy in the Global Ocean Microbiome.”</span> Journal Article. <em>Science</em> 353 (6305): 1272.
</div>
<div id="ref-Louca_High_2016" class="csl-entry">
Louca, Stilianos, Saulo MS Jacques, Aliny PF Pires, Juliana S Leal, Diane S Srivastava, Laura Wegener Parfrey, Vinicius F Farjalla, and Michael Doebeli. 2016. <span>“High Taxonomic Variability Despite Stable Functional Structure Across Microbial Communities.”</span> Journal Article. <em>Nature Ecology &amp; Evolution</em> 1: 0015.
</div>
<div id="ref-Nguyen_FUNGuild_2016" class="csl-entry">
Nguyen, Nhu H., Zewei Song, Scott T. Bates, Sara Branco, Leho Tedersoo, Jon Menke, Jonathan S. Schilling, and Peter G. Kennedy. 2016. <span>“FUNGuild: An Open Annotation Tool for Parsing Fungal Community Datasets by Ecological Guild.”</span> Journal Article. <em>Fungal Ecology</em> 20 (1): 241–48.
</div>
<div id="ref-Polme_FungalTraits_2020" class="csl-entry">
Põlme, Sergei, Kessy Abarenkov, R. Henrik Nilsson, Björn D. Lindahl, Karina Engelbrecht Clemmensen, Havard Kauserud, Nhu Nguyen, et al. 2020. <span>“FungalTraits: A User-Friendly Traits Database of Fungi and Fungus-Like Stramenopiles.”</span> Journal Article. <em>Fungal Diversity</em> 105 (1): 1–16. <a href="https://doi.org/10.1007/s13225-020-00466-2">https://doi.org/10.1007/s13225-020-00466-2</a>.
</div>
<div id="ref-Wemheuer_Tax4Fun2_2020" class="csl-entry">
Wemheuer, Franziska, Jessica A. Taylor, Rolf Daniel, Emma Johnston, Peter Meinicke, Torsten Thomas, and Bernd Wemheuer. 2020. <span>“Tax4Fun2: Prediction of Habitat-Specific Functional Profiles and Functional Redundancy Based on 16S rRNA Gene Sequences.”</span> Journal Article. <em>Environmental Microbiome</em> 15 (1). <a href="https://doi.org/10.1186/s40793-020-00358-7">https://doi.org/10.1186/s40793-020-00358-7</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="model-based-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="file2meco-package.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/06-Explainable_class.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
