<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Extended class | microeco: an R package for data mining in microbial community ecology</title>
  <meta name="description" content="The tutorial for microeco package" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Extended class | microeco: an R package for data mining in microbial community ecology" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The tutorial for microeco package" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Extended class | microeco: an R package for data mining in microbial community ecology" />
  
  <meta name="twitter:description" content="The tutorial for microeco package" />
  

<meta name="author" content="Chi Liu, Felipe Mansoldo, Jarrod J. Scott, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2021-08-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="basic-class.html"/>
<link rel="next" href="other-data.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#features"><i class="fa fa-check"></i><b>2.1</b> Features</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="extended-class.html"><a href="extended-class.html"><i class="fa fa-check"></i><b>4</b> Extended class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="extended-class.html"><a href="extended-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a></li>
<li class="chapter" data-level="4.2" data-path="extended-class.html"><a href="extended-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a></li>
<li class="chapter" data-level="4.3" data-path="extended-class.html"><a href="extended-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>4.3</b> trans_alpha class</a></li>
<li class="chapter" data-level="4.4" data-path="extended-class.html"><a href="extended-class.html#trans_beta-class"><i class="fa fa-check"></i><b>4.4</b> trans_beta class</a></li>
<li class="chapter" data-level="4.5" data-path="extended-class.html"><a href="extended-class.html#trans_diff-class"><i class="fa fa-check"></i><b>4.5</b> trans_diff class</a></li>
<li class="chapter" data-level="4.6" data-path="extended-class.html"><a href="extended-class.html#trans_env-class"><i class="fa fa-check"></i><b>4.6</b> trans_env class</a></li>
<li class="chapter" data-level="4.7" data-path="extended-class.html"><a href="extended-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>4.7</b> trans_nullmodel class</a></li>
<li class="chapter" data-level="4.8" data-path="extended-class.html"><a href="extended-class.html#trans_network-class"><i class="fa fa-check"></i><b>4.8</b> trans_network class</a></li>
<li class="chapter" data-level="4.9" data-path="extended-class.html"><a href="extended-class.html#trans_func-class"><i class="fa fa-check"></i><b>4.9</b> trans_func class</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="other-data.html"><a href="other-data.html"><i class="fa fa-check"></i><b>5</b> Other data</a>
<ul>
<li class="chapter" data-level="5.1" data-path="other-data.html"><a href="other-data.html#fungi-data"><i class="fa fa-check"></i><b>5.1</b> Fungi data</a></li>
<li class="chapter" data-level="5.2" data-path="other-data.html"><a href="other-data.html#metagenomic-data"><i class="fa fa-check"></i><b>5.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="5.3" data-path="other-data.html"><a href="other-data.html#animal-data"><i class="fa fa-check"></i><b>5.3</b> Animal data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>6</b> Notes</a>
<ul>
<li class="chapter" data-level="6.1" data-path="notes.html"><a href="notes.html#clone"><i class="fa fa-check"></i><b>6.1</b> clone</a></li>
<li class="chapter" data-level="6.2" data-path="notes.html"><a href="notes.html#subset-of-samples"><i class="fa fa-check"></i><b>6.2</b> subset of samples</a></li>
<li class="chapter" data-level="6.3" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>6.3</b> change object</a></li>
<li class="chapter" data-level="6.4" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>6.4</b> group order</a></li>
<li class="chapter" data-level="6.5" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>6.5</b> add layers to plot</a></li>
<li class="chapter" data-level="6.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot-1"><i class="fa fa-check"></i><b>6.6</b> add layers to plot</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>7</b> mecodev package</a>
<ul>
<li class="chapter" data-level="7.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>7.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="7.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_convert"><i class="fa fa-check"></i><b>7.2</b> trans_convert</a></li>
<li class="chapter" data-level="7.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_netchord"><i class="fa fa-check"></i><b>7.3</b> trans_netchord</a></li>
<li class="chapter" data-level="7.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>7.4</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="other-examples.html"><a href="other-examples.html"><i class="fa fa-check"></i><b>8</b> Other examples</a>
<ul>
<li class="chapter" data-level="8.1" data-path="other-examples.html"><a href="other-examples.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>8.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="8.2" data-path="other-examples.html"><a href="other-examples.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>8.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="8.3" data-path="other-examples.html"><a href="other-examples.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>8.3</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="8.4" data-path="other-examples.html"><a href="other-examples.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>8.4</b> The use of phylogenetic tree</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">microeco: an R package for data mining in microbial community ecology</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="extended-class" class="section level1" number="4">
<h1><span class="header-section-number">Chapter 4</span> Extended class</h1>
<p>All the extended classes depend on the microtable class.
Generally, one class can work alone.
In some cases, one function of a class may interact with the object from another class to complete a method.</p>
<div id="trans_abund-class" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> trans_abund class</h2>
<p>　The trans_abund class is used to transform taxonomic abundance data for plotting the taxa abundance with the ggplot2 package.
As a premise, the taxa_abund list in the object of microtable class must be first calculated.
We first use this class for the bar plot.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="extended-class.html#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_abund object</span></span>
<span id="cb49-2"><a href="extended-class.html#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="co"># use 10 Phyla with the highest abundance in the dataset.</span></span>
<span id="cb49-3"><a href="extended-class.html#cb49-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>)</span>
<span id="cb49-4"><a href="extended-class.html#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="co"># t1 object now include the transformed abundance data t1$abund_data and other elements for the following plotting</span></span></code></pre></div>
<p>We remove the sample names in x axis and add the facet to show abundance according to groups.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="extended-class.html#cb50-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">others_color =</span> <span class="st">&quot;grey70&quot;</span>, <span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">xtext_keep =</span> <span class="cn">FALSE</span>, <span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span>
<span id="cb50-2"><a href="extended-class.html#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return a ggplot2 object</span></span></code></pre></div>
<p><img src="Images/plot_bar.png" width="750px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="extended-class.html#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The groupmean parameter can be used to obtain the group-mean barplot.</span></span>
<span id="cb51-2"><a href="extended-class.html#cb51-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb51-3"><a href="extended-class.html#cb51-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">others_color =</span> <span class="st">&quot;grey70&quot;</span>, <span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/plot_bar_mean.png" width="400px" style="display: block; margin: auto;" /></p>
<p>Then alluvial plot is implemented in the plot_bar function with the use_alluvium parameter.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="extended-class.html#cb52-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>)</span>
<span id="cb52-2"><a href="extended-class.html#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="co"># use_alluvium = TRUE make the alluvial plot, clustering = TRUE can be used to reorder the samples by clustering</span></span>
<span id="cb52-3"><a href="extended-class.html#cb52-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">use_alluvium =</span> <span class="cn">TRUE</span>, <span class="at">clustering =</span> <span class="cn">TRUE</span>, <span class="at">xtext_type_hor =</span> <span class="cn">FALSE</span>, <span class="at">xtext_size =</span> <span class="dv">6</span>)</span></code></pre></div>
<p><img src="Images/plot_bar_allu.png" width="1440" style="display: block; margin: auto;" /></p>
<p>The box plot is an excellent way to intuitionally show data distribution across groups.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="extended-class.html#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show 15 taxa at Class level</span></span>
<span id="cb53-2"><a href="extended-class.html#cb53-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Class&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">15</span>)</span>
<span id="cb53-3"><a href="extended-class.html#cb53-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_box</span>(<span class="at">group =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_box.png" width="700px" style="display: block; margin: auto;" /></p>
<p>Then we show the heatmap with the high abundant genera.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="extended-class.html#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show 40 taxa at Genus level</span></span>
<span id="cb54-2"><a href="extended-class.html#cb54-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">40</span>)</span>
<span id="cb54-3"><a href="extended-class.html#cb54-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_heatmap</span>(<span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">xtext_keep =</span> <span class="cn">FALSE</span>, <span class="at">withmargin =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/plot_heatmap.png" width="750px" style="display: block; margin: auto;" /></p>
<p>Then, we show the pie chart with the group mean values.</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="extended-class.html#cb55-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">6</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb55-2"><a href="extended-class.html#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="co"># all pie chart in one row</span></span>
<span id="cb55-3"><a href="extended-class.html#cb55-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_pie</span>(<span class="at">facet_nrow =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Images/plot_pie.png" width="600px" style="display: block; margin: auto;" /></p>
</div>
<div id="trans_venn-class" class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> trans_venn class</h2>
<p>The trans_venn class is used for venn analysis.
To analyze the unique and shared OTUs of groups,
we first merge samples according to the “Group” column of sample_table.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="extended-class.html#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># merge samples as one community for each group</span></span>
<span id="cb56-2"><a href="extended-class.html#cb56-2" aria-hidden="true" tabindex="-1"></a>dataset1 <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb56-3"><a href="extended-class.html#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co"># dataset1 is a new microtable object</span></span>
<span id="cb56-4"><a href="extended-class.html#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_venn object</span></span>
<span id="cb56-5"><a href="extended-class.html#cb56-5" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1, <span class="at">ratio =</span> <span class="st">&quot;seqratio&quot;</span>)</span>
<span id="cb56-6"><a href="extended-class.html#cb56-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_venn</span>()</span>
<span id="cb56-7"><a href="extended-class.html#cb56-7" aria-hidden="true" tabindex="-1"></a><span class="co"># The integer data is OTU number</span></span>
<span id="cb56-8"><a href="extended-class.html#cb56-8" aria-hidden="true" tabindex="-1"></a><span class="co"># The percentage data is the sequence number/total sequence number</span></span></code></pre></div>
<p><img src="Images/trans_venn_1.png" width="500px" style="display: block; margin: auto;" /></p>
<p>When the groups are too many to show with venn plot, we can use petal plot.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="extended-class.html#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use &quot;Type&quot; column in sample_table</span></span>
<span id="cb57-2"><a href="extended-class.html#cb57-2" aria-hidden="true" tabindex="-1"></a>dataset1 <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Type&quot;</span>)</span>
<span id="cb57-3"><a href="extended-class.html#cb57-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1)</span>
<span id="cb57-4"><a href="extended-class.html#cb57-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_venn</span>(<span class="at">petal_plot =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/trans_venn_2.png" width="500px" style="display: block; margin: auto;" /></p>
<p>Sometimes, we are interested in the unique and shared species. In another words,
the composition of the unique or shared species may account for the different and similar parts of ecological characteristics across groups<span class="citation">(<a href="#ref-Mendes_Deciphering_2011" role="doc-biblioref">Mendes et al. 2011</a>)</span>.
For this goal, we first transform the results of venn plot to the traditional species-sample table, that is, another object of microtable class.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="extended-class.html#cb58-1" aria-hidden="true" tabindex="-1"></a>dataset1 <span class="ot">&lt;-</span> dataset<span class="sc">$</span><span class="fu">merge_samples</span>(<span class="at">use_group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb58-2"><a href="extended-class.html#cb58-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_venn<span class="sc">$</span><span class="fu">new</span>(dataset1)</span></code></pre></div>
<pre><code>## The result is stored in object$venn_table and object$venn_count_abund ...</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="extended-class.html#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># transform venn results to the sample-species table, here do not consider abundance, only use presence/absence information.</span></span>
<span id="cb60-2"><a href="extended-class.html#cb60-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">trans_venn_com</span>(<span class="at">use_OTUs_frequency =</span> <span class="cn">TRUE</span>)</span>
<span id="cb60-3"><a href="extended-class.html#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t2 is a new microtable class, each part is considered as a sample</span></span>
<span id="cb60-4"><a href="extended-class.html#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(t2)</span></code></pre></div>
<pre><code>## [1] &quot;microtable&quot; &quot;R6&quot;</code></pre>
<p>We use bar plot to show the composition at the Genus level.</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="extended-class.html#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate taxa abundance, that is, the frequency</span></span>
<span id="cb62-2"><a href="extended-class.html#cb62-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb62-3"><a href="extended-class.html#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="co"># transform and plot</span></span>
<span id="cb62-4"><a href="extended-class.html#cb62-4" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> t2, <span class="at">taxrank =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>)</span>
<span id="cb62-5"><a href="extended-class.html#cb62-5" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">bar_type =</span> <span class="st">&quot;part&quot;</span>, <span class="at">legend_text_italic =</span> T, <span class="at">ylab_title =</span> <span class="st">&quot;Frequency (%)&quot;</span>, <span class="at">xtext_type_hor =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/trans_venn_bar.png" width="650px" style="display: block; margin: auto;" /></p>
<p>We also try to use pie chart to show the compositions at the Phylum level.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="extended-class.html#cb63-1" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> t2, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">8</span>)</span>
<span id="cb63-2"><a href="extended-class.html#cb63-2" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">plot_pie</span>(<span class="at">facet_nrow =</span> <span class="dv">3</span>, <span class="at">use_colors =</span> <span class="fu">rev</span>(<span class="fu">c</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">8</span>, <span class="st">&quot;Dark2&quot;</span>), <span class="st">&quot;grey50&quot;</span>)))</span></code></pre></div>
<p><img src="Images/trans_venn_pie.png" width="800px" style="display: block; margin: auto;" /></p>
</div>
<div id="trans_alpha-class" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> trans_alpha class</h2>
<p>　Alpha diversity can be transformed and plotted using trans_alpha class.
Creating the object of trans_alpha class can invoke the alpha_diversity data from the microtable object.
The trans_alpha object have two data frame: alpha_data and alpha_stat.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="extended-class.html#cb64-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_alpha<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb64-2"><a href="extended-class.html#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$alpha_stat</span></span>
<span id="cb64-3"><a href="extended-class.html#cb64-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>alpha_stat[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<pre><code>## The group statistics are stored in object$alpha_stat ...</code></pre>
<pre><code>## The transformed diversity data is stored in object$alpha_data ...</code></pre>
<table style="width:75%;">
<colgroup>
<col width="11%" />
<col width="15%" />
<col width="6%" />
<col width="12%" />
<col width="13%" />
<col width="15%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Group</th>
<th align="center">Measure</th>
<th align="center">N</th>
<th align="center">Mean</th>
<th align="center">SD</th>
<th align="center">SE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">CW</td>
<td align="center">Observed</td>
<td align="center">30</td>
<td align="center">1843</td>
<td align="center">220.6</td>
<td align="center">40.27</td>
</tr>
<tr class="even">
<td align="center">CW</td>
<td align="center">Chao1</td>
<td align="center">30</td>
<td align="center">2553</td>
<td align="center">338.1</td>
<td align="center">61.73</td>
</tr>
<tr class="odd">
<td align="center">CW</td>
<td align="center">ACE</td>
<td align="center">30</td>
<td align="center">2716</td>
<td align="center">367</td>
<td align="center">67.01</td>
</tr>
<tr class="even">
<td align="center">CW</td>
<td align="center">Shannon</td>
<td align="center">30</td>
<td align="center">6.308</td>
<td align="center">0.5355</td>
<td align="center">0.09777</td>
</tr>
<tr class="odd">
<td align="center">CW</td>
<td align="center">Simpson</td>
<td align="center">30</td>
<td align="center">0.9897</td>
<td align="center">0.01305</td>
<td align="center">0.002382</td>
</tr>
</tbody>
</table>
<p>Then, we test the differences among groups using the KW rank sum test and anova with multiple comparisons.</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="extended-class.html#cb67-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;KW&quot;</span>)</span>
<span id="cb67-2"><a href="extended-class.html#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_alpha_diff</span></span>
<span id="cb67-3"><a href="extended-class.html#cb67-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_alpha_diff[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<pre><code>## The result is stored in object$res_alpha_diff ...</code></pre>
<table style="width:94%;">
<colgroup>
<col width="23%" />
<col width="15%" />
<col width="19%" />
<col width="15%" />
<col width="20%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Groups</th>
<th align="center">Measure</th>
<th align="center">Test method</th>
<th align="center">p.value</th>
<th align="center">Significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">IW vs CW</td>
<td align="center">Observed</td>
<td align="center">KW</td>
<td align="center">0.0371</td>
<td align="center">*</td>
</tr>
<tr class="even">
<td align="center">IW vs TW</td>
<td align="center">Observed</td>
<td align="center">KW</td>
<td align="center">0.4553</td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="center">CW vs TW</td>
<td align="center">Observed</td>
<td align="center">KW</td>
<td align="center">0.3912</td>
<td align="center"></td>
</tr>
<tr class="even">
<td align="center">IW vs CW vs TW</td>
<td align="center">Observed</td>
<td align="center">KW</td>
<td align="center">0.155</td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="center">IW vs CW</td>
<td align="center">Chao1</td>
<td align="center">KW</td>
<td align="center">0.002689</td>
<td align="center">**</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="extended-class.html#cb69-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">method =</span> <span class="st">&quot;anova&quot;</span>)</span>
<span id="cb69-2"><a href="extended-class.html#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_alpha_diff</span></span>
<span id="cb69-3"><a href="extended-class.html#cb69-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_alpha_diff</span></code></pre></div>
<pre><code>## Registered S3 methods overwritten by &#39;klaR&#39;:
##   method      from 
##   predict.rda vegan
##   print.rda   vegan
##   plot.rda    vegan</code></pre>
<pre><code>## The result is stored in object$res_alpha_diff ...</code></pre>
<table>
<colgroup>
<col width="10%" />
<col width="12%" />
<col width="9%" />
<col width="6%" />
<col width="11%" />
<col width="11%" />
<col width="14%" />
<col width="10%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">Observed</th>
<th align="center">Chao1</th>
<th align="center">ACE</th>
<th align="center">Shannon</th>
<th align="center">Simpson</th>
<th align="center">InvSimpson</th>
<th align="center">Fisher</th>
<th align="center">Coverage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>IW</strong></td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">b</td>
</tr>
<tr class="even">
<td align="center"><strong>TW</strong></td>
<td align="center">a</td>
<td align="center">ab</td>
<td align="center">b</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
</tr>
<tr class="odd">
<td align="center"><strong>CW</strong></td>
<td align="center">a</td>
<td align="center">b</td>
<td align="center">b</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
<td align="center">a</td>
</tr>
</tbody>
</table>
<p>Now, let us plot the mean and se of alpha diversity for each group, and add the duncan.test (agricolae package) result.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="extended-class.html#cb72-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_alpha</span>(<span class="at">add_letter =</span> T, <span class="at">measure =</span> <span class="st">&quot;Chao1&quot;</span>, <span class="at">use_boxplot =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/plot_alpha_letter.png" width="600px" style="display: block; margin: auto;" /></p>
<p>We can also use the boxplot to show the paired comparisons directly.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="extended-class.html#cb73-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_alpha</span>(<span class="at">pair_compare =</span> <span class="cn">TRUE</span>, <span class="at">measure =</span> <span class="st">&quot;Chao1&quot;</span>, <span class="at">shape =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_alpha.png" width="600px" style="display: block; margin: auto;" /></p>
</div>
<div id="trans_beta-class" class="section level2" number="4.4">
<h2><span class="header-section-number">4.4</span> trans_beta class</h2>
<p>　The distance matrix of beta diversity can be transformed and plotted using trans_beta class.
The analysis referred to the beta diversity in this class mainly include ordination, group distance, clustering and manova.
We first show the ordination using PCoA.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="extended-class.html#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we first create an trans_beta object</span></span>
<span id="cb74-2"><a href="extended-class.html#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="co"># this operation invoke the distance matrix of bray in dataset$beta_diversity</span></span>
<span id="cb74-3"><a href="extended-class.html#cb74-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;bray&quot;</span>)</span></code></pre></div>
<pre><code>## Please also cite the original paper: An et al. (2019). Soil bacterial community structure in Chinese wetlands. Geoderma, 337, 290-299.</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="extended-class.html#cb76-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">ordination =</span> <span class="st">&quot;PCoA&quot;</span>)</span>
<span id="cb76-2"><a href="extended-class.html#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_ordination is the ordination result list</span></span>
<span id="cb76-3"><a href="extended-class.html#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(t1<span class="sc">$</span>res_ordination)</span>
<span id="cb76-4"><a href="extended-class.html#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the PCoA result</span></span>
<span id="cb76-5"><a href="extended-class.html#cb76-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_group_ellipse =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/plot_ordination.png" width="650px" style="display: block; margin: auto;" /></p>
<p>Then we plot and compare the group distances.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="extended-class.html#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate and plot sample distances within groups</span></span>
<span id="cb77-2"><a href="extended-class.html#cb77-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_group_distance</span>()</span>
<span id="cb77-3"><a href="extended-class.html#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_group_distance</span></span>
<span id="cb77-4"><a href="extended-class.html#cb77-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">distance_pair_stat =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/plot_group_distance_within.png" width="500px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="extended-class.html#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate and plot sample distances between groups</span></span>
<span id="cb78-2"><a href="extended-class.html#cb78-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_group_distance</span>(<span class="at">within_group =</span> <span class="cn">FALSE</span>)</span>
<span id="cb78-3"><a href="extended-class.html#cb78-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">distance_pair_stat =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/plot_group_distance_between.png" width="500px" style="display: block; margin: auto;" /></p>
<p>Clustering plot is also a frequently used method.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="extended-class.html#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use replace_name to set the label name, group parameter used to set the color</span></span>
<span id="cb79-2"><a href="extended-class.html#cb79-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_clustering</span>(<span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">replace_name =</span> <span class="fu">c</span>(<span class="st">&quot;Saline&quot;</span>, <span class="st">&quot;Type&quot;</span>))</span></code></pre></div>
<p><img src="Images/plot_clustering.png" width="550px" style="display: block; margin: auto;" /></p>
<p>perMANOVA<span class="citation">(<a href="#ref-Anderson_Austral_2001" role="doc-biblioref">Anderson 2001</a>)</span> is often used in the differential test of distances among groups.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="extended-class.html#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># manova for all groups</span></span>
<span id="cb80-2"><a href="extended-class.html#cb80-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_manova</span>(<span class="at">cal_manova_all =</span> <span class="cn">TRUE</span>)</span>
<span id="cb80-3"><a href="extended-class.html#cb80-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_manova<span class="sc">$</span>aov.tab</span></code></pre></div>
<pre><code>## The result is stored in object$res_manova ...</code></pre>
<table style="width:99%;">
<caption>Permutation: free</caption>
<colgroup>
<col width="22%" />
<col width="6%" />
<col width="16%" />
<col width="13%" />
<col width="13%" />
<col width="12%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">Df</th>
<th align="center">SumsOfSqs</th>
<th align="center">MeanSqs</th>
<th align="center">F.Model</th>
<th align="center">R2</th>
<th align="center">Pr(&gt;F)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Group</strong></td>
<td align="center">2</td>
<td align="center">6.121</td>
<td align="center">3.06</td>
<td align="center">10.57</td>
<td align="center">0.1955</td>
<td align="center">0.001</td>
</tr>
<tr class="even">
<td align="center"><strong>Residuals</strong></td>
<td align="center">87</td>
<td align="center">25.18</td>
<td align="center">0.2895</td>
<td align="center">NA</td>
<td align="center">0.8045</td>
<td align="center">NA</td>
</tr>
<tr class="odd">
<td align="center"><strong>Total</strong></td>
<td align="center">89</td>
<td align="center">31.3</td>
<td align="center">NA</td>
<td align="center">NA</td>
<td align="center">1</td>
<td align="center">NA</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="extended-class.html#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># manova for each paired groups</span></span>
<span id="cb82-2"><a href="extended-class.html#cb82-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_manova</span>(<span class="at">cal_manova_paired =</span> <span class="cn">TRUE</span>)</span>
<span id="cb82-3"><a href="extended-class.html#cb82-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_manova</span></code></pre></div>
<pre><code>## The result is stored in object$res_manova ...</code></pre>
<table style="width:97%;">
<colgroup>
<col width="15%" />
<col width="13%" />
<col width="20%" />
<col width="12%" />
<col width="13%" />
<col width="20%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Groups</th>
<th align="center">measure</th>
<th align="center">permutations</th>
<th align="center">R2</th>
<th align="center">p.value</th>
<th align="center">Significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">IW vs CW</td>
<td align="center">bray</td>
<td align="center">999</td>
<td align="center">0.1595</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="even">
<td align="center">IW vs TW</td>
<td align="center">bray</td>
<td align="center">999</td>
<td align="center">0.147</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="odd">
<td align="center">CW vs TW</td>
<td align="center">bray</td>
<td align="center">999</td>
<td align="center">0.1556</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="extended-class.html#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co"># manova for specified group set: here &quot;Group + Type&quot;</span></span>
<span id="cb84-2"><a href="extended-class.html#cb84-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_manova</span>(<span class="at">cal_manova_set =</span> <span class="st">&quot;Group + Type&quot;</span>)</span>
<span id="cb84-3"><a href="extended-class.html#cb84-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_manova<span class="sc">$</span>aov.tab</span></code></pre></div>
<pre><code>## The result is stored in object$res_manova ...</code></pre>
<table style="width:99%;">
<caption>Permutation: free</caption>
<colgroup>
<col width="22%" />
<col width="6%" />
<col width="16%" />
<col width="13%" />
<col width="13%" />
<col width="12%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">Df</th>
<th align="center">SumsOfSqs</th>
<th align="center">MeanSqs</th>
<th align="center">F.Model</th>
<th align="center">R2</th>
<th align="center">Pr(&gt;F)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>Group</strong></td>
<td align="center">2</td>
<td align="center">6.121</td>
<td align="center">3.06</td>
<td align="center">12.01</td>
<td align="center">0.1955</td>
<td align="center">0.001</td>
</tr>
<tr class="even">
<td align="center"><strong>Type</strong></td>
<td align="center">3</td>
<td align="center">3.783</td>
<td align="center">1.261</td>
<td align="center">4.949</td>
<td align="center">0.1208</td>
<td align="center">0.001</td>
</tr>
<tr class="odd">
<td align="center"><strong>Residuals</strong></td>
<td align="center">84</td>
<td align="center">21.4</td>
<td align="center">0.2548</td>
<td align="center">NA</td>
<td align="center">0.6836</td>
<td align="center">NA</td>
</tr>
<tr class="even">
<td align="center"><strong>Total</strong></td>
<td align="center">89</td>
<td align="center">31.3</td>
<td align="center">NA</td>
<td align="center">NA</td>
<td align="center">1</td>
<td align="center">NA</td>
</tr>
</tbody>
</table>
<p>PERMDISP<span class="citation">(<a href="#ref-Anderson_Navigating_2011" role="doc-biblioref">Anderson et al. 2011</a>)</span> is also implemented to check multivariate homogeneity of groups dispersions (variances).</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="extended-class.html#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PERMDISP for the whole comparison and for each paired groups</span></span>
<span id="cb86-2"><a href="extended-class.html#cb86-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_betadisper</span>()</span></code></pre></div>
<pre><code>## The result is stored in object$res_betadisper ...</code></pre>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="extended-class.html#cb88-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_betadisper</span></code></pre></div>
<pre><code>## 
## Permutation test for homogeneity of multivariate dispersions
## Permutation: free
## Number of permutations: 999
## 
## Response: Distances
##           Df  Sum Sq   Mean Sq      F N.Perm Pr(&gt;F)  
## Groups     2 0.04131 0.0206545 4.1682    999  0.021 *
## Residuals 87 0.43110 0.0049552                       
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## Pairwise comparisons:
## (Observed p-value below diagonal, permuted p-value above diagonal)
##           CW        IW    TW
## CW           0.4690000 0.063
## IW 0.4621193           0.005
## TW 0.0566190 0.0050319</code></pre>
</div>
<div id="trans_diff-class" class="section level2" number="4.5">
<h2><span class="header-section-number">4.5</span> trans_diff class</h2>
<p>　Differential abundance test is a very important part in the microbial community data analysis.
It can be used to find the significant taxa in determining the community differences across groups.
Currently, trans_diff class have four famous approaches to perform this analysis:
metastat<span class="citation">(<a href="#ref-White_Statistical_2009" role="doc-biblioref">White, Nagarajan, and Pop 2009</a>)</span>, LEfSe<span class="citation">(<a href="#ref-Segata_Metagenomic_2011" role="doc-biblioref">Segata et al. 2011</a>)</span>, random forest and metagenomeSeq<span class="citation">(<a href="#ref-Paulson_Differential_2013" role="doc-biblioref">Paulson et al. 2013</a>)</span>.
Metastat depends on the permutations and t-test and performs well on the sparse data.
It is used for the comparisons of taxonomic abundance between two groups at any taxonomic level.
LEfSe and random forest in this class is mainly used for the identification of biomarkers including all taxonomic level.
metagenomeSeq method is implemented to find significant species between two groups at species level (OTU/ASV).</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="extended-class.html#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># metastat analysis at Genus level</span></span>
<span id="cb90-2"><a href="extended-class.html#cb90-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;metastat&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">metastat_taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb90-3"><a href="extended-class.html#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_metastat is the result</span></span>
<span id="cb90-4"><a href="extended-class.html#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_metastat_group_matrix is the group comparisons order for plotting</span></span>
<span id="cb90-5"><a href="extended-class.html#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the first paired groups, choose_group = 1</span></span>
<span id="cb90-6"><a href="extended-class.html#cb90-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_metastat</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">qvalue =</span> <span class="fl">0.05</span>, <span class="at">choose_group =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Images/plot_metastat_1.png" width="650px" style="display: block; margin: auto;" /></p>
<p>LEfSe combines the non-parametric test and linear discriminant analysis.
We implement this approach in this package instead of the python version.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="extended-class.html#cb91-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;lefse&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.01</span>, <span class="at">lefse_subgroup =</span> <span class="cn">NULL</span>)</span>
<span id="cb91-2"><a href="extended-class.html#cb91-2" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_lefse is the LEfSe result</span></span>
<span id="cb91-3"><a href="extended-class.html#cb91-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_abund is the abundance information</span></span>
<span id="cb91-4"><a href="extended-class.html#cb91-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_lefse_bar</span>(<span class="at">LDA_score =</span> <span class="dv">4</span>)</span></code></pre></div>
<p><img src="Images/plot_lefse_bar.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="extended-class.html#cb92-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_lefse[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<table>
<colgroup>
<col width="66%" />
<col width="9%" />
<col width="14%" />
<col width="9%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Taxa</th>
<th align="center">Group</th>
<th align="center">pvalue</th>
<th align="center">LDA</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">k__Bacteria|p__Proteobacteria</td>
<td align="center">CW</td>
<td align="center">3.21e-11</td>
<td align="center">4.834</td>
</tr>
<tr class="even">
<td align="center">k__Bacteria|p__Acidobacteria|c__Acidobacteria</td>
<td align="center">IW</td>
<td align="center">8.559e-13</td>
<td align="center">4.787</td>
</tr>
<tr class="odd">
<td align="center">k__Bacteria|p__Acidobacteria</td>
<td align="center">IW</td>
<td align="center">5.749e-12</td>
<td align="center">4.785</td>
</tr>
<tr class="even">
<td align="center">k__Bacteria|p__Bacteroidetes</td>
<td align="center">TW</td>
<td align="center">1.19e-09</td>
<td align="center">4.776</td>
</tr>
<tr class="odd">
<td align="center">k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria</td>
<td align="center">CW</td>
<td align="center">5.475e-12</td>
<td align="center">4.613</td>
</tr>
</tbody>
</table>
<p>Then, we plot the abundance of biomarkers detected by LEfSe.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="extended-class.html#cb93-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code></pre></div>
<p><img src="Images/plot_lefse_diff_abund.png" width="650px" style="display: block; margin: auto;" /></p>
<p>Then, we show the cladogram of the differential features in the taxonomic tree.
There are too many taxa in this dataset.
As an example, we only use the highest 200 abundant taxa in the tree and 50 differential features.
We only show the full taxonomic label at Phylum level and use letters at other levels to reduce the text overlap.</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="extended-class.html#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clade_label_level 5 represent phylum level in this analysis</span></span>
<span id="cb94-2"><a href="extended-class.html#cb94-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require ggtree package</span></span>
<span id="cb94-3"><a href="extended-class.html#cb94-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_lefse_cladogram</span>(<span class="at">use_taxa_num =</span> <span class="dv">200</span>, <span class="at">use_feature_num =</span> <span class="dv">50</span>, <span class="at">clade_label_level =</span> <span class="dv">5</span>)</span></code></pre></div>
<p><img src="Images/plot_lefse_cladogram.png" width="1000px" style="display: block; margin: auto;" /></p>
<p>There may be a problem related with the taxonomic labels in the plot.
When the levels used are too many, the taxonomic labels may have too much overlap.
However, if you only indicate the Phylum labels, the taxa in the legend with marked letters are too many.
At this time, you can select the taxa that you want to show in the plot manually like the following operation.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="extended-class.html#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose some taxa according to the positions in the previous picture; those taxa labels have minimum overlap</span></span>
<span id="cb95-2"><a href="extended-class.html#cb95-2" aria-hidden="true" tabindex="-1"></a>use_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;c__Deltaproteobacteria&quot;</span>, <span class="st">&quot;c__Actinobacteria&quot;</span>, <span class="st">&quot;o__Rhizobiales&quot;</span>, <span class="st">&quot;p__Proteobacteria&quot;</span>, <span class="st">&quot;p__Bacteroidetes&quot;</span>, </span>
<span id="cb95-3"><a href="extended-class.html#cb95-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;o__Micrococcales&quot;</span>, <span class="st">&quot;p__Acidobacteria&quot;</span>, <span class="st">&quot;p__Verrucomicrobia&quot;</span>, <span class="st">&quot;p__Firmicutes&quot;</span>, </span>
<span id="cb95-4"><a href="extended-class.html#cb95-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;p__Chloroflexi&quot;</span>, <span class="st">&quot;c__Acidobacteria&quot;</span>, <span class="st">&quot;c__Gammaproteobacteria&quot;</span>, <span class="st">&quot;c__Betaproteobacteria&quot;</span>, <span class="st">&quot;c__KD4-96&quot;</span>,</span>
<span id="cb95-5"><a href="extended-class.html#cb95-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;c__Bacilli&quot;</span>, <span class="st">&quot;o__Gemmatimonadales&quot;</span>, <span class="st">&quot;f__Gemmatimonadaceae&quot;</span>, <span class="st">&quot;o__Bacillales&quot;</span>, <span class="st">&quot;o__Rhodobacterales&quot;</span>)</span>
<span id="cb95-6"><a href="extended-class.html#cb95-6" aria-hidden="true" tabindex="-1"></a><span class="co"># then use parameter select_show_labels to show</span></span>
<span id="cb95-7"><a href="extended-class.html#cb95-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_lefse_cladogram</span>(<span class="at">use_taxa_num =</span> <span class="dv">200</span>, <span class="at">use_feature_num =</span> <span class="dv">50</span>, <span class="at">select_show_labels =</span> use_labels)</span>
<span id="cb95-8"><a href="extended-class.html#cb95-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we can see that more taxa names appear in the tree</span></span></code></pre></div>
<p><img src="Images/plot_lefse_cladogram_1.png" width="1000px" style="display: block; margin: auto;" /></p>
<p>If you are interested in taxonomic tree, you can also use metacoder package<span class="citation">(<a href="#ref-Foster_Metacoder_2017" role="doc-biblioref">Foster, Sharpton, and Grunwald 2017</a>)</span> to plot the taxonomic tree based on the selected taxa.
We do not show the usage here.</p>
<p>The third approach is rf, which depends on the random forest<span class="citation">(<a href="#ref-Beck_Machine_2014" role="doc-biblioref">Beck and Foster 2014</a>; <a href="#ref-Yatsunenko_Human_2012" role="doc-biblioref">Yatsunenko et al. 2012</a>)</span> and the non-parametric test.
The current method can calculate random forest by bootstrapping like the method in LEfSe and only use the significant features.
MeanDecreaseGini is selected as the indicator value in the analysis.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="extended-class.html#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Genus level for parameter rf_taxa_level, if you want to use all taxa, change to &quot;all&quot;</span></span>
<span id="cb96-2"><a href="extended-class.html#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="co"># nresam = 1 and boots = 1 represent no bootstrapping and use all samples directly</span></span>
<span id="cb96-3"><a href="extended-class.html#cb96-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;rf&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">rf_taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb96-4"><a href="extended-class.html#cb96-4" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rf is the result stored in the object</span></span>
<span id="cb96-5"><a href="extended-class.html#cb96-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the result</span></span>
<span id="cb96-6"><a href="extended-class.html#cb96-6" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">only_abund_plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb96-7"><a href="extended-class.html#cb96-7" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(t2<span class="sc">$</span>p1, t2<span class="sc">$</span>p2, <span class="at">ncol=</span><span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb96-8"><a href="extended-class.html#cb96-8" aria-hidden="true" tabindex="-1"></a><span class="co"># the middle asterisk represent the significances</span></span></code></pre></div>
<p><img src="Images/plot_diff_abund.png" width="800px" style="display: block; margin: auto;" /></p>
</div>
<div id="trans_env-class" class="section level2" number="4.6">
<h2><span class="header-section-number">4.6</span> trans_env class</h2>
<p>　The environmental variables are very useful in analyzing microbial community structure and assembly mechanisms.
We first show the RDA analysis (db-RDA and RDA).</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="extended-class.html#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add_data is used to add the environmental data</span></span>
<span id="cb97-2"><a href="extended-class.html#cb97-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span></code></pre></div>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="extended-class.html#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use bray-curtis distance to do dbrda</span></span>
<span id="cb98-2"><a href="extended-class.html#cb98-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_rda</span>(<span class="at">use_dbrda =</span> <span class="cn">TRUE</span>, <span class="at">use_measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb98-3"><a href="extended-class.html#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rda is the result list stored in the object</span></span>
<span id="cb98-4"><a href="extended-class.html#cb98-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">trans_rda</span>(<span class="at">adjust_arrow_length =</span> <span class="cn">TRUE</span>, <span class="at">max_perc_env =</span> <span class="dv">10</span>)</span>
<span id="cb98-5"><a href="extended-class.html#cb98-5" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rda_trans is the transformed result for plotting</span></span>
<span id="cb98-6"><a href="extended-class.html#cb98-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_rda</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_rda_dbrda.png" width="650px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="extended-class.html#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Genus</span></span>
<span id="cb99-2"><a href="extended-class.html#cb99-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_rda</span>(<span class="at">use_dbrda =</span> <span class="cn">FALSE</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb99-3"><a href="extended-class.html#cb99-3" aria-hidden="true" tabindex="-1"></a><span class="co"># As the main results of RDA are related with the projection and angles between different arrows,</span></span>
<span id="cb99-4"><a href="extended-class.html#cb99-4" aria-hidden="true" tabindex="-1"></a><span class="co"># we adjust the length of the arrow to show them clearly using several parameters.</span></span>
<span id="cb99-5"><a href="extended-class.html#cb99-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">trans_rda</span>(<span class="at">show_taxa =</span> <span class="dv">10</span>, <span class="at">adjust_arrow_length =</span> <span class="cn">TRUE</span>, <span class="at">max_perc_env =</span> <span class="dv">1500</span>, <span class="at">max_perc_tax =</span> <span class="dv">3000</span>, <span class="at">min_perc_env =</span> <span class="dv">200</span>, <span class="at">min_perc_tax =</span> <span class="dv">300</span>)</span>
<span id="cb99-6"><a href="extended-class.html#cb99-6" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rda_trans is the transformed result for plotting</span></span>
<span id="cb99-7"><a href="extended-class.html#cb99-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_rda</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="Images/plot_rda_genus.png" width="650px" style="display: block; margin: auto;" /></p>
<p>Mantel test can be used to check whether there is significant correlations between environmental variables and distance matrix.</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="extended-class.html#cb100-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_mantel</span>(<span class="at">use_measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb100-2"><a href="extended-class.html#cb100-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_mantel</span></span>
<span id="cb100-3"><a href="extended-class.html#cb100-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_mantel</span></code></pre></div>
<pre><code>## The result is stored in object$res_mantel or object$res_mantel_partial ...</code></pre>
<table style="width:88%;">
<colgroup>
<col width="22%" />
<col width="18%" />
<col width="15%" />
<col width="11%" />
<col width="20%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">variable_name</th>
<th align="center">cor_method</th>
<th align="center">corr_res</th>
<th align="center">p_res</th>
<th align="center">significance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Temperature</td>
<td align="center">pearson</td>
<td align="center">0.452</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="even">
<td align="center">Precipitation</td>
<td align="center">pearson</td>
<td align="center">0.2791</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="odd">
<td align="center">TOC</td>
<td align="center">pearson</td>
<td align="center">0.13</td>
<td align="center">0.003</td>
<td align="center">**</td>
</tr>
<tr class="even">
<td align="center">NH4</td>
<td align="center">pearson</td>
<td align="center">-0.05539</td>
<td align="center">0.926</td>
<td align="center"></td>
</tr>
<tr class="odd">
<td align="center">NO3</td>
<td align="center">pearson</td>
<td align="center">0.06758</td>
<td align="center">0.05</td>
<td align="center">*</td>
</tr>
<tr class="even">
<td align="center">pH</td>
<td align="center">pearson</td>
<td align="center">0.4085</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="odd">
<td align="center">Conductivity</td>
<td align="center">pearson</td>
<td align="center">0.2643</td>
<td align="center">0.001</td>
<td align="center">***</td>
</tr>
<tr class="even">
<td align="center">TN</td>
<td align="center">pearson</td>
<td align="center">0.1321</td>
<td align="center">0.002</td>
<td align="center">**</td>
</tr>
</tbody>
</table>
<p>The correlations between environmental variables and taxa are important in analyzing and inferring the factors affecting community structure.
In this example, we first perform the differential abundance test and random forest analysis to obtain the important genera.
Then we use those taxa to perform correlation analysis.</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="extended-class.html#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first create trans_diff object</span></span>
<span id="cb102-2"><a href="extended-class.html#cb102-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;rf&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">rf_taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span></code></pre></div>
<pre><code>## Start differential test for Group ...</code></pre>
<pre><code>## Total 432 biomarkers found ...</code></pre>
<pre><code>## The result is stored in object$res_rf ...</code></pre>
<pre><code>## The abundance is stored in object$res_abund ...</code></pre>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="extended-class.html#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># then create trans_env object</span></span>
<span id="cb107-2"><a href="extended-class.html#cb107-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb107-3"><a href="extended-class.html#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate correlation</span></span>
<span id="cb107-4"><a href="extended-class.html#cb107-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">use_data =</span> <span class="st">&quot;other&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;fdr&quot;</span>, <span class="at">other_taxa =</span> t2<span class="sc">$</span>res_rf<span class="sc">$</span>Taxa[<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>])</span></code></pre></div>
<pre><code>## The correlation result is stored in object$res_cor ...</code></pre>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="extended-class.html#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_cor </span></span></code></pre></div>
<p>Then, we can plot the correlation results using ggplot2 or pheatmap.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="extended-class.html#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co"># default ggplot2 method</span></span>
<span id="cb110-2"><a href="extended-class.html#cb110-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/plot_corr_ggplot.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="extended-class.html#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clustering heatmap; require pheatmap package</span></span>
<span id="cb111-2"><a href="extended-class.html#cb111-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">pheatmap =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/plot_corr_pheatmap.png" width="700px" style="display: block; margin: auto;" /></p>
<p>Sometimes, it is necessary to study the correlations between environmental variables and taxa for different groups.</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="extended-class.html#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate correlations for different groups using parameter by_group</span></span>
<span id="cb112-2"><a href="extended-class.html#cb112-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">by_group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">use_data =</span> <span class="st">&quot;other&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;fdr&quot;</span>, <span class="at">other_taxa =</span> t2<span class="sc">$</span>res_rf<span class="sc">$</span>Taxa[<span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>])</span>
<span id="cb112-3"><a href="extended-class.html#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_cor</span></span>
<span id="cb112-4"><a href="extended-class.html#cb112-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/plot_corr_ggplot_groups.png" width="700px" style="display: block; margin: auto;" /></p>
<p>If you are concerned with the relationship between environmental factors and alpha diversity, you can also use this function.</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="extended-class.html#cb113-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb113-2"><a href="extended-class.html#cb113-2" aria-hidden="true" tabindex="-1"></a><span class="co"># use add_abund_table parameter to add the extra data table</span></span>
<span id="cb113-3"><a href="extended-class.html#cb113-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">add_abund_table =</span> dataset<span class="sc">$</span>alpha_diversity)</span>
<span id="cb113-4"><a href="extended-class.html#cb113-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/plot_corr_alpha_diversity.png" width="600px" style="display: block; margin: auto;" /></p>
<p>The function plot_scatterfit() in trans_env class is designed for the scatter plot and adding the fitted line.</p>
</div>
<div id="trans_nullmodel-class" class="section level2" number="4.7">
<h2><span class="header-section-number">4.7</span> trans_nullmodel class</h2>
<p>In recent decades,
the integration of phylogenetic analysis and null model promotes the inference of niche and neutral influences on community assembly more powerfully
by adding a phylogeny dimension <span class="citation">(<a href="#ref-Webb_Phylogenies_2002" role="doc-biblioref">Webb et al. 2002</a>; <a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>)</span>.
The trans_nullmodel class provides an encapsulation, including the calculation of the phylogenetic signal,
beta mean pairwise phylogenetic distance (betaMPD), beta mean nearest taxon distance (betaMNTD),
beta nearest taxon index (betaNTI), beta net relatedness index (betaNRI) and Bray-Curtis-based Raup-Crick (RCbray).
The approach for phylogenetic signal analysis is based on the mantel correlogram <span class="citation">(<a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>,
in which the change of phylogenetic signal is intuitional and clear compared to other approaches.
The algorithms of betaMNTD and betaMPD have been optimized to be faster than those in the picante package <span class="citation">(<a href="#ref-Picante_Kembel_2010" role="doc-biblioref">Kembel et al. 2010</a>)</span>.
The combinations between RCbray and betaNTI (or betaNRI) can be used to infer the strength of each ecological process dominating the community assembly
under the specific hypothesis <span class="citation">(<a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>)</span>.
This can be achievable by the function cal_process() to parse the percentage of each inferred process.
We first check the phylogenetic signal.</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="extended-class.html#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generate trans_nullmodel object; use 1000 OTUs as example</span></span>
<span id="cb114-2"><a href="extended-class.html#cb114-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_nullmodel<span class="sc">$</span><span class="fu">new</span>(dataset, <span class="at">taxa_number =</span> <span class="dv">1000</span>, <span class="at">add_data =</span> env_data_16S)</span></code></pre></div>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="extended-class.html#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use pH as the test variable</span></span>
<span id="cb115-2"><a href="extended-class.html#cb115-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_mantel_corr</span>(<span class="at">use_env =</span> <span class="st">&quot;pH&quot;</span>)</span>
<span id="cb115-3"><a href="extended-class.html#cb115-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_mantel_corr</span></span>
<span id="cb115-4"><a href="extended-class.html#cb115-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the mantel correlogram</span></span>
<span id="cb115-5"><a href="extended-class.html#cb115-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_mantel_corr</span>()</span></code></pre></div>
<p><img src="Images/plot_mantel_corr.png" width="600px" style="display: block; margin: auto;" /></p>
<p>betaNRI(ses.betampd) is used to show the ‘basal’ phylogenetic turnover<span class="citation">(<a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>.
Compared to betaNTI, it can capture more turnover information associated with the deep phylogeny.
It is noted that there are many null models with the development in the several decades.
In the trans_nullmodel class, we randomized the phylogenetic relatedness of species.
This shuffling approach fix the observed levels of species α-diversity and β-diversity to
explore whether the observed phylogenetic turnover significantly differ from null model that phylogenetic relatedness among species are random.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="extended-class.html#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># null model run 500 times</span></span>
<span id="cb116-2"><a href="extended-class.html#cb116-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ses_betampd</span>(<span class="at">runs=</span><span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb116-3"><a href="extended-class.html#cb116-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_ses_betampd</span></span></code></pre></div>
<p>If we want to plot the betaNRI, we can use plot_group_distance function in trans_beta class.
For example, the results showed that the mean betaNRI of TW is extremely and significantly larger that those in CW and IW,
revealing that the basal phylogenetic turnover in TW is high.</p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="extended-class.html#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add betaNRI matrix to beta_diversity list</span></span>
<span id="cb117-2"><a href="extended-class.html#cb117-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>beta_diversity[[<span class="st">&quot;betaNRI&quot;</span>]] <span class="ot">&lt;-</span> t1<span class="sc">$</span>res_ses_betampd</span>
<span id="cb117-3"><a href="extended-class.html#cb117-3" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_beta class, use measure &quot;betaNRI&quot;</span></span>
<span id="cb117-4"><a href="extended-class.html#cb117-4" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;betaNRI&quot;</span>)</span>
<span id="cb117-5"><a href="extended-class.html#cb117-5" aria-hidden="true" tabindex="-1"></a><span class="co"># transform the distance for each group</span></span>
<span id="cb117-6"><a href="extended-class.html#cb117-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance</span>()</span>
<span id="cb117-7"><a href="extended-class.html#cb117-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the results</span></span>
<span id="cb117-8"><a href="extended-class.html#cb117-8" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t2<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">distance_pair_stat =</span> <span class="cn">TRUE</span>)</span>
<span id="cb117-9"><a href="extended-class.html#cb117-9" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/plot_betaNRI_one_dataset.png" width="550px" style="display: block; margin: auto;" /></p>
<p>Sometimes, if you want to perform null model analysis for each group individually, such as one group as one species pool,
you can calculate the results for each group, respectively.
We can find that, when we perform betaNRI for each group respectively,
mean betaNRI between CW and TW are not significantly different, and they are both significantly higher than that in IW,
revealing that the strength of variable selection in CW and TW may be similar under the condition that each area is considered as a specific species pool.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="extended-class.html#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we create a list to store the trans_nullmodel results.</span></span>
<span id="cb118-2"><a href="extended-class.html#cb118-2" aria-hidden="true" tabindex="-1"></a>sesbeta_each <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb118-3"><a href="extended-class.html#cb118-3" aria-hidden="true" tabindex="-1"></a>group_col <span class="ot">&lt;-</span> <span class="st">&quot;Group&quot;</span></span>
<span id="cb118-4"><a href="extended-class.html#cb118-4" aria-hidden="true" tabindex="-1"></a>all_groups <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>sample_table[, group_col])</span>
<span id="cb118-5"><a href="extended-class.html#cb118-5" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate for each group, respectively</span></span>
<span id="cb118-6"><a href="extended-class.html#cb118-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> all_groups){</span>
<span id="cb118-7"><a href="extended-class.html#cb118-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># like the above operation, but need provide &#39;group&#39; and &#39;select_group&#39;</span></span>
<span id="cb118-8"><a href="extended-class.html#cb118-8" aria-hidden="true" tabindex="-1"></a>    test <span class="ot">&lt;-</span> trans_nullmodel<span class="sc">$</span><span class="fu">new</span>(dataset, <span class="at">group =</span> group_col, <span class="at">select_group =</span> i, <span class="at">taxa_number =</span> <span class="dv">1000</span>, <span class="at">add_data =</span> env_data_16S)</span>
<span id="cb118-9"><a href="extended-class.html#cb118-9" aria-hidden="true" tabindex="-1"></a>    test<span class="sc">$</span><span class="fu">cal_ses_betampd</span>(<span class="at">runs =</span> <span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb118-10"><a href="extended-class.html#cb118-10" aria-hidden="true" tabindex="-1"></a>    sesbeta_each[[i]] <span class="ot">&lt;-</span> test<span class="sc">$</span>res_ses_betampd</span>
<span id="cb118-11"><a href="extended-class.html#cb118-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb118-12"><a href="extended-class.html#cb118-12" aria-hidden="true" tabindex="-1"></a><span class="co"># merge and reshape to generate one symmetrical matrix</span></span>
<span id="cb118-13"><a href="extended-class.html#cb118-13" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">lapply</span>(sesbeta_each, melt) <span class="sc">%&gt;%</span> <span class="fu">do.call</span>(rbind, .) <span class="sc">%&gt;%</span></span>
<span id="cb118-14"><a href="extended-class.html#cb118-14" aria-hidden="true" tabindex="-1"></a>    reshape2<span class="sc">::</span><span class="fu">dcast</span>(., Var1<span class="sc">~</span>Var2, <span class="at">value.var =</span> <span class="st">&quot;value&quot;</span>) <span class="sc">%&gt;%</span> <span class="st">`</span><span class="at">row.names&lt;-</span><span class="st">`</span>(.[,<span class="dv">1</span>]) <span class="sc">%&gt;%</span> .[, <span class="sc">-</span><span class="dv">1</span>, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb118-15"><a href="extended-class.html#cb118-15" aria-hidden="true" tabindex="-1"></a><span class="co"># like the above operation</span></span>
<span id="cb118-16"><a href="extended-class.html#cb118-16" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>beta_diversity[[<span class="st">&quot;betaNRI&quot;</span>]] <span class="ot">&lt;-</span> test</span>
<span id="cb118-17"><a href="extended-class.html#cb118-17" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;betaNRI&quot;</span>)</span>
<span id="cb118-18"><a href="extended-class.html#cb118-18" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance</span>()</span>
<span id="cb118-19"><a href="extended-class.html#cb118-19" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t2<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">distance_pair_stat =</span> <span class="cn">TRUE</span>)</span>
<span id="cb118-20"><a href="extended-class.html#cb118-20" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/plot_betaNRI_each_dataset.png" width="550px" style="display: block; margin: auto;" /></p>
<p>BetaNTI(ses.betamntd) can be used to indicate the phylogenetic terminal turnover <span class="citation">(<a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>)</span>.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="extended-class.html#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co"># null model run 500 times</span></span>
<span id="cb119-2"><a href="extended-class.html#cb119-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ses_betamntd</span>(<span class="at">runs=</span><span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb119-3"><a href="extended-class.html#cb119-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_ses_betamntd</span></span></code></pre></div>
<table style="width:75%;">
<colgroup>
<col width="12%" />
<col width="12%" />
<col width="12%" />
<col width="12%" />
<col width="12%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">S1</th>
<th align="center">S2</th>
<th align="center">S3</th>
<th align="center">S4</th>
<th align="center">S5</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">0</td>
<td align="center">-6.554</td>
<td align="center">-6.563</td>
<td align="center">-6.308</td>
<td align="center">-6.153</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">-6.554</td>
<td align="center">0</td>
<td align="center">-6.678</td>
<td align="center">-6.675</td>
<td align="center">-6.124</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">-6.563</td>
<td align="center">-6.678</td>
<td align="center">0</td>
<td align="center">-6.544</td>
<td align="center">-6.46</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">-6.308</td>
<td align="center">-6.675</td>
<td align="center">-6.544</td>
<td align="center">0</td>
<td align="center">-6.356</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">-6.153</td>
<td align="center">-6.124</td>
<td align="center">-6.46</td>
<td align="center">-6.356</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>RCbray (Bray-Curtis-based Raup-Crick) can be calculated using function cal_rcbray()
to assess whether the compositional turnover was governed primarily by drift <span class="citation">(<a href="#ref-Chase_null_2011" role="doc-biblioref">Chase et al. 2011</a>)</span>.
We applied null model to simulate species distribution by randomly sampling individuals from each
species pool with preserving species occurrence frequency and sample species richness <span class="citation">(<a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="extended-class.html#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># result stored in t1$res_rcbray</span></span>
<span id="cb120-2"><a href="extended-class.html#cb120-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_rcbray</span>(<span class="at">runs =</span> <span class="dv">1000</span>)</span>
<span id="cb120-3"><a href="extended-class.html#cb120-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_rcbray</span></span></code></pre></div>
<p>As an example, we also calculate the proportion of the inferred processes on the community assembly as shown in the references <span class="citation">(<a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>; <a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>.
In the example, the fraction of pairwise comparisons with significant betaNTI values (|βNTI| &gt; 2) is the estimated influence of Selection;
βNTI &gt; 2 represents the heterogeneous selection; βNTI &lt; -2 represents the homogeneous selection.
The value of RCbray characterizes the magnitude of deviation between observed Bray–Curtis and Bray–Curtis expected under the randomization;
a value of |RCbray| &gt; 0.95 was considered significant.
The fraction of all pairwise comparisons with |βNTI| &lt; 2 and RCbray &gt; +0.95 was taken as the influence of Dispersal Limitation combined with Drift.
The fraction of all pairwise comparisons with |βNTI| &lt; 2 and RCbray &lt; -0.95 was taken as an estimate for the influence of Homogenizing Dispersal.
The fraction of all pairwise comparisons with |βNTI| &lt; 2 and |RCbray| &lt; 0.95 estimates the influence of Drift acting alone.</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="extended-class.html#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use betaNTI and rcbray to evaluate processes</span></span>
<span id="cb121-2"><a href="extended-class.html#cb121-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_process</span>(<span class="at">use_betamntd =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## The result is stored in object$res_process ...</code></pre>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="extended-class.html#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_process</span></span></code></pre></div>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="extended-class.html#cb124-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_process</span></code></pre></div>
<table style="width:51%;">
<colgroup>
<col width="33%" />
<col width="18%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">process</th>
<th align="center">percentage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">variable selection</td>
<td align="center">3.995</td>
</tr>
<tr class="even">
<td align="center">homogeneous selection</td>
<td align="center">48.34</td>
</tr>
<tr class="odd">
<td align="center">dispersal limitation</td>
<td align="center">0.02497</td>
</tr>
<tr class="even">
<td align="center">homogeneous dispersal</td>
<td align="center">8.539</td>
</tr>
<tr class="odd">
<td align="center">drift</td>
<td align="center">39.1</td>
</tr>
</tbody>
</table>
</div>
<div id="trans_network-class" class="section level2" number="4.8">
<h2><span class="header-section-number">4.8</span> trans_network class</h2>
<p>　Network is a frequently used approach to study the co-occurrence patterns in microbial ecology<span class="citation">(<a href="#ref-Deng_Molecular_2012" role="doc-biblioref">Deng et al. 2012</a>; <a href="#ref-Faust_Microbial_2012" role="doc-biblioref">Faust and Raes 2012</a>; <a href="#ref-Coyte_Theecology_2015" role="doc-biblioref">Coyte, Schluter, and Foster 2015</a>)</span>.
In this part, we describe all the core contents in the trans_network class.
The network construction approaches can be classified into two types: correlation-based and non correlation-based.
Several approaches can be used to calculate correlations and significances.</p>
<p>We first introduce the correlation-based network. The parameter cal_cor in trans_network is used for selecting the correlation calculation method.</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="extended-class.html#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use R base cor.test, slow</span></span>
<span id="cb125-2"><a href="extended-class.html#cb125-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="st">&quot;base&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb125-3"><a href="extended-class.html#cb125-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_cor_p list; one table: correlation; another: p value</span></span></code></pre></div>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="extended-class.html#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="co"># SparCC method, require SpiecEasi package</span></span>
<span id="cb126-2"><a href="extended-class.html#cb126-2" aria-hidden="true" tabindex="-1"></a><span class="co"># SparCC is very slow, so consider filtering more species with low abundance</span></span>
<span id="cb126-3"><a href="extended-class.html#cb126-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="st">&quot;SparCC&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>, <span class="at">SparCC_simu_num =</span> <span class="dv">100</span>)</span></code></pre></div>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="extended-class.html#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="co"># When the OTU number is large, use R WGCNA package to replace R base to calculate correlations</span></span>
<span id="cb127-2"><a href="extended-class.html#cb127-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require WGCNA package</span></span>
<span id="cb127-3"><a href="extended-class.html#cb127-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="st">&quot;WGCNA&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span></code></pre></div>
<p>The parameter COR_cut can be used to select the correlation threshold.
Furthermore, COR_optimization = TRUE represent using RMT theory to find the optimized correlation threshold instead of the COR_cut<span class="citation">(<a href="#ref-Deng_Molecular_2012" role="doc-biblioref">Deng et al. 2012</a>)</span>.</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="extended-class.html#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="co"># construct network; require igraph package</span></span>
<span id="cb128-2"><a href="extended-class.html#cb128-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">p_thres =</span> <span class="fl">0.01</span>, <span class="at">COR_optimization =</span> <span class="cn">TRUE</span>)</span>
<span id="cb128-3"><a href="extended-class.html#cb128-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_network</span></span></code></pre></div>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="extended-class.html#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use arbitrary coefficient threshold to contruct network</span></span>
<span id="cb129-2"><a href="extended-class.html#cb129-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">p_thres =</span> <span class="fl">0.01</span>, <span class="at">COR_cut =</span> <span class="fl">0.7</span>)</span></code></pre></div>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="extended-class.html#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add modules in the network</span></span>
<span id="cb130-2"><a href="extended-class.html#cb130-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_module</span>()</span></code></pre></div>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="extended-class.html#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save network</span></span>
<span id="cb131-2"><a href="extended-class.html#cb131-2" aria-hidden="true" tabindex="-1"></a><span class="co"># open the gexf file using Gephi(https://gephi.org/)</span></span>
<span id="cb131-3"><a href="extended-class.html#cb131-3" aria-hidden="true" tabindex="-1"></a><span class="co"># require rgexf package</span></span>
<span id="cb131-4"><a href="extended-class.html#cb131-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">save_network</span>(<span class="at">filepath =</span> <span class="st">&quot;network.gexf&quot;</span>)</span></code></pre></div>
<p>We plot the network and present the node colors according to the calculated modules in Gephi.
<img src="Images/network1_spearman.png" width="550px" style="display: block; margin: auto;" /></p>
<p>Now, we show the node colors with the Phylum information and the edges colors with the positive and negative correlations.
All the data used has been stored in the network.gexf file, including modules classifications, Phylum information and edges classifications.</p>
<p><img src="Images/network2_spearman.png" width="550px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="extended-class.html#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate network attributes</span></span>
<span id="cb132-2"><a href="extended-class.html#cb132-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network_attr</span>()</span>
<span id="cb132-3"><a href="extended-class.html#cb132-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_network_attr</span></span></code></pre></div>
<table style="width:49%;">
<colgroup>
<col width="34%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Property</th>
<th align="center">Value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Vertex</td>
<td align="center">407</td>
</tr>
<tr class="even">
<td align="center">Edge</td>
<td align="center">1989</td>
</tr>
<tr class="odd">
<td align="center">Average_degree</td>
<td align="center">9.774</td>
</tr>
<tr class="even">
<td align="center">Average_path_length</td>
<td align="center">3.878</td>
</tr>
<tr class="odd">
<td align="center">Network_diameter</td>
<td align="center">9</td>
</tr>
<tr class="even">
<td align="center">Clustering_coefficient</td>
<td align="center">0.4698</td>
</tr>
<tr class="odd">
<td align="center">Density</td>
<td align="center">0.02407</td>
</tr>
<tr class="even">
<td align="center">Heterogeneity</td>
<td align="center">1.194</td>
</tr>
<tr class="odd">
<td align="center">Centralization</td>
<td align="center">0.09908</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="extended-class.html#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="co"># classify the node; return t1$res_node_type</span></span>
<span id="cb133-2"><a href="extended-class.html#cb133-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_node_type</span>()</span>
<span id="cb133-3"><a href="extended-class.html#cb133-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_node_type</span></span>
<span id="cb133-4"><a href="extended-class.html#cb133-4" aria-hidden="true" tabindex="-1"></a><span class="co"># we retain the file for the following example in trans_func part</span></span>
<span id="cb133-5"><a href="extended-class.html#cb133-5" aria-hidden="true" tabindex="-1"></a>network_node_type <span class="ot">&lt;-</span> t1<span class="sc">$</span>res_node_type</span></code></pre></div>
<table style="width:89%;">
<colgroup>
<col width="22%" />
<col width="15%" />
<col width="12%" />
<col width="12%" />
<col width="26%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">z</th>
<th align="center">module</th>
<th align="center">p</th>
<th align="center">taxa_roles</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>OTU_50</strong></td>
<td align="center">-1.305</td>
<td align="center">M2</td>
<td align="center">0</td>
<td align="center">Peripheral nodes</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_1</strong></td>
<td align="center">-0.04067</td>
<td align="center">M2</td>
<td align="center">0</td>
<td align="center">Peripheral nodes</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_55</strong></td>
<td align="center">-1.239</td>
<td align="center">M2</td>
<td align="center">0</td>
<td align="center">Peripheral nodes</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_13824</strong></td>
<td align="center">-0.2403</td>
<td align="center">M2</td>
<td align="center">0</td>
<td align="center">Peripheral nodes</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_151</strong></td>
<td align="center">-1.372</td>
<td align="center">M2</td>
<td align="center">0.4444</td>
<td align="center">Peripheral nodes</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="extended-class.html#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot node roles in terms of the within-module connectivity and among-module connectivity</span></span>
<span id="cb134-2"><a href="extended-class.html#cb134-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_taxa_roles</span>(<span class="at">use_type =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Images/plot_taxa_roles.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="extended-class.html#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot node roles with phylum information</span></span>
<span id="cb135-2"><a href="extended-class.html#cb135-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_taxa_roles</span>(<span class="at">use_type =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/plot_taxa_roles_2.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Now, we show the eigengene analysis of modules.
The eigengene of a module, i.e. the first principal component of PCA, represents the main variance of the abundance in the species of the module.</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="extended-class.html#cb136-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_eigen</span>()</span>
<span id="cb136-2"><a href="extended-class.html#cb136-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_eigen</span></span></code></pre></div>
<p>Then we perform correlation heatmap to show the relationships between eigengenes and environmental factors.</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="extended-class.html#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_env object like the above operation</span></span>
<span id="cb137-2"><a href="extended-class.html#cb137-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb137-3"><a href="extended-class.html#cb137-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate correlations</span></span>
<span id="cb137-4"><a href="extended-class.html#cb137-4" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">add_abund_table =</span> t1<span class="sc">$</span>res_eigen)</span>
<span id="cb137-5"><a href="extended-class.html#cb137-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the correlation heatmap</span></span>
<span id="cb137-6"><a href="extended-class.html#cb137-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/Env_module_eigen.png" width="600px" style="display: block; margin: auto;" /></p>
<p>The subset_network() function can be used to extract a part of nodes and edges among these nodes from the network.
In this function, you should provide the nodes you need using the node parameter.</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="extended-class.html#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract a sub network that contains all nodes in module M1</span></span>
<span id="cb138-2"><a href="extended-class.html#cb138-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">subset_network</span>(<span class="at">node =</span> t1<span class="sc">$</span>res_node_type <span class="sc">%&gt;%</span> .[.<span class="sc">$</span>module <span class="sc">==</span> <span class="st">&quot;M1&quot;</span>, ] <span class="sc">%&gt;%</span> rownames, <span class="at">rm_single =</span> <span class="cn">TRUE</span>)</span>
<span id="cb138-3"><a href="extended-class.html#cb138-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return a new network with igraph class</span></span>
<span id="cb138-4"><a href="extended-class.html#cb138-4" aria-hidden="true" tabindex="-1"></a><span class="co"># extract sub network in which all edge labels are &quot;+&quot;, i.e. positive edges</span></span>
<span id="cb138-5"><a href="extended-class.html#cb138-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">subset_network</span>(<span class="at">edge =</span> <span class="st">&quot;+&quot;</span>)</span></code></pre></div>
<p>Then we show the next implemented network construction approach:
SPIEC-EASI (SParse InversE Covariance Estimation for Ecological Association Inference) network in SpiecEasi R package <span class="citation">(<a href="#ref-Kurtz_Sparse_2015" role="doc-biblioref">Kurtz et al. 2015</a>)</span>.</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="extended-class.html#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cal_cor select NA</span></span>
<span id="cb139-2"><a href="extended-class.html#cb139-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="cn">NA</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0005</span>)</span>
<span id="cb139-3"><a href="extended-class.html#cb139-3" aria-hidden="true" tabindex="-1"></a><span class="co"># require SpiecEasi package  https://github.com/zdk123/SpiecEasi</span></span>
<span id="cb139-4"><a href="extended-class.html#cb139-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">network_method =</span> <span class="st">&quot;SpiecEasi&quot;</span>)</span>
<span id="cb139-5"><a href="extended-class.html#cb139-5" aria-hidden="true" tabindex="-1"></a><span class="co"># see t1$res_network</span></span></code></pre></div>
<p>We also introduce the third network construction approach: Probabilistic Graphical Models (PGM), which is implemented in julia package FlashWeave<span class="citation">(<a href="#ref-Tackmann_Rapid_2019" role="doc-biblioref">Tackmann, Matias Rodrigues, and Mering 2019</a>)</span>.
If you want to use this method like the following code, you should first install julia language in your computer and the FlashWeave package,
and add the julia in the computer path (see FlashWeave part in <a href="https://github.com/ChiLiubio/microeco" class="uri">https://github.com/ChiLiubio/microeco</a>).</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="extended-class.html#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cal_cor select NA</span></span>
<span id="cb140-2"><a href="extended-class.html#cb140-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="cn">NA</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>)</span>
<span id="cb140-3"><a href="extended-class.html#cb140-3" aria-hidden="true" tabindex="-1"></a><span class="co"># require Julia in the computer path, and the package FlashWeave</span></span>
<span id="cb140-4"><a href="extended-class.html#cb140-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">network_method =</span> <span class="st">&quot;PGM&quot;</span>)</span>
<span id="cb140-5"><a href="extended-class.html#cb140-5" aria-hidden="true" tabindex="-1"></a><span class="co"># see t1$res_network</span></span></code></pre></div>
</div>
<div id="trans_func-class" class="section level2" number="4.9">
<h2><span class="header-section-number">4.9</span> trans_func class</h2>
<p>　Ecological researchers are usually interested in the the funtional profiles of microbial communities,
because functional or metabolic data is powerful to explain the structure and dynamics of microbial communities and to infer the underlying mechanisms.
As metagenomic sequencing is complicated and expensive, using amplicon sequencing data to predict functional profiles is a good choice.
Several software are often used for this goal, such as PICRUSt<span class="citation">(<a href="#ref-Langille_Predictive_2013" role="doc-biblioref">Langille et al. 2013</a>)</span>, Tax4Fun<span class="citation">(<a href="#ref-Aßhauer_Tax4Fun_2015" role="doc-biblioref">Aßhauer et al. 2015</a>)</span> and FAPROTAX<span class="citation">(<a href="#ref-Louca_High_2016" role="doc-biblioref">Stilianos Louca et al. 2016</a>; <a href="#ref-Louca_Decoupling_2016" role="doc-biblioref">S. Louca, Parfrey, and Doebeli 2016</a>)</span>.
These tools are great to be used for the prediction of functional profiles based on the prokaryotic communities from sequencing results.
In addition, it is also important to obtain the functions for each taxa or OTU, not just the whole profile of communities.
But it is hard to know exact functions of each OTU.
FAPROTAX database is a collection of the traits and characteristics of prokaryotes based on the known research results published in books and literatures.
We match the taxonomic information of prokaryotes against this database to identify the traits of prokaryotes on biogeochemical roles.
The NJC19 database<span class="citation">(<a href="#ref-Lim_Large_2020" role="doc-biblioref">Lim et al. 2020</a>)</span> is also available for animal associated prokaryotic data, such as human gut microbiota.
We also implement the FUNGuild <span class="citation">(<a href="#ref-Nguyen_FUNGuild_2016" role="doc-biblioref">Nguyen et al. 2016</a>)</span> and FungalTraits <span class="citation">(<a href="#ref-Polme_FungalTraits_2020" role="doc-biblioref">Põlme et al. 2020</a>)</span> databases to identify the fungal traits.</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="extended-class.html#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify microbial traits</span></span>
<span id="cb141-2"><a href="extended-class.html#cb141-2" aria-hidden="true" tabindex="-1"></a><span class="co"># create object of trans_func</span></span>
<span id="cb141-3"><a href="extended-class.html#cb141-3" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(dataset)</span>
<span id="cb141-4"><a href="extended-class.html#cb141-4" aria-hidden="true" tabindex="-1"></a><span class="co"># mapping the taxonomy to the database</span></span>
<span id="cb141-5"><a href="extended-class.html#cb141-5" aria-hidden="true" tabindex="-1"></a><span class="co"># this can recognize prokaryotes or fungi automatically.</span></span>
<span id="cb141-6"><a href="extended-class.html#cb141-6" aria-hidden="true" tabindex="-1"></a><span class="co"># default database for prokaryotes is FAPROTAX database</span></span>
<span id="cb141-7"><a href="extended-class.html#cb141-7" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func</span>()</span></code></pre></div>
<pre><code>## Please also cite the original FAPROTAX paper: Louca et al. (2016).</code></pre>
<pre><code>## Decoupling function and taxonomy in the global ocean microbiome. Science, 353(6305), 1272.</code></pre>
<pre><code>## The functional table is stored in object$res_spe_func ...</code></pre>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="extended-class.html#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return t2$res_spe_func, 1 represent function exists, 0 represent no or cannot confirmed.</span></span></code></pre></div>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="extended-class.html#cb146-1" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>res_spe_func[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code></pre></div>
<table style="width:85%;">
<colgroup>
<col width="20%" />
<col width="22%" />
<col width="41%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">methanotrophy</th>
<th align="center">acetoclastic_methanogenesis</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>OTU_4272</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_236</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_399</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_1556</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_32</strong></td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>The percentages of the OTUs having the same trait can reflect the functional redundancy of this function in the community or the module in the network.</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="extended-class.html#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the percentages of OTUs for each trait in each module of network</span></span>
<span id="cb147-2"><a href="extended-class.html#cb147-2" aria-hidden="true" tabindex="-1"></a><span class="co"># use_community = FALSE represent calculating module, not community, node_type_table provide the module information</span></span>
<span id="cb147-3"><a href="extended-class.html#cb147-3" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">use_community =</span> <span class="cn">FALSE</span>, <span class="at">node_type_table =</span> network_node_type)</span>
<span id="cb147-4"><a href="extended-class.html#cb147-4" aria-hidden="true" tabindex="-1"></a><span class="co"># return t2$res_spe_func_perc</span></span>
<span id="cb147-5"><a href="extended-class.html#cb147-5" aria-hidden="true" tabindex="-1"></a><span class="co"># we only plot some important traits, so we use the default group list to filter and show the traits.</span></span>
<span id="cb147-6"><a href="extended-class.html#cb147-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_spe_func_perc</span>(<span class="at">select_samples =</span> <span class="fu">paste0</span>(<span class="st">&quot;M&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>))</span>
<span id="cb147-7"><a href="extended-class.html#cb147-7" aria-hidden="true" tabindex="-1"></a><span class="co"># M represents module, ordered by the nodes number from high to low</span></span></code></pre></div>
<p><img src="Images/plot_func_perc.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="extended-class.html#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If you want to change the group list, reset the list t2$func_group_list</span></span>
<span id="cb148-2"><a href="extended-class.html#cb148-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>func_group_list</span>
<span id="cb148-3"><a href="extended-class.html#cb148-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use show_prok_func to see the detailed information of prokaryotic traits</span></span>
<span id="cb148-4"><a href="extended-class.html#cb148-4" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">show_prok_func</span>(<span class="st">&quot;methanotrophy&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="extended-class.html#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the percentages for communities</span></span>
<span id="cb149-2"><a href="extended-class.html#cb149-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">use_community =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## The result table is stored in object$res_spe_func_perc ...</code></pre>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="extended-class.html#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="co"># t2$res_spe_func_perc[1:5, 1:2]</span></span></code></pre></div>
<table style="width:76%;">
<colgroup>
<col width="12%" />
<col width="22%" />
<col width="41%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">methanotrophy</th>
<th align="center">acetoclastic_methanogenesis</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">0.39</td>
<td align="center">0.04</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">0.27</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">0.48</td>
<td align="center">0</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">0.48</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">0.56</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="extended-class.html#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="co"># then we try to correlate the res_spe_func_perc of communities to environmental variables</span></span>
<span id="cb152-2"><a href="extended-class.html#cb152-2" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb152-3"><a href="extended-class.html#cb152-3" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">add_abund_table =</span> t2<span class="sc">$</span>res_spe_func_perc, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb152-4"><a href="extended-class.html#cb152-4" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span><span class="fu">plot_cor</span>(<span class="at">pheatmap =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="Images/plot_func_perc_corr.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Tax4Fun requires a strict input file demand on the taxonomic information.
To analyze the trimmed or changed OTU data in R with Tax4Fun, we provide a link to the Tax4Fun functional prediction.</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="extended-class.html#cb153-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(dataset)</span>
<span id="cb153-2"><a href="extended-class.html#cb153-2" aria-hidden="true" tabindex="-1"></a><span class="co"># install Tax4Fun package and download SILVA123 ref data from  http://tax4fun.gobics.de/</span></span>
<span id="cb153-3"><a href="extended-class.html#cb153-3" aria-hidden="true" tabindex="-1"></a><span class="co"># decompress SILVA123; provide path in folderReferenceData as you put</span></span>
<span id="cb153-4"><a href="extended-class.html#cb153-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_tax4fun</span>(<span class="at">folderReferenceData =</span> <span class="st">&quot;./SILVA123&quot;</span>)</span></code></pre></div>
<pre><code>## 载入需要的程辑包：Tax4Fun</code></pre>
<pre><code>## 载入需要的程辑包：Matrix</code></pre>
<pre><code>## 载入需要的程辑包：qiimer</code></pre>
<pre><code>## 载入需要的程辑包：biom</code></pre>
<pre><code>## Warning in write.table(otu_file, file = output, append = TRUE, quote = FALSE, :
## appending column names to file</code></pre>
<pre><code>## The KO abundance result is stored in object$tax4fun_KO ...</code></pre>
<pre><code>## The pathway abundance result is stored in object$tax4fun_path ...</code></pre>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="extended-class.html#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return two files: t1$tax4fun_KO: KO file; t1$tax4fun_path: pathway file.</span></span>
<span id="cb161-2"><a href="extended-class.html#cb161-2" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$tax4fun_KO$Tax4FunProfile[1:5, 1:2]</span></span></code></pre></div>
<table>
<colgroup>
<col width="12%" />
<col width="43%" />
<col width="43%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">K00001; alcohol dehydrogenase
[EC:1.1.1.1]</th>
<th align="center">K00002; alcohol dehydrogenase
(NADP+) [EC:1.1.1.2]</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">0.0004823</td>
<td align="center">5.942e-06</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">0.0005266</td>
<td align="center">4.017e-06</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">0.0005054</td>
<td align="center">6.168e-06</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">0.0005109</td>
<td align="center">5.888e-06</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">0.0005083</td>
<td align="center">5.547e-06</td>
</tr>
</tbody>
</table>
<p>Now, we use pathway file to analyze the abundance of pathway.</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="extended-class.html#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="co"># must transpose to taxa row, sample column</span></span>
<span id="cb162-2"><a href="extended-class.html#cb162-2" aria-hidden="true" tabindex="-1"></a>pathway_file <span class="ot">&lt;-</span> t1<span class="sc">$</span>tax4fun_path<span class="sc">$</span>Tax4FunProfile <span class="sc">%&gt;%</span> t <span class="sc">%&gt;%</span> as.data.frame</span>
<span id="cb162-3"><a href="extended-class.html#cb162-3" aria-hidden="true" tabindex="-1"></a><span class="co"># filter rownames, only keep ko+number</span></span>
<span id="cb162-4"><a href="extended-class.html#cb162-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(pathway_file) <span class="sc">%&lt;&gt;%</span> <span class="fu">gsub</span>(<span class="st">&quot;(^.*);</span><span class="sc">\\</span><span class="st">s.*&quot;</span>, <span class="st">&quot;</span><span class="sc">\\</span><span class="st">1&quot;</span>, .)</span>
<span id="cb162-5"><a href="extended-class.html#cb162-5" aria-hidden="true" tabindex="-1"></a><span class="co"># load the pathway hierarchical metadata</span></span>
<span id="cb162-6"><a href="extended-class.html#cb162-6" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(ko_map)</span>
<span id="cb162-7"><a href="extended-class.html#cb162-7" aria-hidden="true" tabindex="-1"></a><span class="co"># create a microtable object, familiar?</span></span>
<span id="cb162-8"><a href="extended-class.html#cb162-8" aria-hidden="true" tabindex="-1"></a>func1 <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">otu_table =</span> pathway_file, <span class="at">tax_table =</span> ko_map, <span class="at">sample_table =</span> t1<span class="sc">$</span>sample_table)</span>
<span id="cb162-9"><a href="extended-class.html#cb162-9" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(func1)</span></code></pre></div>
<pre><code>## microtable class:
## sample_table have 90 rows and 4 columns
## otu_table have 284 rows and 90 columns
## tax_table have 341 rows and 3 columns</code></pre>
<p>Now, we need to trim data and calculate abundance.</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="extended-class.html#cb164-1" aria-hidden="true" tabindex="-1"></a>func1<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb164-2"><a href="extended-class.html#cb164-2" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate abundance automatically at three levels: level_1, level_2, level_3</span></span>
<span id="cb164-3"><a href="extended-class.html#cb164-3" aria-hidden="true" tabindex="-1"></a>func1<span class="sc">$</span><span class="fu">cal_abund</span>()</span></code></pre></div>
<pre><code>## The result is stored in object$taxa_abund ...</code></pre>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="extended-class.html#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(func1)</span></code></pre></div>
<pre><code>## microtable class:
## sample_table have 90 rows and 4 columns
## otu_table have 284 rows and 90 columns
## tax_table have 284 rows and 3 columns
## Taxa abundance: calculated for level_1,level_2,level_3</code></pre>
<p>Then, we can plot the abundance.</p>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="extended-class.html#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="co"># bar plot at level_1</span></span>
<span id="cb168-2"><a href="extended-class.html#cb168-2" aria-hidden="true" tabindex="-1"></a>func2 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(func1, <span class="at">taxrank =</span> <span class="st">&quot;level_1&quot;</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb168-3"><a href="extended-class.html#cb168-3" aria-hidden="true" tabindex="-1"></a>func2<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/plot_bar_tax4fun1.png" width="600px" style="display: block; margin: auto;" /></p>
<p>We can also do something else. For example, we can use lefse to test the differences of the abundances and find the important enriched pathways across groups.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="extended-class.html#cb169-1" aria-hidden="true" tabindex="-1"></a>func2 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> func1, <span class="at">method =</span> <span class="st">&quot;lefse&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.05</span>, <span class="at">lefse_subgroup =</span> <span class="cn">NULL</span>)</span>
<span id="cb169-2"><a href="extended-class.html#cb169-2" aria-hidden="true" tabindex="-1"></a>func2<span class="sc">$</span><span class="fu">plot_lefse_bar</span>(<span class="at">LDA_score =</span> <span class="dv">3</span>, <span class="at">width =</span> <span class="fl">0.8</span>)</span></code></pre></div>
<p><img src="Images/plot_lefse_bar_tax4fun.png" width="600px" style="display: block; margin: auto;" /></p>
<p>Tax4Fun2 <span class="citation">(<a href="#ref-Wemheuer_Tax4Fun2_2020" role="doc-biblioref">Wemheuer et al. 2020</a>)</span> is another R package for the prediction of functional profiles of prokaryotic communities from 16S rRNA gene sequences.
It can provides two indexes for the evaluation of functional gene redundancies.
We provide two functions cal_tax4fun2() and cal_tax4fun2_FRI() in trans_func class for the Tax4Fun2 analysis.<br />
If you want to use Tax4Fun2 method, you must add the fasta file to the microtable object.</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="extended-class.html#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a microtable object with the fasta file</span></span>
<span id="cb170-2"><a href="extended-class.html#cb170-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(sample_info_16S)</span>
<span id="cb170-3"><a href="extended-class.html#cb170-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(otu_table_16S)</span>
<span id="cb170-4"><a href="extended-class.html#cb170-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(taxonomy_table_16S)</span>
<span id="cb170-5"><a href="extended-class.html#cb170-5" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(rep_fasta_16S)</span>
<span id="cb170-6"><a href="extended-class.html#cb170-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-7"><a href="extended-class.html#cb170-7" aria-hidden="true" tabindex="-1"></a>use_dataset <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">sample_table =</span> sample_info_16S, <span class="at">otu_table =</span> otu_table_16S, <span class="at">tax_table =</span> taxonomy_table_16S, <span class="at">rep_fasta =</span> rep_fasta_16S)</span>
<span id="cb170-8"><a href="extended-class.html#cb170-8" aria-hidden="true" tabindex="-1"></a>use_dataset<span class="sc">$</span><span class="fu">filter_pollution</span>(<span class="at">taxa =</span> <span class="fu">c</span>(<span class="st">&quot;mitochondria&quot;</span>, <span class="st">&quot;chloroplast&quot;</span>))</span>
<span id="cb170-9"><a href="extended-class.html#cb170-9" aria-hidden="true" tabindex="-1"></a>use_dataset<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb170-10"><a href="extended-class.html#cb170-10" aria-hidden="true" tabindex="-1"></a>use_dataset</span>
<span id="cb170-11"><a href="extended-class.html#cb170-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-12"><a href="extended-class.html#cb170-12" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(use_dataset)</span>
<span id="cb170-13"><a href="extended-class.html#cb170-13" aria-hidden="true" tabindex="-1"></a><span class="co"># create a directory for result and log files</span></span>
<span id="cb170-14"><a href="extended-class.html#cb170-14" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;test_prediction&quot;</span>)</span>
<span id="cb170-15"><a href="extended-class.html#cb170-15" aria-hidden="true" tabindex="-1"></a><span class="co"># see https://github.com/ChiLiubio/microeco for downloading ncbi-blast and Ref99NR/Ref100NR</span></span>
<span id="cb170-16"><a href="extended-class.html#cb170-16" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_tax4fun2</span>(<span class="at">blast_tool_path =</span> <span class="st">&quot;ncbi-blast-2.11.0+/bin&quot;</span>, <span class="at">path_to_reference_data =</span> <span class="st">&quot;test_ReferenceData_v2&quot;</span>,</span>
<span id="cb170-17"><a href="extended-class.html#cb170-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">path_to_temp_folder =</span> <span class="st">&quot;test_prediction&quot;</span>, <span class="at">database_mode =</span> <span class="st">&quot;Ref99NR&quot;</span>)</span>
<span id="cb170-18"><a href="extended-class.html#cb170-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-19"><a href="extended-class.html#cb170-19" aria-hidden="true" tabindex="-1"></a><span class="co"># functional gene redundancies</span></span>
<span id="cb170-20"><a href="extended-class.html#cb170-20" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_tax4fun2_FRI</span>()</span></code></pre></div>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Anderson_Austral_2001" class="csl-entry">
Anderson, M. J. 2001. <span>“A New Method for Non-Parametric Multivariate Analysis of Variance.”</span> Journal Article. <em>Austral Ecology</em>, no. 26: 32–46.
</div>
<div id="ref-Anderson_Navigating_2011" class="csl-entry">
Anderson, M. J., T. O. Crist, J. M. Chase, M. Vellend, B. D. Inouye, A. L. Freestone, N. J. Sanders, et al. 2011. <span>“Navigating the Multiple Meanings of Beta Diversity: A Roadmap for the Practicing Ecologist.”</span> Journal Article. <em>Ecology Letters</em> 14 (1): 19–28. <a href="https://doi.org/10.1111/j.1461-0248.2010.01552.x">https://doi.org/10.1111/j.1461-0248.2010.01552.x</a>.
</div>
<div id="ref-Aßhauer_Tax4Fun_2015" class="csl-entry">
Aßhauer, Kathrin P, Bernd Wemheuer, Rolf Daniel, and Peter Meinicke. 2015. <span>“Tax4Fun: Predicting Functional Profiles from Metagenomic 16s rRNA Data.”</span> Journal Article. <em>Bioinformatics</em> 31 (17): 2882–84.
</div>
<div id="ref-Beck_Machine_2014" class="csl-entry">
Beck, Daniel, and James A Foster. 2014. <span>“Machine Learning Techniques Accurately Classify Microbial Communities by Bacterial Vaginosis Characteristics.”</span> Journal Article. <em>PloS One</em> 9 (2): e87830.
</div>
<div id="ref-Chase_null_2011" class="csl-entry">
Chase, Jonathan M, Nathan JB Kraft, Kevin G Smith, Mark Vellend, and Brian D Inouye. 2011. <span>“Using Null Models to Disentangle Variation in Community Dissimilarity from Variation in α-Diversity.”</span> Journal Article. <em>Ecosphere</em> 2 (2): art24. <a href="https://doi.org/10.1890/ES10-00117.1">https://doi.org/10.1890/ES10-00117.1</a>.
</div>
<div id="ref-Coyte_Theecology_2015" class="csl-entry">
Coyte, Katharine Z, Jonas Schluter, and Kevin R Foster. 2015. <span>“The Ecology of the Microbiome: Networks, Competition, and Stability.”</span> Journal Article. <em>Science</em> 350 (6261): 663–66.
</div>
<div id="ref-Deng_Molecular_2012" class="csl-entry">
Deng, Ye, Yi-Huei Jiang, Yunfeng Yang, Zhili He, Feng Luo, and Jizhong Zhou. 2012. <span>“Molecular Ecological Network Analyses.”</span> Journal Article. <em>BMC Bioinformatics</em> 13 (1): 113.
</div>
<div id="ref-Faust_Microbial_2012" class="csl-entry">
Faust, Karoline, and Jeroen Raes. 2012. <span>“Microbial Interactions: From Networks to Models.”</span> Journal Article. <em>Nature Reviews Microbiology</em> 10 (8): 538–50.
</div>
<div id="ref-Foster_Metacoder_2017" class="csl-entry">
Foster, Zachary, Thomas J Sharpton, and Niklaus J Grunwald. 2017. <span>“Metacoder: An r Package for Visualization and Manipulation of Community Taxonomic Diversity Data.”</span> Journal Article. <em>PLOS Computational Biology</em> 13 (2): e1005404.
</div>
<div id="ref-Picante_Kembel_2010" class="csl-entry">
Kembel, S. W., P. D. Cowan, M. R. Helmus, W. K. Cornwell, H. Morlon, D. D. Ackerly, S. P. Blomberg, and C. O. Webb. 2010. <span>“Picante: <span>R</span> Tools for Integrating Phylogenies and Ecology.”</span> <em>Bioinformatics</em> 26: 1463–64.
</div>
<div id="ref-Kurtz_Sparse_2015" class="csl-entry">
Kurtz, Z. D., C. L. Muller, E. R. Miraldi, D. R. Littman, M. J. Blaser, and R. A. Bonneau. 2015. <span>“Sparse and Compositionally Robust Inference of Microbial Ecological Networks.”</span> Journal Article. <em>PLoS Comput Biol</em> 11 (5): e1004226. <a href="https://doi.org/10.1371/journal.pcbi.1004226">https://doi.org/10.1371/journal.pcbi.1004226</a>.
</div>
<div id="ref-Langille_Predictive_2013" class="csl-entry">
Langille, M. G., J. Zaneveld, J. G. Caporaso, D. McDonald, D. Knights, J. A. Reyes, J. C. Clemente, et al. 2013. <span>“Predictive Functional Profiling of Microbial Communities Using 16s rRNA Marker Gene Sequences.”</span> Journal Article. <em>Nature Biotechnology</em> 31 (9): 814–21. <a href="https://doi.org/10.1038/nbt.2676">https://doi.org/10.1038/nbt.2676</a>.
</div>
<div id="ref-Lim_Large_2020" class="csl-entry">
Lim, Roktaek, Josephine Jill T. Cabatbat, Thomas L. P. Martin, Haneul Kim, Seunghyeon Kim, Jaeyun Sung, Cheol-Min Ghim, and Pan-Jun Kim. 2020. <span>“Large-Scale Metabolic Interaction Network of the Mouse and Human Gut Microbiota.”</span> Journal Article. <em>Scientific Data</em> 7 (1). <a href="https://doi.org/10.1038/s41597-020-0516-5">https://doi.org/10.1038/s41597-020-0516-5</a>.
</div>
<div id="ref-Liu_Long_term_2017" class="csl-entry">
Liu, Chi, Minjie Yao, James C Stegen, Junpeng Rui, Jiabao Li, and Xiangzhen Li. 2017. <span>“Long-Term Nitrogen Addition Affects the Phylogenetic Turnover of Soil Microbial Community Responding to Moisture Pulse.”</span> Journal Article. <em>Scientific Reports</em> 7 (1): 17492.
</div>
<div id="ref-Louca_Decoupling_2016" class="csl-entry">
Louca, S, L. W. Parfrey, and M Doebeli. 2016. <span>“Decoupling Function and Taxonomy in the Global Ocean Microbiome.”</span> Journal Article. <em>Science</em> 353 (6305): 1272.
</div>
<div id="ref-Louca_High_2016" class="csl-entry">
Louca, Stilianos, Saulo MS Jacques, Aliny PF Pires, Juliana S Leal, Diane S Srivastava, Laura Wegener Parfrey, Vinicius F Farjalla, and Michael Doebeli. 2016. <span>“High Taxonomic Variability Despite Stable Functional Structure Across Microbial Communities.”</span> Journal Article. <em>Nature Ecology &amp; Evolution</em> 1: 0015.
</div>
<div id="ref-Mendes_Deciphering_2011" class="csl-entry">
Mendes, R, M Kruijt, Bruijn I De, E Dekkers, der Voort M Van, J. H. Schneider, Y. M. Piceno, T. Z. Desantis, G. L. Andersen, and P. A. Bakker. 2011. <span>“Deciphering the Rhizosphere Microbiome for Disease-Suppressive Bacteria.”</span> Journal Article. <em>Science</em> 332 (6033): 1097–1100.
</div>
<div id="ref-Nguyen_FUNGuild_2016" class="csl-entry">
Nguyen, Nhu H., Zewei Song, Scott T. Bates, Sara Branco, Leho Tedersoo, Jon Menke, Jonathan S. Schilling, and Peter G. Kennedy. 2016. <span>“FUNGuild: An Open Annotation Tool for Parsing Fungal Community Datasets by Ecological Guild.”</span> Journal Article. <em>Fungal Ecology</em> 20 (1): 241–48.
</div>
<div id="ref-Paulson_Differential_2013" class="csl-entry">
Paulson, Joseph N, O Colin Stine, Héctor Corrada Bravo, and Mihai Pop. 2013. <span>“Differential Abundance Analysis for Microbial Marker-Gene Surveys.”</span> Journal Article. <em>Nature Methods</em> 10 (12): 1200–1202.
</div>
<div id="ref-Polme_FungalTraits_2020" class="csl-entry">
Põlme, Sergei, Kessy Abarenkov, R. Henrik Nilsson, Björn D. Lindahl, Karina Engelbrecht Clemmensen, Havard Kauserud, Nhu Nguyen, et al. 2020. <span>“FungalTraits: A User-Friendly Traits Database of Fungi and Fungus-Like Stramenopiles.”</span> Journal Article. <em>Fungal Diversity</em> 105 (1): 1–16. <a href="https://doi.org/10.1007/s13225-020-00466-2">https://doi.org/10.1007/s13225-020-00466-2</a>.
</div>
<div id="ref-Segata_Metagenomic_2011" class="csl-entry">
Segata, Nicola, Jacques Izard, Levi Waldron, Dirk Gevers, Larisa Miropolsky, Wendy S Garrett, and Curtis Huttenhower. 2011. <span>“Metagenomic Biomarker Discovery and Explanation.”</span> Journal Article. <em>Genome Biology</em> 12 (6): R60.
</div>
<div id="ref-Stegen_Quantifying_2013" class="csl-entry">
Stegen, James C, Xueju Lin, Jim K Fredrickson, Xingyuan Chen, David W Kennedy, Christopher J Murray, Mark L Rockhold, and Allan Konopka. 2013. <span>“Quantifying Community Assembly Processes and Identifying Features That Impose Them.”</span> Journal Article. <em>The ISME Journal</em> 7 (11): 2069–79.
</div>
<div id="ref-Tackmann_Rapid_2019" class="csl-entry">
Tackmann, J., J. F. Matias Rodrigues, and C. von Mering. 2019. <span>“Rapid Inference of Direct Interactions in Large-Scale Ecological Networks from Heterogeneous Microbial Sequencing Data.”</span> Journal Article. <em>Cell Systems</em> 9 (3): 286–296 e8. <a href="https://doi.org/10.1016/j.cels.2019.08.002">https://doi.org/10.1016/j.cels.2019.08.002</a>.
</div>
<div id="ref-Webb_Phylogenies_2002" class="csl-entry">
Webb, Campbell O, David D Ackerly, Mark A McPeek, and Michael J Donoghue. 2002. <span>“Phylogenies and Community Ecology.”</span> Journal Article. <em>Annu. Rev. Ecol. Syst.</em> 33: 475–505. <a href="https://doi.org/10.1146/annurev.ecolsys.33.010802.150448">https://doi.org/10.1146/annurev.ecolsys.33.010802.150448</a>.
</div>
<div id="ref-Wemheuer_Tax4Fun2_2020" class="csl-entry">
Wemheuer, Franziska, Jessica A. Taylor, Rolf Daniel, Emma Johnston, Peter Meinicke, Torsten Thomas, and Bernd Wemheuer. 2020. <span>“Tax4Fun2: Prediction of Habitat-Specific Functional Profiles and Functional Redundancy Based on 16s rRNA Gene Sequences.”</span> Journal Article. <em>Environmental Microbiome</em> 15 (1). <a href="https://doi.org/10.1186/s40793-020-00358-7">https://doi.org/10.1186/s40793-020-00358-7</a>.
</div>
<div id="ref-White_Statistical_2009" class="csl-entry">
White, JR, N Nagarajan, and M Pop. 2009. <span>“Statistical Methods for Detecting Differentially Abundant Features in Clinical Metagenomic Samples.”</span> Journal Article. <em>PLoS Computational Biology</em> 5 (4): e1000352.
</div>
<div id="ref-Yatsunenko_Human_2012" class="csl-entry">
Yatsunenko, Tanya, Federico E Rey, Mark J Manary, Indi Trehan, Maria Gloria Dominguez-Bello, Monica Contreras, Magda Magris, Glida Hidalgo, Robert N Baldassano, and Andrey P Anokhin. 2012. <span>“Human Gut Microbiome Viewed Across Age and Geography.”</span> Journal Article. <em>Nature</em> 486 (7402): 222–27.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="basic-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="other-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/03-extended_class.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
