<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 file2meco package | Tutorial for R microeco package (v1.6.0)</title>
  <meta name="description" content="The tutorial for R microeco (v1.6.0), file2meco (v0.7.1), meconetcomp (v0.5.0) and mecoturn packages" />
  <meta name="generator" content="bookdown 0.38 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 file2meco package | Tutorial for R microeco package (v1.6.0)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The tutorial for R microeco (v1.6.0), file2meco (v0.7.1), meconetcomp (v0.5.0) and mecoturn packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 file2meco package | Tutorial for R microeco package (v1.6.0)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco (v1.6.0), file2meco (v0.7.1), meconetcomp (v0.5.0) and mecoturn packages" />
  



<meta name="date" content="2024-04-03" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="explainable-class.html"/>
<link rel="next" href="other-dataset.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#rtools"><i class="fa fa-check"></i><b>2.4</b> RTools</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.5</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#description"><i class="fa fa-check"></i><b>2.5.1</b> Description</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#cran-packages"><i class="fa fa-check"></i><b>2.5.2</b> CRAN packages</a></li>
<li class="chapter" data-level="2.5.3" data-path="intro.html"><a href="intro.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor packages</a></li>
<li class="chapter" data-level="2.5.4" data-path="intro.html"><a href="intro.html#github-packages"><i class="fa fa-check"></i><b>2.5.4</b> Github packages</a></li>
<li class="chapter" data-level="2.5.5" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.5.5</b> Gephi</a></li>
<li class="chapter" data-level="2.5.6" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.5.6</b> Tax4Fun</a></li>
<li class="chapter" data-level="2.5.7" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.5.7</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#plot"><i class="fa fa-check"></i><b>2.6</b> Plot</a></li>
<li class="chapter" data-level="2.7" data-path="intro.html"><a href="intro.html#rstudio"><i class="fa fa-check"></i><b>2.7</b> Rstudio</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#prepare-the-example-data"><i class="fa fa-check"></i><b>3.1.1</b> Prepare the example data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#how-to-read-your-files-to-microtable-object"><i class="fa fa-check"></i><b>3.1.2</b> How to read your files to microtable object?</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#functions-in-microtable-class"><i class="fa fa-check"></i><b>3.1.3</b> Functions in microtable class</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#merge-taxa-or-samples"><i class="fa fa-check"></i><b>3.1.4</b> merge taxa or samples</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-class.html"><a href="basic-class.html#subset-of-samples"><i class="fa fa-check"></i><b>3.1.5</b> subset of samples</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-class.html"><a href="basic-class.html#subset-of-taxa"><i class="fa fa-check"></i><b>3.1.6</b> subset of taxa</a></li>
<li class="chapter" data-level="3.1.7" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.7</b> Other examples</a></li>
<li class="chapter" data-level="3.1.8" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.8</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-2"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-4"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#more-options"><i class="fa fa-check"></i><b>6.1.2</b> More options</a></li>
<li class="chapter" data-level="6.1.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.3</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#network-comparison"><i class="fa fa-check"></i><b>6.2.2</b> Network comparison</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.3</b> Key points</a></li>
<li class="chapter" data-level="6.2.4" data-path="model-based-class.html"><a href="model-based-class.html#other-functions"><i class="fa fa-check"></i><b>6.2.4</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies</a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples</a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points</a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-7"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME</a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2</a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann"><i class="fa fa-check"></i><b>8.3</b> HUMAnN</a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn</a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2braken"><i class="fa fa-check"></i><b>8.5</b> Kraken2/Braken</a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdbpcycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB/PCycDB</a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq</a></li>
<li class="chapter" data-level="8.8" data-path="file2meco-package.html"><a href="file2meco-package.html#picrust2"><i class="fa fa-check"></i><b>8.8</b> PICRUSt2</a></li>
<li class="chapter" data-level="8.9" data-path="file2meco-package.html"><a href="file2meco-package.html#viromescan"><i class="fa fa-check"></i><b>8.9</b> ViromeScan</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset</a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data</a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> <code>clone</code> function</a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> <code>save</code> function</a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#github-issues"><i class="fa fa-check"></i><b>10.3</b> Github Issues</a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.4</b> group order</a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.5</b> add layers to plot</a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#colors-for-more-groups"><i class="fa fa-check"></i><b>10.6</b> colors for more groups</a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.7</b> change object</a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html"><i class="fa fa-check"></i><b>11</b> meconetcomp package</a>
<ul>
<li class="chapter" data-level="11.1" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-modularity-for-all-networks"><i class="fa fa-check"></i><b>11.1</b> Network modularity for all networks</a></li>
<li class="chapter" data-level="11.2" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-topological-attributes-for-all-networks"><i class="fa fa-check"></i><b>11.2</b> Network topological attributes for all networks</a></li>
<li class="chapter" data-level="11.3" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#node-and-edge-properties-extraction-for-all-networks"><i class="fa fa-check"></i><b>11.3</b> Node and edge properties extraction for all networks</a></li>
<li class="chapter" data-level="11.4" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-nodes-across-networks"><i class="fa fa-check"></i><b>11.4</b> Compare nodes across networks</a></li>
<li class="chapter" data-level="11.5" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-edges-across-networks"><i class="fa fa-check"></i><b>11.5</b> Compare edges across networks</a></li>
<li class="chapter" data-level="11.6" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#extract-overlapped-edges-of-networks-to-a-new-network"><i class="fa fa-check"></i><b>11.6</b> Extract overlapped edges of networks to a new network</a></li>
<li class="chapter" data-level="11.7" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-phylogenetic-distances-of-paired-nodes-in-edges"><i class="fa fa-check"></i><b>11.7</b> Compare phylogenetic distances of paired nodes in edges</a></li>
<li class="chapter" data-level="11.8" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-node-sources-of-edges-across-networks"><i class="fa fa-check"></i><b>11.8</b> Compare node sources of edges across networks</a></li>
<li class="chapter" data-level="11.9" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-topological-properties-of-sub-networks"><i class="fa fa-check"></i><b>11.9</b> Compare topological properties of sub-networks</a></li>
<li class="chapter" data-level="11.10" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#robustness-of-network"><i class="fa fa-check"></i><b>11.10</b> Robustness of network</a></li>
<li class="chapter" data-level="11.11" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#vulnerability-of-nodes"><i class="fa fa-check"></i><b>11.11</b> Vulnerability of nodes</a></li>
<li class="chapter" data-level="11.12" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#cohesion"><i class="fa fa-check"></i><b>11.12</b> Cohesion</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>12</b> mecodev package</a>
<ul>
<li class="chapter" data-level="12.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>12.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="12.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>12.2</b> trans_ts</a></li>
<li class="chapter" data-level="12.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>12.3</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mecoturn-package.html"><a href="mecoturn-package.html"><i class="fa fa-check"></i><b>13</b> mecoturn package</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mecoturn-package.html"><a href="mecoturn-package.html#betaturn-class"><i class="fa fa-check"></i><b>13.1</b> betaturn class</a></li>
<li class="chapter" data-level="13.2" data-path="mecoturn-package.html"><a href="mecoturn-package.html#taxaturn-class"><i class="fa fa-check"></i><b>13.2</b> taxaturn class</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>14</b> Other examples</a>
<ul>
<li class="chapter" data-level="14.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>14.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="14.2" data-path="other-examples-1.html"><a href="other-examples-1.html#tidy_taxonomy-function"><i class="fa fa-check"></i><b>14.2</b> tidy_taxonomy function</a></li>
<li class="chapter" data-level="14.3" data-path="other-examples-1.html"><a href="other-examples-1.html#show-the-abundance-of-unknown-taxa"><i class="fa fa-check"></i><b>14.3</b> Show the abundance of unknown taxa</a></li>
<li class="chapter" data-level="14.4" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>14.4</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="14.5" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>14.5</b> The use of phylogenetic tree</a></li>
<li class="chapter" data-level="14.6" data-path="other-examples-1.html"><a href="other-examples-1.html#trans_norm"><i class="fa fa-check"></i><b>14.6</b> trans_norm</a></li>
<li class="chapter" data-level="14.7" data-path="other-examples-1.html"><a href="other-examples-1.html#mantel-test-correlation-heatmap"><i class="fa fa-check"></i><b>14.7</b> Mantel test + Correlation heatmap</a></li>
<li class="chapter" data-level="14.8" data-path="other-examples-1.html"><a href="other-examples-1.html#pcoa-boxplot"><i class="fa fa-check"></i><b>14.8</b> PCoA + boxplot</a></li>
<li class="chapter" data-level="14.9" data-path="other-examples-1.html"><a href="other-examples-1.html#faprotax-differential-test"><i class="fa fa-check"></i><b>14.9</b> FAPROTAX + differential test</a></li>
<li class="chapter" data-level="14.10" data-path="other-examples-1.html"><a href="other-examples-1.html#rda-results-output"><i class="fa fa-check"></i><b>14.10</b> RDA results output</a></li>
<li class="chapter" data-level="14.11" data-path="other-examples-1.html"><a href="other-examples-1.html#betanribetanti-for-each-group"><i class="fa fa-check"></i><b>14.11</b> betaNRI/betaNTI for each group</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v1.6.0)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="file2meco-package" class="section level1 hasAnchor" number="8">
<h1><span class="header-section-number">Chapter 8</span> file2meco package<a href="file2meco-package.html#file2meco-package" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In the microtable class part, we showed the basic way about how to create microtable object with the example data.
Actually, constructing the microtable object from other tools/platforms (e.g., QIIME, QIIME2, HUMAnN, Kraken2 and phyloseq)
can be easily achieved with the package file2meco (<a href="https://github.com/ChiLiubio/file2meco" class="uri">https://github.com/ChiLiubio/file2meco</a>).
The idea of creating file2meco package comes from a study involved in complex metagenomic analysis <span class="citation">(<a href="#ref-Liu_Microbial_2022" role="doc-biblioref">C. Liu et al. 2022</a>)</span>.
Note that the <code>sample_table</code> parameter in each function of file2meco package supports various metadata input format, including<br />
1) comma seperated file with the suffix .csv or tab seperated file with the suffix .tsv or .txt;<br />
2) Excel file with the suffix .xlsx or .xls;<br />
3) data.frame object in R session.</p>
<div class="sourceCode" id="cb289"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb289-1"><a href="file2meco-package.html#cb289-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install file2meco package (&gt;= 0.7.1)</span></span>
<span id="cb289-2"><a href="file2meco-package.html#cb289-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb289-3"><a href="file2meco-package.html#cb289-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;file2meco&quot;</span>)) <span class="fu">install.packages</span>(<span class="st">&quot;file2meco&quot;</span>, <span class="at">repos =</span> BiocManager<span class="sc">::</span><span class="fu">repositories</span>())</span></code></pre></div>
<p>▲ Trouble shooting:<br />
• The files in the following examples all come from the package file2meco and are found by the function <code>system.file</code> automatically irrespective of Operating System.
When the user imports a file, please donot use this function.
The first thing should be to make sure that R can find your input file.
The user should either provide a full path to the import function or only the file name after putting the file into the R working directory.
For the latter one, if the user does not know where the working directory is,
please use the function <code>getwd</code> to find the working directory or directly create a new project in the target directory with RStudio [File –&gt; New Project].</p>
<div id="qiime" class="section level2 hasAnchor" number="8.1">
<h2><span class="header-section-number">8.1</span> QIIME<a href="file2meco-package.html#qiime" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The qiime1meco() function can be used to construct the microtable object using the raw OTU file from QIIME 1 <span class="citation">(<a href="#ref-Caporaso_QIIME_2010" role="doc-biblioref">Caporaso et al. 2010</a>)</span>.</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="file2meco-package.html#cb290-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb290-2"><a href="file2meco-package.html#cb290-2" aria-hidden="true" tabindex="-1"></a><span class="co"># see the help document</span></span>
<span id="cb290-3"><a href="file2meco-package.html#cb290-3" aria-hidden="true" tabindex="-1"></a>?qiime1meco</span>
<span id="cb290-4"><a href="file2meco-package.html#cb290-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Let&#39;s run the examples</span></span>
<span id="cb290-5"><a href="file2meco-package.html#cb290-5" aria-hidden="true" tabindex="-1"></a><span class="co"># use the raw data files stored inside the package</span></span>
<span id="cb290-6"><a href="file2meco-package.html#cb290-6" aria-hidden="true" tabindex="-1"></a>otu_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;otu_table_raw.txt&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb290-7"><a href="file2meco-package.html#cb290-7" aria-hidden="true" tabindex="-1"></a><span class="co"># csv file of metadata</span></span>
<span id="cb290-8"><a href="file2meco-package.html#cb290-8" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sample_info.csv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb290-9"><a href="file2meco-package.html#cb290-9" aria-hidden="true" tabindex="-1"></a>phylo_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;rep_phylo.tre&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb290-10"><a href="file2meco-package.html#cb290-10" aria-hidden="true" tabindex="-1"></a><span class="co"># if you want to use Tax4Fun2 approach, you need read the representative sequences and add it to the microtable object.</span></span>
<span id="cb290-11"><a href="file2meco-package.html#cb290-11" aria-hidden="true" tabindex="-1"></a>rep_fasta_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;rep.fna&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb290-12"><a href="file2meco-package.html#cb290-12" aria-hidden="true" tabindex="-1"></a><span class="co"># contruct microtable object</span></span>
<span id="cb290-13"><a href="file2meco-package.html#cb290-13" aria-hidden="true" tabindex="-1"></a><span class="fu">qiime1meco</span>(otu_file_path)</span>
<span id="cb290-14"><a href="file2meco-package.html#cb290-14" aria-hidden="true" tabindex="-1"></a><span class="fu">qiime1meco</span>(otu_file_path, <span class="at">sample_table =</span> sample_file_path)</span>
<span id="cb290-15"><a href="file2meco-package.html#cb290-15" aria-hidden="true" tabindex="-1"></a><span class="fu">qiime1meco</span>(otu_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">phylo_tree =</span> phylo_file_path)</span>
<span id="cb290-16"><a href="file2meco-package.html#cb290-16" aria-hidden="true" tabindex="-1"></a><span class="fu">qiime1meco</span>(otu_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">phylo_tree =</span> phylo_file_path, <span class="at">rep_fasta =</span> rep_fasta_path)</span></code></pre></div>
</div>
<div id="qiime2" class="section level2 hasAnchor" number="8.2">
<h2><span class="header-section-number">8.2</span> QIIME2<a href="file2meco-package.html#qiime2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The qiime2meco() function is designed to create the microtable object using files from QIIME2 <span class="citation">(<a href="#ref-Bolyen_Reproducible_2019" role="doc-biblioref">Bolyen et al. 2019</a>)</span>.
The example data is the ASV (amplicon sequence variant) abundance table based on DADA2 <span class="citation">(<a href="#ref-Callahan_DADA2_2016" role="doc-biblioref">Callahan et al. 2016</a>)</span>.</p>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="file2meco-package.html#cb291-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb291-2"><a href="file2meco-package.html#cb291-2" aria-hidden="true" tabindex="-1"></a>?qiime2meco</span>
<span id="cb291-3"><a href="file2meco-package.html#cb291-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use data files inside the package which were downloaded from (https://docs.qiime2.org/2022.2/tutorials/pd-mice/).</span></span>
<span id="cb291-4"><a href="file2meco-package.html#cb291-4" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;dada2_table.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb291-5"><a href="file2meco-package.html#cb291-5" aria-hidden="true" tabindex="-1"></a><span class="co"># tsv file of metadata</span></span>
<span id="cb291-6"><a href="file2meco-package.html#cb291-6" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sample-metadata.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb291-7"><a href="file2meco-package.html#cb291-7" aria-hidden="true" tabindex="-1"></a>taxonomy_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;taxonomy.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb291-8"><a href="file2meco-package.html#cb291-8" aria-hidden="true" tabindex="-1"></a><span class="co"># construct microtable object</span></span>
<span id="cb291-9"><a href="file2meco-package.html#cb291-9" aria-hidden="true" tabindex="-1"></a><span class="fu">qiime2meco</span>(abund_file_path)</span>
<span id="cb291-10"><a href="file2meco-package.html#cb291-10" aria-hidden="true" tabindex="-1"></a><span class="fu">qiime2meco</span>(abund_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">taxonomy_table =</span> taxonomy_file_path)</span>
<span id="cb291-11"><a href="file2meco-package.html#cb291-11" aria-hidden="true" tabindex="-1"></a><span class="co"># add phylogenetic tree and fasta for more demonstrations</span></span>
<span id="cb291-12"><a href="file2meco-package.html#cb291-12" aria-hidden="true" tabindex="-1"></a><span class="co"># please download tree from https://docs.qiime2.org/2022.2/data/tutorials/pd-mice/tree.qza</span></span>
<span id="cb291-13"><a href="file2meco-package.html#cb291-13" aria-hidden="true" tabindex="-1"></a><span class="co"># the file name is &#39;tree.qza&#39;; put it into the R working directory</span></span>
<span id="cb291-14"><a href="file2meco-package.html#cb291-14" aria-hidden="true" tabindex="-1"></a>tree_data <span class="ot">&lt;-</span> <span class="st">&quot;tree.qza&quot;</span></span>
<span id="cb291-15"><a href="file2meco-package.html#cb291-15" aria-hidden="true" tabindex="-1"></a><span class="co"># please download fasta from https://docs.qiime2.org/2022.2/data/tutorials/pd-mice/dada2_rep_set.qza</span></span>
<span id="cb291-16"><a href="file2meco-package.html#cb291-16" aria-hidden="true" tabindex="-1"></a><span class="co"># the file name is &#39;dada2_rep_set.qza&#39;; put it into the R working directory</span></span>
<span id="cb291-17"><a href="file2meco-package.html#cb291-17" aria-hidden="true" tabindex="-1"></a>rep_data <span class="ot">&lt;-</span> <span class="st">&quot;dada2_rep_set.qza&quot;</span></span>
<span id="cb291-18"><a href="file2meco-package.html#cb291-18" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> <span class="fu">qiime2meco</span>(abund_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">taxonomy_table =</span> taxonomy_file_path, <span class="at">phylo_tree =</span> tree_data, <span class="at">rep_fasta =</span> rep_data, <span class="at">auto_tidy =</span> <span class="cn">TRUE</span>)</span>
<span id="cb291-19"><a href="file2meco-package.html#cb291-19" aria-hidden="true" tabindex="-1"></a>test1</span></code></pre></div>
</div>
<div id="humann" class="section level2 hasAnchor" number="8.3">
<h2><span class="header-section-number">8.3</span> HUMAnN<a href="file2meco-package.html#humann" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Many methods in microeco package can be used not only for the traditional species abundance data, i.e. species-sample table,
but also for other data, such as metagenomic data.
HUMAnN <span class="citation">(<a href="#ref-Franzosa_Species_2018" role="doc-biblioref">Franzosa et al. 2018</a>)</span> is an excellent tool for functional profiling analysis of metagenomes and metatranscriptomes at species-level.
The humann2meco() function can be used to create the microtable object using metagenomic analysis files from HUMAnN3 (<a href="https://huttenhower.sph.harvard.edu/humann" class="uri">https://huttenhower.sph.harvard.edu/humann</a>).
Certainly, it can also be used for the whole community profile of metabolic pathways when needed.
Currently, it supports both the MetaCyc (<a href="https://metacyc.org/" class="uri">https://metacyc.org/</a>) and KEGG pathway abundance file input directly.</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="file2meco-package.html#cb292-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb292-2"><a href="file2meco-package.html#cb292-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb292-3"><a href="file2meco-package.html#cb292-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb292-4"><a href="file2meco-package.html#cb292-4" aria-hidden="true" tabindex="-1"></a>?humann2meco</span>
<span id="cb292-5"><a href="file2meco-package.html#cb292-5" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_metagenome_sample_info.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb292-6"><a href="file2meco-package.html#cb292-6" aria-hidden="true" tabindex="-1"></a>match_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_metagenome_match_table.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb292-7"><a href="file2meco-package.html#cb292-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb292-8"><a href="file2meco-package.html#cb292-8" aria-hidden="true" tabindex="-1"></a><span class="co"># MetaCyc pathway database based analysis</span></span>
<span id="cb292-9"><a href="file2meco-package.html#cb292-9" aria-hidden="true" tabindex="-1"></a><span class="co"># use the raw data files stored inside the package for MetaCyc pathway database based analysis</span></span>
<span id="cb292-10"><a href="file2meco-package.html#cb292-10" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_HUMAnN_MetaCyc_abund.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb292-11"><a href="file2meco-package.html#cb292-11" aria-hidden="true" tabindex="-1"></a><span class="co"># the default db is &quot;MetaCyc&quot;</span></span>
<span id="cb292-12"><a href="file2meco-package.html#cb292-12" aria-hidden="true" tabindex="-1"></a><span class="fu">humann2meco</span>(abund_file_path, <span class="at">db =</span> <span class="st">&quot;MetaCyc&quot;</span>)</span>
<span id="cb292-13"><a href="file2meco-package.html#cb292-13" aria-hidden="true" tabindex="-1"></a><span class="fu">humann2meco</span>(abund_file_path, <span class="at">db =</span> <span class="st">&quot;MetaCyc&quot;</span>, <span class="at">sample_table =</span> sample_file_path, <span class="at">match_table =</span> match_file_path)</span>
<span id="cb292-14"><a href="file2meco-package.html#cb292-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Let&#39;s try more interesting usages with microeco</span></span>
<span id="cb292-15"><a href="file2meco-package.html#cb292-15" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">humann2meco</span>(abund_file_path, <span class="at">db =</span> <span class="st">&quot;MetaCyc&quot;</span>, <span class="at">sample_table =</span> sample_file_path, <span class="at">match_table =</span> match_file_path)</span>
<span id="cb292-16"><a href="file2meco-package.html#cb292-16" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb292-17"><a href="file2meco-package.html#cb292-17" aria-hidden="true" tabindex="-1"></a><span class="co"># rel = FALSE sum original abundance instead of relative abundance</span></span>
<span id="cb292-18"><a href="file2meco-package.html#cb292-18" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">select_cols =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">rel =</span> <span class="cn">FALSE</span>)</span>
<span id="cb292-19"><a href="file2meco-package.html#cb292-19" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>taxa_abund<span class="sc">$</span>Superclass1 <span class="sc">%&lt;&gt;%</span> .[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;unclass&quot;</span>, <span class="fu">rownames</span>(.)), ]</span>
<span id="cb292-20"><a href="file2meco-package.html#cb292-20" aria-hidden="true" tabindex="-1"></a><span class="co"># use_percentage = FALSE disable percentage for relative abundance</span></span>
<span id="cb292-21"><a href="file2meco-package.html#cb292-21" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(test, <span class="at">taxrank =</span> <span class="st">&quot;Superclass1&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">use_percentage =</span> <span class="cn">FALSE</span>)</span>
<span id="cb292-22"><a href="file2meco-package.html#cb292-22" aria-hidden="true" tabindex="-1"></a><span class="co"># reassign ylab title instead of default &#39;Relative Abundance&#39;</span></span>
<span id="cb292-23"><a href="file2meco-package.html#cb292-23" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span>ylabname <span class="ot">&lt;-</span> <span class="st">&quot;Abundance (RPK)&quot;</span></span>
<span id="cb292-24"><a href="file2meco-package.html#cb292-24" aria-hidden="true" tabindex="-1"></a><span class="co"># bar_type = &quot;notfull&quot; show original abundance instead of normalized 0-1</span></span>
<span id="cb292-25"><a href="file2meco-package.html#cb292-25" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">bar_type =</span> <span class="st">&quot;notfull&quot;</span>)</span>
<span id="cb292-26"><a href="file2meco-package.html#cb292-26" aria-hidden="true" tabindex="-1"></a><span class="co"># select both function and taxa</span></span>
<span id="cb292-27"><a href="file2meco-package.html#cb292-27" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">select_cols =</span> <span class="fu">c</span>(<span class="st">&quot;Superclass1&quot;</span>, <span class="st">&quot;Phylum&quot;</span>, <span class="st">&quot;Genus&quot;</span>), <span class="at">rel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb292-28"><a href="file2meco-package.html#cb292-28" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(test, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">delete_taxonomy_lineage =</span> T)</span>
<span id="cb292-29"><a href="file2meco-package.html#cb292-29" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb292-30"><a href="file2meco-package.html#cb292-30" aria-hidden="true" tabindex="-1"></a><span class="co"># functional biomarker</span></span>
<span id="cb292-31"><a href="file2meco-package.html#cb292-31" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">select_cols =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">rel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb292-32"><a href="file2meco-package.html#cb292-32" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>taxa_abund<span class="sc">$</span>Superclass1 <span class="sc">%&lt;&gt;%</span> .[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;unclass&quot;</span>, <span class="fu">rownames</span>(.)), ]</span>
<span id="cb292-33"><a href="file2meco-package.html#cb292-33" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(test, <span class="at">method =</span> <span class="st">&quot;lefse&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb292-34"><a href="file2meco-package.html#cb292-34" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb292-35"><a href="file2meco-package.html#cb292-35" aria-hidden="true" tabindex="-1"></a><span class="co"># taxa biomarker</span></span>
<span id="cb292-36"><a href="file2meco-package.html#cb292-36" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">select_cols =</span> <span class="dv">4</span><span class="sc">:</span><span class="dv">9</span>, <span class="at">rel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb292-37"><a href="file2meco-package.html#cb292-37" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>taxa_abund<span class="sc">$</span>Phylum <span class="sc">%&lt;&gt;%</span> .[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;unclass&quot;</span>, <span class="fu">rownames</span>(.)), ]</span>
<span id="cb292-38"><a href="file2meco-package.html#cb292-38" aria-hidden="true" tabindex="-1"></a><span class="co"># p_adjust_method = &quot;none&quot; shut down the p value adjustment</span></span>
<span id="cb292-39"><a href="file2meco-package.html#cb292-39" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(test, <span class="at">method =</span> <span class="st">&quot;lefse&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">p_adjust_method =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb292-40"><a href="file2meco-package.html#cb292-40" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">threshold =</span> <span class="dv">2</span>)</span></code></pre></div>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="file2meco-package.html#cb293-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use KEGG pathway based HUMAnN result</span></span>
<span id="cb293-2"><a href="file2meco-package.html#cb293-2" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_HUMAnN_KEGG_abund.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb293-3"><a href="file2meco-package.html#cb293-3" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">humann2meco</span>(abund_file_path, <span class="at">db =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">sample_table =</span> sample_file_path, <span class="at">match_table =</span> match_file_path)</span>
<span id="cb293-4"><a href="file2meco-package.html#cb293-4" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>tax_table <span class="sc">%&lt;&gt;%</span> <span class="fu">subset</span>(Level<span class="fl">.1</span> <span class="sc">!=</span> <span class="st">&quot;unclassified&quot;</span>)</span>
<span id="cb293-5"><a href="file2meco-package.html#cb293-5" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span></code></pre></div>
</div>
<div id="metaphlan" class="section level2 hasAnchor" number="8.4">
<h2><span class="header-section-number">8.4</span> MetaPhlAn<a href="file2meco-package.html#metaphlan" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>MetaPhlAn is an software used for metagenomic taxonomic profiling <span class="citation">(<a href="#ref-Truong_MeTApHLaN2_2015" role="doc-biblioref">Truong et al. 2015</a>)</span>.
The format of MetaPhlAn classification results is usually called ‘mpa’ format.
The mpa2meco function is developed for this format conversion to microtable object.
See the following example of Kraken2 part.</p>
</div>
<div id="kraken2braken" class="section level2 hasAnchor" number="8.5">
<h2><span class="header-section-number">8.5</span> Kraken2/Braken<a href="file2meco-package.html#kraken2braken" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Kraken is a taxonomic sequence classifier that assigns taxonomic labels to DNA sequences.
Kraken examines the k-mers within a query sequence and uses the information within those k-mers to query a database.
That database maps k-mers to the lowest common ancestor (LCA) of all genomes known to contain a given k-mer.
Kraken2 <span class="citation">(<a href="#ref-Wood_Improved_2019" role="doc-biblioref">Wood, Lu, and Langmead 2019</a>)</span> is the newest version.
Braken <span class="citation">(<a href="#ref-Lu_Bracken_2017" role="doc-biblioref">Lu et al. 2017</a>)</span> can be applied to estimate species abundance following the Kraken analysis.
The merged Kraken2/Braken results can be obtained by merge_metaphlan_tables.py from MetaPhlAn or combine_mpa.py from KrakenTools (<a href="https://ccb.jhu.edu/software/krakentools/" class="uri">https://ccb.jhu.edu/software/krakentools/</a>).</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="file2meco-package.html#cb294-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use the raw data files inside the package</span></span>
<span id="cb294-2"><a href="file2meco-package.html#cb294-2" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_kraken2_merge.txt&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb294-3"><a href="file2meco-package.html#cb294-3" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_metagenome_sample_info.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb294-4"><a href="file2meco-package.html#cb294-4" aria-hidden="true" tabindex="-1"></a>match_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_metagenome_match_table.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb294-5"><a href="file2meco-package.html#cb294-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mpa2meco</span>(abund_file_path)</span>
<span id="cb294-6"><a href="file2meco-package.html#cb294-6" aria-hidden="true" tabindex="-1"></a><span class="co"># &#39;rel = FALSE&#39; means raw abundance in taxa_abund</span></span>
<span id="cb294-7"><a href="file2meco-package.html#cb294-7" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> <span class="fu">mpa2meco</span>(abund_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">match_table =</span> match_file_path, <span class="at">rel =</span> <span class="cn">FALSE</span>, <span class="at">use_level =</span> <span class="st">&quot;s__&quot;</span>)</span>
<span id="cb294-8"><a href="file2meco-package.html#cb294-8" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span>taxa_abund<span class="sc">$</span>Kingdom[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb294-9"><a href="file2meco-package.html#cb294-9" aria-hidden="true" tabindex="-1"></a><span class="co"># &#39;rel = TRUE&#39; means relative abundance in taxa_abund</span></span>
<span id="cb294-10"><a href="file2meco-package.html#cb294-10" aria-hidden="true" tabindex="-1"></a>test2 <span class="ot">&lt;-</span> <span class="fu">mpa2meco</span>(abund_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">match_table =</span> match_file_path, <span class="at">rel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb294-11"><a href="file2meco-package.html#cb294-11" aria-hidden="true" tabindex="-1"></a>test2<span class="sc">$</span>taxa_abund<span class="sc">$</span>Kingdom[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb294-12"><a href="file2meco-package.html#cb294-12" aria-hidden="true" tabindex="-1"></a><span class="co"># The relative abundance in test2 is different with that in test3</span></span>
<span id="cb294-13"><a href="file2meco-package.html#cb294-13" aria-hidden="true" tabindex="-1"></a><span class="co"># The taxonomic abundances in taxa_abund of test3 is calculated based on the otu_table (species abundance) and tax_table</span></span>
<span id="cb294-14"><a href="file2meco-package.html#cb294-14" aria-hidden="true" tabindex="-1"></a><span class="co"># So when the user need to use raw or relative taxonomic abundance coming from original file, please donot run cal_abund function.</span></span>
<span id="cb294-15"><a href="file2meco-package.html#cb294-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb294-16"><a href="file2meco-package.html#cb294-16" aria-hidden="true" tabindex="-1"></a>test3 <span class="ot">&lt;-</span> <span class="fu">clone</span>(test2)</span>
<span id="cb294-17"><a href="file2meco-package.html#cb294-17" aria-hidden="true" tabindex="-1"></a>test3<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb294-18"><a href="file2meco-package.html#cb294-18" aria-hidden="true" tabindex="-1"></a>test3<span class="sc">$</span>taxa_abund<span class="sc">$</span>Kingdom[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code></pre></div>
</div>
<div id="ncycdbpcycdb" class="section level2 hasAnchor" number="8.6">
<h2><span class="header-section-number">8.6</span> NCycDB/PCycDB<a href="file2meco-package.html#ncycdbpcycdb" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>NCycDB database <span class="citation">(<a href="#ref-Tu_NCycDB_2018" role="doc-biblioref">Tu et al. 2018</a>)</span> is a curated integrative database for fast and accurate metagenomic profiling of nitrogen cycling genes.
The <code>ncyc2meco</code> function is designed for construct the microtable object using gene abundance files from NCycDB.
This function can also be used to parse the output of PCycDB <span class="citation">(<a href="#ref-Zeng_PCycDB_2022" role="doc-biblioref">Zeng et al. 2022</a>)</span> database benefiting from implemented mapping database from v0.7.0.
The <code>ncyc2meco</code> function can identify the database and invoke the internal mapping data automatically according to the gene names of input features.</p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="file2meco-package.html#cb295-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb295-2"><a href="file2meco-package.html#cb295-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb295-3"><a href="file2meco-package.html#cb295-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb295-4"><a href="file2meco-package.html#cb295-4" aria-hidden="true" tabindex="-1"></a>?ncyc2meco</span>
<span id="cb295-5"><a href="file2meco-package.html#cb295-5" aria-hidden="true" tabindex="-1"></a><span class="co"># use the raw data files stored inside the package</span></span>
<span id="cb295-6"><a href="file2meco-package.html#cb295-6" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_Ncyc_table.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb295-7"><a href="file2meco-package.html#cb295-7" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_metagenome_sample_info.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb295-8"><a href="file2meco-package.html#cb295-8" aria-hidden="true" tabindex="-1"></a>match_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_metagenome_match_table.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb295-9"><a href="file2meco-package.html#cb295-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ncyc2meco</span>(abund_file_path)</span>
<span id="cb295-10"><a href="file2meco-package.html#cb295-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ncyc2meco</span>(abund_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">match_table =</span> match_file_path)</span></code></pre></div>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="file2meco-package.html#cb296-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let&#39;s try more interesting usages with microeco</span></span>
<span id="cb296-2"><a href="file2meco-package.html#cb296-2" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">ncyc2meco</span>(abund_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">match_table =</span> match_file_path)</span>
<span id="cb296-3"><a href="file2meco-package.html#cb296-3" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb296-4"><a href="file2meco-package.html#cb296-4" aria-hidden="true" tabindex="-1"></a><span class="co"># use split_group = TRUE to calculate the pathway abundance with multipe map correspondance</span></span>
<span id="cb296-5"><a href="file2meco-package.html#cb296-5" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">select_cols =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="at">rel =</span> <span class="cn">TRUE</span>, <span class="at">split_group =</span> <span class="cn">TRUE</span>, <span class="at">split_column =</span> <span class="st">&quot;Pathway&quot;</span>)</span>
<span id="cb296-6"><a href="file2meco-package.html#cb296-6" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>taxa_abund<span class="sc">$</span>Pathway <span class="sc">%&lt;&gt;%</span> .[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;unclass&quot;</span>, <span class="fu">rownames</span>(.)), ]</span>
<span id="cb296-7"><a href="file2meco-package.html#cb296-7" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(test, <span class="at">taxrank =</span> <span class="st">&quot;Pathway&quot;</span>)</span>
<span id="cb296-8"><a href="file2meco-package.html#cb296-8" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">bar_type =</span> <span class="st">&quot;notfull&quot;</span>)</span>
<span id="cb296-9"><a href="file2meco-package.html#cb296-9" aria-hidden="true" tabindex="-1"></a><span class="co"># for gene abundance, no splitting on the pathways</span></span>
<span id="cb296-10"><a href="file2meco-package.html#cb296-10" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">select_cols =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="at">rel =</span> <span class="cn">TRUE</span>, <span class="at">split_group =</span> <span class="cn">FALSE</span>)</span>
<span id="cb296-11"><a href="file2meco-package.html#cb296-11" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>taxa_abund<span class="sc">$</span>Gene <span class="sc">%&lt;&gt;%</span> .[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;unclass&quot;</span>, <span class="fu">rownames</span>(.)), ]</span>
<span id="cb296-12"><a href="file2meco-package.html#cb296-12" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(test, <span class="at">taxrank =</span> <span class="st">&quot;Gene&quot;</span>)</span>
<span id="cb296-13"><a href="file2meco-package.html#cb296-13" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">bar_type =</span> <span class="st">&quot;notfull&quot;</span>)</span></code></pre></div>
</div>
<div id="phyloseq" class="section level2 hasAnchor" number="8.7">
<h2><span class="header-section-number">8.7</span> phyloseq<a href="file2meco-package.html#phyloseq" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Two functions meco2phyloseq() and phyloseq2meco() were provided for the conversion between microtable object and phyloseq object of phyloseq package <span class="citation">(<a href="#ref-Mcmurdie_phyloseq_2013" role="doc-biblioref">Mcmurdie and Holmes 2013</a>)</span>.</p>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="file2meco-package.html#cb297-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Please first install phyloseq</span></span>
<span id="cb297-2"><a href="file2meco-package.html#cb297-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb297-3"><a href="file2meco-package.html#cb297-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;phyloseq&quot;</span>)</span>
<span id="cb297-4"><a href="file2meco-package.html#cb297-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phyloseq)</span></code></pre></div>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="file2meco-package.html#cb298-1" aria-hidden="true" tabindex="-1"></a><span class="co"># from microtable to phyloseq object</span></span>
<span id="cb298-2"><a href="file2meco-package.html#cb298-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;dataset&quot;</span>)</span>
<span id="cb298-3"><a href="file2meco-package.html#cb298-3" aria-hidden="true" tabindex="-1"></a>physeq <span class="ot">&lt;-</span> <span class="fu">meco2phyloseq</span>(dataset)</span>
<span id="cb298-4"><a href="file2meco-package.html#cb298-4" aria-hidden="true" tabindex="-1"></a>physeq</span></code></pre></div>
<div class="sourceCode" id="cb299"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb299-1"><a href="file2meco-package.html#cb299-1" aria-hidden="true" tabindex="-1"></a><span class="co"># from phyloseq to microtable object</span></span>
<span id="cb299-2"><a href="file2meco-package.html#cb299-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;GlobalPatterns&quot;</span>)</span>
<span id="cb299-3"><a href="file2meco-package.html#cb299-3" aria-hidden="true" tabindex="-1"></a>meco_dataset <span class="ot">&lt;-</span> <span class="fu">phyloseq2meco</span>(GlobalPatterns)</span>
<span id="cb299-4"><a href="file2meco-package.html#cb299-4" aria-hidden="true" tabindex="-1"></a>meco_dataset</span></code></pre></div>
</div>
<div id="picrust2" class="section level2 hasAnchor" number="8.8">
<h2><span class="header-section-number">8.8</span> PICRUSt2<a href="file2meco-package.html#picrust2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>PICRUSt2 <span class="citation">(<a href="#ref-Douglas_PICRUSt2_2020" role="doc-biblioref">Douglas et al. 2020</a>)</span> contains an updated and larger database of gene families and reference genomes compared to the original version of PICRUSt.
We do not create a special file conversion function for PICRUSt2
as it is very easy to convert the output pathway files of PICRUSt2 to microtable object.
Two example files of PICRUSt2 output in file2meco package were used to show the operation.</p>
<div class="sourceCode" id="cb300"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb300-1"><a href="file2meco-package.html#cb300-1" aria-hidden="true" tabindex="-1"></a><span class="co"># MetaCyc pathway output</span></span>
<span id="cb300-2"><a href="file2meco-package.html#cb300-2" aria-hidden="true" tabindex="-1"></a>tmp_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_PICRUSt2_MetaCyc_path_abun_unstrat.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb300-3"><a href="file2meco-package.html#cb300-3" aria-hidden="true" tabindex="-1"></a>pathway_table <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(tmp_file_path, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb300-4"><a href="file2meco-package.html#cb300-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;MetaCyc_pathway_map&quot;</span>)</span>
<span id="cb300-5"><a href="file2meco-package.html#cb300-5" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">otu_table =</span> pathway_table, <span class="at">tax_table =</span> MetaCyc_pathway_map)</span>
<span id="cb300-6"><a href="file2meco-package.html#cb300-6" aria-hidden="true" tabindex="-1"></a>tmp<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb300-7"><a href="file2meco-package.html#cb300-7" aria-hidden="true" tabindex="-1"></a>tmp</span></code></pre></div>
<div class="sourceCode" id="cb301"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb301-1"><a href="file2meco-package.html#cb301-1" aria-hidden="true" tabindex="-1"></a><span class="co"># KEGG pathway output</span></span>
<span id="cb301-2"><a href="file2meco-package.html#cb301-2" aria-hidden="true" tabindex="-1"></a>tmp_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_PICRUSt2_KEGG_path_abun_unstrat.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb301-3"><a href="file2meco-package.html#cb301-3" aria-hidden="true" tabindex="-1"></a>pathway_table <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(tmp_file_path, <span class="at">row.names =</span> <span class="dv">1</span>)</span>
<span id="cb301-4"><a href="file2meco-package.html#cb301-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;Tax4Fun2_KEGG&quot;</span>)</span>
<span id="cb301-5"><a href="file2meco-package.html#cb301-5" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">otu_table =</span> pathway_table, <span class="at">tax_table =</span> Tax4Fun2_KEGG<span class="sc">$</span>ptw_desc)</span>
<span id="cb301-6"><a href="file2meco-package.html#cb301-6" aria-hidden="true" tabindex="-1"></a>tmp<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb301-7"><a href="file2meco-package.html#cb301-7" aria-hidden="true" tabindex="-1"></a>tmp</span></code></pre></div>
</div>
<div id="viromescan" class="section level2 hasAnchor" number="8.9">
<h2><span class="header-section-number">8.9</span> ViromeScan<a href="file2meco-package.html#viromescan" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>ViromeScan <span class="citation">(<a href="#ref-Rampelli_ViromeScan_2016" role="doc-biblioref">Rampelli et al. 2016</a>)</span> is a tool for metagenomic viral community profiling.
The input of <code>vs2meco</code> function must be a folder containing all the directories named by sample names.
Each sample directory should have the original output files generated by ViromeScan software.</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="file2meco-package.html#cb302-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb302-2"><a href="file2meco-package.html#cb302-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb302-3"><a href="file2meco-package.html#cb302-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use viromescan directory inside the package</span></span>
<span id="cb302-4"><a href="file2meco-package.html#cb302-4" aria-hidden="true" tabindex="-1"></a>dir_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;viromescan&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb302-5"><a href="file2meco-package.html#cb302-5" aria-hidden="true" tabindex="-1"></a>d1 <span class="ot">&lt;-</span> <span class="fu">vs2meco</span>(dir_path)</span>
<span id="cb302-6"><a href="file2meco-package.html#cb302-6" aria-hidden="true" tabindex="-1"></a>d1<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">rel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb302-7"><a href="file2meco-package.html#cb302-7" aria-hidden="true" tabindex="-1"></a><span class="co"># d1$taxa_abund$Family is same with the percentage output of viromescan at </span></span>
<span id="cb302-8"><a href="file2meco-package.html#cb302-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Family level, i.e. Family_level_results-%.txt file</span></span>
<span id="cb302-9"><a href="file2meco-package.html#cb302-9" aria-hidden="true" tabindex="-1"></a>d1<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">rel =</span> <span class="cn">FALSE</span>)</span>
<span id="cb302-10"><a href="file2meco-package.html#cb302-10" aria-hidden="true" tabindex="-1"></a><span class="co"># d1$taxa_abund$Family is same with the count output of viromescan at </span></span>
<span id="cb302-11"><a href="file2meco-package.html#cb302-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Family level, i.e. Family_level_results-Counts.txt file</span></span></code></pre></div>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bolyen_Reproducible_2019" class="csl-entry">
Bolyen, E., J. R. Rideout, M. R. Dillon, N. A. Bokulich, C. C. Abnet, G. A. Al-Ghalith, H. Alexander, et al. 2019. <span>“Reproducible, Interactive, Scalable and Extensible Microbiome Data Science Using QIIME 2.”</span> Journal Article. <em>Nat Biotechnol</em> 37 (8): 852–57. <a href="https://doi.org/10.1038/s41587-019-0209-9">https://doi.org/10.1038/s41587-019-0209-9</a>.
</div>
<div id="ref-Callahan_DADA2_2016" class="csl-entry">
Callahan, Benjamin J, Paul J McMurdie, Michael J Rosen, Andrew W Han, Amy Jo A Johnson, and Susan P Holmes. 2016. <span>“DADA2: High-Resolution Sample Inference from Illumina Amplicon Data.”</span> Journal Article. <em>Nature Methods</em> 13 (7): 581–83. <a href="https://doi.org/10.1038/nmeth.3869">https://doi.org/10.1038/nmeth.3869</a>.
</div>
<div id="ref-Caporaso_QIIME_2010" class="csl-entry">
Caporaso, J. Gregory, Justin Kuczynski, Jesse Stombaugh, Kyle Bittinger, Frederic D. Bushman, Elizabeth K. Costello, Noah Fierer, et al. 2010. <span>“QIIME Allows Analysis of High-Throughput Community Sequencing Data.”</span> Journal Article. <em>Nature Methods</em> 7 (5): 335–36. https://doi.org/<a href="http://www.nature.com/nmeth/journal/v7/n5/suppinfo/nmeth.f.303_S1.html">http://www.nature.com/nmeth/journal/v7/n5/suppinfo/nmeth.f.303_S1.html</a>.
</div>
<div id="ref-Douglas_PICRUSt2_2020" class="csl-entry">
Douglas, G. M., V. J. Maffei, J. R. Zaneveld, S. N. Yurgel, J. R. Brown, C. M. Taylor, C. Huttenhower, and M. G. I. Langille. 2020. <span>“PICRUSt2 for Prediction of Metagenome Functions.”</span> Journal Article. <em>Nat Biotechnol</em> 38 (6): 685–88. <a href="https://doi.org/10.1038/s41587-020-0548-6">https://doi.org/10.1038/s41587-020-0548-6</a>.
</div>
<div id="ref-Franzosa_Species_2018" class="csl-entry">
Franzosa, Eric A., Lauren J. McIver, Gholamali Rahnavard, Luke R. Thompson, Melanie Schirmer, George Weingart, Karen Schwarzberg Lipson, et al. 2018. <span>“Species-Level Functional Profiling of Metagenomes and Metatranscriptomes.”</span> Journal Article. <em>Nature Methods</em> 15 (11): 962–68. <a href="https://doi.org/10.1038/s41592-018-0176-y">https://doi.org/10.1038/s41592-018-0176-y</a>.
</div>
<div id="ref-Liu_Microbial_2022" class="csl-entry">
Liu, Chi, Xiangzhen Li, Felipe R. P. Mansoldo, Jiaxing An, Yongping Kou, Xiao Zhang, Junming Wang, Jianxiong Zeng, Alane B. Vermelho, and Minjie Yao. 2022. <span>“Microbial Habitat Specificity Largely Affects Microbial Co-Occurrence Patterns and Functional Profiles in Wetland Soils.”</span> Journal Article. <em>Geoderma</em> 418: 115866. <a href="https://doi.org/10.1016/j.geoderma.2022.115866">https://doi.org/10.1016/j.geoderma.2022.115866</a>.
</div>
<div id="ref-Lu_Bracken_2017" class="csl-entry">
Lu, Jennifer, Florian P Breitwieser, Peter Thielen, and Steven L Salzberg. 2017. <span>“Bracken: Estimating Species Abundance in Metagenomics Data.”</span> Journal Article. <em>PeerJ Computer Science</em> 3: e104. <a href="https://doi.org/10.7717/peerj-cs.104">https://doi.org/10.7717/peerj-cs.104</a>.
</div>
<div id="ref-Mcmurdie_phyloseq_2013" class="csl-entry">
Mcmurdie, Paul J, and Susan Holmes. 2013. <span>“Phyloseq: An r Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.”</span> Journal Article. <em>Plos One</em> 8 (4): e61217.
</div>
<div id="ref-Rampelli_ViromeScan_2016" class="csl-entry">
Rampelli, Simone, Matteo Soverini, Silvia Turroni, Sara Quercia, Elena Biagi, Patrizia Brigidi, and Marco Candela. 2016. <span>“ViromeScan: A New Tool for Metagenomic Viral Community Profiling.”</span> Journal Article. <em>Bmc Genomics</em> 17 (1): 1–9. <a href="https://doi.org/10.1186/s12864-016-2446-3">https://doi.org/10.1186/s12864-016-2446-3</a>.
</div>
<div id="ref-Truong_MeTApHLaN2_2015" class="csl-entry">
Truong, Duy Tin, Eric A Franzosa, Timothy L Tickle, Matthias Scholz, George Weingart, Edoardo Pasolli, Adrian Tett, Curtis Huttenhower, and Nicola Segata. 2015. <span>“MetaPhlAn2 for Enhanced Metagenomic Taxonomic Profiling.”</span> Journal Article. <em>Nature Methods</em> 12 (10): 902–3.
</div>
<div id="ref-Tu_NCycDB_2018" class="csl-entry">
Tu, Qichao, Lu Lin, Lei Cheng, Ye Deng, and Zhili He. 2018. <span>“NCycDB: A Curated Integrative Database for Fast and Accurate Metagenomic Profiling of Nitrogen Cycling Genes.”</span> Journal Article. <em>Bioinformatics</em> 35 (6): 1040–48. <a href="https://doi.org/10.1093/bioinformatics/bty741">https://doi.org/10.1093/bioinformatics/bty741</a>.
</div>
<div id="ref-Wood_Improved_2019" class="csl-entry">
Wood, D. E., J. Lu, and B. Langmead. 2019. <span>“Improved Metagenomic Analysis with Kraken 2.”</span> Journal Article. <em>Genome Biol</em> 20 (1): 257. <a href="https://doi.org/10.1186/s13059-019-1891-0">https://doi.org/10.1186/s13059-019-1891-0</a>.
</div>
<div id="ref-Zeng_PCycDB_2022" class="csl-entry">
Zeng, Jiaxiong, Qichao Tu, Xiaoli Yu, Lu Qian, Cheng Wang, Longfei Shu, Fei Liu, et al. 2022. <span>“<span class="nocase">PCycDB: a comprehensive and accurate database for fast analysis of phosphorus cycling genes</span>.”</span> Journal Article. <em>Microbiome</em> 10 (1): 101. <a href="https://doi.org/10.1186/s40168-022-01292-1">https://doi.org/10.1186/s40168-022-01292-1</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="explainable-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="other-dataset.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/07-file2meco.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
