<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>microeco: an R package for data mining in microbial community ecology</title>
  <meta name="description" content="The tutorial for microeco package" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="microeco: an R package for data mining in microbial community ecology" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The tutorial for microeco package" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="microeco: an R package for data mining in microbial community ecology" />
  
  <meta name="twitter:description" content="The tutorial for microeco package" />
  

<meta name="author" content="Chi Liu, Felipe Mansoldo, Chenhao Li, Yang Cao, Jarrod J. Scott, Alane B. Vermelho, Xiangzhen Li, Minjie Yao" />


<meta name="date" content="2021-08-27" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="intro.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#features"><i class="fa fa-check"></i><b>2.1</b> Features</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="extended-class.html"><a href="extended-class.html"><i class="fa fa-check"></i><b>4</b> Extended class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="extended-class.html"><a href="extended-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a></li>
<li class="chapter" data-level="4.2" data-path="extended-class.html"><a href="extended-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a></li>
<li class="chapter" data-level="4.3" data-path="extended-class.html"><a href="extended-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>4.3</b> trans_alpha class</a></li>
<li class="chapter" data-level="4.4" data-path="extended-class.html"><a href="extended-class.html#trans_beta-class"><i class="fa fa-check"></i><b>4.4</b> trans_beta class</a></li>
<li class="chapter" data-level="4.5" data-path="extended-class.html"><a href="extended-class.html#trans_diff-class"><i class="fa fa-check"></i><b>4.5</b> trans_diff class</a></li>
<li class="chapter" data-level="4.6" data-path="extended-class.html"><a href="extended-class.html#trans_env-class"><i class="fa fa-check"></i><b>4.6</b> trans_env class</a></li>
<li class="chapter" data-level="4.7" data-path="extended-class.html"><a href="extended-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>4.7</b> trans_nullmodel class</a></li>
<li class="chapter" data-level="4.8" data-path="extended-class.html"><a href="extended-class.html#trans_network-class"><i class="fa fa-check"></i><b>4.8</b> trans_network class</a></li>
<li class="chapter" data-level="4.9" data-path="extended-class.html"><a href="extended-class.html#trans_func-class"><i class="fa fa-check"></i><b>4.9</b> trans_func class</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="other-data.html"><a href="other-data.html"><i class="fa fa-check"></i><b>5</b> Other data</a>
<ul>
<li class="chapter" data-level="5.1" data-path="other-data.html"><a href="other-data.html#fungi-data"><i class="fa fa-check"></i><b>5.1</b> Fungi data</a></li>
<li class="chapter" data-level="5.2" data-path="other-data.html"><a href="other-data.html#metagenomic-data"><i class="fa fa-check"></i><b>5.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="5.3" data-path="other-data.html"><a href="other-data.html#animal-data"><i class="fa fa-check"></i><b>5.3</b> Animal data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>6</b> Notes</a>
<ul>
<li class="chapter" data-level="6.1" data-path="notes.html"><a href="notes.html#clone"><i class="fa fa-check"></i><b>6.1</b> clone</a></li>
<li class="chapter" data-level="6.2" data-path="notes.html"><a href="notes.html#subset-of-samples"><i class="fa fa-check"></i><b>6.2</b> subset of samples</a></li>
<li class="chapter" data-level="6.3" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>6.3</b> change object</a></li>
<li class="chapter" data-level="6.4" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>6.4</b> group order</a></li>
<li class="chapter" data-level="6.5" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>6.5</b> add layers to plot</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>7</b> mecodev package</a>
<ul>
<li class="chapter" data-level="7.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>7.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="7.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_convert"><i class="fa fa-check"></i><b>7.2</b> trans_convert</a></li>
<li class="chapter" data-level="7.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_netchord"><i class="fa fa-check"></i><b>7.3</b> trans_netchord</a></li>
<li class="chapter" data-level="7.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>7.4</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="other-examples.html"><a href="other-examples.html"><i class="fa fa-check"></i><b>8</b> Other examples</a>
<ul>
<li class="chapter" data-level="8.1" data-path="other-examples.html"><a href="other-examples.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>8.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="8.2" data-path="other-examples.html"><a href="other-examples.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>8.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="8.3" data-path="other-examples.html"><a href="other-examples.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>8.3</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="8.4" data-path="other-examples.html"><a href="other-examples.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>8.4</b> The use of phylogenetic tree</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">microeco: an R package for data mining in microbial community ecology</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">microeco: an R package for data mining in microbial community ecology</h1>
<p class="author"><em>Chi Liu, Felipe Mansoldo, Chenhao Li, Yang Cao, Jarrod J. Scott, Alane B. Vermelho, Xiangzhen Li, Minjie Yao</em></p>
<p class="date" style="margin-top: 1.5em;"><em>2021-08-27</em></p>
</div>
<div id="background" class="section level1" number="1">
<h1><span class="header-section-number">Chapter 1</span> Background</h1>
<p>　R language <span class="citation">(<a href="#ref-R-base" role="doc-biblioref">R Core Team 2016</a>)</span> and its packages ecosystem are wonderful tools for data analysis.
In microbial community ecology field, many packages can be used for the data analysis,
such as vegan<span class="citation">(<a href="#ref-Jari_vegan_2019" role="doc-biblioref">Oksanen et al. 2019</a>)</span>, ape<span class="citation">(<a href="#ref-Paradis_ape_2018" role="doc-biblioref">Paradis and Schliep 2018</a>)</span> and picante<span class="citation">(<a href="#ref-Picante_Kembel_2010" role="doc-biblioref">Kembel et al. 2010</a>)</span>.
However, with the development of the high-throughput sequencing techniques,
the increasing data amount and complexity make the data mining a challenge.
There have been some R packages created for the statistics and visualization in microbial community ecology, such as phyloseq<span class="citation">(<a href="#ref-Mcmurdie_phyloseq_2013" role="doc-biblioref">Mcmurdie and Holmes 2013</a>)</span>,
microbiome (<a href="https://github.com/microbiome/microbiome" class="uri">https://github.com/microbiome/microbiome</a>), microbiomeSeq (<a href="http://www.github.com/umerijaz/microbiomeSeq" class="uri">http://www.github.com/umerijaz/microbiomeSeq</a>),
ampvis2 (<a href="https://madsalbertsen.github.io/ampvis2/reference/index.html" class="uri">https://madsalbertsen.github.io/ampvis2/reference/index.html</a>), MicrobiomeR(<a href="https://github.com/vallenderlab/MicrobiomeR" class="uri">https://github.com/vallenderlab/MicrobiomeR</a>),
theseus<span class="citation">(<a href="#ref-Price_theseus_2018" role="doc-biblioref">Price et al. 2018</a>)</span>, rANOMALY<span class="citation">(<a href="#ref-Theil_rANOMALY_2021" role="doc-biblioref">Theil and Rifa 2021</a>)</span>,
MicrobiomeExplorer<span class="citation">(<a href="#ref-Reeder_MicrobiomeExplorer_2021" role="doc-biblioref">Reeder et al. 2021</a>)</span>,
tidyMicro<span class="citation">(<a href="#ref-Carpenter_tidyMicro_2021" role="doc-biblioref">Carpenter et al. 2021</a>)</span>
and so on.
However, we lack a flexible, comprehensive and modularized R package to analyze and manage the data.
So we create the microeco R package<span class="citation">(<a href="#ref-Liu_microeco_2021" role="doc-biblioref">Liu et al. 2021</a>)</span> for this goal (<a href="https://github.com/ChiLiubio/microeco" class="uri">https://github.com/ChiLiubio/microeco</a>).
Besides, we also develop the file2meco package (<a href="https://github.com/ChiLiubio/file2meco" class="uri">https://github.com/ChiLiubio/file2meco</a>) for the data input from other tools easily
and mecodev package (<a href="https://github.com/ChiLiubio/mecodev" class="uri">https://github.com/ChiLiubio/mecodev</a>) for some extending approaches in the area.</p>

</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Carpenter_tidyMicro_2021" class="csl-entry">
Carpenter, C. M., D. N. Frank, K. Williamson, J. Arbet, B. D. Wagner, K. Kechris, and M. E. Kroehl. 2021. <span>“tidyMicro: A Pipeline for Microbiome Data Analysis and Visualization Using the Tidyverse in r.”</span> Journal Article. <em>BMC Bioinformatics</em> 22 (1): 41. <a href="https://doi.org/10.1186/s12859-021-03967-2">https://doi.org/10.1186/s12859-021-03967-2</a>.
</div>
<div id="ref-Picante_Kembel_2010" class="csl-entry">
Kembel, S. W., P. D. Cowan, M. R. Helmus, W. K. Cornwell, H. Morlon, D. D. Ackerly, S. P. Blomberg, and C. O. Webb. 2010. <span>“Picante: <span>R</span> Tools for Integrating Phylogenies and Ecology.”</span> <em>Bioinformatics</em> 26: 1463–64.
</div>
<div id="ref-Liu_microeco_2021" class="csl-entry">
Liu, Chi, Yaoming Cui, Xiangzhen Li, and Minjie Yao. 2021. <span>“<span class="nocase">microeco: An R package for data mining in microbial community ecology</span>.”</span> Journal Article. <em>FEMS Microbiol. Ecol.</em> 97 (2): fiaa255.
</div>
<div id="ref-Mcmurdie_phyloseq_2013" class="csl-entry">
Mcmurdie, Paul J, and Susan Holmes. 2013. <span>“Phyloseq: An r Package for Reproducible Interactive Analysis and Graphics of Microbiome Census Data.”</span> Journal Article. <em>Plos One</em> 8 (4): e61217.
</div>
<div id="ref-Jari_vegan_2019" class="csl-entry">
Oksanen, Jari, F. Guillaume Blanchet, Michael Friendly, Roeland Kindt, Pierre Legendre, Dan McGlinn, Peter R. Minchin, et al. 2019. <em>Vegan: Community Ecology Package</em>. <a href="https://CRAN.R-project.org/package=vegan">https://CRAN.R-project.org/package=vegan</a>.
</div>
<div id="ref-Paradis_ape_2018" class="csl-entry">
Paradis, E., and K. Schliep. 2018. <span>“Ape 5.0: An Environment for Modern Phylogenetics and Evolutionary Analyses in <span>R</span>.”</span> <em>Bioinformatics</em> 35: 526–28.
</div>
<div id="ref-Price_theseus_2018" class="csl-entry">
Price, Jacob R., Stephen Woloszynek, Gail Rosen, and Christopher M. Sales. 2018. <span>“Theseus - an r Package for the Analysis and Visualization.”</span> Journal Article. <em>bioRxiv</em>. <a href="https://doi.org/10.1101/295675">https://doi.org/10.1101/295675</a>.
</div>
<div id="ref-R-base" class="csl-entry">
R Core Team. 2016. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-Reeder_MicrobiomeExplorer_2021" class="csl-entry">
Reeder, J., M. Huang, J. S. Kaminker, and J. N. Paulson. 2021. <span>“MicrobiomeExplorer: An r Package for the Analysis and Visualization of Microbial Communities.”</span> Journal Article. <em>Bioinformatics</em> 37 (9): 1317–18. <a href="https://doi.org/10.1093/bioinformatics/btaa838">https://doi.org/10.1093/bioinformatics/btaa838</a>.
</div>
<div id="ref-Theil_rANOMALY_2021" class="csl-entry">
Theil, S., and E. Rifa. 2021. <span>“rANOMALY: AmplicoN wOrkflow for Microbial Community AnaLYsis.”</span> Journal Article. <em>F1000Res</em> 10: 7. <a href="https://doi.org/10.12688/f1000research.27268.1">https://doi.org/10.12688/f1000research.27268.1</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="intro.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/index.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
