<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Introduction | Tutorial for R microeco package (v0.6.5)</title>
  <meta name="description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Introduction | Tutorial for R microeco package (v0.6.5)" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Introduction | Tutorial for R microeco package (v0.6.5)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  

<meta name="author" content="Chi Liu, Felipe R. P. Mansoldo, Umer Zeeshan Ijaz, Chenhao Li, Yang Cao, Jarrod J. Scott, Yaoming Cui, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2022-02-28" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="basic-class.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.1</b> R6 Class</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.2</b> Help</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.3</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="intro.html"><a href="intro.html#important-packages"><i class="fa fa-check"></i><b>2.3.1</b> Important packages</a></li>
<li class="chapter" data-level="2.3.2" data-path="intro.html"><a href="intro.html#ggtree"><i class="fa fa-check"></i><b>2.3.2</b> ggtree</a></li>
<li class="chapter" data-level="2.3.3" data-path="intro.html"><a href="intro.html#spieceasi"><i class="fa fa-check"></i><b>2.3.3</b> SpiecEasi</a></li>
<li class="chapter" data-level="2.3.4" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.3.4</b> Gephi</a></li>
<li class="chapter" data-level="2.3.5" data-path="intro.html"><a href="intro.html#wgcna"><i class="fa fa-check"></i><b>2.3.5</b> WGCNA</a></li>
<li class="chapter" data-level="2.3.6" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.3.6</b> Tax4Fun</a></li>
<li class="chapter" data-level="2.3.7" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.3.7</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#plotting"><i class="fa fa-check"></i><b>2.4</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#example"><i class="fa fa-check"></i><b>3.1.1</b> Example</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.2</b> Key points</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#other-functions"><i class="fa fa-check"></i><b>3.1.3</b> Other functions</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-2"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-4"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.2</b> Key points</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#other-functions-1"><i class="fa fa-check"></i><b>6.2.3</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-7"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-8"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-9"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>8</b> Other dataset</a>
<ul>
<li class="chapter" data-level="8.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>8.1</b> Fungi data</a></li>
<li class="chapter" data-level="8.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>8.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="8.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>8.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>9</b> Notes</a>
<ul>
<li class="chapter" data-level="9.1" data-path="notes.html"><a href="notes.html#clone"><i class="fa fa-check"></i><b>9.1</b> clone</a></li>
<li class="chapter" data-level="9.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>9.2</b> save() function</a></li>
<li class="chapter" data-level="9.3" data-path="notes.html"><a href="notes.html#subset-of-samples"><i class="fa fa-check"></i><b>9.3</b> subset of samples</a></li>
<li class="chapter" data-level="9.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>9.4</b> change object</a></li>
<li class="chapter" data-level="9.5" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>9.5</b> group order</a></li>
<li class="chapter" data-level="9.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>9.6</b> add layers to plot</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>10</b> mecodev package</a>
<ul>
<li class="chapter" data-level="10.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>10.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="10.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_convert"><i class="fa fa-check"></i><b>10.2</b> trans_convert</a></li>
<li class="chapter" data-level="10.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_netchord"><i class="fa fa-check"></i><b>10.3</b> trans_netchord</a></li>
<li class="chapter" data-level="10.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>10.4</b> trans_ts</a></li>
<li class="chapter" data-level="10.5" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>10.5</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="other-examples.html"><a href="other-examples.html"><i class="fa fa-check"></i><b>11</b> Other examples</a>
<ul>
<li class="chapter" data-level="11.1" data-path="other-examples.html"><a href="other-examples.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>11.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="11.2" data-path="other-examples.html"><a href="other-examples.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>11.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="11.3" data-path="other-examples.html"><a href="other-examples.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>11.3</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="11.4" data-path="other-examples.html"><a href="other-examples.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>11.4</b> The use of phylogenetic tree</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v0.6.5)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="intro" class="section level1" number="2">
<h1><span class="header-section-number">Chapter 2</span> Introduction</h1>
<p>The microeco package has several advantages compared to other packages in R.
The main goal of developing this package is to help users analyse communtiy ecology data fast.
So a series of commonly-used and cutting-edge approaches are implemented.
To facilitate the data mining, the whole structure of microeco package are highly modularized to
make users conveniently remember, search and use.
It is notable that, beside the demonstration in the tutorial, users can also save the intermediate files in each object and
apply those files to other tools according to the format requirement.
Main files stored in the objects of microeco package is the commonly used data.frame format.
So the intermediate and result files are easily saved, modified and used for other tools in microbial ecology.</p>
<div id="r6-class" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> R6 Class</h2>
<p>All the classes in microeco package depend on the R6 class <span class="citation">(<a href="#ref-R6_Winston" role="doc-biblioref">Chang 2020</a>)</span>.
R6 uses the encapsulated object-oriented programming paradigm,
which means that R6 is a profoundly different OO system from S3 and S4 because it is built on encapsulated objects, rather than generic functions.
If you are interested in the class features, read more from ‘Advanced R’ book (<a href="https://adv-r.hadley.nz/" class="uri">https://adv-r.hadley.nz/</a>).</p>
<ul>
<li><p>A generic is a regular function, so it lives in the global namespace. An R6 method belongs to an object so it lives in a local namespace.
This influences how we think about naming. The methods belong to objects, not generics, and you can call them like object$method().</p></li>
<li><p>R6’s reference semantics allow methods to simultaneously return a value and modify an object.</p></li>
<li><p>Every R6 object has an S3 class that reflects its hierarchy of R6 class.</p></li>
</ul>
</div>
<div id="help" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Help</h2>
<p>The use of help documents in the microeco package may be a little different from other packages we often used.
If you wish to see one of help documents of functions, you should search the name of the class (not the name of the function)
and click the link of each function.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="intro.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first install microeco, see https://github.com/ChiLiubio/microeco</span></span>
<span id="cb1-2"><a href="intro.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># load package microeco</span></span>
<span id="cb1-3"><a href="intro.html#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="intro.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this can show all the functions and the detailed descriptions in the microtable class</span></span>
<span id="cb2-2"><a href="intro.html#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># same with: help(microtable)</span></span>
<span id="cb2-3"><a href="intro.html#cb2-3" aria-hidden="true" tabindex="-1"></a>?microtable</span></code></pre></div>
</div>
<div id="dependence" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Dependence</h2>
<div id="important-packages" class="section level3" number="2.3.1">
<h3><span class="header-section-number">2.3.1</span> Important packages</h3>
<p>To keep the start and use of microeco package simplified,
the installation of microeco only depend on several packages, which are compulsory-installed from CRAN and frequently used in the data analysis.
So the question is that you may encounter an error when using a class or function that invoke an additional package like this:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="intro.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb3-2"><a href="intro.html#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataset)</span>
<span id="cb3-3"><a href="intro.html#cb3-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="cn">NA</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0005</span>)</span>
<span id="cb3-4"><a href="intro.html#cb3-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">network_method =</span> <span class="st">&quot;SpiecEasi&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode html"><code class="sourceCode html"><span id="cb4-1"><a href="intro.html#cb4-1" aria-hidden="true" tabindex="-1"></a>Error in t1$cal_network(network_method = &quot;SpiecEasi&quot;): igraph package not installed ...</span></code></pre></div>
<p><br>
The reason is that network construction require igraph package. We donot put the igraph and some other packages (e.g. SpiecEasi in github) on the “Imports” part of microeco package.</p>
<p>The solutions:</p>
<ol style="list-style-type: decimal">
<li><p>install the package when encounter such an error. Actually, it’s very easy to install the packages from CRAN or bioconductor or github. Just try it.</p></li>
<li><p>install the packages in advance. We recommend this solution if you are interest in most of the methods in the microeco package and want to run a large number of examples in tutorial.</p></li>
</ol>
<p>We show several packages that are published in CRAN and not installed automatically.</p>
<table>
<colgroup>
<col width="14%" />
<col width="33%" />
<col width="52%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Package</th>
<th align="left">where</th>
<th align="left">description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">MASS</td>
<td align="left">trans_diff class</td>
<td align="left">linear discriminant analysis</td>
</tr>
<tr class="even">
<td align="left">GUniFrac</td>
<td align="left">cal_betadiv()</td>
<td align="left">UniFrac distance matrix</td>
</tr>
<tr class="odd">
<td align="left">ggpubr</td>
<td align="left">plot_alpha()</td>
<td align="left">some plotting functions</td>
</tr>
<tr class="even">
<td align="left">randomForest</td>
<td align="left">trans_diff class</td>
<td align="left">random forest analysis</td>
</tr>
<tr class="odd">
<td align="left">ggdendro</td>
<td align="left">plot_clustering()</td>
<td align="left">plotting clustering dendrogram</td>
</tr>
<tr class="even">
<td align="left">ggrepel</td>
<td align="left">trans_rda class</td>
<td align="left">reduce the text overlap in the plot</td>
</tr>
<tr class="odd">
<td align="left">agricolae</td>
<td align="left">cal_diff()</td>
<td align="left">multiple comparisons in anova</td>
</tr>
<tr class="even">
<td align="left">gridExtra</td>
<td align="left">trans_diff class</td>
<td align="left">merge plots</td>
</tr>
<tr class="odd">
<td align="left">picante</td>
<td align="left">cal_alphadiv()</td>
<td align="left">Faith’s phylogenetic alpha diversity</td>
</tr>
<tr class="even">
<td align="left">pheatmap</td>
<td align="left">plot_corr(pheatmap = TRUE)</td>
<td align="left">correlation heatmap with clustering dendrogram</td>
</tr>
<tr class="odd">
<td align="left">tidytree</td>
<td align="left">trans_diff class</td>
<td align="left">plot the taxonomic tree</td>
</tr>
<tr class="even">
<td align="left">igraph</td>
<td align="left">trans_network class</td>
<td align="left">network related operations</td>
</tr>
<tr class="odd">
<td align="left">rgexf</td>
<td align="left">save_network()</td>
<td align="left">save network with gexf style</td>
</tr>
<tr class="even">
<td align="left">ggalluvial</td>
<td align="left">plot_bar(use_alluvium = TRUE)</td>
<td align="left">alluvial plot</td>
</tr>
</tbody>
</table>
<p>Then, if you want to install these packages or some of them, you can do like this:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="intro.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If a package is not installed, it will be installed from CRAN.</span></span>
<span id="cb5-2"><a href="intro.html#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># First select the packages of interest</span></span>
<span id="cb5-3"><a href="intro.html#cb5-3" aria-hidden="true" tabindex="-1"></a>packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;MASS&quot;</span>, <span class="st">&quot;GUniFrac&quot;</span>, <span class="st">&quot;ggpubr&quot;</span>, <span class="st">&quot;randomForest&quot;</span>, <span class="st">&quot;ggdendro&quot;</span>, <span class="st">&quot;ggrepel&quot;</span>, <span class="st">&quot;agricolae&quot;</span>, <span class="st">&quot;gridExtra&quot;</span>, <span class="st">&quot;picante&quot;</span>, <span class="st">&quot;pheatmap&quot;</span>, <span class="st">&quot;igraph&quot;</span>, <span class="st">&quot;rgexf&quot;</span>, <span class="st">&quot;ggalluvial&quot;</span>, <span class="st">&quot;ggh4x&quot;</span>)</span>
<span id="cb5-4"><a href="intro.html#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Now check or install</span></span>
<span id="cb5-5"><a href="intro.html#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> packages){</span>
<span id="cb5-6"><a href="intro.html#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(x, <span class="at">character.only =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb5-7"><a href="intro.html#cb5-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">install.packages</span>(x, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-8"><a href="intro.html#cb5-8" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb5-9"><a href="intro.html#cb5-9" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>There are also some packages that are useful in some functions. These packages may be R packages published in bioconductor or github,
or packages written by other languages.</p>
</div>
<div id="ggtree" class="section level3" number="2.3.2">
<h3><span class="header-section-number">2.3.2</span> ggtree</h3>
<p>Plotting the cladogram from LEfSe result requires the ggtree package in bioconductor (<a href="https://bioconductor.org/packages/release/bioc/html/ggtree.html" class="uri">https://bioconductor.org/packages/release/bioc/html/ggtree.html</a>).</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="intro.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb6-2"><a href="intro.html#cb6-2" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;ggtree&quot;</span>)</span></code></pre></div>
</div>
<div id="spieceasi" class="section level3" number="2.3.3">
<h3><span class="header-section-number">2.3.3</span> SpiecEasi</h3>
<p>The R package SpiecEasi can be used for the network construction using SPIEC-EASI (SParse InversE Covariance Estimation for Ecological Association Inference) approach.
The package can be installed from Github <a href="https://github.com/zdk123/SpiecEasi" class="uri">https://github.com/zdk123/SpiecEasi</a></p>
</div>
<div id="gephi" class="section level3" number="2.3.4">
<h3><span class="header-section-number">2.3.4</span> Gephi</h3>
<p>Gephi is an excellent network visualization tool and used to open the saved network file,
i.e. network.gexf in the <a href="https://chiliubio.github.io/microeco_tutorial/model-based-class.html#trans_network-class">tutorial</a>.
You can download Gephi and learn how to use it from <a href="https://gephi.org/users/download/" class="uri">https://gephi.org/users/download/</a></p>
</div>
<div id="wgcna" class="section level3" number="2.3.5">
<h3><span class="header-section-number">2.3.5</span> WGCNA</h3>
<p>In the correlation-based network, when the species number is very large,
the correlation algorithm in WGCNA is very fast compared to the ‘cor’ option in trans_network.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="intro.html#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;WGCNA&quot;</span>, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="tax4fun" class="section level3" number="2.3.6">
<h3><span class="header-section-number">2.3.6</span> Tax4Fun</h3>
<p>Tax4Fun is an R package used for the prediction of functional potential of prokaryotic communities.</p>
<ol style="list-style-type: decimal">
<li>install Tax4Fun package</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="intro.html#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;RJSONIO&quot;</span>)</span>
<span id="cb8-2"><a href="intro.html#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;biom_0.3.12.tar.gz&quot;</span>, <span class="at">package=</span><span class="st">&quot;microeco&quot;</span>), <span class="at">repos =</span> <span class="cn">NULL</span>, <span class="at">type =</span> <span class="st">&quot;source&quot;</span>)</span>
<span id="cb8-3"><a href="intro.html#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;qiimer_0.9.4.tar.gz&quot;</span>, <span class="at">package=</span><span class="st">&quot;microeco&quot;</span>), <span class="at">repos =</span> <span class="cn">NULL</span>, <span class="at">type =</span> <span class="st">&quot;source&quot;</span>)</span>
<span id="cb8-4"><a href="intro.html#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;Tax4Fun_0.3.1.tar.gz&quot;</span>, <span class="at">package=</span><span class="st">&quot;microeco&quot;</span>), <span class="at">repos =</span> <span class="cn">NULL</span>, <span class="at">type =</span> <span class="st">&quot;source&quot;</span>)</span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>download SILVA123 reference data from <a href="http://tax4fun.gobics.de/" class="uri">http://tax4fun.gobics.de/</a><br />
　unzip SILVA123.zip and provide this path to the folderReferenceData parameter of cal_tax4fun function in trans_func class.</li>
</ol>
</div>
<div id="tax4fun2" class="section level3" number="2.3.7">
<h3><span class="header-section-number">2.3.7</span> Tax4Fun2</h3>
<p>Tax4Fun2 is another R package for the the prediction of functional profiles and functional gene redundancies of prokaryotic communities <span class="citation">(<a href="#ref-Wemheuer_Tax4Fun2_2020" role="doc-biblioref">Wemheuer et al. 2020</a>)</span>.
It has higher accuracies than PICRUSt and Tax4Fun. The Tax4Fun2 approach implemented in microeco is a little different from the original package.
Using Tax4Fun2 approach require the representative fasta file.
The user do not need to install Tax4Fun2 R package.
The only thing need to do is to download the blast tool (<strong>ignore this if the blast tool has been in the path</strong>) and Ref99NR/Ref100NR database (select one).
Downlaod blast tools from “<a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+" class="uri">ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+</a>” ; e.g. ncbi-blast-****-x64-win64.tar.gz for windows system.
Note that some errors can come from the latest versions because of memory issue (<a href="https://www.biostars.org/p/413294/" class="uri">https://www.biostars.org/p/413294/</a>).
An easy solution is to use previous version (such as 2.5.0).
Downlaod Ref99NR.zip from “<a href="https://cloudstor.aarnet.edu.au/plus/s/DkoZIyZpMNbrzSw/download" class="uri">https://cloudstor.aarnet.edu.au/plus/s/DkoZIyZpMNbrzSw/download</a>” or Ref100NR.zip from “<a href="https://cloudstor.aarnet.edu.au/plus/s/jIByczak9ZAFUB4/download" class="uri">https://cloudstor.aarnet.edu.au/plus/s/jIByczak9ZAFUB4/download</a>” .
Uncompress all the folders. The final folders should be like these structures:</p>
<p>blast tools:<br />
　|– ncbi-blast-2.5.0+<br />
　　|—- bin<br />
　　　|—— blastn.exe<br />
　　　|—— makeblastdb.exe<br />
　　　|—— ……</p>
<p>Ref99NR:<br />
　|– Tax4Fun2_ReferenceData_v2<br />
　　|—- Ref99NR<br />
　　　|—— otu000001.tbl.gz<br />
　　　|—— ……<br />
　　　|—— Ref99NR.fasta<br />
　　　|—— Ref99NR.tre</p>
<p>The path “Tax4Fun2_ReferenceData_v2” will be required in the trans_func$cal_tax4fun2() function.
The blast tool path “ncbi-blast-2.5.0+/bin” is also required if it is not added to the path (environmental variable).</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="intro.html#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Either seqinr or Biostrings package should be installed for reading and writing fasta file</span></span>
<span id="cb9-2"><a href="intro.html#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;seqinr&quot;</span>, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-3"><a href="intro.html#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># or install Biostrings from bioconductor https://bioconductor.org/packages/release/bioc/html/Biostrings.html</span></span>
<span id="cb9-4"><a href="intro.html#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we show how to read the fasta file</span></span>
<span id="cb9-5"><a href="intro.html#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># see https://github.com/ChiLiubio/file2meco if you do not have installed file2meco</span></span>
<span id="cb9-6"><a href="intro.html#cb9-6" aria-hidden="true" tabindex="-1"></a>rep_fasta_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;rep.fna&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb9-7"><a href="intro.html#cb9-7" aria-hidden="true" tabindex="-1"></a>rep_fasta <span class="ot">&lt;-</span> seqinr<span class="sc">::</span><span class="fu">read.fasta</span>(rep_fasta_path)</span>
<span id="cb9-8"><a href="intro.html#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># or use Biostrings::readDNAStringSet(rep_fasta_path)</span></span>
<span id="cb9-9"><a href="intro.html#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co"># then see the help document of microtable class about the rep_fasta in microtable$new().</span></span></code></pre></div>
</div>
</div>
<div id="plotting" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Plotting</h2>
<p>Most of the plotting in the package rely on the ggplot2 package system.
We provide some parameters to change the corresponding plot, but it may be far from enough.
The user can also assign the output a name and use the ggplot2-style grammers to modify it.
Each data table used for plotting is stored in the object and can be saved for the customized analysis and plotting.
Of course, the user can also directly modify the class and reload them to use.
Any contribution of a modified class is also appreciated via github pushing or email (<a href="mailto:liuchi0426@126.com" class="email">liuchi0426@126.com</a>).</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-R6_Winston" class="csl-entry">
Chang, Winston. 2020. <em>R6: Encapsulated Classes with Reference Semantics</em>. <a href="https://CRAN.R-project.org/package=R6">https://CRAN.R-project.org/package=R6</a>.
</div>
<div id="ref-Wemheuer_Tax4Fun2_2020" class="csl-entry">
Wemheuer, Franziska, Jessica A. Taylor, Rolf Daniel, Emma Johnston, Peter Meinicke, Torsten Thomas, and Bernd Wemheuer. 2020. <span>“Tax4Fun2: Prediction of Habitat-Specific Functional Profiles and Functional Redundancy Based on 16s rRNA Gene Sequences.”</span> Journal Article. <em>Environmental Microbiome</em> 15 (1). <a href="https://doi.org/10.1186/s40793-020-00358-7">https://doi.org/10.1186/s40793-020-00358-7</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="basic-class.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/01-intro.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
