<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 mecodev package | microeco: an R package for data mining in microbial community ecology</title>
  <meta name="description" content="The tutorial for microeco package" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 mecodev package | microeco: an R package for data mining in microbial community ecology" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The tutorial for microeco package" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 mecodev package | microeco: an R package for data mining in microbial community ecology" />
  
  <meta name="twitter:description" content="The tutorial for microeco package" />
  

<meta name="author" content="Chi Liu, Felipe Mansoldo, Jarrod J. Scott, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2021-08-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="notes.html"/>
<link rel="next" href="other-examples.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#features"><i class="fa fa-check"></i><b>2.1</b> Features</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="extended-class.html"><a href="extended-class.html"><i class="fa fa-check"></i><b>4</b> Extended class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="extended-class.html"><a href="extended-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a></li>
<li class="chapter" data-level="4.2" data-path="extended-class.html"><a href="extended-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a></li>
<li class="chapter" data-level="4.3" data-path="extended-class.html"><a href="extended-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>4.3</b> trans_alpha class</a></li>
<li class="chapter" data-level="4.4" data-path="extended-class.html"><a href="extended-class.html#trans_beta-class"><i class="fa fa-check"></i><b>4.4</b> trans_beta class</a></li>
<li class="chapter" data-level="4.5" data-path="extended-class.html"><a href="extended-class.html#trans_diff-class"><i class="fa fa-check"></i><b>4.5</b> trans_diff class</a></li>
<li class="chapter" data-level="4.6" data-path="extended-class.html"><a href="extended-class.html#trans_env-class"><i class="fa fa-check"></i><b>4.6</b> trans_env class</a></li>
<li class="chapter" data-level="4.7" data-path="extended-class.html"><a href="extended-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>4.7</b> trans_nullmodel class</a></li>
<li class="chapter" data-level="4.8" data-path="extended-class.html"><a href="extended-class.html#trans_network-class"><i class="fa fa-check"></i><b>4.8</b> trans_network class</a></li>
<li class="chapter" data-level="4.9" data-path="extended-class.html"><a href="extended-class.html#trans_func-class"><i class="fa fa-check"></i><b>4.9</b> trans_func class</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="other-data.html"><a href="other-data.html"><i class="fa fa-check"></i><b>5</b> Other data</a>
<ul>
<li class="chapter" data-level="5.1" data-path="other-data.html"><a href="other-data.html#fungi-data"><i class="fa fa-check"></i><b>5.1</b> Fungi data</a></li>
<li class="chapter" data-level="5.2" data-path="other-data.html"><a href="other-data.html#metagenomic-data"><i class="fa fa-check"></i><b>5.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="5.3" data-path="other-data.html"><a href="other-data.html#animal-data"><i class="fa fa-check"></i><b>5.3</b> Animal data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>6</b> Notes</a>
<ul>
<li class="chapter" data-level="6.1" data-path="notes.html"><a href="notes.html#clone"><i class="fa fa-check"></i><b>6.1</b> clone</a></li>
<li class="chapter" data-level="6.2" data-path="notes.html"><a href="notes.html#subset-of-samples"><i class="fa fa-check"></i><b>6.2</b> subset of samples</a></li>
<li class="chapter" data-level="6.3" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>6.3</b> change object</a></li>
<li class="chapter" data-level="6.4" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>6.4</b> group order</a></li>
<li class="chapter" data-level="6.5" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>6.5</b> add layers to plot</a></li>
<li class="chapter" data-level="6.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot-1"><i class="fa fa-check"></i><b>6.6</b> add layers to plot</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>7</b> mecodev package</a>
<ul>
<li class="chapter" data-level="7.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>7.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="7.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_convert"><i class="fa fa-check"></i><b>7.2</b> trans_convert</a></li>
<li class="chapter" data-level="7.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_netchord"><i class="fa fa-check"></i><b>7.3</b> trans_netchord</a></li>
<li class="chapter" data-level="7.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>7.4</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="other-examples.html"><a href="other-examples.html"><i class="fa fa-check"></i><b>8</b> Other examples</a>
<ul>
<li class="chapter" data-level="8.1" data-path="other-examples.html"><a href="other-examples.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>8.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="8.2" data-path="other-examples.html"><a href="other-examples.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>8.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="8.3" data-path="other-examples.html"><a href="other-examples.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>8.3</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="8.4" data-path="other-examples.html"><a href="other-examples.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>8.4</b> The use of phylogenetic tree</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">microeco: an R package for data mining in microbial community ecology</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="mecodev-package" class="section level1" number="7">
<h1><span class="header-section-number">Chapter 7</span> mecodev package</h1>
<p>The mecodev package (<a href="https://github.com/ChiLiubio/mecodev/" class="uri">https://github.com/ChiLiubio/mecodev/</a>) is designed to develop more classes for data analysis based on the microeco package.</p>
<div id="trans_rarefy" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> trans_rarefy</h2>
<p>The class trans_rarefy in mecodev package can be used for the rarefaction and the following plotting to see whether
the sequencing depth is enough to cover all the so-called species in the microbial community.</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="mecodev-package.html#cb193-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mecodev)</span>
<span id="cb193-2"><a href="mecodev-package.html#cb193-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb193-3"><a href="mecodev-package.html#cb193-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(sample_info_16S)</span>
<span id="cb193-4"><a href="mecodev-package.html#cb193-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(otu_table_16S)</span>
<span id="cb193-5"><a href="mecodev-package.html#cb193-5" aria-hidden="true" tabindex="-1"></a><span class="co"># set.seed is used to fix the random number generation to make the results repeatable</span></span>
<span id="cb193-6"><a href="mecodev-package.html#cb193-6" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb193-7"><a href="mecodev-package.html#cb193-7" aria-hidden="true" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">sample_table =</span> sample_info_16S, <span class="at">otu_table =</span> otu_table_16S)</span>
<span id="cb193-8"><a href="mecodev-package.html#cb193-8" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb193-9"><a href="mecodev-package.html#cb193-9" aria-hidden="true" tabindex="-1"></a><span class="co"># trans_rarefy class</span></span>
<span id="cb193-10"><a href="mecodev-package.html#cb193-10" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_rarefy<span class="sc">$</span><span class="fu">new</span>(dataset, <span class="at">alphadiv =</span> <span class="st">&quot;Shannon&quot;</span>, <span class="at">depth =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">10</span>, <span class="dv">50</span>, <span class="dv">500</span>, <span class="dv">2000</span>, <span class="dv">4000</span>, <span class="dv">6000</span>, <span class="dv">8000</span>))</span>
<span id="cb193-11"><a href="mecodev-package.html#cb193-11" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_rarefy</span>(<span class="at">color_values =</span> <span class="fu">rep</span>(<span class="st">&quot;grey&quot;</span>, <span class="dv">100</span>), <span class="at">show_point =</span> <span class="cn">TRUE</span>, <span class="at">add_fitting =</span> <span class="cn">FALSE</span>, <span class="at">show_legend =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p><img src="Images/plot_trans_rarefy.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<div id="trans_convert" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> trans_convert</h2>
<p>The class trans_convert provide several data transformation approaches for the microtable object.
The output will also be a microtable object.</p>
</div>
<div id="trans_netchord" class="section level2" number="7.3">
<h2><span class="header-section-number">7.3</span> trans_netchord</h2>
<p>The class trans_netchord is developed to sum and plot the links number from one taxa to another or in the same taxa in the network.
The input dataset must be a trans_network object.
Creating the trans_netchord object can sum the links (edge) number from one taxa to another or in the same taxa.
The function plot_sum_links() is used to show the result from the function cal_sum_links().
This is very useful to fast see how many nodes are connected between different taxa or within one taxa.
In terms of “Phylum” level in the tutorial,
the function cal_sum_links() sum the linkages number from one Phylum to another Phylum or the linkages in the same Phylum.
So the numbers along the outside of the circular plot represent how many edges or linkages are related with the Phylum.
For example, in terms of Proteobacteria,
there are roughly total 900 edges associated with the OTUs in Proteobacteria,
in which roughly 200 edges connect both OTUs in Proteobacteria and roughly 150 edges connect the OTUs from Proteobacteria with the OTUs from Chloroflexi.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="mecodev-package.html#cb194-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let&#39;s first create a network</span></span>
<span id="cb194-2"><a href="mecodev-package.html#cb194-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="st">&quot;WGCNA&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb194-3"><a href="mecodev-package.html#cb194-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">p_thres =</span> <span class="fl">0.01</span>, <span class="at">COR_cut =</span> <span class="fl">0.7</span>)</span>
<span id="cb194-4"><a href="mecodev-package.html#cb194-4" aria-hidden="true" tabindex="-1"></a><span class="co"># trans_netchord</span></span>
<span id="cb194-5"><a href="mecodev-package.html#cb194-5" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_netchord<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> t1, <span class="at">taxa_level =</span> <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb194-6"><a href="mecodev-package.html#cb194-6" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_sum_links</span>(<span class="at">plot_pos =</span> <span class="cn">TRUE</span>, <span class="at">plot_num =</span> <span class="dv">10</span>)</span></code></pre></div>
<p><img src="Images/plot_sum_links.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<div id="trans_gamma" class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> trans_gamma</h2>
<p>The class trans_gamma is developed to explore the relationship between beta diversity and gamma diversity
based on the method of a biogeographic study<span class="citation">(<a href="#ref-Zhang_Local_2020" role="doc-biblioref">Zhang et al. 2020</a>)</span>.
The contents include the observed beta-gamma diversity relationship, simulated beta-gamma diversity relationship and the following plotting.
The main aim of this class is to judge whether the observed gamma-beta diversity pattern are similar with the tendency of simulated
gamma-beta diversity relationship. If so, species pool have overwhelming effect on the community assembly and
should be first considered to explain community assembly patterns compared to other factors.</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Zhang_Local_2020" class="csl-entry">
Zhang, X., S. Liu, J. Wang, Y. Huang, Z. Freedman, S. Fu, K. Liu, et al. 2020. <span>“Local Community Assembly Mechanisms Shape Soil Bacterial Beta Diversity Patterns Along a Latitudinal Gradient.”</span> Journal Article. <em>Nat Commun</em> 11 (1): 5428. <a href="https://doi.org/10.1038/s41467-020-19228-4">https://doi.org/10.1038/s41467-020-19228-4</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="notes.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="other-examples.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/06-mecodev.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
