<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Model-based class | Tutorial for R microeco package (v0.7.1)</title>
  <meta name="description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  <meta name="generator" content="bookdown 0.25 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Model-based class | Tutorial for R microeco package (v0.7.1)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Model-based class | Tutorial for R microeco package (v0.7.1)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco, file2meco and mecodev packages" />
  

<meta name="author" content="Chi Liu, Felipe R. P. Mansoldo, Umer Zeeshan Ijaz, Chenhao Li, Yang Cao, Jarrod J. Scott, Yaoming Cui, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2022-03-31" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="diversity-based-class.html"/>
<link rel="next" href="explainable-class.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background<span></span></a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction<span></span></a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework<span></span></a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class<span></span></a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help<span></span></a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.4</b> Dependence<span></span></a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="intro.html"><a href="intro.html#important-packages"><i class="fa fa-check"></i><b>2.4.1</b> Important packages<span></span></a></li>
<li class="chapter" data-level="2.4.2" data-path="intro.html"><a href="intro.html#ggtree"><i class="fa fa-check"></i><b>2.4.2</b> ggtree<span></span></a></li>
<li class="chapter" data-level="2.4.3" data-path="intro.html"><a href="intro.html#spieceasi"><i class="fa fa-check"></i><b>2.4.3</b> SpiecEasi<span></span></a></li>
<li class="chapter" data-level="2.4.4" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.4.4</b> Gephi<span></span></a></li>
<li class="chapter" data-level="2.4.5" data-path="intro.html"><a href="intro.html#wgcna"><i class="fa fa-check"></i><b>2.4.5</b> WGCNA<span></span></a></li>
<li class="chapter" data-level="2.4.6" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.4.6</b> Tax4Fun<span></span></a></li>
<li class="chapter" data-level="2.4.7" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.4.7</b> Tax4Fun2<span></span></a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#plotting"><i class="fa fa-check"></i><b>2.5</b> Plotting<span></span></a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class<span></span></a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class<span></span></a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#example"><i class="fa fa-check"></i><b>3.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.2</b> Other examples<span></span></a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.3</b> Key points<span></span></a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#other-functions"><i class="fa fa-check"></i><b>3.1.4</b> Other functions<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class<span></span></a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class<span></span></a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points<span></span></a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class<span></span></a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-2"><i class="fa fa-check"></i><b>4.2.1</b> Example<span></span></a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class<span></span></a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class<span></span></a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points<span></span></a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class<span></span></a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-4"><i class="fa fa-check"></i><b>5.2.1</b> Example<span></span></a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class<span></span></a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class<span></span></a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.2</b> Key points<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class<span></span></a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.2.1</b> Example<span></span></a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.2</b> Key points<span></span></a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#other-functions-1"><i class="fa fa-check"></i><b>6.2.3</b> Other functions<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class<span></span></a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-7"><i class="fa fa-check"></i><b>6.3.1</b> Example<span></span></a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points<span></span></a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function<span></span></a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class<span></span></a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies<span></span></a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples<span></span></a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points<span></span></a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class<span></span></a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class<span></span></a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.1.1</b> Example<span></span></a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points<span></span></a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class<span></span></a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-9"><i class="fa fa-check"></i><b>7.2.1</b> Example<span></span></a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points<span></span></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package<span></span></a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME<span></span></a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2<span></span></a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann-metagenomic-results"><i class="fa fa-check"></i><b>8.3</b> HUMAnN metagenomic results<span></span></a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn<span></span></a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2"><i class="fa fa-check"></i><b>8.5</b> Kraken2<span></span></a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB<span></span></a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq<span></span></a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset<span></span></a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data<span></span></a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data<span></span></a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome<span></span></a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes<span></span></a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> clone() function<span></span></a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> save() function<span></span></a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#subset-of-samples"><i class="fa fa-check"></i><b>10.3</b> subset of samples<span></span></a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.4</b> change object<span></span></a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.5</b> group order<span></span></a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.6</b> add layers to plot<span></span></a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#colors-for-many-groups"><i class="fa fa-check"></i><b>10.7</b> colors for many groups<span></span></a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input<span></span></a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>11</b> mecodev package<span></span></a>
<ul>
<li class="chapter" data-level="11.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>11.1</b> trans_rarefy<span></span></a></li>
<li class="chapter" data-level="11.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_convert"><i class="fa fa-check"></i><b>11.2</b> trans_convert<span></span></a></li>
<li class="chapter" data-level="11.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_netchord"><i class="fa fa-check"></i><b>11.3</b> trans_netchord<span></span></a></li>
<li class="chapter" data-level="11.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>11.4</b> trans_ts<span></span></a></li>
<li class="chapter" data-level="11.5" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>11.5</b> trans_gamma<span></span></a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>12</b> Other examples<span></span></a>
<ul>
<li class="chapter" data-level="12.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>12.1</b> Custom taxa order in bar plot<span></span></a></li>
<li class="chapter" data-level="12.2" data-path="other-examples-1.html"><a href="other-examples-1.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>12.2</b> The importance of tidy_taxonomy function<span></span></a></li>
<li class="chapter" data-level="12.3" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>12.3</b> Question of prefix in the taxa<span></span></a></li>
<li class="chapter" data-level="12.4" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>12.4</b> The use of phylogenetic tree<span></span></a></li>
</ul></li>
<li><a href="references.html#references">References<span></span></a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v0.7.1)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="model-based-class" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Chapter 6</span> Model-based class<a href="model-based-class.html#model-based-class" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>We group all the classes with complex models into the type ‘Model-based class.’</p>
<div id="trans_diff-class" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> trans_diff class<a href="model-based-class.html#trans_diff-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　Differential abundance test is a very important part in the microbial community data analysis.
It can be used to find the significant taxa in determining the community differences across groups.
With the field development, there are some combined methods to better capture the important biomarker information.
Currently, trans_diff class have several famous approaches to perform this analysis:
metastat<span class="citation">(<a href="#ref-White_Statistical_2009" role="doc-biblioref">White, Nagarajan, and Pop 2009</a>)</span>, LEfSe<span class="citation">(<a href="#ref-Segata_Metagenomic_2011" role="doc-biblioref">Segata et al. 2011</a>)</span>, random forest and metagenomeSeq<span class="citation">(<a href="#ref-Paulson_Differential_2013" role="doc-biblioref">Paulson et al. 2013</a>)</span>.
Metastat depends on the permutations and t-test and performs well on the sparse data.
It is used for the comparisons of taxonomic abundance between two groups at any taxonomic level.
LEfSe and random forest in this class is mainly used for the identification of biomarkers at all taxonomic levels.
metagenomeSeq method is implemented to find significant species between two groups at species level (OTU/ASV).
In addition, Kruskal-Wallis Rank Sum Test (for groups &gt; 2), Wilcoxon Rank Sum Tests (for groups = 2) and
Dunn’s Kruskal-Wallis Multiple Comparisons for cases with groups &gt; 2 are also available.</p>
<div id="example-5" class="section level3 hasAnchor" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> Example<a href="model-based-class.html#example-5" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let’s first run metastat example.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="model-based-class.html#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="co"># metastat analysis at Genus level</span></span>
<span id="cb115-2"><a href="model-based-class.html#cb115-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;metastat&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">metastat_taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb115-3"><a href="model-based-class.html#cb115-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_metastat is the result</span></span>
<span id="cb115-4"><a href="model-based-class.html#cb115-4" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_metastat_group_matrix is the group comparisons order for plotting</span></span>
<span id="cb115-5"><a href="model-based-class.html#cb115-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the first paired groups, choose_group = 1</span></span>
<span id="cb115-6"><a href="model-based-class.html#cb115-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_metastat</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">qvalue =</span> <span class="fl">0.05</span>, <span class="at">choose_group =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Images/plot_metastat_1.png" width="650px" style="display: block; margin: auto;" /></p>
<p>LEfSe combines the non-parametric test and linear discriminant analysis <span class="citation">(<a href="#ref-Segata_Metagenomic_2011" role="doc-biblioref">Segata et al. 2011</a>)</span>.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="model-based-class.html#cb116-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;lefse&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.01</span>, <span class="at">lefse_subgroup =</span> <span class="cn">NULL</span>)</span>
<span id="cb116-2"><a href="model-based-class.html#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_lefse is the LEfSe result</span></span>
<span id="cb116-3"><a href="model-based-class.html#cb116-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_abund is the abundance information</span></span>
<span id="cb116-4"><a href="model-based-class.html#cb116-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_lefse_bar</span>(<span class="at">LDA_score =</span> <span class="dv">4</span>)</span></code></pre></div>
<p><img src="Images/plot_lefse_bar.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="model-based-class.html#cb117-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_lefse[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, ]</span></code></pre></div>
<table>
<colgroup>
<col width="66%" />
<col width="9%" />
<col width="14%" />
<col width="9%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Taxa</th>
<th align="center">Group</th>
<th align="center">pvalue</th>
<th align="center">LDA</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">k__Bacteria|p__Proteobacteria</td>
<td align="center">CW</td>
<td align="center">3.21e-11</td>
<td align="center">4.834</td>
</tr>
<tr class="even">
<td align="center">k__Bacteria|p__Acidobacteria|c__Acidobacteria</td>
<td align="center">IW</td>
<td align="center">8.559e-13</td>
<td align="center">4.787</td>
</tr>
<tr class="odd">
<td align="center">k__Bacteria|p__Acidobacteria</td>
<td align="center">IW</td>
<td align="center">5.749e-12</td>
<td align="center">4.785</td>
</tr>
<tr class="even">
<td align="center">k__Bacteria|p__Bacteroidetes</td>
<td align="center">TW</td>
<td align="center">1.19e-09</td>
<td align="center">4.776</td>
</tr>
<tr class="odd">
<td align="center">k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria</td>
<td align="center">CW</td>
<td align="center">5.475e-12</td>
<td align="center">4.613</td>
</tr>
</tbody>
</table>
<p>Then, we plot the abundance of biomarkers detected by LEfSe.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="model-based-class.html#cb118-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code></pre></div>
<p><img src="Images/plot_lefse_diff_abund.png" width="650px" style="display: block; margin: auto;" /></p>
<p>Then, we show the cladogram of the differential features in the taxonomic tree.
There are too many taxa in this dataset.
As an example, we only use the highest 200 abundant taxa in the tree and 50 differential features.
We only show the full taxonomic label at Phylum level and use letters at other levels to reduce the text overlap.</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="model-based-class.html#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clade_label_level 5 represent phylum level in this analysis</span></span>
<span id="cb119-2"><a href="model-based-class.html#cb119-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require ggtree package</span></span>
<span id="cb119-3"><a href="model-based-class.html#cb119-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_lefse_cladogram</span>(<span class="at">use_taxa_num =</span> <span class="dv">200</span>, <span class="at">use_feature_num =</span> <span class="dv">50</span>, <span class="at">clade_label_level =</span> <span class="dv">5</span>)</span></code></pre></div>
<p><img src="Images/plot_lefse_cladogram.png" width="1000px" style="display: block; margin: auto;" /></p>
<p>There may be a problem related with the taxonomic labels in the plot.
When the levels used are too many, the taxonomic labels may have too much overlap.
However, if you only indicate the Phylum labels, the taxa in the legend with marked letters are too many.
At this time, you can select the taxa that you want to show in the plot manually like the following operation.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="model-based-class.html#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose some taxa according to the positions in the previous picture; those taxa labels have minimum overlap</span></span>
<span id="cb120-2"><a href="model-based-class.html#cb120-2" aria-hidden="true" tabindex="-1"></a>use_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;c__Deltaproteobacteria&quot;</span>, <span class="st">&quot;c__Actinobacteria&quot;</span>, <span class="st">&quot;o__Rhizobiales&quot;</span>, <span class="st">&quot;p__Proteobacteria&quot;</span>, <span class="st">&quot;p__Bacteroidetes&quot;</span>, </span>
<span id="cb120-3"><a href="model-based-class.html#cb120-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;o__Micrococcales&quot;</span>, <span class="st">&quot;p__Acidobacteria&quot;</span>, <span class="st">&quot;p__Verrucomicrobia&quot;</span>, <span class="st">&quot;p__Firmicutes&quot;</span>, </span>
<span id="cb120-4"><a href="model-based-class.html#cb120-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;p__Chloroflexi&quot;</span>, <span class="st">&quot;c__Acidobacteria&quot;</span>, <span class="st">&quot;c__Gammaproteobacteria&quot;</span>, <span class="st">&quot;c__Betaproteobacteria&quot;</span>, <span class="st">&quot;c__KD4-96&quot;</span>,</span>
<span id="cb120-5"><a href="model-based-class.html#cb120-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;c__Bacilli&quot;</span>, <span class="st">&quot;o__Gemmatimonadales&quot;</span>, <span class="st">&quot;f__Gemmatimonadaceae&quot;</span>, <span class="st">&quot;o__Bacillales&quot;</span>, <span class="st">&quot;o__Rhodobacterales&quot;</span>)</span>
<span id="cb120-6"><a href="model-based-class.html#cb120-6" aria-hidden="true" tabindex="-1"></a><span class="co"># then use parameter select_show_labels to show</span></span>
<span id="cb120-7"><a href="model-based-class.html#cb120-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_lefse_cladogram</span>(<span class="at">use_taxa_num =</span> <span class="dv">200</span>, <span class="at">use_feature_num =</span> <span class="dv">50</span>, <span class="at">select_show_labels =</span> use_labels)</span>
<span id="cb120-8"><a href="model-based-class.html#cb120-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we can see that more taxa names appear in the tree</span></span></code></pre></div>
<p><img src="Images/plot_lefse_cladogram_1.png" width="1000px" style="display: block; margin: auto;" /></p>
<p>If you are interested in taxonomic tree, you can also use metacoder package<span class="citation">(<a href="#ref-Foster_Metacoder_2017" role="doc-biblioref">Foster, Sharpton, and Grunwald 2017</a>)</span> to plot the taxonomic tree based on the selected taxa.
We do not show the usage here.</p>
<p>The third approach is rf, which depends on the random forest<span class="citation">(<a href="#ref-Beck_Machine_2014" role="doc-biblioref">Beck and Foster 2014</a>; <a href="#ref-Yatsunenko_Human_2012" role="doc-biblioref">Yatsunenko et al. 2012</a>)</span> and the non-parametric test.
The current method can calculate random forest by bootstrapping like the method in LEfSe and only use the significant features.
MeanDecreaseGini is selected as the indicator value in the analysis.</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="model-based-class.html#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Genus level for parameter taxa_level, if you want to use all taxa, change to &quot;all&quot;</span></span>
<span id="cb121-2"><a href="model-based-class.html#cb121-2" aria-hidden="true" tabindex="-1"></a><span class="co"># nresam = 1 and boots = 1 represent no bootstrapping and use all samples directly</span></span>
<span id="cb121-3"><a href="model-based-class.html#cb121-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;rf&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb121-4"><a href="model-based-class.html#cb121-4" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_rf is the result stored in the object</span></span>
<span id="cb121-5"><a href="model-based-class.html#cb121-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the result</span></span>
<span id="cb121-6"><a href="model-based-class.html#cb121-6" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">only_abund_plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb121-7"><a href="model-based-class.html#cb121-7" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(t2<span class="sc">$</span>p1, t2<span class="sc">$</span>p2, <span class="at">ncol=</span><span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb121-8"><a href="model-based-class.html#cb121-8" aria-hidden="true" tabindex="-1"></a><span class="co"># the middle asterisk represent the significances</span></span></code></pre></div>
<p><img src="Images/plot_diff_abund.png" width="800px" style="display: block; margin: auto;" /></p>
<p>If the user only want to get the taxa with significant different abundance, both the method parameter “KW” and “KW_dunn” are available.
The method = “KW” represents Kruskal-Wallis Rank Sum Test (for groups &gt; 2) and Wilcoxon Rank Sum Tests (for groups = 2)
for a specific taxonomic level or all levels of microtable$taxa_abund.
The method = “KW_dunn” represents Dunn’s Kruskal-Wallis Multiple Comparisons for cases with groups &gt; 2.</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="model-based-class.html#cb122-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;KW&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;all&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb122-2"><a href="model-based-class.html#cb122-2" aria-hidden="true" tabindex="-1"></a><span class="co"># &quot;KW_dunn&quot; method depends on the FSA package</span></span>
<span id="cb122-3"><a href="model-based-class.html#cb122-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> d1, <span class="at">method =</span> <span class="st">&quot;KW_dunn&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb122-4"><a href="model-based-class.html#cb122-4" aria-hidden="true" tabindex="-1"></a><span class="co"># see t1$res_diff for the result</span></span></code></pre></div>
</div>
<div id="key-points-5" class="section level3 hasAnchor" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Key points<a href="model-based-class.html#key-points-5" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>trans_diff$new: this class have a strict requirement on the taxonomic information, make sure tidy_taxonomy() function has been performed before the analysis</li>
<li>trans_diff$new: creating this class will invoke taxa_abund stored in microtable object</li>
<li>trans_diff$plot_lefse_cladogram: clade_label_size, clade_label_size_add and clade_label_size_log are used to control the text size in the cladogram</li>
</ul>
</div>
</div>
<div id="trans_network-class" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> trans_network class<a href="model-based-class.html#trans_network-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　Network is a frequently used approach to study the co-occurrence patterns in microbial ecology <span class="citation">(<a href="#ref-Deng_Molecular_2012" role="doc-biblioref">Deng et al. 2012</a>; <a href="#ref-Faust_Microbial_2012" role="doc-biblioref">Faust and Raes 2012</a>; <a href="#ref-Coyte_Theecology_2015" role="doc-biblioref">Coyte, Schluter, and Foster 2015</a>)</span>.
In this part, we describe all the core contents in the trans_network class.
The network construction approaches can be classified into two types: correlation-based and non correlation-based.
Several approaches can be used to calculate correlations and significances.</p>
<div id="example-6" class="section level3 hasAnchor" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> Example<a href="model-based-class.html#example-6" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We first introduce the correlation-based network. The parameter cal_cor in trans_network is used for selecting the correlation calculation method.</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="model-based-class.html#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use R base cor.test, a little slow</span></span>
<span id="cb123-2"><a href="model-based-class.html#cb123-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="st">&quot;base&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb123-3"><a href="model-based-class.html#cb123-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_cor_p list; one table: correlation; another: p value</span></span></code></pre></div>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="model-based-class.html#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="co"># SparCC method, require SpiecEasi package, see https://chiliubio.github.io/microeco_tutorial/intro.html#spieceasi for the installation</span></span>
<span id="cb124-2"><a href="model-based-class.html#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="co"># SparCC is very slow, so consider filtering more species with low abundance</span></span>
<span id="cb124-3"><a href="model-based-class.html#cb124-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="st">&quot;SparCC&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>, <span class="at">SparCC_simu_num =</span> <span class="dv">100</span>)</span></code></pre></div>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="model-based-class.html#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="co"># When the OTU number is large, using R WGCNA package to replace &quot;base&quot; will be faster</span></span>
<span id="cb125-2"><a href="model-based-class.html#cb125-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require WGCNA package, see https://chiliubio.github.io/microeco_tutorial/intro.html#wgcna for the installation</span></span>
<span id="cb125-3"><a href="model-based-class.html#cb125-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="st">&quot;WGCNA&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span></code></pre></div>
<p>The parameter COR_cut can be used to select the correlation threshold.
Furthermore, COR_optimization = TRUE represent using RMT theory to find the optimized correlation threshold instead of the COR_cut<span class="citation">(<a href="#ref-Deng_Molecular_2012" role="doc-biblioref">Deng et al. 2012</a>)</span>.</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="model-based-class.html#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="co"># construct network; require igraph package</span></span>
<span id="cb126-2"><a href="model-based-class.html#cb126-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">p_thres =</span> <span class="fl">0.01</span>, <span class="at">COR_optimization =</span> <span class="cn">TRUE</span>)</span>
<span id="cb126-3"><a href="model-based-class.html#cb126-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_network</span></span></code></pre></div>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="model-based-class.html#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use arbitrary coefficient threshold to contruct network</span></span>
<span id="cb127-2"><a href="model-based-class.html#cb127-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">p_thres =</span> <span class="fl">0.01</span>, <span class="at">COR_cut =</span> <span class="fl">0.7</span>)</span></code></pre></div>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="model-based-class.html#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add modules in the network</span></span>
<span id="cb128-2"><a href="model-based-class.html#cb128-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_module</span>()</span></code></pre></div>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="model-based-class.html#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save network</span></span>
<span id="cb129-2"><a href="model-based-class.html#cb129-2" aria-hidden="true" tabindex="-1"></a><span class="co"># open the gexf file using Gephi(https://gephi.org/)</span></span>
<span id="cb129-3"><a href="model-based-class.html#cb129-3" aria-hidden="true" tabindex="-1"></a><span class="co"># require rgexf package</span></span>
<span id="cb129-4"><a href="model-based-class.html#cb129-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">save_network</span>(<span class="at">filepath =</span> <span class="st">&quot;network.gexf&quot;</span>)</span></code></pre></div>
<p>We plot the network and present the node colors according to the calculated modules in Gephi.
<img src="Images/network1_spearman.png" width="550px" style="display: block; margin: auto;" /></p>
<p>Now, we show the node colors with the Phylum information and the edges colors with the positive and negative correlations.
All the data used has been stored in the network.gexf file, including modules classifications, Phylum information and edges classifications.</p>
<p><img src="Images/network2_spearman.png" width="550px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="model-based-class.html#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate network attributes</span></span>
<span id="cb130-2"><a href="model-based-class.html#cb130-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network_attr</span>()</span>
<span id="cb130-3"><a href="model-based-class.html#cb130-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_network_attr</span></span></code></pre></div>
<table style="width:49%;">
<colgroup>
<col width="34%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Property</th>
<th align="center">Value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Vertex</td>
<td align="center">407</td>
</tr>
<tr class="even">
<td align="center">Edge</td>
<td align="center">1989</td>
</tr>
<tr class="odd">
<td align="center">Average_degree</td>
<td align="center">9.774</td>
</tr>
<tr class="even">
<td align="center">Average_path_length</td>
<td align="center">3.878</td>
</tr>
<tr class="odd">
<td align="center">Network_diameter</td>
<td align="center">9</td>
</tr>
<tr class="even">
<td align="center">Clustering_coefficient</td>
<td align="center">0.4698</td>
</tr>
<tr class="odd">
<td align="center">Density</td>
<td align="center">0.02407</td>
</tr>
<tr class="even">
<td align="center">Heterogeneity</td>
<td align="center">1.194</td>
</tr>
<tr class="odd">
<td align="center">Centralization</td>
<td align="center">0.09908</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="model-based-class.html#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># classify the node; return t1$res_node_type</span></span>
<span id="cb131-2"><a href="model-based-class.html#cb131-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_node_type</span>()</span>
<span id="cb131-3"><a href="model-based-class.html#cb131-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_node_type</span></span>
<span id="cb131-4"><a href="model-based-class.html#cb131-4" aria-hidden="true" tabindex="-1"></a><span class="co"># we retain the file for the following example in trans_func part</span></span>
<span id="cb131-5"><a href="model-based-class.html#cb131-5" aria-hidden="true" tabindex="-1"></a>network_node_type <span class="ot">&lt;-</span> t1<span class="sc">$</span>res_node_type</span></code></pre></div>
<table style="width:89%;">
<colgroup>
<col width="22%" />
<col width="15%" />
<col width="12%" />
<col width="12%" />
<col width="26%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">z</th>
<th align="center">module</th>
<th align="center">p</th>
<th align="center">taxa_roles</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>OTU_50</strong></td>
<td align="center">-1.305</td>
<td align="center">M2</td>
<td align="center">0</td>
<td align="center">Peripheral nodes</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_1</strong></td>
<td align="center">-0.04067</td>
<td align="center">M2</td>
<td align="center">0</td>
<td align="center">Peripheral nodes</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_55</strong></td>
<td align="center">-1.239</td>
<td align="center">M2</td>
<td align="center">0</td>
<td align="center">Peripheral nodes</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_13824</strong></td>
<td align="center">-0.2403</td>
<td align="center">M2</td>
<td align="center">0</td>
<td align="center">Peripheral nodes</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_151</strong></td>
<td align="center">-1.372</td>
<td align="center">M2</td>
<td align="center">0.4444</td>
<td align="center">Peripheral nodes</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="model-based-class.html#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot node roles in terms of the within-module connectivity and among-module connectivity</span></span>
<span id="cb132-2"><a href="model-based-class.html#cb132-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_taxa_roles</span>(<span class="at">use_type =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Images/plot_taxa_roles.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="model-based-class.html#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot node roles with phylum information</span></span>
<span id="cb133-2"><a href="model-based-class.html#cb133-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_taxa_roles</span>(<span class="at">use_type =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/plot_taxa_roles_2.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Now, we show the eigengene analysis of modules.
The eigengene of a module, i.e. the first principal component of PCA, represents the main variance of the abundance in the species of the module.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="model-based-class.html#cb134-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_eigen</span>()</span>
<span id="cb134-2"><a href="model-based-class.html#cb134-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_eigen</span></span></code></pre></div>
<p>Then we perform correlation heatmap to show the relationships between eigengenes and environmental factors.</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="model-based-class.html#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_env object like the above operation</span></span>
<span id="cb135-2"><a href="model-based-class.html#cb135-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb135-3"><a href="model-based-class.html#cb135-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate correlations</span></span>
<span id="cb135-4"><a href="model-based-class.html#cb135-4" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">add_abund_table =</span> t1<span class="sc">$</span>res_eigen)</span>
<span id="cb135-5"><a href="model-based-class.html#cb135-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the correlation heatmap</span></span>
<span id="cb135-6"><a href="model-based-class.html#cb135-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/Env_module_eigen.png" width="600px" style="display: block; margin: auto;" /></p>
<p>The subset_network() function can be used to extract a part of nodes and edges among these nodes from the network.
In this function, you should provide the nodes you need using the node parameter.</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="model-based-class.html#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract a sub network that contains all nodes in module M1</span></span>
<span id="cb136-2"><a href="model-based-class.html#cb136-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">subset_network</span>(<span class="at">node =</span> t1<span class="sc">$</span>res_node_type <span class="sc">%&gt;%</span> .[.<span class="sc">$</span>module <span class="sc">==</span> <span class="st">&quot;M1&quot;</span>, ] <span class="sc">%&gt;%</span> rownames, <span class="at">rm_single =</span> <span class="cn">TRUE</span>)</span>
<span id="cb136-3"><a href="model-based-class.html#cb136-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return a new network with igraph class</span></span>
<span id="cb136-4"><a href="model-based-class.html#cb136-4" aria-hidden="true" tabindex="-1"></a><span class="co"># extract sub network in which all edge labels are &quot;+&quot;, i.e. positive edges</span></span>
<span id="cb136-5"><a href="model-based-class.html#cb136-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">subset_network</span>(<span class="at">edge =</span> <span class="st">&quot;+&quot;</span>)</span></code></pre></div>
<p><strong>Then let’s show how to extract sub-network for samples and the whole analysis flow.</strong></p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="model-based-class.html#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the sub-network of sample &#39;S1&#39;</span></span>
<span id="cb137-2"><a href="model-based-class.html#cb137-2" aria-hidden="true" tabindex="-1"></a>sub1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">subset_network</span>(<span class="at">node =</span> dataset<span class="sc">$</span>otu_table[, <span class="st">&quot;S1&quot;</span>, <span class="at">drop =</span> <span class="cn">FALSE</span>] <span class="sc">%&gt;%</span> .[.[, <span class="dv">1</span>] <span class="sc">!=</span><span class="dv">0</span>, , <span class="at">drop =</span> <span class="cn">FALSE</span>] <span class="sc">%&gt;%</span> rownames, <span class="at">rm_single =</span> <span class="cn">TRUE</span>)</span>
<span id="cb137-3"><a href="model-based-class.html#cb137-3" aria-hidden="true" tabindex="-1"></a><span class="co"># see https://chiliubio.github.io/microeco_tutorial/notes.html#clone for the &#39;clone&#39; function explanation</span></span>
<span id="cb137-4"><a href="model-based-class.html#cb137-4" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">clone</span>(t1)</span>
<span id="cb137-5"><a href="model-based-class.html#cb137-5" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>res_network <span class="ot">&lt;-</span> sub1</span>
<span id="cb137-6"><a href="model-based-class.html#cb137-6" aria-hidden="true" tabindex="-1"></a><span class="co"># then t2 have a network for &#39;S1&#39; and can be used for further analysis</span></span>
<span id="cb137-7"><a href="model-based-class.html#cb137-7" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_module</span>()</span>
<span id="cb137-8"><a href="model-based-class.html#cb137-8" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">save_network</span>(<span class="st">&quot;S1.gexf&quot;</span>)</span>
<span id="cb137-9"><a href="model-based-class.html#cb137-9" aria-hidden="true" tabindex="-1"></a><span class="co"># please use a loop for more samples</span></span></code></pre></div>
<p>Then we show the next implemented network construction approach:
SPIEC-EASI (SParse InversE Covariance Estimation for Ecological Association Inference) network in SpiecEasi R package <span class="citation">(<a href="#ref-Kurtz_Sparse_2015" role="doc-biblioref">Kurtz et al. 2015</a>)</span>.
See <a href="https://chiliubio.github.io/microeco_tutorial/intro.html#spieceasi" class="uri">https://chiliubio.github.io/microeco_tutorial/intro.html#spieceasi</a> for the package installation.</p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="model-based-class.html#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cal_cor select NA</span></span>
<span id="cb138-2"><a href="model-based-class.html#cb138-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="cn">NA</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0005</span>)</span>
<span id="cb138-3"><a href="model-based-class.html#cb138-3" aria-hidden="true" tabindex="-1"></a><span class="co"># require SpiecEasi package  https://github.com/zdk123/SpiecEasi</span></span>
<span id="cb138-4"><a href="model-based-class.html#cb138-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">network_method =</span> <span class="st">&quot;SpiecEasi&quot;</span>)</span>
<span id="cb138-5"><a href="model-based-class.html#cb138-5" aria-hidden="true" tabindex="-1"></a><span class="co"># see t1$res_network</span></span></code></pre></div>
<p>We also introduce the third network construction approach: Probabilistic Graphical Models (PGM), which is implemented in julia package FlashWeave<span class="citation">(<a href="#ref-Tackmann_Rapid_2019" role="doc-biblioref">Tackmann, Matias Rodrigues, and Mering 2019</a>)</span>.
It predicts ecological interactions among microbes from large-scale compositional abundance data (i.e. OTU tables constructed from sequencing data)
through statistical co-occurrence.
If you want to use this method like the following code, you should first install julia language in your computer and the FlashWeave package,
and add the julia in the computer path.</p>
<ol style="list-style-type: decimal">
<li>download and install julia from <a href="https://julialang.org/downloads/" class="uri">https://julialang.org/downloads/</a><br />
</li>
<li>Put julia in the computer env PATH, such as your_directory_path</li>
<li>Open terminal or cmd or Powershell, open julia, install FlashWeave following the operation in <a href="https://github.com/meringlab/FlashWeave.jl" class="uri">https://github.com/meringlab/FlashWeave.jl</a></li>
</ol>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="model-based-class.html#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cal_cor select NA</span></span>
<span id="cb139-2"><a href="model-based-class.html#cb139-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="cn">NA</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>)</span>
<span id="cb139-3"><a href="model-based-class.html#cb139-3" aria-hidden="true" tabindex="-1"></a><span class="co"># require Julia in the computer path, and the package FlashWeave</span></span>
<span id="cb139-4"><a href="model-based-class.html#cb139-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">network_method =</span> <span class="st">&quot;PGM&quot;</span>)</span>
<span id="cb139-5"><a href="model-based-class.html#cb139-5" aria-hidden="true" tabindex="-1"></a><span class="co"># see t1$res_network</span></span></code></pre></div>
</div>
<div id="key-points-6" class="section level3 hasAnchor" number="6.2.2">
<h3><span class="header-section-number">6.2.2</span> Key points<a href="model-based-class.html#key-points-6" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>subset_network(): this function can extract any sub-network according to the input nodes, e.g. sub-network for modules or samples</li>
</ul>
</div>
<div id="other-functions-1" class="section level3 hasAnchor" number="6.2.3">
<h3><span class="header-section-number">6.2.3</span> Other functions<a href="model-based-class.html#other-functions-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>get_edge_table(): get the table of edges, including connected nodes, labels and weight.</li>
<li>cal_powerlaw_p(): perform a bootstrapping hypothesis test to determine whether degrees follows a power law distribution</li>
<li>cal_powerlaw_p(): fit degrees to a power law distribution.</li>
</ul>
</div>
</div>
<div id="trans_nullmodel-class" class="section level2 hasAnchor" number="6.3">
<h2><span class="header-section-number">6.3</span> trans_nullmodel class<a href="model-based-class.html#trans_nullmodel-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In recent decades,
the integration of phylogenetic analysis and null model promotes the inference of niche and neutral influences on community assembly more powerfully
by adding a phylogeny dimension <span class="citation">(<a href="#ref-Webb_Phylogenies_2002" role="doc-biblioref">Webb et al. 2002</a>; <a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>)</span>.
The trans_nullmodel class provides an encapsulation, including the calculation of the phylogenetic signal,
beta mean pairwise phylogenetic distance (betaMPD), beta mean nearest taxon distance (betaMNTD),
beta nearest taxon index (betaNTI), beta net relatedness index (betaNRI) and Bray-Curtis-based Raup-Crick (RCbray).
The approach for phylogenetic signal analysis is based on the mantel correlogram <span class="citation">(<a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>,
in which the change of phylogenetic signal is intuitional and clear compared to other approaches.
The algorithms of betaMNTD and betaMPD have been optimized to be faster than those in the picante package <span class="citation">(<a href="#ref-Picante_Kembel_2010" role="doc-biblioref">Kembel et al. 2010</a>)</span>.
The combinations between RCbray and betaNTI can be used to infer the strength of each ecological process dominating the community assembly
under the specific hypothesis <span class="citation">(<a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>)</span>.
This can be achievable by the function cal_process() to parse the percentage of each inferred process.</p>
<div id="example-7" class="section level3 hasAnchor" number="6.3.1">
<h3><span class="header-section-number">6.3.1</span> Example<a href="model-based-class.html#example-7" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We first check the phylogenetic signal.</p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="model-based-class.html#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generate trans_nullmodel object; use 1000 OTUs as example</span></span>
<span id="cb140-2"><a href="model-based-class.html#cb140-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_nullmodel<span class="sc">$</span><span class="fu">new</span>(dataset, <span class="at">taxa_number =</span> <span class="dv">1000</span>, <span class="at">add_data =</span> env_data_16S)</span></code></pre></div>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="model-based-class.html#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use pH as the test variable</span></span>
<span id="cb141-2"><a href="model-based-class.html#cb141-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_mantel_corr</span>(<span class="at">use_env =</span> <span class="st">&quot;pH&quot;</span>)</span>
<span id="cb141-3"><a href="model-based-class.html#cb141-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_mantel_corr</span></span>
<span id="cb141-4"><a href="model-based-class.html#cb141-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the mantel correlogram</span></span>
<span id="cb141-5"><a href="model-based-class.html#cb141-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_mantel_corr</span>()</span></code></pre></div>
<p><img src="Images/plot_mantel_corr.png" width="600px" style="display: block; margin: auto;" /></p>
<p>betaNRI(ses.betampd) is used to show the ‘basal’ phylogenetic turnover<span class="citation">(<a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>.
Compared to betaNTI, it can capture more turnover information associated with the deep phylogeny.
It is noted that there are many null models with the development in the several decades.
In the trans_nullmodel class, we randomized the phylogenetic relatedness of species.
This shuffling approach fix the observed levels of species α-diversity and β-diversity to
explore whether the observed phylogenetic turnover significantly differ from null model that phylogenetic relatedness among species are random.</p>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="model-based-class.html#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="co"># null model run 500 times</span></span>
<span id="cb142-2"><a href="model-based-class.html#cb142-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ses_betampd</span>(<span class="at">runs=</span><span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb142-3"><a href="model-based-class.html#cb142-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_ses_betampd</span></span></code></pre></div>
<p>If we want to plot the betaNRI, we can use plot_group_distance function in trans_beta class.
For example, the results showed that the mean betaNRI of TW is extremely and significantly larger that those in CW and IW,
revealing that the basal phylogenetic turnover in TW is high.</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="model-based-class.html#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add betaNRI matrix to beta_diversity list</span></span>
<span id="cb143-2"><a href="model-based-class.html#cb143-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>beta_diversity[[<span class="st">&quot;betaNRI&quot;</span>]] <span class="ot">&lt;-</span> t1<span class="sc">$</span>res_ses_betampd</span>
<span id="cb143-3"><a href="model-based-class.html#cb143-3" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_beta class, use measure &quot;betaNRI&quot;</span></span>
<span id="cb143-4"><a href="model-based-class.html#cb143-4" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;betaNRI&quot;</span>)</span>
<span id="cb143-5"><a href="model-based-class.html#cb143-5" aria-hidden="true" tabindex="-1"></a><span class="co"># transform the distance for each group</span></span>
<span id="cb143-6"><a href="model-based-class.html#cb143-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance</span>()</span>
<span id="cb143-7"><a href="model-based-class.html#cb143-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the results</span></span>
<span id="cb143-8"><a href="model-based-class.html#cb143-8" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t2<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">distance_pair_stat =</span> <span class="cn">TRUE</span>)</span>
<span id="cb143-9"><a href="model-based-class.html#cb143-9" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/plot_betaNRI_one_dataset.png" width="550px" style="display: block; margin: auto;" /></p>
<p>Sometimes, if you want to perform null model analysis for each group individually, such as one group as one species pool,
you can calculate the results for each group, respectively.
The results show that, when we perform betaNRI for each group respectively,
mean betaNRI between CW and TW are not significantly different, and they are both significantly higher than that in IW,
revealing that the strength of variable selection in CW and TW may be similar under the condition that each area is considered as a specific species pool.</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="model-based-class.html#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we create a list to store the trans_nullmodel results.</span></span>
<span id="cb144-2"><a href="model-based-class.html#cb144-2" aria-hidden="true" tabindex="-1"></a>sesbeta_each <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb144-3"><a href="model-based-class.html#cb144-3" aria-hidden="true" tabindex="-1"></a>group_col <span class="ot">&lt;-</span> <span class="st">&quot;Group&quot;</span></span>
<span id="cb144-4"><a href="model-based-class.html#cb144-4" aria-hidden="true" tabindex="-1"></a>all_groups <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>sample_table[, group_col])</span>
<span id="cb144-5"><a href="model-based-class.html#cb144-5" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate for each group, respectively</span></span>
<span id="cb144-6"><a href="model-based-class.html#cb144-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> all_groups){</span>
<span id="cb144-7"><a href="model-based-class.html#cb144-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># like the above operation, but need provide &#39;group&#39; and &#39;select_group&#39;</span></span>
<span id="cb144-8"><a href="model-based-class.html#cb144-8" aria-hidden="true" tabindex="-1"></a>    test <span class="ot">&lt;-</span> trans_nullmodel<span class="sc">$</span><span class="fu">new</span>(dataset, <span class="at">group =</span> group_col, <span class="at">select_group =</span> i, <span class="at">taxa_number =</span> <span class="dv">1000</span>, <span class="at">add_data =</span> env_data_16S)</span>
<span id="cb144-9"><a href="model-based-class.html#cb144-9" aria-hidden="true" tabindex="-1"></a>    test<span class="sc">$</span><span class="fu">cal_ses_betampd</span>(<span class="at">runs =</span> <span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb144-10"><a href="model-based-class.html#cb144-10" aria-hidden="true" tabindex="-1"></a>    sesbeta_each[[i]] <span class="ot">&lt;-</span> test<span class="sc">$</span>res_ses_betampd</span>
<span id="cb144-11"><a href="model-based-class.html#cb144-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb144-12"><a href="model-based-class.html#cb144-12" aria-hidden="true" tabindex="-1"></a><span class="co"># merge and reshape to generate one symmetrical matrix</span></span>
<span id="cb144-13"><a href="model-based-class.html#cb144-13" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">lapply</span>(sesbeta_each, reshape2<span class="sc">::</span>melt) <span class="sc">%&gt;%</span> </span>
<span id="cb144-14"><a href="model-based-class.html#cb144-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">do.call</span>(rbind, .) <span class="sc">%&gt;%</span></span>
<span id="cb144-15"><a href="model-based-class.html#cb144-15" aria-hidden="true" tabindex="-1"></a>    reshape2<span class="sc">::</span><span class="fu">dcast</span>(., Var1<span class="sc">~</span>Var2, <span class="at">value.var =</span> <span class="st">&quot;value&quot;</span>)</span>
<span id="cb144-16"><a href="model-based-class.html#cb144-16" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(test) <span class="ot">&lt;-</span> test[, <span class="dv">1</span>]</span>
<span id="cb144-17"><a href="model-based-class.html#cb144-17" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> test[, <span class="sc">-</span><span class="dv">1</span>, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb144-18"><a href="model-based-class.html#cb144-18" aria-hidden="true" tabindex="-1"></a><span class="co"># like the above operation</span></span>
<span id="cb144-19"><a href="model-based-class.html#cb144-19" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>beta_diversity[[<span class="st">&quot;betaNRI&quot;</span>]] <span class="ot">&lt;-</span> test</span>
<span id="cb144-20"><a href="model-based-class.html#cb144-20" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;betaNRI&quot;</span>)</span>
<span id="cb144-21"><a href="model-based-class.html#cb144-21" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance</span>()</span>
<span id="cb144-22"><a href="model-based-class.html#cb144-22" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t2<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">distance_pair_stat =</span> <span class="cn">TRUE</span>)</span>
<span id="cb144-23"><a href="model-based-class.html#cb144-23" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/plot_betaNRI_each_dataset.png" width="550px" style="display: block; margin: auto;" /></p>
<p>BetaNTI(ses.betamntd) can be used to indicate the phylogenetic terminal turnover <span class="citation">(<a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>)</span>.</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="model-based-class.html#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="co"># null model run 500 times</span></span>
<span id="cb145-2"><a href="model-based-class.html#cb145-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ses_betamntd</span>(<span class="at">runs=</span><span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb145-3"><a href="model-based-class.html#cb145-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_ses_betamntd</span></span></code></pre></div>
<table style="width:75%;">
<colgroup>
<col width="12%" />
<col width="12%" />
<col width="12%" />
<col width="12%" />
<col width="12%" />
<col width="12%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">S1</th>
<th align="center">S2</th>
<th align="center">S3</th>
<th align="center">S4</th>
<th align="center">S5</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">0</td>
<td align="center">-6.554</td>
<td align="center">-6.563</td>
<td align="center">-6.308</td>
<td align="center">-6.153</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">-6.554</td>
<td align="center">0</td>
<td align="center">-6.678</td>
<td align="center">-6.675</td>
<td align="center">-6.124</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">-6.563</td>
<td align="center">-6.678</td>
<td align="center">0</td>
<td align="center">-6.544</td>
<td align="center">-6.46</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">-6.308</td>
<td align="center">-6.675</td>
<td align="center">-6.544</td>
<td align="center">0</td>
<td align="center">-6.356</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">-6.153</td>
<td align="center">-6.124</td>
<td align="center">-6.46</td>
<td align="center">-6.356</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>RCbray (Bray-Curtis-based Raup-Crick) can be calculated using function cal_rcbray()
to assess whether the compositional turnover was governed primarily by drift <span class="citation">(<a href="#ref-Chase_null_2011" role="doc-biblioref">Chase et al. 2011</a>)</span>.
We applied null model to simulate species distribution by randomly sampling individuals from each
species pool with preserving species occurrence frequency and sample species richness <span class="citation">(<a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>.</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="model-based-class.html#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="co"># result stored in t1$res_rcbray</span></span>
<span id="cb146-2"><a href="model-based-class.html#cb146-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_rcbray</span>(<span class="at">runs =</span> <span class="dv">1000</span>)</span>
<span id="cb146-3"><a href="model-based-class.html#cb146-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_rcbray</span></span></code></pre></div>
<p>As an example, we also calculate the proportion of the inferred processes on the community assembly as shown in the references <span class="citation">(<a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>; <a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>.
In the example, the fraction of pairwise comparisons with significant betaNTI values (|βNTI| &gt; 2) is the estimated influence of Selection;
βNTI &gt; 2 represents the heterogeneous selection; βNTI &lt; -2 represents the homogeneous selection.
The value of RCbray characterizes the magnitude of deviation between observed Bray–Curtis and Bray–Curtis expected under the randomization;
a value of |RCbray| &gt; 0.95 was considered significant.
The fraction of all pairwise comparisons with |βNTI| &lt; 2 and RCbray &gt; +0.95 was taken as the influence of Dispersal Limitation combined with Drift.
The fraction of all pairwise comparisons with |βNTI| &lt; 2 and RCbray &lt; -0.95 was taken as an estimate for the influence of Homogenizing Dispersal.
The fraction of all pairwise comparisons with |βNTI| &lt; 2 and |RCbray| &lt; 0.95 estimates the influence of Drift acting alone.</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="model-based-class.html#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use betaNTI and rcbray to evaluate processes</span></span>
<span id="cb147-2"><a href="model-based-class.html#cb147-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_process</span>(<span class="at">use_betamntd =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## The result is stored in object$res_process ...</code></pre>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="model-based-class.html#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_process</span></span></code></pre></div>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="model-based-class.html#cb150-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_process</span></code></pre></div>
<table style="width:51%;">
<colgroup>
<col width="33%" />
<col width="18%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">process</th>
<th align="center">percentage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">variable selection</td>
<td align="center">3.995</td>
</tr>
<tr class="even">
<td align="center">homogeneous selection</td>
<td align="center">48.34</td>
</tr>
<tr class="odd">
<td align="center">dispersal limitation</td>
<td align="center">0.02497</td>
</tr>
<tr class="even">
<td align="center">homogeneous dispersal</td>
<td align="center">8.539</td>
</tr>
<tr class="odd">
<td align="center">drift</td>
<td align="center">39.1</td>
</tr>
</tbody>
</table>
</div>
<div id="key-points-7" class="section level3 hasAnchor" number="6.3.2">
<h3><span class="header-section-number">6.3.2</span> Key points<a href="model-based-class.html#key-points-7" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>trans_nullmodel$new: filter_thres parameter for the filtering of taxa with relative low abundance</li>
<li>cal_rcbray(): if you just need rcbray, ignore other phylogenetic operations</li>
</ul>
</div>
<div id="other-function" class="section level3 hasAnchor" number="6.3.3">
<h3><span class="header-section-number">6.3.3</span> Other function<a href="model-based-class.html#other-function" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>cal_Cscore(): calculates the (normalised) mean number of checkerboard combinations (C-score) using C.score</li>
<li>cal_NRI(): calculates Nearest Relative Index (NRI)</li>
<li>cal_NTI(): calculates Nearest Taxon Index (NTI)</li>
</ul>
</div>
</div>
<div id="trans_classifier-class" class="section level2 hasAnchor" number="6.4">
<h2><span class="header-section-number">6.4</span> trans_classifier class<a href="model-based-class.html#trans_classifier-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The trans_classifier class is a wrapper for methods of machine-learning-based classification models.
Microbiome-based supervised machine-learning has been successful in predicting human health status <span class="citation">(<a href="#ref-Poore_Microbiome_2020" role="doc-biblioref">Poore et al. 2020</a>)</span>
and soil categories <span class="citation">(<a href="#ref-Wilhelm_Predicting_2021" role="doc-biblioref">Wilhelm, Es, and Buckley 2021</a>)</span>.</p>
<div id="dependencies" class="section level3 hasAnchor" number="6.4.1">
<h3><span class="header-section-number">6.4.1</span> Dependencies<a href="model-based-class.html#dependencies" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Before starting the examples, make sure those packages have been installed.</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="model-based-class.html#cb151-1" aria-hidden="true" tabindex="-1"></a>packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Boruta&quot;</span>, <span class="st">&quot;parallel&quot;</span>, <span class="st">&quot;rsample&quot;</span>, <span class="st">&quot;randomForest&quot;</span>, <span class="st">&quot;caret&quot;</span>, <span class="st">&quot;gridExtra&quot;</span>, <span class="st">&quot;ROCR&quot;</span>)</span>
<span id="cb151-2"><a href="model-based-class.html#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Now check or install</span></span>
<span id="cb151-3"><a href="model-based-class.html#cb151-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> packages){</span>
<span id="cb151-4"><a href="model-based-class.html#cb151-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(x, <span class="at">character.only =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb151-5"><a href="model-based-class.html#cb151-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">install.packages</span>(x, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb151-6"><a href="model-based-class.html#cb151-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb151-7"><a href="model-based-class.html#cb151-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="examples" class="section level3 hasAnchor" number="6.4.2">
<h3><span class="header-section-number">6.4.2</span> Examples<a href="model-based-class.html#examples" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this tutorial, we use the example data in file2meco to demonstrate the feature selection, data training and prediction with random forest algorithm.</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="model-based-class.html#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb152-2"><a href="model-based-class.html#cb152-2" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;dada2_table.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb152-3"><a href="model-based-class.html#cb152-3" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sample-metadata.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb152-4"><a href="model-based-class.html#cb152-4" aria-hidden="true" tabindex="-1"></a>taxonomy_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;taxonomy.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb152-5"><a href="model-based-class.html#cb152-5" aria-hidden="true" tabindex="-1"></a><span class="co"># construct microtable object</span></span>
<span id="cb152-6"><a href="model-based-class.html#cb152-6" aria-hidden="true" tabindex="-1"></a>d1 <span class="ot">&lt;-</span> <span class="fu">qiime2meco</span>(<span class="at">ASV_data =</span> abund_file_path, <span class="at">sample_data =</span> sample_file_path, <span class="at">taxonomy_data =</span> taxonomy_file_path)</span>
<span id="cb152-7"><a href="model-based-class.html#cb152-7" aria-hidden="true" tabindex="-1"></a>d1<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb152-8"><a href="model-based-class.html#cb152-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb152-9"><a href="model-based-class.html#cb152-9" aria-hidden="true" tabindex="-1"></a><span class="co"># initialize: use &quot;genotype&quot; as response variable</span></span>
<span id="cb152-10"><a href="model-based-class.html#cb152-10" aria-hidden="true" tabindex="-1"></a><span class="co"># x.predictors parameter is used to select the taxa; here we use all the taxa data in d1$taxa_abund</span></span>
<span id="cb152-11"><a href="model-based-class.html#cb152-11" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_classifier<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> d1, <span class="at">y.response =</span> <span class="st">&quot;genotype&quot;</span>, <span class="at">x.predictors =</span> <span class="st">&quot;All&quot;</span>)</span></code></pre></div>
<p>Then, we attempt to use Boruta package <span class="citation">(<a href="#ref-Kursa_Feature_2010" role="doc-biblioref">Kursa and Rudnicki 2010</a>)</span> to do feature selection.</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="model-based-class.html#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require Boruta package</span></span>
<span id="cb153-2"><a href="model-based-class.html#cb153-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_feature_sel</span>(<span class="at">boruta.maxRuns =</span> <span class="dv">300</span>, <span class="at">boruta.pValue =</span> <span class="fl">0.01</span>)</span></code></pre></div>
<p>We silit the data into training and testing set.</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="model-based-class.html#cb154-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generate train and test set</span></span>
<span id="cb154-2"><a href="model-based-class.html#cb154-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_split</span>(<span class="at">prop.train =</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>)</span></code></pre></div>
<p>Before training the model, we run the set_trainControl to invoke the trainControl function of caret package to generate the parameters used for training.
Here we use the default parameters in trainControl function.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="model-based-class.html#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require caret package</span></span>
<span id="cb155-2"><a href="model-based-class.html#cb155-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">set_trainControl</span>()</span></code></pre></div>
<p>Now let’s start model training with rf method.</p>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="model-based-class.html#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use default parameter method = &quot;rf&quot;</span></span>
<span id="cb156-2"><a href="model-based-class.html#cb156-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_train</span>(<span class="at">max.ntree =</span> <span class="dv">500</span>)</span></code></pre></div>
<p>We can use cal_predict function to predict the testing data set.</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="model-based-class.html#cb157-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_predict</span>()</span>
<span id="cb157-2"><a href="model-based-class.html#cb157-2" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the confusionMatrix to check out the performance</span></span>
<span id="cb157-3"><a href="model-based-class.html#cb157-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_confusionMatrix</span>()</span></code></pre></div>
<p><img src="Images/plot_confusionMatrix_with_selection.png" width="600px" style="display: block; margin: auto;" /></p>
<p>Using cal_ROC and plot_ROC can get the ROC curve.</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="model-based-class.html#cb158-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ROC</span>()</span>
<span id="cb158-2"><a href="model-based-class.html#cb158-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ROC</span>(<span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>)</span></code></pre></div>
<p><img src="Images/plot_ROC_with_selection.png" width="500px" style="display: block; margin: auto;" /></p>
<p>To compare the results between the procedure with feature selection and that without feature selection,
we also perfom all the analysis without feature selection to show the results.</p>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="model-based-class.html#cb159-1" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_classifier<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> d1, <span class="at">y.response =</span> <span class="st">&quot;genotype&quot;</span>, <span class="at">x.predictors =</span> <span class="st">&quot;All&quot;</span>)</span>
<span id="cb159-2"><a href="model-based-class.html#cb159-2" aria-hidden="true" tabindex="-1"></a><span class="co"># no feature selection, others same with previous steps</span></span>
<span id="cb159-3"><a href="model-based-class.html#cb159-3" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_split</span>(<span class="at">prop.train =</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>)</span>
<span id="cb159-4"><a href="model-based-class.html#cb159-4" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">set_trainControl</span>()</span>
<span id="cb159-5"><a href="model-based-class.html#cb159-5" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_train</span>(<span class="at">max.ntree =</span> <span class="dv">500</span>)</span>
<span id="cb159-6"><a href="model-based-class.html#cb159-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_predict</span>()</span>
<span id="cb159-7"><a href="model-based-class.html#cb159-7" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_confusionMatrix</span>()</span>
<span id="cb159-8"><a href="model-based-class.html#cb159-8" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_ROC</span>()</span>
<span id="cb159-9"><a href="model-based-class.html#cb159-9" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_ROC</span>(<span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>)</span></code></pre></div>
<p><img src="Images/plot_confusionMatrix_without_selection.png" width="600px" style="display: block; margin: auto;" />
<img src="Images/plot_ROC_without_selection.png" width="500px" style="display: block; margin: auto;" /></p>
</div>
<div id="key-points-8" class="section level3 hasAnchor" number="6.4.3">
<h3><span class="header-section-number">6.4.3</span> Key points<a href="model-based-class.html#key-points-8" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>cal_feature_sel(): perform feature selection</li>
</ul>
</div>
<div id="other-function-1" class="section level3 hasAnchor" number="6.4.4">
<h3><span class="header-section-number">6.4.4</span> Other function<a href="model-based-class.html#other-function-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>cal_feature_imp(): get feature importance from the training model when method is “rf”</li>
</ul>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Beck_Machine_2014" class="csl-entry">
Beck, Daniel, and James A Foster. 2014. <span>“Machine Learning Techniques Accurately Classify Microbial Communities by Bacterial Vaginosis Characteristics.”</span> Journal Article. <em>PloS One</em> 9 (2): e87830.
</div>
<div id="ref-Chase_null_2011" class="csl-entry">
Chase, Jonathan M, Nathan JB Kraft, Kevin G Smith, Mark Vellend, and Brian D Inouye. 2011. <span>“Using Null Models to Disentangle Variation in Community Dissimilarity from Variation in α-Diversity.”</span> Journal Article. <em>Ecosphere</em> 2 (2): art24. <a href="https://doi.org/10.1890/ES10-00117.1">https://doi.org/10.1890/ES10-00117.1</a>.
</div>
<div id="ref-Coyte_Theecology_2015" class="csl-entry">
Coyte, Katharine Z, Jonas Schluter, and Kevin R Foster. 2015. <span>“The Ecology of the Microbiome: Networks, Competition, and Stability.”</span> Journal Article. <em>Science</em> 350 (6261): 663–66.
</div>
<div id="ref-Deng_Molecular_2012" class="csl-entry">
Deng, Ye, Yi-Huei Jiang, Yunfeng Yang, Zhili He, Feng Luo, and Jizhong Zhou. 2012. <span>“Molecular Ecological Network Analyses.”</span> Journal Article. <em>BMC Bioinformatics</em> 13 (1): 113.
</div>
<div id="ref-Faust_Microbial_2012" class="csl-entry">
Faust, Karoline, and Jeroen Raes. 2012. <span>“Microbial Interactions: From Networks to Models.”</span> Journal Article. <em>Nature Reviews Microbiology</em> 10 (8): 538–50.
</div>
<div id="ref-Foster_Metacoder_2017" class="csl-entry">
Foster, Zachary, Thomas J Sharpton, and Niklaus J Grunwald. 2017. <span>“Metacoder: An r Package for Visualization and Manipulation of Community Taxonomic Diversity Data.”</span> Journal Article. <em>PLOS Computational Biology</em> 13 (2): e1005404.
</div>
<div id="ref-Picante_Kembel_2010" class="csl-entry">
Kembel, S. W., P. D. Cowan, M. R. Helmus, W. K. Cornwell, H. Morlon, D. D. Ackerly, S. P. Blomberg, and C. O. Webb. 2010. <span>“Picante: <span>R</span> Tools for Integrating Phylogenies and Ecology.”</span> <em>Bioinformatics</em> 26: 1463–64.
</div>
<div id="ref-Kursa_Feature_2010" class="csl-entry">
Kursa, Miron B, and Witold R Rudnicki. 2010. <span>“Feature Selection with the Boruta Package.”</span> Journal Article. <em>J Stat Softw</em> 36 (11): 1–13.
</div>
<div id="ref-Kurtz_Sparse_2015" class="csl-entry">
Kurtz, Z. D., C. L. Muller, E. R. Miraldi, D. R. Littman, M. J. Blaser, and R. A. Bonneau. 2015. <span>“Sparse and Compositionally Robust Inference of Microbial Ecological Networks.”</span> Journal Article. <em>PLoS Comput Biol</em> 11 (5): e1004226. <a href="https://doi.org/10.1371/journal.pcbi.1004226">https://doi.org/10.1371/journal.pcbi.1004226</a>.
</div>
<div id="ref-Liu_Long_term_2017" class="csl-entry">
Liu, Chi, Minjie Yao, James C Stegen, Junpeng Rui, Jiabao Li, and Xiangzhen Li. 2017. <span>“Long-Term Nitrogen Addition Affects the Phylogenetic Turnover of Soil Microbial Community Responding to Moisture Pulse.”</span> Journal Article. <em>Scientific Reports</em> 7 (1): 17492.
</div>
<div id="ref-Paulson_Differential_2013" class="csl-entry">
Paulson, Joseph N, O Colin Stine, Héctor Corrada Bravo, and Mihai Pop. 2013. <span>“Differential Abundance Analysis for Microbial Marker-Gene Surveys.”</span> Journal Article. <em>Nature Methods</em> 10 (12): 1200–1202.
</div>
<div id="ref-Poore_Microbiome_2020" class="csl-entry">
Poore, Gregory D, Evguenia Kopylova, Qiyun Zhu, Carolina Carpenter, Serena Fraraccio, Stephen Wandro, Tomasz Kosciolek, Stefan Janssen, Jessica Metcalf, and Se Jin Song. 2020. <span>“Microbiome Analyses of Blood and Tissues Suggest Cancer Diagnostic Approach.”</span> Journal Article. <em>Nature</em> 579 (7800): 567–74.
</div>
<div id="ref-Segata_Metagenomic_2011" class="csl-entry">
Segata, Nicola, Jacques Izard, Levi Waldron, Dirk Gevers, Larisa Miropolsky, Wendy S Garrett, and Curtis Huttenhower. 2011. <span>“Metagenomic Biomarker Discovery and Explanation.”</span> Journal Article. <em>Genome Biology</em> 12 (6): R60.
</div>
<div id="ref-Stegen_Quantifying_2013" class="csl-entry">
Stegen, James C, Xueju Lin, Jim K Fredrickson, Xingyuan Chen, David W Kennedy, Christopher J Murray, Mark L Rockhold, and Allan Konopka. 2013. <span>“Quantifying Community Assembly Processes and Identifying Features That Impose Them.”</span> Journal Article. <em>The ISME Journal</em> 7 (11): 2069–79.
</div>
<div id="ref-Tackmann_Rapid_2019" class="csl-entry">
Tackmann, J., J. F. Matias Rodrigues, and C. von Mering. 2019. <span>“Rapid Inference of Direct Interactions in Large-Scale Ecological Networks from Heterogeneous Microbial Sequencing Data.”</span> Journal Article. <em>Cell Systems</em> 9 (3): 286–296 e8. <a href="https://doi.org/10.1016/j.cels.2019.08.002">https://doi.org/10.1016/j.cels.2019.08.002</a>.
</div>
<div id="ref-Webb_Phylogenies_2002" class="csl-entry">
Webb, Campbell O, David D Ackerly, Mark A McPeek, and Michael J Donoghue. 2002. <span>“Phylogenies and Community Ecology.”</span> Journal Article. <em>Annu. Rev. Ecol. Syst.</em> 33: 475–505. <a href="https://doi.org/10.1146/annurev.ecolsys.33.010802.150448">https://doi.org/10.1146/annurev.ecolsys.33.010802.150448</a>.
</div>
<div id="ref-White_Statistical_2009" class="csl-entry">
White, JR, N Nagarajan, and M Pop. 2009. <span>“Statistical Methods for Detecting Differentially Abundant Features in Clinical Metagenomic Samples.”</span> Journal Article. <em>PLoS Computational Biology</em> 5 (4): e1000352.
</div>
<div id="ref-Wilhelm_Predicting_2021" class="csl-entry">
Wilhelm, Roland C, Harold M van Es, and Daniel H Buckley. 2021. <span>“Predicting Measures of Soil Health Using the Microbiome and Supervised Machine Learning.”</span> Journal Article. <em>Soil Biology and Biochemistry</em>, 108472.
</div>
<div id="ref-Yatsunenko_Human_2012" class="csl-entry">
Yatsunenko, Tanya, Federico E Rey, Mark J Manary, Indi Trehan, Maria Gloria Dominguez-Bello, Monica Contreras, Magda Magris, Glida Hidalgo, Robert N Baldassano, and Andrey P Anokhin. 2012. <span>“Human Gut Microbiome Viewed Across Age and Geography.”</span> Journal Article. <em>Nature</em> 486 (7402): 222–27.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="diversity-based-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="explainable-class.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/05-Model-based_class.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
