<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Model-based class | Tutorial for R microeco package (v0.17.0)</title>
  <meta name="description" content="The tutorial for R microeco, file2meco, meconetcomp and mecoturn packages" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Model-based class | Tutorial for R microeco package (v0.17.0)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecoturn packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Model-based class | Tutorial for R microeco package (v0.17.0)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecoturn packages" />
  

<meta name="author" content="Chi Liu, Felipe R. P. Mansoldo, Umer Zeeshan Ijaz, Chenhao Li, Yang Cao, Jarrod J. Scott, Yaoming Cui, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2023-05-03" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="diversity-based-class.html"/>
<link rel="next" href="explainable-class.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#rtools"><i class="fa fa-check"></i><b>2.4</b> RTools</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.5</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#description"><i class="fa fa-check"></i><b>2.5.1</b> Description</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#cran-packages"><i class="fa fa-check"></i><b>2.5.2</b> CRAN packages</a></li>
<li class="chapter" data-level="2.5.3" data-path="intro.html"><a href="intro.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor packages</a></li>
<li class="chapter" data-level="2.5.4" data-path="intro.html"><a href="intro.html#github-packages"><i class="fa fa-check"></i><b>2.5.4</b> Github packages</a></li>
<li class="chapter" data-level="2.5.5" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.5.5</b> Gephi</a></li>
<li class="chapter" data-level="2.5.6" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.5.6</b> Tax4Fun</a></li>
<li class="chapter" data-level="2.5.7" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.5.7</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#plot"><i class="fa fa-check"></i><b>2.6</b> Plot</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#prepare-the-example-data"><i class="fa fa-check"></i><b>3.1.1</b> Prepare the example data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#how-to-read-your-files-to-microtable-object"><i class="fa fa-check"></i><b>3.1.2</b> How to read your files to microtable object?</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#functions-in-microtable-class"><i class="fa fa-check"></i><b>3.1.3</b> Functions in microtable class</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#merge-taxa-or-samples"><i class="fa fa-check"></i><b>3.1.4</b> merge taxa or samples</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-class.html"><a href="basic-class.html#subset-of-samples"><i class="fa fa-check"></i><b>3.1.5</b> subset of samples</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-class.html"><a href="basic-class.html#subset-of-taxa"><i class="fa fa-check"></i><b>3.1.6</b> subset of taxa</a></li>
<li class="chapter" data-level="3.1.7" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.7</b> Other examples</a></li>
<li class="chapter" data-level="3.1.8" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.8</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-2"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-4"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#network-comparison"><i class="fa fa-check"></i><b>6.2.2</b> Network comparison</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.3</b> Key points</a></li>
<li class="chapter" data-level="6.2.4" data-path="model-based-class.html"><a href="model-based-class.html#other-functions"><i class="fa fa-check"></i><b>6.2.4</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies</a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples</a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points</a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-7"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME</a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2</a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann"><i class="fa fa-check"></i><b>8.3</b> HUMAnN</a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn</a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2"><i class="fa fa-check"></i><b>8.5</b> Kraken2</a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB</a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq</a></li>
<li class="chapter" data-level="8.8" data-path="file2meco-package.html"><a href="file2meco-package.html#picrust2"><i class="fa fa-check"></i><b>8.8</b> PICRUSt2</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset</a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data</a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> <code>clone</code> function</a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> <code>save</code> function</a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#github-issues"><i class="fa fa-check"></i><b>10.3</b> Github Issues</a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.4</b> change object</a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.5</b> group order</a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.6</b> add layers to plot</a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#colors-for-more-groups"><i class="fa fa-check"></i><b>10.7</b> colors for more groups</a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>11</b> mecodev package</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>11.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="11.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_norm"><i class="fa fa-check"></i><b>11.2</b> trans_norm</a></li>
<li class="chapter" data-level="11.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>11.3</b> trans_ts</a></li>
<li class="chapter" data-level="11.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>11.4</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html"><i class="fa fa-check"></i><b>12</b> meconetcomp package</a>
<ul>
<li class="chapter" data-level="12.1" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-modularity-for-all-networks"><i class="fa fa-check"></i><b>12.1</b> Network modularity for all networks</a></li>
<li class="chapter" data-level="12.2" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-topological-attributes-for-all-networks"><i class="fa fa-check"></i><b>12.2</b> Network topological attributes for all networks</a></li>
<li class="chapter" data-level="12.3" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#node-and-edge-properties-extraction-for-all-networks"><i class="fa fa-check"></i><b>12.3</b> Node and edge properties extraction for all networks</a></li>
<li class="chapter" data-level="12.4" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-nodes-across-networks"><i class="fa fa-check"></i><b>12.4</b> Compare nodes across networks</a></li>
<li class="chapter" data-level="12.5" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-edges-across-networks"><i class="fa fa-check"></i><b>12.5</b> Compare edges across networks</a></li>
<li class="chapter" data-level="12.6" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#extract-overlapped-edges-of-networks-to-a-new-network"><i class="fa fa-check"></i><b>12.6</b> Extract overlapped edges of networks to a new network</a></li>
<li class="chapter" data-level="12.7" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-phylogenetic-distances-of-paired-nodes-in-edges"><i class="fa fa-check"></i><b>12.7</b> Compare phylogenetic distances of paired nodes in edges</a></li>
<li class="chapter" data-level="12.8" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-node-sources-of-edges-across-networks"><i class="fa fa-check"></i><b>12.8</b> Compare node sources of edges across networks</a></li>
<li class="chapter" data-level="12.9" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-topological-properties-of-sub-networks"><i class="fa fa-check"></i><b>12.9</b> Compare topological properties of sub-networks</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mecoturn-package.html"><a href="mecoturn-package.html"><i class="fa fa-check"></i><b>13</b> mecoturn package</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mecoturn-package.html"><a href="mecoturn-package.html#betaturn-class"><i class="fa fa-check"></i><b>13.1</b> betaturn class</a></li>
<li class="chapter" data-level="13.2" data-path="mecoturn-package.html"><a href="mecoturn-package.html#taxaturn-class"><i class="fa fa-check"></i><b>13.2</b> taxaturn class</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>14</b> Other examples</a>
<ul>
<li class="chapter" data-level="14.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>14.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="14.2" data-path="other-examples-1.html"><a href="other-examples-1.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>14.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="14.3" data-path="other-examples-1.html"><a href="other-examples-1.html#show-the-abundance-of-unknown-taxa"><i class="fa fa-check"></i><b>14.3</b> Show the abundance of unknown taxa</a></li>
<li class="chapter" data-level="14.4" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>14.4</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="14.5" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>14.5</b> The use of phylogenetic tree</a></li>
<li class="chapter" data-level="14.6" data-path="other-examples-1.html"><a href="other-examples-1.html#mantel-test-correlation-heatmap"><i class="fa fa-check"></i><b>14.6</b> Mantel test + Correlation heatmap</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v0.17.0)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="model-based-class" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Chapter 6</span> Model-based class<a href="model-based-class.html#model-based-class" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>All the classes with complex models are grouped into this section.</p>
<div id="trans_diff-class" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> trans_diff class<a href="model-based-class.html#trans_diff-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　Differential abundance test is an important part in the microbiome profiling analysis.
It can find the significant taxa in determining community differences across groups.
Different approaches may produce inconsistent results since the underlying models/hypothesis are different <span class="citation">(<a href="#ref-Nearing_Microbiome_2022" role="doc-biblioref">Nearing et al. 2022</a>)</span>.
Currently, trans_diff class has multiple famous differential test approaches or wrapped methods to better capture the important biomarkers:
metastat<span class="citation">(<a href="#ref-White_Statistical_2009" role="doc-biblioref">White, Nagarajan, and Pop 2009</a>)</span>, LEfSe<span class="citation">(<a href="#ref-Segata_Metagenomic_2011" role="doc-biblioref">Segata et al. 2011</a>)</span>, RF (random forest + differential test), metagenomeSeq<span class="citation">(<a href="#ref-Paulson_Differential_2013" role="doc-biblioref">Paulson et al. 2013</a>)</span>,
Kruskal-Wallis Rank Sum Test (for groups &gt; 2), Wilcoxon Rank Sum Tests (for each paired group) and
Dunn’s Kruskal-Wallis Multiple Comparisons (for paired group in cases groups &gt; 2),
t.test, ANOVA, Scheirer Ray Hare test,
ANCOMBC <span class="citation">(<a href="#ref-Lin_Analysis_2020" role="doc-biblioref">Lin and Peddada 2020</a>)</span>, ALDEx2 <span class="citation">(<a href="#ref-Fernandes_Unifying_2014" role="doc-biblioref">Fernandes et al. 2014</a>)</span>,
DESeq2 <span class="citation">(<a href="#ref-Michael_Moderated_2014" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span>, LinDA <span class="citation">(<a href="#ref-LinDA_Zhou_2022" role="doc-biblioref">Zhou et al. 2022</a>)</span>, beta regression <span class="citation">(<a href="#ref-Betaregression_2010" role="doc-biblioref">Cribari-Neto and Zeileis 2010</a>)</span>, etc.
Given that multiple approaches are available,
it is easy and feasible to compare the results from different approaches and extract a part of biomarkers with high confidence for a specific dataset.</p>
<div id="example-4" class="section level3 hasAnchor" number="6.1.1">
<h3><span class="header-section-number">6.1.1</span> Example<a href="model-based-class.html#example-4" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>All the differential test result is stored in the <code>object$res_diff</code>.
LEfSe combines the non-parametric test and linear discriminant analysis <span class="citation">(<a href="#ref-Segata_Metagenomic_2011" role="doc-biblioref">Segata et al. 2011</a>)</span>.</p>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="model-based-class.html#cb146-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;lefse&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">alpha =</span> <span class="fl">0.01</span>, <span class="at">lefse_subgroup =</span> <span class="cn">NULL</span>)</span>
<span id="cb146-2"><a href="model-based-class.html#cb146-2" aria-hidden="true" tabindex="-1"></a><span class="co"># see t1$res_diff for the result</span></span>
<span id="cb146-3"><a href="model-based-class.html#cb146-3" aria-hidden="true" tabindex="-1"></a><span class="co"># From v0.8.0, threshold is used for the LDA score selection.</span></span>
<span id="cb146-4"><a href="model-based-class.html#cb146-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">threshold =</span> <span class="dv">4</span>)</span>
<span id="cb146-5"><a href="model-based-class.html#cb146-5" aria-hidden="true" tabindex="-1"></a><span class="co"># we show 20 taxa with the highest LDA (log10)</span></span>
<span id="cb146-6"><a href="model-based-class.html#cb146-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">width =</span> <span class="fl">0.8</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>, <span class="st">&quot;TW&quot;</span>))</span></code></pre></div>
<p><img src="Images/trans_diff_lefse_bar.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="model-based-class.html#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show part of the table</span></span>
<span id="cb147-2"><a href="model-based-class.html#cb147-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_diff[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">6</span>)]</span></code></pre></div>
<table>
<colgroup>
<col width="66%" />
<col width="9%" />
<col width="9%" />
<col width="14%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Taxa</th>
<th align="center">Group</th>
<th align="center">LDA</th>
<th align="center">P.adj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">k__Bacteria|p__Proteobacteria</td>
<td align="center">CW</td>
<td align="center">4.837</td>
<td align="center">1.076e-09</td>
</tr>
<tr class="even">
<td align="center">k__Bacteria|p__Acidobacteria</td>
<td align="center">IW</td>
<td align="center">4.797</td>
<td align="center">2.955e-10</td>
</tr>
<tr class="odd">
<td align="center">k__Bacteria|p__Acidobacteria|c__Acidobacteria</td>
<td align="center">IW</td>
<td align="center">4.797</td>
<td align="center">6.946e-11</td>
</tr>
<tr class="even">
<td align="center">k__Bacteria|p__Bacteroidetes</td>
<td align="center">TW</td>
<td align="center">4.782</td>
<td align="center">1.529e-08</td>
</tr>
<tr class="odd">
<td align="center">k__Bacteria|p__Proteobacteria|c__Gammaproteobacteria</td>
<td align="center">CW</td>
<td align="center">4.613</td>
<td align="center">2.911e-10</td>
</tr>
</tbody>
</table>
<p>Then, we plot the abundance of biomarkers detected by LEfSe.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="model-based-class.html#cb148-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>, <span class="st">&quot;TW&quot;</span>))</span></code></pre></div>
<p><img src="Images/trans_diff_lefse_abund.png" width="650px" style="display: block; margin: auto;" /></p>
<p>Then, we show the cladogram of the differential features in the taxonomic tree.
There are too many taxa in this dataset.
As an example, we only select the highest 200 abundant taxa in the tree and 50 differential features.
We only show the full taxonomic label at Phylum level and use letters at other levels to reduce the text overlap.
<strong>Note that if an error occurs in this function, the reason with a high probability is the chaotic taxonomy in the user’s data</strong>.
<strong>Please see the <code>tidy_taxonomy</code> function of microtable class part to solve this issue.</strong></p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="model-based-class.html#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clade_label_level 5 represent phylum level in this analysis</span></span>
<span id="cb149-2"><a href="model-based-class.html#cb149-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require ggtree package</span></span>
<span id="cb149-3"><a href="model-based-class.html#cb149-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_cladogram</span>(<span class="at">use_taxa_num =</span> <span class="dv">200</span>, <span class="at">use_feature_num =</span> <span class="dv">50</span>, <span class="at">clade_label_level =</span> <span class="dv">5</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>, <span class="st">&quot;TW&quot;</span>))</span></code></pre></div>
<p><img src="Images/trans_diff_lefse_cladogram.png" width="1000px" style="display: block; margin: auto;" /></p>
<p>There may be a problem related with the taxonomic labels in the plot.
When there are too many levels shown, the taxonomic labels can have too much overlap.
However, if only Phylum labels are indicated, the taxa in the legend with marked letters are too many.
At this time, taxa can be manually choosed to show like the following operation.</p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="model-based-class.html#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="co"># choose some taxa according to the positions in the previous picture; those taxa labels have minimum overlap</span></span>
<span id="cb150-2"><a href="model-based-class.html#cb150-2" aria-hidden="true" tabindex="-1"></a>use_labels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;c__Deltaproteobacteria&quot;</span>, <span class="st">&quot;c__Actinobacteria&quot;</span>, <span class="st">&quot;o__Rhizobiales&quot;</span>, <span class="st">&quot;p__Proteobacteria&quot;</span>, <span class="st">&quot;p__Bacteroidetes&quot;</span>, </span>
<span id="cb150-3"><a href="model-based-class.html#cb150-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;o__Micrococcales&quot;</span>, <span class="st">&quot;p__Acidobacteria&quot;</span>, <span class="st">&quot;p__Verrucomicrobia&quot;</span>, <span class="st">&quot;p__Firmicutes&quot;</span>, </span>
<span id="cb150-4"><a href="model-based-class.html#cb150-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;p__Chloroflexi&quot;</span>, <span class="st">&quot;c__Acidobacteria&quot;</span>, <span class="st">&quot;c__Gammaproteobacteria&quot;</span>, <span class="st">&quot;c__Betaproteobacteria&quot;</span>, <span class="st">&quot;c__KD4-96&quot;</span>,</span>
<span id="cb150-5"><a href="model-based-class.html#cb150-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">&quot;c__Bacilli&quot;</span>, <span class="st">&quot;o__Gemmatimonadales&quot;</span>, <span class="st">&quot;f__Gemmatimonadaceae&quot;</span>, <span class="st">&quot;o__Bacillales&quot;</span>, <span class="st">&quot;o__Rhodobacterales&quot;</span>)</span>
<span id="cb150-6"><a href="model-based-class.html#cb150-6" aria-hidden="true" tabindex="-1"></a><span class="co"># then use parameter select_show_labels to show</span></span>
<span id="cb150-7"><a href="model-based-class.html#cb150-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_cladogram</span>(<span class="at">use_taxa_num =</span> <span class="dv">200</span>, <span class="at">use_feature_num =</span> <span class="dv">50</span>, <span class="at">select_show_labels =</span> use_labels)</span>
<span id="cb150-8"><a href="model-based-class.html#cb150-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we can see that more taxa names appear in the tree</span></span></code></pre></div>
<p><img src="Images/trans_diff_lefse_cladogram_1.png" width="1000px" style="display: block; margin: auto;" /></p>
<p>The ‘rf’ method depends on the random forest<span class="citation">(<a href="#ref-Beck_Machine_2014" role="doc-biblioref">Beck and Foster 2014</a>; <a href="#ref-Yatsunenko_Human_2012" role="doc-biblioref">Yatsunenko et al. 2012</a>)</span> and the non-parametric test.
The current method can calculate random forest by bootstrapping like the method in LEfSe and only use the significant features.
MeanDecreaseGini is selected as the indicator value in the analysis.</p>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="model-based-class.html#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Genus level for parameter taxa_level, if you want to use all taxa, change to &quot;all&quot;</span></span>
<span id="cb151-2"><a href="model-based-class.html#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="co"># nresam = 1 and boots = 1 represent no bootstrapping and use all samples directly</span></span>
<span id="cb151-3"><a href="model-based-class.html#cb151-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;rf&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb151-4"><a href="model-based-class.html#cb151-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the MeanDecreaseGini bar</span></span>
<span id="cb151-5"><a href="model-based-class.html#cb151-5" aria-hidden="true" tabindex="-1"></a><span class="co"># group_order is designed to sort the groups</span></span>
<span id="cb151-6"><a href="model-based-class.html#cb151-6" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>))</span>
<span id="cb151-7"><a href="model-based-class.html#cb151-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the abundance using same taxa in g1</span></span>
<span id="cb151-8"><a href="model-based-class.html#cb151-8" aria-hidden="true" tabindex="-1"></a>g2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>), <span class="at">select_taxa =</span> t1<span class="sc">$</span>plot_diff_bar_taxa)</span>
<span id="cb151-9"><a href="model-based-class.html#cb151-9" aria-hidden="true" tabindex="-1"></a><span class="co"># now the y axis in g1 and g2 is same, so we can merge them</span></span>
<span id="cb151-10"><a href="model-based-class.html#cb151-10" aria-hidden="true" tabindex="-1"></a><span class="co"># remove g1 legend; remove g2 y axis text and ticks</span></span>
<span id="cb151-11"><a href="model-based-class.html#cb151-11" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> g1 <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>)</span>
<span id="cb151-12"><a href="model-based-class.html#cb151-12" aria-hidden="true" tabindex="-1"></a>g2 <span class="ot">&lt;-</span> g2 <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>())</span>
<span id="cb151-13"><a href="model-based-class.html#cb151-13" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(g1, g2, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>, <span class="at">widths =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="fl">1.7</span>))</span></code></pre></div>
<p><img src="Images/trans_diff_rf_diff_abund.png" width="800px" style="display: block; margin: auto;" /></p>
<p>The significance label can also be added in the abundance plot controlled by add_sig parameter and other related parameters.
Now adding labels supports all the differential test methods.</p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="model-based-class.html#cb152-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb152-2"><a href="model-based-class.html#cb152-2" aria-hidden="true" tabindex="-1"></a><span class="co"># filter something not needed to show</span></span>
<span id="cb152-3"><a href="model-based-class.html#cb152-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_diff <span class="sc">%&lt;&gt;%</span> <span class="fu">subset</span>(Significance <span class="sc">%in%</span> <span class="st">&quot;***&quot;</span>)</span>
<span id="cb152-4"><a href="model-based-class.html#cb152-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">add_sig =</span> T, <span class="at">add_sig_label =</span> <span class="st">&quot;Significance&quot;</span>)</span></code></pre></div>
<p><img src="Images/trans_diff_siglabel_wilcox.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="model-based-class.html#cb153-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;anova&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb153-2"><a href="model-based-class.html#cb153-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">add_sig =</span> T, <span class="at">coord_flip =</span> F)</span></code></pre></div>
<p><img src="Images/trans_diff_siglabel_anova.png" width="700px" style="display: block; margin: auto;" /></p>
<p>Metastat depends on the permutations and t-test and performs well on the sparse data for paired groups test.</p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="model-based-class.html#cb154-1" aria-hidden="true" tabindex="-1"></a><span class="co"># metastat analysis at Genus level</span></span>
<span id="cb154-2"><a href="model-based-class.html#cb154-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;metastat&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb154-3"><a href="model-based-class.html#cb154-3" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_diff is the differential test result</span></span>
<span id="cb154-4"><a href="model-based-class.html#cb154-4" aria-hidden="true" tabindex="-1"></a><span class="co"># t1$res_abund is the group abundance</span></span></code></pre></div>
<p>Because the example ‘Group’ in sample_table has three groups,
the metastat can run the comparisons for each paired group. So there are three pairs in t1$res_diff$Comparison.
For the abundance plotting, the user should use select_group to select the required pair.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="model-based-class.html#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select_group should be one of groups in t1$res_diff$Comparison</span></span>
<span id="cb155-2"><a href="model-based-class.html#cb155-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">select_group =</span> <span class="st">&quot;CW - TW&quot;</span>, <span class="at">coord_flip =</span> F)</span></code></pre></div>
<p><img src="Images/trans_diff_metastat_1.png" width="650px" style="display: block; margin: auto;" /></p>
<p>The following are the examples for the methods ‘KW’, ‘KW_dunn’, ‘wilcox’, ‘t.test’ and ‘anova’.</p>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="model-based-class.html#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Kruskal-Wallis Rank Sum Test for all groups (&gt;= 2)</span></span>
<span id="cb156-2"><a href="model-based-class.html#cb156-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;KW&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;all&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb156-3"><a href="model-based-class.html#cb156-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb156-4"><a href="model-based-class.html#cb156-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Dunn&#39;s Kruskal-Wallis Multiple Comparisons when group number &gt; 2; require FSA package</span></span>
<span id="cb156-5"><a href="model-based-class.html#cb156-5" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;KW_dunn&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>)</span>
<span id="cb156-6"><a href="model-based-class.html#cb156-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">add_sig =</span> T, <span class="at">coord_flip =</span> F)</span>
<span id="cb156-7"><a href="model-based-class.html#cb156-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Wilcoxon Rank Sum and Signed Rank Tests for all paired groups</span></span>
<span id="cb156-8"><a href="model-based-class.html#cb156-8" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb156-9"><a href="model-based-class.html#cb156-9" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">select_group =</span> <span class="st">&quot;CW - TW&quot;</span>)</span>
<span id="cb156-10"><a href="model-based-class.html#cb156-10" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">select_group =</span> <span class="st">&quot;CW - TW&quot;</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>))</span>
<span id="cb156-11"><a href="model-based-class.html#cb156-11" aria-hidden="true" tabindex="-1"></a><span class="co"># t.test</span></span>
<span id="cb156-12"><a href="model-based-class.html#cb156-12" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;t.test&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;all&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb156-13"><a href="model-based-class.html#cb156-13" aria-hidden="true" tabindex="-1"></a><span class="co"># anova</span></span>
<span id="cb156-14"><a href="model-based-class.html#cb156-14" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;anova&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb156-15"><a href="model-based-class.html#cb156-15" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(t1<span class="sc">$</span>res_diff)</span></code></pre></div>
<p>The method ‘metagenomeSeq’ <span class="citation">(<a href="#ref-Paulson_Differential_2013" role="doc-biblioref">Paulson et al. 2013</a>)</span> and ‘ancombc2’ <span class="citation">(<a href="#ref-Lin_Analysis_2020" role="doc-biblioref">Lin and Peddada 2020</a>)</span> depend on the metagenomeSeq package and ANCOMBC package, respectively.
The method ‘ALDEx2_t’ and ‘ALDEx2_kw’ depend on the ALDEx2 package <span class="citation">(<a href="#ref-Fernandes_Unifying_2014" role="doc-biblioref">Fernandes et al. 2014</a>)</span>.
These three packages are all deposited on the Bioconductor.</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="model-based-class.html#cb157-1" aria-hidden="true" tabindex="-1"></a><span class="co"># zero-inflated log-normal model-based differential test method from metagenomeSeq package</span></span>
<span id="cb157-2"><a href="model-based-class.html#cb157-2" aria-hidden="true" tabindex="-1"></a><span class="co"># If metagenomeSeq package is not installed, please first run: BiocManager::install(&quot;metagenomeSeq&quot;)</span></span>
<span id="cb157-3"><a href="model-based-class.html#cb157-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;metagenomeSeq&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb157-4"><a href="model-based-class.html#cb157-4" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;metagenomeSeq&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>)</span>
<span id="cb157-5"><a href="model-based-class.html#cb157-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>))</span>
<span id="cb157-6"><a href="model-based-class.html#cb157-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code></pre></div>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="model-based-class.html#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="co"># &#39;ALDEx2_t&#39; and &#39;ALDEx2_kw&#39; methods; use ?trans_diff to see detailed description of the methods</span></span>
<span id="cb158-2"><a href="model-based-class.html#cb158-2" aria-hidden="true" tabindex="-1"></a><span class="co"># If ALDEx2 package is not installed, please first run: BiocManager::install(&quot;ALDEx2&quot;)</span></span>
<span id="cb158-3"><a href="model-based-class.html#cb158-3" aria-hidden="true" tabindex="-1"></a><span class="co"># &#39;ALDEx2_t&#39;</span></span>
<span id="cb158-4"><a href="model-based-class.html#cb158-4" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;ALDEx2_t&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb158-5"><a href="model-based-class.html#cb158-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>))</span>
<span id="cb158-6"><a href="model-based-class.html#cb158-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">select_group =</span> <span class="st">&quot;CW - TW&quot;</span>)</span>
<span id="cb158-7"><a href="model-based-class.html#cb158-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">select_group =</span> <span class="st">&quot;CW - TW&quot;</span>, <span class="at">add_sig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb158-8"><a href="model-based-class.html#cb158-8" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;ALDEx2_t&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.0005</span>)</span>
<span id="cb158-9"><a href="model-based-class.html#cb158-9" aria-hidden="true" tabindex="-1"></a><span class="co"># ALDEx2_kw</span></span>
<span id="cb158-10"><a href="model-based-class.html#cb158-10" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;ALDEx2_kw&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb158-11"><a href="model-based-class.html#cb158-11" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>))</span>
<span id="cb158-12"><a href="model-based-class.html#cb158-12" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb158-13"><a href="model-based-class.html#cb158-13" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>), <span class="at">add_sig =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb159-1"><a href="model-based-class.html#cb159-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ANCOMBC method</span></span>
<span id="cb159-2"><a href="model-based-class.html#cb159-2" aria-hidden="true" tabindex="-1"></a><span class="co"># If ANCOMBC package is not installed, please first run: BiocManager::install(&quot;ANCOMBC&quot;)</span></span>
<span id="cb159-3"><a href="model-based-class.html#cb159-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;ancombc2&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Family&quot;</span>)</span>
<span id="cb159-4"><a href="model-based-class.html#cb159-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">select_group =</span> <span class="st">&quot;CW - TW&quot;</span>)</span>
<span id="cb159-5"><a href="model-based-class.html#cb159-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>), <span class="at">add_sig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb159-6"><a href="model-based-class.html#cb159-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span></code></pre></div>
<p>The method ‘linda’ <span class="citation">(<a href="#ref-LinDA_Zhou_2022" role="doc-biblioref">Zhou et al. 2022</a>)</span> depends on the MicrobiomeStat package.</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="model-based-class.html#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="co"># LinDA method</span></span>
<span id="cb160-2"><a href="model-based-class.html#cb160-2" aria-hidden="true" tabindex="-1"></a><span class="co"># If MicrobiomeStat package is not installed, please first run: install.packages(&quot;MicrobiomeStat&quot;)</span></span>
<span id="cb160-3"><a href="model-based-class.html#cb160-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;linda&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>)</span>
<span id="cb160-4"><a href="model-based-class.html#cb160-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">group_order =</span> <span class="fu">c</span>(<span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>, <span class="st">&quot;IW&quot;</span>), <span class="at">add_sig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb160-5"><a href="model-based-class.html#cb160-5" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">method =</span> <span class="st">&quot;linda&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group+Type&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb160-6"><a href="model-based-class.html#cb160-6" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(t1<span class="sc">$</span>res_diff)</span></code></pre></div>
</div>
<div id="key-points-5" class="section level3 hasAnchor" number="6.1.2">
<h3><span class="header-section-number">6.1.2</span> Key points<a href="model-based-class.html#key-points-5" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>trans_diff$new: In trans_diff$new, p_adjust_method = “none” or NULL can close the p value adjustment.
This is useful in cases where very few significant taxa are found (generally no significant feature found after adjustment) and
where LEfSe result is needed to be compared with that from Galaxy server or other LEfSe python version.</li>
<li>trans_diff$new: this class has a strict requirement on the taxonomic information,
make sure <code>tidy_taxonomy()</code> function has been performed for the <code>tax_table</code> in <code>microtable</code> object.</li>
<li>trans_diff$new: creating this class invokes one or more tables in <code>taxa_abund</code> list, which is stored in <code>microtable</code> object.</li>
<li>trans_diff$plot_diff_cladogram: <code>clade_label_size</code>, <code>clade_label_size_add</code> and <code>clade_label_size_log</code> can control the text size all together in the cladogram.</li>
</ul>
</div>
</div>
<div id="trans_network-class" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> trans_network class<a href="model-based-class.html#trans_network-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>　Network analysis has been frequently used to study microbial co-occurrence patterns <span class="citation">(<a href="#ref-Deng_Molecular_2012" role="doc-biblioref">Deng et al. 2012</a>; <a href="#ref-Faust_Microbial_2012" role="doc-biblioref">Faust and Raes 2012</a>; <a href="#ref-Coyte_Theecology_2015" role="doc-biblioref">Coyte, Schluter, and Foster 2015</a>)</span>.
In this part, we describe part of the implemented methods in the trans_network class.</p>
<p><img src="Images/network_framework.png" width="8000px" style="display: block; margin: auto;" /></p>
<p>The objects inside the rectangle with full line represent functions.
The dashed line denotes the key objects (input or output). The <code>res_network</code> inside the ellipse with dashed line means it is a hub object for other analysis.</p>
<div id="example-5" class="section level3 hasAnchor" number="6.2.1">
<h3><span class="header-section-number">6.2.1</span> Example<a href="model-based-class.html#example-5" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The correlation-based network is selected to show the main operations.
This is only intended to show some operations conveniently.
<strong>Do not mean we are suggesting this approach in any case</strong>.
Please check the final part for other network construction methods.</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="model-based-class.html#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The parameter cor_method in trans_network is used to select correlation calculation method.</span></span>
<span id="cb161-2"><a href="model-based-class.html#cb161-2" aria-hidden="true" tabindex="-1"></a><span class="co"># default pearson or spearman correlation invoke R base cor.test, a little slow</span></span>
<span id="cb161-3"><a href="model-based-class.html#cb161-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb161-4"><a href="model-based-class.html#cb161-4" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_cor_p list, containing two tables: correlation coefficient table and p value table</span></span></code></pre></div>
<p>Spearman correlation based on WGCNA package is applied to show all the following operations.</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="model-based-class.html#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require WGCNA package</span></span>
<span id="cb162-2"><a href="model-based-class.html#cb162-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;WGCNA&quot;</span>)) <span class="fu">install.packages</span>(<span class="st">&quot;WGCNA&quot;</span>, <span class="at">repos =</span> BiocManager<span class="sc">::</span><span class="fu">repositories</span>())</span>
<span id="cb162-3"><a href="model-based-class.html#cb162-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>, <span class="at">use_WGCNA_pearson_spearman =</span> <span class="cn">TRUE</span>, <span class="at">filter_thres =</span> <span class="fl">0.0001</span>)</span></code></pre></div>
<p>The parameter <code>COR_cut</code> can be used to select the correlation threshold.
Furthermore, <code>COR_optimization = TRUE</code> can be used to find the optimized coefficient threshold (potential transition point of network eigenvalues)
instead of the <code>COR_cut</code> based on the RMT theory <span class="citation">(<a href="#ref-Deng_Molecular_2012" role="doc-biblioref">Deng et al. 2012</a>)</span>.</p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="model-based-class.html#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="co"># construct network; require igraph package</span></span>
<span id="cb163-2"><a href="model-based-class.html#cb163-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">COR_p_thres =</span> <span class="fl">0.01</span>, <span class="at">COR_optimization =</span> <span class="cn">TRUE</span>)</span>
<span id="cb163-3"><a href="model-based-class.html#cb163-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use arbitrary coefficient threshold to contruct network</span></span>
<span id="cb163-4"><a href="model-based-class.html#cb163-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">COR_p_thres =</span> <span class="fl">0.01</span>, <span class="at">COR_cut =</span> <span class="fl">0.7</span>)</span>
<span id="cb163-5"><a href="model-based-class.html#cb163-5" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_network</span></span></code></pre></div>
<p>Then, let’s partition modules for the network.</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="model-based-class.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="co"># invoke igraph cluster_fast_greedy function for this undirected network </span></span>
<span id="cb164-2"><a href="model-based-class.html#cb164-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_module</span>(<span class="at">method =</span> <span class="st">&quot;cluster_fast_greedy&quot;</span>)</span></code></pre></div>
<p>The following operation is to save network to gexf format for the visualization with Gephi(<a href="https://gephi.org/" class="uri">https://gephi.org/</a>).
The file ‘network.gexf’ can be directly opened by Gephi.</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="model-based-class.html#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require rgexf package to be installed</span></span>
<span id="cb165-2"><a href="model-based-class.html#cb165-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">save_network</span>(<span class="at">filepath =</span> <span class="st">&quot;network.gexf&quot;</span>)</span></code></pre></div>
<p>Then, we plot the network in Gephi and present the node colors according to the partitioned modules.</p>
<p><img src="Images/network1_spearman.png" width="550px" style="display: block; margin: auto;" /></p>
<p>Now, we show the node colors with the Phylum information and the edges colors with the positive and negative correlations.
All the data used has been stored in the network.gexf file, including modules classifications, Phylum information and edge labels.</p>
<p><img src="Images/network2_spearman.png" width="550px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb166"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb166-1"><a href="model-based-class.html#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate network attributes</span></span>
<span id="cb166-2"><a href="model-based-class.html#cb166-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network_attr</span>()</span>
<span id="cb166-3"><a href="model-based-class.html#cb166-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_network_attr</span></code></pre></div>
<table style="width:49%;">
<colgroup>
<col width="34%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Property</th>
<th align="center">Value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Vertex</td>
<td align="center">407</td>
</tr>
<tr class="even">
<td align="center">Edge</td>
<td align="center">1989</td>
</tr>
<tr class="odd">
<td align="center">Average_degree</td>
<td align="center">9.774</td>
</tr>
<tr class="even">
<td align="center">Average_path_length</td>
<td align="center">3.878</td>
</tr>
<tr class="odd">
<td align="center">Network_diameter</td>
<td align="center">9</td>
</tr>
<tr class="even">
<td align="center">Clustering_coefficient</td>
<td align="center">0.4698</td>
</tr>
<tr class="odd">
<td align="center">Density</td>
<td align="center">0.02407</td>
</tr>
<tr class="even">
<td align="center">Heterogeneity</td>
<td align="center">1.194</td>
</tr>
<tr class="odd">
<td align="center">Centralization</td>
<td align="center">0.09908</td>
</tr>
</tbody>
</table>
<p>The function get_node_table, get_edge_table and get_adjacency_matrix are designed to
get node properties table, edge properties table and adjacency matrix from network, respectively.</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="model-based-class.html#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get node properties</span></span>
<span id="cb167-2"><a href="model-based-class.html#cb167-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">get_node_table</span>(<span class="at">node_roles =</span> <span class="cn">TRUE</span>)</span>
<span id="cb167-3"><a href="model-based-class.html#cb167-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_node_table</span></span></code></pre></div>
<table style="width:100%;">
<colgroup>
<col width="17%" />
<col width="12%" />
<col width="11%" />
<col width="17%" />
<col width="15%" />
<col width="11%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">name</th>
<th align="center">degree</th>
<th align="center">betweenness</th>
<th align="center">Abundance</th>
<th align="center">module</th>
<th align="center">z</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>OTU_50</strong></td>
<td align="center">OTU_50</td>
<td align="center">2</td>
<td align="center">0</td>
<td align="center">0.2301</td>
<td align="center">M2</td>
<td align="center">-1.305</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_305</strong></td>
<td align="center">OTU_305</td>
<td align="center">3</td>
<td align="center">4</td>
<td align="center">0.1394</td>
<td align="center">M8</td>
<td align="center">1.118</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_1</strong></td>
<td align="center">OTU_1</td>
<td align="center">21</td>
<td align="center">2</td>
<td align="center">1.409</td>
<td align="center">M2</td>
<td align="center">-0.04067</td>
</tr>
<tr class="even">
<td align="center"><strong>OTU_41</strong></td>
<td align="center">OTU_41</td>
<td align="center">1</td>
<td align="center">0</td>
<td align="center">0.1603</td>
<td align="center">M14</td>
<td align="center">-0.5774</td>
</tr>
<tr class="odd">
<td align="center"><strong>OTU_59</strong></td>
<td align="center">OTU_59</td>
<td align="center">3</td>
<td align="center">2306</td>
<td align="center">0.5588</td>
<td align="center">M6</td>
<td align="center">0.7771</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb168"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb168-1"><a href="model-based-class.html#cb168-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get edge properties</span></span>
<span id="cb168-2"><a href="model-based-class.html#cb168-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">get_edge_table</span>()</span>
<span id="cb168-3"><a href="model-based-class.html#cb168-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_edge_table </span></span>
<span id="cb168-4"><a href="model-based-class.html#cb168-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">get_adjacency_matrix</span>()</span>
<span id="cb168-5"><a href="model-based-class.html#cb168-5" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_adjacency_matrix</span></span></code></pre></div>
<p>Then, let’s plot the node classification in terms of the within-module connectivity and among-module connectivity.</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="model-based-class.html#cb169-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add_label = TRUE can be used to directly add text label for points</span></span>
<span id="cb169-2"><a href="model-based-class.html#cb169-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_taxa_roles</span>(<span class="at">use_type =</span> <span class="dv">1</span>)</span></code></pre></div>
<p><img src="Images/trans_network_taxa_roles.png" width="700px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="model-based-class.html#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot node roles with phylum information</span></span>
<span id="cb170-2"><a href="model-based-class.html#cb170-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_taxa_roles</span>(<span class="at">use_type =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/trans_network_taxa_roles_2.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Now, we show the eigengene analysis of modules.
The eigengene of a module, i.e. the first principal component of PCA, represents the main variance of the abundance in the species of the module.</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="model-based-class.html#cb171-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_eigen</span>()</span>
<span id="cb171-2"><a href="model-based-class.html#cb171-2" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_eigen</span></span></code></pre></div>
<p>Then we perform correlation heatmap to show the associations between eigengenes and environmental factors.</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="model-based-class.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_env object</span></span>
<span id="cb172-2"><a href="model-based-class.html#cb172-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb172-3"><a href="model-based-class.html#cb172-3" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate correlations</span></span>
<span id="cb172-4"><a href="model-based-class.html#cb172-4" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_cor</span>(<span class="at">add_abund_table =</span> t1<span class="sc">$</span>res_eigen)</span>
<span id="cb172-5"><a href="model-based-class.html#cb172-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the correlation heatmap</span></span>
<span id="cb172-6"><a href="model-based-class.html#cb172-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_cor</span>()</span></code></pre></div>
<p><img src="Images/trans_network_env_module_eigen.png" width="600px" style="display: block; margin: auto;" /></p>
<p>The <code>subset_network</code> function can be applied to extract a part of nodes and edges among these nodes from the network.
In this function, you should provide the nodes you need using the node parameter.</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="model-based-class.html#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract a sub network that contains all nodes in module M1</span></span>
<span id="cb173-2"><a href="model-based-class.html#cb173-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">subset_network</span>(<span class="at">node =</span> t1<span class="sc">$</span>res_node_table <span class="sc">%&gt;%</span> base<span class="sc">::</span><span class="fu">subset</span>(module <span class="sc">==</span> <span class="st">&quot;M1&quot;</span>) <span class="sc">%&gt;%</span> rownames, <span class="at">rm_single =</span> <span class="cn">TRUE</span>)</span>
<span id="cb173-3"><a href="model-based-class.html#cb173-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return a new network with igraph class</span></span>
<span id="cb173-4"><a href="model-based-class.html#cb173-4" aria-hidden="true" tabindex="-1"></a><span class="co"># extract sub network in which all edge labels are &quot;+&quot;, i.e. positive edges</span></span>
<span id="cb173-5"><a href="model-based-class.html#cb173-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">subset_network</span>(<span class="at">edge =</span> <span class="st">&quot;+&quot;</span>)</span></code></pre></div>
<p><strong>Then let’s show how to extract sub-network for samples and the whole analysis flow.</strong></p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="model-based-class.html#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the sub-network of sample &#39;S1&#39;</span></span>
<span id="cb174-2"><a href="model-based-class.html#cb174-2" aria-hidden="true" tabindex="-1"></a>sub1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">subset_network</span>(<span class="at">node =</span> dataset<span class="sc">$</span>otu_table <span class="sc">%&gt;%</span> .[.[, <span class="st">&quot;S1&quot;</span>] <span class="sc">!=</span> <span class="dv">0</span>, ] <span class="sc">%&gt;%</span> rownames, <span class="at">rm_single =</span> <span class="cn">TRUE</span>)</span>
<span id="cb174-3"><a href="model-based-class.html#cb174-3" aria-hidden="true" tabindex="-1"></a><span class="co"># see https://chiliubio.github.io/microeco_tutorial/notes.html#clone for the &#39;clone&#39; function explanation</span></span>
<span id="cb174-4"><a href="model-based-class.html#cb174-4" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">clone</span>(t1)</span>
<span id="cb174-5"><a href="model-based-class.html#cb174-5" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>res_network <span class="ot">&lt;-</span> sub1</span>
<span id="cb174-6"><a href="model-based-class.html#cb174-6" aria-hidden="true" tabindex="-1"></a><span class="co"># then t2 have a network for &#39;S1&#39; and can be used for further analysis</span></span>
<span id="cb174-7"><a href="model-based-class.html#cb174-7" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_module</span>()</span>
<span id="cb174-8"><a href="model-based-class.html#cb174-8" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">save_network</span>(<span class="st">&quot;S1.gexf&quot;</span>)</span>
<span id="cb174-9"><a href="model-based-class.html#cb174-9" aria-hidden="true" tabindex="-1"></a><span class="co"># please use a loop for more samples</span></span></code></pre></div>
<p>We also add the function <code>plot_network</code> to directly plot the network in R, including the static network and dynamic network.
The static network is suitable for the case with relatively few nodes, while dynamic network can be better applied to a large network.
See <a href="https://yunranchen.github.io/intro-net-r/advanced-network-visualization.html" class="uri">https://yunranchen.github.io/intro-net-r/advanced-network-visualization.html</a> and <a href="https://kateto.net/network-visualization" class="uri">https://kateto.net/network-visualization</a> for more
details on the network visualization in R.</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="model-based-class.html#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="co"># default parameter represents using igraph plot.igraph function</span></span>
<span id="cb175-2"><a href="model-based-class.html#cb175-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_network</span>()</span>
<span id="cb175-3"><a href="model-based-class.html#cb175-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use ggraph method; require ggraph package</span></span>
<span id="cb175-4"><a href="model-based-class.html#cb175-4" aria-hidden="true" tabindex="-1"></a><span class="co"># If ggraph is not installed; first install it with command: install.packages(&quot;ggraph&quot;)</span></span>
<span id="cb175-5"><a href="model-based-class.html#cb175-5" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_network</span>(<span class="at">method =</span> <span class="st">&quot;ggraph&quot;</span>, <span class="at">node_color =</span> <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb175-6"><a href="model-based-class.html#cb175-6" aria-hidden="true" tabindex="-1"></a><span class="co"># use networkD3 package method for the dynamic network visualization in R</span></span>
<span id="cb175-7"><a href="model-based-class.html#cb175-7" aria-hidden="true" tabindex="-1"></a><span class="co"># If networkD3 is not installed; first install it with command: install.packages(&quot;networkD3&quot;)</span></span>
<span id="cb175-8"><a href="model-based-class.html#cb175-8" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_network</span>(<span class="at">method =</span> <span class="st">&quot;networkD3&quot;</span>, <span class="at">node_color =</span> <span class="st">&quot;module&quot;</span>)</span>
<span id="cb175-9"><a href="model-based-class.html#cb175-9" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_network</span>(<span class="at">method =</span> <span class="st">&quot;networkD3&quot;</span>, <span class="at">node_color =</span> <span class="st">&quot;Phylum&quot;</span>)</span></code></pre></div>
<p>The <code>trans_comm</code> function can be used to convert the node classification to a new microtable object for other analysis.</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="model-based-class.html#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use_col is used to select a column of t1$res_node_table</span></span>
<span id="cb176-2"><a href="model-based-class.html#cb176-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">trans_comm</span>(<span class="at">use_col =</span> <span class="st">&quot;module&quot;</span>, <span class="at">abundance =</span> <span class="cn">FALSE</span>)</span>
<span id="cb176-3"><a href="model-based-class.html#cb176-3" aria-hidden="true" tabindex="-1"></a>tmp</span>
<span id="cb176-4"><a href="model-based-class.html#cb176-4" aria-hidden="true" tabindex="-1"></a>tmp<span class="sc">$</span>otu_table[tmp<span class="sc">$</span>otu_table <span class="sc">&gt;</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb176-5"><a href="model-based-class.html#cb176-5" aria-hidden="true" tabindex="-1"></a>tmp<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb176-6"><a href="model-based-class.html#cb176-6" aria-hidden="true" tabindex="-1"></a>tmp<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb176-7"><a href="model-based-class.html#cb176-7" aria-hidden="true" tabindex="-1"></a>tmp2 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(tmp, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>)</span>
<span id="cb176-8"><a href="model-based-class.html#cb176-8" aria-hidden="true" tabindex="-1"></a>tmp2<span class="sc">$</span>data_abund<span class="sc">$</span>Sample <span class="sc">%&lt;&gt;%</span> <span class="fu">factor</span>(., <span class="at">levels =</span> <span class="fu">rownames</span>(tmp<span class="sc">$</span>sample_table))</span>
<span id="cb176-9"><a href="model-based-class.html#cb176-9" aria-hidden="true" tabindex="-1"></a>tmp2<span class="sc">$</span><span class="fu">plot_line</span>(<span class="at">xtext_angle =</span> <span class="dv">30</span>, <span class="at">color_values =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">12</span>, <span class="st">&quot;Paired&quot;</span>)) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;OTUs ratio (%)&quot;</span>)</span></code></pre></div>
<p>The function <code>cal_sum_links</code> can sum the links (edge) number from one taxa to another or within the same taxa.
The function <code>plot_sum_links</code> is used to show the result from the function <code>cal_sum_links</code>.
This is very useful to fast see how many nodes are connected between different taxa or within one taxa.
In terms of ‘Phylum’ level in the tutorial, the function cal_sum_links() sum the linkages number from one Phylum to another Phylum or the linkages in the same Phylum.
So the numbers along the outside of the circular plot represent how many edges or linkages are related with the Phylum.
For example, in terms of Proteobacteria, there are roughly total 900 edges associated with the OTUs in Proteobacteria,
in which roughly 200 edges connect both OTUs in Proteobacteria and roughly 150 edges connect the OTUs from Proteobacteria with the OTUs from Chloroflexi.</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="model-based-class.html#cb177-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_sum_links</span>(<span class="at">taxa_level =</span> <span class="st">&quot;Phylum&quot;</span>)</span>
<span id="cb177-2"><a href="model-based-class.html#cb177-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require chorddiag package; see https://github.com/mattflor/chorddiag</span></span>
<span id="cb177-3"><a href="model-based-class.html#cb177-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_sum_links</span>(<span class="at">plot_pos =</span> <span class="cn">TRUE</span>, <span class="at">plot_num =</span> <span class="dv">10</span>, <span class="at">color_values =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">10</span>, <span class="st">&quot;Paired&quot;</span>))</span></code></pre></div>
<p><img src="Images/trans_network_sum_links.png" width="700px" style="display: block; margin: auto;" /></p>
<p>For the correlation network, there are also other available correlation/association calculation options,
such as Bray–Curtis (1-dissimilarity), SparCC <span class="citation">(<a href="#ref-Friedman_Inferring_2012" role="doc-biblioref">Friedman and Alm 2012</a>)</span>, CCLasso <span class="citation">(<a href="#ref-Fang_CCLasso_2015" role="doc-biblioref">Fang et al. 2015</a>)</span>,
Pearson or Spearman with data normalization based on NetCoMi package <span class="citation">(<a href="#ref-Peschel_NetCoMi_2021" role="doc-biblioref">Peschel et al. 2021</a>)</span>.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="model-based-class.html#cb178-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use Bray–Curtis index (1-dissimilarity)</span></span>
<span id="cb178-2"><a href="model-based-class.html#cb178-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;bray&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb178-3"><a href="model-based-class.html#cb178-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Pearson correlation</span></span>
<span id="cb178-4"><a href="model-based-class.html#cb178-4" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;pearson&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb178-5"><a href="model-based-class.html#cb178-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Pearson correlation using WGCNA package</span></span>
<span id="cb178-6"><a href="model-based-class.html#cb178-6" aria-hidden="true" tabindex="-1"></a><span class="co"># install WGCNA package</span></span>
<span id="cb178-7"><a href="model-based-class.html#cb178-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;WGCNA&quot;</span>)) <span class="fu">install.packages</span>(<span class="st">&quot;WGCNA&quot;</span>, <span class="at">repos =</span> BiocManager<span class="sc">::</span><span class="fu">repositories</span>())</span>
<span id="cb178-8"><a href="model-based-class.html#cb178-8" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;pearson&quot;</span>, <span class="at">use_WGCNA_pearson_spearman =</span> <span class="cn">TRUE</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb178-9"><a href="model-based-class.html#cb178-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Pearson correlation using NetCoMi package; install it from https://github.com/stefpeschel/NetCoMi</span></span>
<span id="cb178-10"><a href="model-based-class.html#cb178-10" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;pearson&quot;</span>, <span class="at">use_NetCoMi_pearson_spearman =</span> <span class="cn">TRUE</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb178-11"><a href="model-based-class.html#cb178-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Spearman correlation using WGCNA package</span></span>
<span id="cb178-12"><a href="model-based-class.html#cb178-12" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>, <span class="at">use_WGCNA_pearson_spearman =</span> <span class="cn">TRUE</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb178-13"><a href="model-based-class.html#cb178-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Spearman correlation using NetCoMi package</span></span>
<span id="cb178-14"><a href="model-based-class.html#cb178-14" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>, <span class="at">use_NetCoMi_pearson_spearman =</span> <span class="cn">TRUE</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb178-15"><a href="model-based-class.html#cb178-15" aria-hidden="true" tabindex="-1"></a><span class="co"># SparCC method, from SpiecEasi package, see https://github.com/zdk123/SpiecEasi for the installation</span></span>
<span id="cb178-16"><a href="model-based-class.html#cb178-16" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;sparcc&quot;</span>, <span class="at">use_sparcc_method =</span> <span class="st">&quot;SpiecEasi&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.003</span>)</span>
<span id="cb178-17"><a href="model-based-class.html#cb178-17" aria-hidden="true" tabindex="-1"></a><span class="co"># SparCC method, from NetCoMi package; https://github.com/stefpeschel/NetCoMi</span></span>
<span id="cb178-18"><a href="model-based-class.html#cb178-18" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;sparcc&quot;</span>, <span class="at">use_sparcc_method =</span> <span class="st">&quot;NetCoMi&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb178-19"><a href="model-based-class.html#cb178-19" aria-hidden="true" tabindex="-1"></a><span class="co"># CCLasso method based on NetCoMi package</span></span>
<span id="cb178-20"><a href="model-based-class.html#cb178-20" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;cclasso&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb178-21"><a href="model-based-class.html#cb178-21" aria-hidden="true" tabindex="-1"></a><span class="co"># CCREPE method based on NetCoMi package</span></span>
<span id="cb178-22"><a href="model-based-class.html#cb178-22" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="st">&quot;ccrepe&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span></code></pre></div>
<p><strong>Then let’s show other implemented network construction approaches</strong>:<br />
SPIEC-EASI (SParse InversE Covariance Estimation for Ecological Association Inference) approach of SpiecEasi R package <span class="citation">(<a href="#ref-Kurtz_Sparse_2015" role="doc-biblioref">Kurtz et al. 2015</a>)</span>
has two network construction approaches based on graph model, which relies on algorithms for sparse neighborhood and inverse covariance selection.
See <a href="https://github.com/zdk123/SpiecEasi" class="uri">https://github.com/zdk123/SpiecEasi</a> for the package installation.
It is very slow for SpiecEasi_method = ‘glasso’ when there is a large number (such as hundreds to thousands) according to our test experience.</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="model-based-class.html#cb179-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="cn">NULL</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb179-2"><a href="model-based-class.html#cb179-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require SpiecEasi package installed https://github.com/zdk123/SpiecEasi</span></span>
<span id="cb179-3"><a href="model-based-class.html#cb179-3" aria-hidden="true" tabindex="-1"></a><span class="co"># also see SpiecEasi::spiec.easi for available model parameters</span></span>
<span id="cb179-4"><a href="model-based-class.html#cb179-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">network_method =</span> <span class="st">&quot;SpiecEasi&quot;</span>, <span class="at">SpiecEasi_method =</span> <span class="st">&quot;mb&quot;</span>)</span>
<span id="cb179-5"><a href="model-based-class.html#cb179-5" aria-hidden="true" tabindex="-1"></a><span class="co"># see t1$res_network</span></span></code></pre></div>
<p>Another network construction approach comes from julia package FlashWeave <span class="citation">(<a href="#ref-Tackmann_Rapid_2019" role="doc-biblioref">Tackmann, Matias Rodrigues, and Mering 2019</a>)</span>.
This is a probabilistic graph-based method to obtain the conditional independence.
It predicts direct associations among microbes from large-scale compositional abundance data through statistical co-occurrence.
To repeat the following code, please first install julia language in your computer and the FlashWeave package, and add the julia in the computer path.</p>
<ol style="list-style-type: decimal">
<li>download and install julia from <a href="https://julialang.org/downloads/" class="uri">https://julialang.org/downloads/</a><br />
</li>
<li>Put julia in the computer env PATH, such as your_directory_path</li>
<li>Open terminal or cmd or Powershell, open julia, install FlashWeave following the operation in <a href="https://github.com/meringlab/FlashWeave.jl" class="uri">https://github.com/meringlab/FlashWeave.jl</a></li>
</ol>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="model-based-class.html#cb180-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="cn">NULL</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="dv">0</span>)</span>
<span id="cb180-2"><a href="model-based-class.html#cb180-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require Julia in the computer path, and the package FlashWeave</span></span>
<span id="cb180-3"><a href="model-based-class.html#cb180-3" aria-hidden="true" tabindex="-1"></a><span class="co"># different with the direct parameter passing of &#39;SpiecEasi&#39; network_method, FlashWeave_other_para is used to pass parameters to Julia FlashWeave</span></span>
<span id="cb180-4"><a href="model-based-class.html#cb180-4" aria-hidden="true" tabindex="-1"></a><span class="co"># assign FlashWeave_tempdir parameter can change the temporary working directory</span></span>
<span id="cb180-5"><a href="model-based-class.html#cb180-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">network_method =</span> <span class="st">&quot;FlashWeave&quot;</span>, <span class="at">FlashWeave_other_para =</span> <span class="st">&quot;alpha=0.01,sensitive=true,heterogeneous=true&quot;</span>)</span>
<span id="cb180-6"><a href="model-based-class.html#cb180-6" aria-hidden="true" tabindex="-1"></a><span class="co"># see t1$res_network</span></span></code></pre></div>
<p>The final method we want to show comes from beemStatic package <span class="citation">(<a href="#ref-Li_BEEMStatic_2021" role="doc-biblioref">Li et al. 2021</a>)</span>.
This method can be applied to cross-sectional datasets to infer interaction network based on the generalized Lotka-Volterra model,
which is typically used in the microbial time-series data.
So the network from this approach is a directed network.
Please see <a href="https://github.com/CSB5/BEEM-static" class="uri">https://github.com/CSB5/BEEM-static</a> for installing the R beemStatic package.</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="model-based-class.html#cb181-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cor_method =</span> <span class="cn">NULL</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.001</span>)</span>
<span id="cb181-2"><a href="model-based-class.html#cb181-2" aria-hidden="true" tabindex="-1"></a><span class="co"># require beemStatic package installed</span></span>
<span id="cb181-3"><a href="model-based-class.html#cb181-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">network_method =</span> <span class="st">&quot;beemStatic&quot;</span>)</span>
<span id="cb181-4"><a href="model-based-class.html#cb181-4" aria-hidden="true" tabindex="-1"></a><span class="co"># use cluster_walktrap method for the directed network</span></span>
<span id="cb181-5"><a href="model-based-class.html#cb181-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_module</span>(<span class="at">method =</span> <span class="st">&quot;cluster_walktrap&quot;</span>)</span></code></pre></div>
</div>
<div id="network-comparison" class="section level3 hasAnchor" number="6.2.2">
<h3><span class="header-section-number">6.2.2</span> Network comparison<a href="model-based-class.html#network-comparison" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To compare different networks from trans_network class,
please see the meconetcomp package chapter (<a href="https://chiliubio.github.io/microeco_tutorial/meconetcomp-package.html" class="uri">https://chiliubio.github.io/microeco_tutorial/meconetcomp-package.html</a>).</p>
</div>
<div id="key-points-6" class="section level3 hasAnchor" number="6.2.3">
<h3><span class="header-section-number">6.2.3</span> Key points<a href="model-based-class.html#key-points-6" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>cal_network(): get a network named <code>res_network</code> based on different methods</li>
<li>get_node_table(): get node properties table</li>
<li>subset_network(): this function can extract any sub-network according to the input nodes, e.g. sub-network for modules or samples</li>
</ul>
</div>
<div id="other-functions" class="section level3 hasAnchor" number="6.2.4">
<h3><span class="header-section-number">6.2.4</span> Other functions<a href="model-based-class.html#other-functions" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>cal_powerlaw(): perform bootstrapping hypothesis test to determine whether degrees follow a power law distribution and fit degrees to a power law distribution.<br />
</li>
<li>random_network(): generate random networks, compare them with the empirical network and get the p value of topological properties.</li>
</ul>
</div>
</div>
<div id="trans_nullmodel-class" class="section level2 hasAnchor" number="6.3">
<h2><span class="header-section-number">6.3</span> trans_nullmodel class<a href="model-based-class.html#trans_nullmodel-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In recent decades,
the integration of phylogenetic analysis and null model promotes the inference of niche and neutral influences on community assembly more powerfully
by adding a phylogeny dimension <span class="citation">(<a href="#ref-Webb_Phylogenies_2002" role="doc-biblioref">Webb et al. 2002</a>; <a href="#ref-Picante_Kembel_2010" role="doc-biblioref">Kembel et al. 2010</a>; <a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>)</span>.
The trans_nullmodel class provides an encapsulation, including the calculation of the phylogenetic signal,
beta mean pairwise phylogenetic distance (betaMPD), beta mean nearest taxon distance (betaMNTD),
beta nearest taxon index (betaNTI), beta net relatedness index (betaNRI) and Bray-Curtis-based Raup-Crick (RCbray).
The approach for phylogenetic signal analysis is based on the mantel correlogram <span class="citation">(<a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>,
in which the change of phylogenetic signal is intuitional and clear compared to other approaches.
The combinations between RCbray and betaNTI can be used to infer the strength of each ecological process dominating the community assembly
under the specific hypothesis <span class="citation">(<a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>)</span>.</p>
<div id="example-6" class="section level3 hasAnchor" number="6.3.1">
<h3><span class="header-section-number">6.3.1</span> Example<a href="model-based-class.html#example-6" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We first check the phylogenetic signal.</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="model-based-class.html#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generate trans_nullmodel object</span></span>
<span id="cb182-2"><a href="model-based-class.html#cb182-2" aria-hidden="true" tabindex="-1"></a><span class="co"># as an example, we only use high abundance OTU with mean relative abundance &gt; 0.0005</span></span>
<span id="cb182-3"><a href="model-based-class.html#cb182-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_nullmodel<span class="sc">$</span><span class="fu">new</span>(dataset, <span class="at">filter_thres =</span> <span class="fl">0.0005</span>, <span class="at">add_data =</span> env_data_16S)</span></code></pre></div>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="model-based-class.html#cb183-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use pH as the test variable</span></span>
<span id="cb183-2"><a href="model-based-class.html#cb183-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_mantel_corr</span>(<span class="at">use_env =</span> <span class="st">&quot;pH&quot;</span>)</span>
<span id="cb183-3"><a href="model-based-class.html#cb183-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_mantel_corr</span></span>
<span id="cb183-4"><a href="model-based-class.html#cb183-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the mantel correlogram</span></span>
<span id="cb183-5"><a href="model-based-class.html#cb183-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_mantel_corr</span>()</span></code></pre></div>
<p><img src="Images/trans_nullmodel_mantel_corr.png" width="600px" style="display: block; margin: auto;" /></p>
<p>betaNRI(ses.betampd) is used to show the ‘basal’ phylogenetic turnover.
Compared to betaNTI, it can capture more turnover information associated with the deep phylogeny.
It is noted that there are many null models with the development in the several decades of community ecology.
In the trans_nullmodel class,
the default null mode of betaNTI and betaNRI is the randomization of the phylogenetic relatedness among species.
This shuffling approach fix the observed levels of species α-diversity and β-diversity to
explore whether the observed phylogenetic turnover significantly differ from null model that phylogenetic relatedness among species are random.</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="model-based-class.html#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="co"># see null.model parameter for other null models</span></span>
<span id="cb184-2"><a href="model-based-class.html#cb184-2" aria-hidden="true" tabindex="-1"></a><span class="co"># null model run 500 times for the example</span></span>
<span id="cb184-3"><a href="model-based-class.html#cb184-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ses_betampd</span>(<span class="at">runs =</span> <span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb184-4"><a href="model-based-class.html#cb184-4" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_ses_betampd</span></span></code></pre></div>
<p>If we want to plot the betaNRI, we can use plot_group_distance function in trans_beta class.
For example, the results showed that the mean betaNRI of TW is extremely and significantly larger that those in CW and IW,
revealing that the basal phylogenetic turnover in TW is high.</p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="model-based-class.html#cb185-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add betaNRI matrix to beta_diversity list</span></span>
<span id="cb185-2"><a href="model-based-class.html#cb185-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>beta_diversity[[<span class="st">&quot;betaNRI&quot;</span>]] <span class="ot">&lt;-</span> t1<span class="sc">$</span>res_ses_betampd</span>
<span id="cb185-3"><a href="model-based-class.html#cb185-3" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_beta class, use measure &quot;betaNRI&quot;</span></span>
<span id="cb185-4"><a href="model-based-class.html#cb185-4" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;betaNRI&quot;</span>)</span>
<span id="cb185-5"><a href="model-based-class.html#cb185-5" aria-hidden="true" tabindex="-1"></a><span class="co"># transform the distance for each group</span></span>
<span id="cb185-6"><a href="model-based-class.html#cb185-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance</span>()</span>
<span id="cb185-7"><a href="model-based-class.html#cb185-7" aria-hidden="true" tabindex="-1"></a><span class="co"># see the help document for more methods, e.g. &quot;anova&quot; and &quot;KW_dunn&quot;</span></span>
<span id="cb185-8"><a href="model-based-class.html#cb185-8" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance_diff</span>(<span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>)</span>
<span id="cb185-9"><a href="model-based-class.html#cb185-9" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the results</span></span>
<span id="cb185-10"><a href="model-based-class.html#cb185-10" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t2<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">boxplot_add =</span> <span class="st">&quot;mean&quot;</span>)</span>
<span id="cb185-11"><a href="model-based-class.html#cb185-11" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/trans_nullmodel_betaNRI_one_dataset.png" width="550px" style="display: block; margin: auto;" /></p>
<p>Sometimes, if you want to perform null model analysis for each group individually, such as one group as one species pool,
you should calculate the results for each group, respectively.
The results show that, when we perform betaNRI for each group respectively,
mean betaNRI between CW and TW are not significantly different, and they are both significantly higher than that in IW,
revealing that the strength of variable selection in CW and TW may be similar under the condition that each area is considered as a specific species pool.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="model-based-class.html#cb186-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we create a list to store the trans_nullmodel results.</span></span>
<span id="cb186-2"><a href="model-based-class.html#cb186-2" aria-hidden="true" tabindex="-1"></a>sesbeta_each <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb186-3"><a href="model-based-class.html#cb186-3" aria-hidden="true" tabindex="-1"></a>group_col <span class="ot">&lt;-</span> <span class="st">&quot;Group&quot;</span></span>
<span id="cb186-4"><a href="model-based-class.html#cb186-4" aria-hidden="true" tabindex="-1"></a>all_groups <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>sample_table[, group_col])</span>
<span id="cb186-5"><a href="model-based-class.html#cb186-5" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate for each group, respectively</span></span>
<span id="cb186-6"><a href="model-based-class.html#cb186-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> all_groups){</span>
<span id="cb186-7"><a href="model-based-class.html#cb186-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># like the above operation, but need provide &#39;group&#39; and &#39;select_group&#39;</span></span>
<span id="cb186-8"><a href="model-based-class.html#cb186-8" aria-hidden="true" tabindex="-1"></a>    test <span class="ot">&lt;-</span> trans_nullmodel<span class="sc">$</span><span class="fu">new</span>(dataset, <span class="at">group =</span> group_col, <span class="at">select_group =</span> i, <span class="at">filter_thres =</span> <span class="fl">0.0005</span>)</span>
<span id="cb186-9"><a href="model-based-class.html#cb186-9" aria-hidden="true" tabindex="-1"></a>    test<span class="sc">$</span><span class="fu">cal_ses_betampd</span>(<span class="at">runs =</span> <span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb186-10"><a href="model-based-class.html#cb186-10" aria-hidden="true" tabindex="-1"></a>    sesbeta_each[[i]] <span class="ot">&lt;-</span> test<span class="sc">$</span>res_ses_betampd</span>
<span id="cb186-11"><a href="model-based-class.html#cb186-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb186-12"><a href="model-based-class.html#cb186-12" aria-hidden="true" tabindex="-1"></a><span class="co"># merge and reshape to generate one symmetrical matrix</span></span>
<span id="cb186-13"><a href="model-based-class.html#cb186-13" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">lapply</span>(sesbeta_each, reshape2<span class="sc">::</span>melt) <span class="sc">%&gt;%</span> </span>
<span id="cb186-14"><a href="model-based-class.html#cb186-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">do.call</span>(rbind, .) <span class="sc">%&gt;%</span></span>
<span id="cb186-15"><a href="model-based-class.html#cb186-15" aria-hidden="true" tabindex="-1"></a>    reshape2<span class="sc">::</span><span class="fu">dcast</span>(., Var1<span class="sc">~</span>Var2, <span class="at">value.var =</span> <span class="st">&quot;value&quot;</span>)</span>
<span id="cb186-16"><a href="model-based-class.html#cb186-16" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(test) <span class="ot">&lt;-</span> test[, <span class="dv">1</span>]</span>
<span id="cb186-17"><a href="model-based-class.html#cb186-17" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> test[, <span class="sc">-</span><span class="dv">1</span>, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb186-18"><a href="model-based-class.html#cb186-18" aria-hidden="true" tabindex="-1"></a><span class="co"># like the above operation</span></span>
<span id="cb186-19"><a href="model-based-class.html#cb186-19" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>beta_diversity[[<span class="st">&quot;betaNRI&quot;</span>]] <span class="ot">&lt;-</span> test</span>
<span id="cb186-20"><a href="model-based-class.html#cb186-20" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;betaNRI&quot;</span>)</span>
<span id="cb186-21"><a href="model-based-class.html#cb186-21" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance</span>()</span>
<span id="cb186-22"><a href="model-based-class.html#cb186-22" aria-hidden="true" tabindex="-1"></a><span class="co"># statistical analysis</span></span>
<span id="cb186-23"><a href="model-based-class.html#cb186-23" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance_diff</span>(<span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>)</span>
<span id="cb186-24"><a href="model-based-class.html#cb186-24" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t2<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">boxplot_add =</span> <span class="st">&quot;mean&quot;</span>)</span>
<span id="cb186-25"><a href="model-based-class.html#cb186-25" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/trans_nullmodel_betaNRI_each_dataset.png" width="550px" style="display: block; margin: auto;" /></p>
<p>BetaNTI(ses.betamntd) can be used to indicate the phylogenetic terminal turnover <span class="citation">(<a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>)</span>.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="model-based-class.html#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="co"># null model run 500 times</span></span>
<span id="cb187-2"><a href="model-based-class.html#cb187-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ses_betamntd</span>(<span class="at">runs =</span> <span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>, <span class="at">null.model =</span> <span class="st">&quot;taxa.labels&quot;</span>)</span>
<span id="cb187-3"><a href="model-based-class.html#cb187-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_ses_betamntd</span></span></code></pre></div>
<table style="width:81%;">
<colgroup>
<col width="12%" />
<col width="13%" />
<col width="13%" />
<col width="13%" />
<col width="12%" />
<col width="13%" />
</colgroup>
<thead>
<tr class="header">
<th align="center"> </th>
<th align="center">S1</th>
<th align="center">S2</th>
<th align="center">S3</th>
<th align="center">S4</th>
<th align="center">S5</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><strong>S1</strong></td>
<td align="center">0</td>
<td align="center">-0.432</td>
<td align="center">-0.7318</td>
<td align="center">0.6482</td>
<td align="center">1.236</td>
</tr>
<tr class="even">
<td align="center"><strong>S2</strong></td>
<td align="center">-0.432</td>
<td align="center">0</td>
<td align="center">-0.9817</td>
<td align="center">0.1372</td>
<td align="center">0.8991</td>
</tr>
<tr class="odd">
<td align="center"><strong>S3</strong></td>
<td align="center">-0.7318</td>
<td align="center">-0.9817</td>
<td align="center">0</td>
<td align="center">1.098</td>
<td align="center">0.02069</td>
</tr>
<tr class="even">
<td align="center"><strong>S4</strong></td>
<td align="center">0.6482</td>
<td align="center">0.1372</td>
<td align="center">1.098</td>
<td align="center">0</td>
<td align="center">-1.085</td>
</tr>
<tr class="odd">
<td align="center"><strong>S5</strong></td>
<td align="center">1.236</td>
<td align="center">0.8991</td>
<td align="center">0.02069</td>
<td align="center">-1.085</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>If the dataset is large, it is faster with use_iCAMP = TRUE for betaNTI.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="model-based-class.html#cb188-1" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="st">&quot;./test1&quot;</span>; <span class="fu">dir.create</span>(tmp)</span>
<span id="cb188-2"><a href="model-based-class.html#cb188-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ses_betamntd</span>(<span class="at">runs =</span> <span class="dv">1000</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>, <span class="at">use_iCAMP =</span> <span class="cn">TRUE</span>, <span class="at">iCAMP_tempdir =</span> tmp)</span></code></pre></div>
<p>RCbray (Bray-Curtis-based Raup-Crick) can be calculated using function cal_rcbray()
to assess whether the compositional turnover was governed primarily by drift <span class="citation">(<a href="#ref-Chase_null_2011" role="doc-biblioref">Chase et al. 2011</a>)</span>.
We applied null model to simulate species distribution by randomly sampling individuals from each
species pool with preserving species occurrence frequency and sample species richness <span class="citation">(<a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>.</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="model-based-class.html#cb189-1" aria-hidden="true" tabindex="-1"></a><span class="co"># result stored in t1$res_rcbray</span></span>
<span id="cb189-2"><a href="model-based-class.html#cb189-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_rcbray</span>(<span class="at">runs =</span> <span class="dv">1000</span>)</span>
<span id="cb189-3"><a href="model-based-class.html#cb189-3" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_rcbray</span></span></code></pre></div>
<p>As an example, we also calculate the proportion of the inferred processes on the community assembly as shown in the references <span class="citation">(<a href="#ref-Stegen_Quantifying_2013" role="doc-biblioref">Stegen et al. 2013</a>; <a href="#ref-Liu_Long_term_2017" role="doc-biblioref">Liu et al. 2017</a>)</span>.
In the example, the fraction of pairwise comparisons with significant betaNTI values (|βNTI| &gt; 2) is the estimated influence of Selection;
βNTI &gt; 2 represents the heterogeneous selection; βNTI &lt; -2 represents the homogeneous selection.
The value of RCbray characterizes the magnitude of deviation between observed Bray–Curtis and Bray–Curtis expected under the randomization;
a value of |RCbray| &gt; 0.95 was considered significant.
The fraction of all pairwise comparisons with |βNTI| &lt; 2 and RCbray &gt; +0.95 was taken as the influence of Dispersal Limitation combined with Drift.
The fraction of all pairwise comparisons with |βNTI| &lt; 2 and RCbray &lt; -0.95 was taken as an estimate for the influence of Homogenizing Dispersal.
The fraction of all pairwise comparisons with |βNTI| &lt; 2 and |RCbray| &lt; 0.95 estimates the influence of Drift acting alone.</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="model-based-class.html#cb190-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use betaNTI and rcbray to evaluate processes</span></span>
<span id="cb190-2"><a href="model-based-class.html#cb190-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_process</span>(<span class="at">use_betamntd =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>## The result is stored in object$res_process ...</code></pre>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="model-based-class.html#cb192-1" aria-hidden="true" tabindex="-1"></a><span class="co"># return t1$res_process</span></span></code></pre></div>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="model-based-class.html#cb193-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_process</span></code></pre></div>
<table style="width:51%;">
<colgroup>
<col width="33%" />
<col width="18%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">process</th>
<th align="center">percentage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">variable selection</td>
<td align="center">9.238</td>
</tr>
<tr class="even">
<td align="center">homogeneous selection</td>
<td align="center">0</td>
</tr>
<tr class="odd">
<td align="center">dispersal limitation</td>
<td align="center">8.964</td>
</tr>
<tr class="even">
<td align="center">homogeneous dispersal</td>
<td align="center">11.54</td>
</tr>
<tr class="odd">
<td align="center">drift</td>
<td align="center">70.26</td>
</tr>
</tbody>
</table>
<p>The <code>cal_NST</code> function can be used to calculate normalized stochasticity ratio based on the NST package <span class="citation">(<a href="#ref-Ning_general_2019" role="doc-biblioref">Ning et al. 2019</a>)</span>,
including ‘tNST’ and ‘pNST’ methods.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="model-based-class.html#cb194-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require NST package to be installed</span></span>
<span id="cb194-2"><a href="model-based-class.html#cb194-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_NST</span>(<span class="at">method =</span> <span class="st">&quot;tNST&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">dist.method =</span> <span class="st">&quot;bray&quot;</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>, <span class="at">output.rand =</span> <span class="cn">TRUE</span>, <span class="at">SES =</span> <span class="cn">TRUE</span>)</span>
<span id="cb194-3"><a href="model-based-class.html#cb194-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_NST<span class="sc">$</span>index.grp</span></code></pre></div>
<pre><code>## Perform tNST analysis ...</code></pre>
<pre><code>## All match very well.</code></pre>
<pre><code>## Now randomizing by parallel computing. Begin at Wed May  3 16:38:22 2023. Please wait...</code></pre>
<pre><code>## The result is stored in object$res_NST ...</code></pre>
<table style="width:92%;">
<colgroup>
<col width="11%" />
<col width="9%" />
<col width="16%" />
<col width="18%" />
<col width="18%" />
<col width="18%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">group</th>
<th align="center">size</th>
<th align="center">ST.i.bray</th>
<th align="center">NST.i.bray</th>
<th align="center">MST.i.bray</th>
<th align="center">SES.i.bray</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">IW</td>
<td align="center">435</td>
<td align="center">0.6539</td>
<td align="center">0.3875</td>
<td align="center">0.364</td>
<td align="center">9.074</td>
</tr>
<tr class="even">
<td align="center">CW</td>
<td align="center">435</td>
<td align="center">0.5627</td>
<td align="center">0.2955</td>
<td align="center">0.2864</td>
<td align="center">14.42</td>
</tr>
<tr class="odd">
<td align="center">TW</td>
<td align="center">435</td>
<td align="center">0.6243</td>
<td align="center">0.4145</td>
<td align="center">0.3998</td>
<td align="center">9.854</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="model-based-class.html#cb199-1" aria-hidden="true" tabindex="-1"></a><span class="co"># test the NST difference between each pair of groups</span></span>
<span id="cb199-2"><a href="model-based-class.html#cb199-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_NST_test</span>(<span class="at">method =</span> <span class="st">&quot;nst.boot&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="model-based-class.html#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert long format table to square matrix</span></span>
<span id="cb200-2"><a href="model-based-class.html#cb200-2" aria-hidden="true" tabindex="-1"></a><span class="co"># the 10th column: MST.ij.bray in t1$res_NST$index.pair</span></span>
<span id="cb200-3"><a href="model-based-class.html#cb200-3" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">cal_NST_convert</span>(<span class="dv">10</span>)</span></code></pre></div>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="model-based-class.html#cb201-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for pNST method, phylogenetic tree is needed</span></span>
<span id="cb201-2"><a href="model-based-class.html#cb201-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_NST</span>(<span class="at">method =</span> <span class="st">&quot;pNST&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">output.rand =</span> <span class="cn">TRUE</span>, <span class="at">SES =</span> <span class="cn">TRUE</span>)</span>
<span id="cb201-3"><a href="model-based-class.html#cb201-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_NST_test</span>(<span class="at">method =</span> <span class="st">&quot;nst.boot&quot;</span>)</span></code></pre></div>
<p>For nearest Taxon Index (NTI) and nearest Relative Index (NRI), please use cal_NTI and cal_NRI, respectively.</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="model-based-class.html#cb202-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_NRI</span>(<span class="at">null.model =</span> <span class="st">&quot;taxa.labels&quot;</span>, <span class="at">abundance.weighted =</span> <span class="cn">FALSE</span>, <span class="at">runs =</span> <span class="dv">999</span>)</span>
<span id="cb202-2"><a href="model-based-class.html#cb202-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_NTI</span>(<span class="at">null.model =</span> <span class="st">&quot;taxa.labels&quot;</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>, <span class="at">runs =</span> <span class="dv">999</span>)</span></code></pre></div>
</div>
<div id="key-points-7" class="section level3 hasAnchor" number="6.3.2">
<h3><span class="header-section-number">6.3.2</span> Key points<a href="model-based-class.html#key-points-7" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>trans_nullmodel$new: filter_thres parameter for the filtering of taxa with relative low abundance</li>
<li>cal_rcbray(): if only need rcbray, ignore other phylogenetic operations</li>
</ul>
</div>
<div id="other-function" class="section level3 hasAnchor" number="6.3.3">
<h3><span class="header-section-number">6.3.3</span> Other function<a href="model-based-class.html#other-function" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>cal_Cscore(): calculates the (normalised) mean number of checkerboard combinations (C-score) using C.score</li>
</ul>
</div>
</div>
<div id="trans_classifier-class" class="section level2 hasAnchor" number="6.4">
<h2><span class="header-section-number">6.4</span> trans_classifier class<a href="model-based-class.html#trans_classifier-class" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The trans_classifier class is a wrapper for methods of machine-learning-based classification models.
Microbiome-based supervised machine-learning has been successful in predicting human health status <span class="citation">(<a href="#ref-Poore_Microbiome_2020" role="doc-biblioref">Poore et al. 2020</a>)</span>
and soil categories <span class="citation">(<a href="#ref-Wilhelm_Predicting_2021" role="doc-biblioref">Wilhelm, Es, and Buckley 2021</a>)</span>.</p>
<div id="dependencies" class="section level3 hasAnchor" number="6.4.1">
<h3><span class="header-section-number">6.4.1</span> Dependencies<a href="model-based-class.html#dependencies" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Before starting the examples, make sure those packages have been installed.</p>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="model-based-class.html#cb203-1" aria-hidden="true" tabindex="-1"></a>packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Boruta&quot;</span>, <span class="st">&quot;parallel&quot;</span>, <span class="st">&quot;rsample&quot;</span>, <span class="st">&quot;randomForest&quot;</span>, <span class="st">&quot;caret&quot;</span>, <span class="st">&quot;gridExtra&quot;</span>, <span class="st">&quot;multiROC&quot;</span>)</span>
<span id="cb203-2"><a href="model-based-class.html#cb203-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Now check or install</span></span>
<span id="cb203-3"><a href="model-based-class.html#cb203-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(x <span class="cf">in</span> packages){</span>
<span id="cb203-4"><a href="model-based-class.html#cb203-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(x, <span class="at">character.only =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb203-5"><a href="model-based-class.html#cb203-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">install.packages</span>(x, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span>
<span id="cb203-6"><a href="model-based-class.html#cb203-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb203-7"><a href="model-based-class.html#cb203-7" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="examples" class="section level3 hasAnchor" number="6.4.2">
<h3><span class="header-section-number">6.4.2</span> Examples<a href="model-based-class.html#examples" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In this section, we use the example data in file2meco package (<a href="https://chiliubio.github.io/microeco_tutorial/file2meco-package.html" class="uri">https://chiliubio.github.io/microeco_tutorial/file2meco-package.html</a>) to demonstrate the feature selection,
data training and prediction with random forest algorithm.</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="model-based-class.html#cb204-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb204-2"><a href="model-based-class.html#cb204-2" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;dada2_table.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb204-3"><a href="model-based-class.html#cb204-3" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sample-metadata.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb204-4"><a href="model-based-class.html#cb204-4" aria-hidden="true" tabindex="-1"></a>taxonomy_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;taxonomy.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb204-5"><a href="model-based-class.html#cb204-5" aria-hidden="true" tabindex="-1"></a><span class="co"># construct microtable object</span></span>
<span id="cb204-6"><a href="model-based-class.html#cb204-6" aria-hidden="true" tabindex="-1"></a>d1 <span class="ot">&lt;-</span> <span class="fu">qiime2meco</span>(<span class="at">feature_table =</span> abund_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">taxonomy_table =</span> taxonomy_file_path)</span>
<span id="cb204-7"><a href="model-based-class.html#cb204-7" aria-hidden="true" tabindex="-1"></a>d1<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb204-8"><a href="model-based-class.html#cb204-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb204-9"><a href="model-based-class.html#cb204-9" aria-hidden="true" tabindex="-1"></a><span class="co"># initialize: use &quot;genotype&quot; as response variable</span></span>
<span id="cb204-10"><a href="model-based-class.html#cb204-10" aria-hidden="true" tabindex="-1"></a><span class="co"># x.predictors parameter is used to select the taxa; here we use all the taxa data in d1$taxa_abund</span></span>
<span id="cb204-11"><a href="model-based-class.html#cb204-11" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_classifier<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> d1, <span class="at">y.response =</span> <span class="st">&quot;genotype&quot;</span>, <span class="at">x.predictors =</span> <span class="st">&quot;All&quot;</span>)</span></code></pre></div>
<p>We silit the data into training and testing set.</p>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="model-based-class.html#cb205-1" aria-hidden="true" tabindex="-1"></a><span class="co"># generate train and test set</span></span>
<span id="cb205-2"><a href="model-based-class.html#cb205-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_split</span>(<span class="at">prop.train =</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>)</span></code></pre></div>
<p>Before training the model, we run the set_trainControl to invoke the trainControl function of caret package to generate the parameters used for training.
Here we use the default parameters in trainControl function.</p>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="model-based-class.html#cb206-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require caret package</span></span>
<span id="cb206-2"><a href="model-based-class.html#cb206-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">set_trainControl</span>()</span></code></pre></div>
<p>Now let’s start model training with rf method.</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="model-based-class.html#cb207-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use default parameter method = &quot;rf&quot;</span></span>
<span id="cb207-2"><a href="model-based-class.html#cb207-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_train</span>(<span class="at">max.ntree =</span> <span class="dv">500</span>)</span></code></pre></div>
<p>We can use cal_predict function to predict the testing data set.</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="model-based-class.html#cb208-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_predict</span>()</span>
<span id="cb208-2"><a href="model-based-class.html#cb208-2" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the confusionMatrix to check out the performance</span></span>
<span id="cb208-3"><a href="model-based-class.html#cb208-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_confusionMatrix</span>()</span></code></pre></div>
<p><img src="Images/plot_confusionMatrix_without_selection.png" width="600px" style="display: block; margin: auto;" /></p>
<p>Using cal_ROC and plot_ROC can get the ROC (Receiver Operator Characteristic) curve.</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="model-based-class.html#cb209-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ROC</span>()</span>
<span id="cb209-2"><a href="model-based-class.html#cb209-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ROC</span>(<span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>)</span></code></pre></div>
<p><img src="Images/plot_ROC_without_selection.png" width="500px" style="display: block; margin: auto;" /></p>
<p>While building a machine learning model for microbiome data,
the huge diversity of microbial community and/or associated relationships among taxa accross phylogeny can lead to a large number of unnecessary features,
which can reduce the overall accuracy, increase the complexity and overfit of the model and decrease the generalization capability of the model.
So, feature selection is one important step in building machine-learning model.
Then, we attempt to use Boruta package <span class="citation">(<a href="#ref-Kursa_Feature_2010" role="doc-biblioref">Kursa and Rudnicki 2010</a>)</span> to do feature selection.</p>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="model-based-class.html#cb210-1" aria-hidden="true" tabindex="-1"></a><span class="co"># require Boruta package</span></span>
<span id="cb210-2"><a href="model-based-class.html#cb210-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_feature_sel</span>(<span class="at">boruta.maxRuns =</span> <span class="dv">300</span>, <span class="at">boruta.pValue =</span> <span class="fl">0.01</span>)</span></code></pre></div>
<p>To compare the results between the procedure with feature selection and that without feature selection,
we also perfom all the analysis with feature selection to show the whole results.</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="model-based-class.html#cb211-1" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_classifier<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> d1, <span class="at">y.response =</span> <span class="st">&quot;genotype&quot;</span>, <span class="at">x.predictors =</span> <span class="st">&quot;All&quot;</span>)</span>
<span id="cb211-2"><a href="model-based-class.html#cb211-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_feature_sel</span>(<span class="at">boruta.maxRuns =</span> <span class="dv">300</span>, <span class="at">boruta.pValue =</span> <span class="fl">0.01</span>)</span>
<span id="cb211-3"><a href="model-based-class.html#cb211-3" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_split</span>(<span class="at">prop.train =</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>)</span>
<span id="cb211-4"><a href="model-based-class.html#cb211-4" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">set_trainControl</span>()</span>
<span id="cb211-5"><a href="model-based-class.html#cb211-5" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_train</span>(<span class="at">max.ntree =</span> <span class="dv">500</span>)</span>
<span id="cb211-6"><a href="model-based-class.html#cb211-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_predict</span>()</span>
<span id="cb211-7"><a href="model-based-class.html#cb211-7" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_confusionMatrix</span>()</span>
<span id="cb211-8"><a href="model-based-class.html#cb211-8" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_ROC</span>()</span>
<span id="cb211-9"><a href="model-based-class.html#cb211-9" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_ROC</span>(<span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>)</span></code></pre></div>
<p><img src="Images/plot_confusionMatrix_with_selection.png" width="600px" style="display: block; margin: auto;" />
<img src="Images/plot_ROC_with_selection.png" width="500px" style="display: block; margin: auto;" /></p>
<p>To plot the Precision-Recall curve (PR curve), please make plot_type = “PR” in plot_ROC function.</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="model-based-class.html#cb212-1" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_ROC</span>(<span class="at">plot_type =</span> <span class="st">&quot;PR&quot;</span>, <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>)</span></code></pre></div>
<p><img src="Images/plot_PR_with_selection.png" width="500px" style="display: block; margin: auto;" /></p>
<p>To show the ROC curve or PR curve of the training result, please make input = “train” in plot_ROC function.</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="model-based-class.html#cb213-1" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_ROC</span>(<span class="at">input =</span> <span class="st">&quot;train&quot;</span>)</span>
<span id="cb213-2"><a href="model-based-class.html#cb213-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_ROC</span>(<span class="at">plot_type =</span> <span class="st">&quot;ROC&quot;</span>, <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>)</span></code></pre></div>
<p><img src="Images/plot_ROC_with_selection_training.png" width="500px" style="display: block; margin: auto;" /></p>
<p>For other machine-learning models, please use method parameter in cal_train function.</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="model-based-class.html#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use SVM method</span></span>
<span id="cb214-2"><a href="model-based-class.html#cb214-2" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">al_train</span>(<span class="at">method =</span> <span class="st">&quot;svmRadial&quot;</span>, <span class="at">tuneLength =</span> <span class="dv">15</span>)</span></code></pre></div>
<p>The regression is supported from v0.15.0.</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="model-based-class.html#cb215-1" aria-hidden="true" tabindex="-1"></a><span class="co"># prepare data</span></span>
<span id="cb215-2"><a href="model-based-class.html#cb215-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(env_data_16S)</span>
<span id="cb215-3"><a href="model-based-class.html#cb215-3" aria-hidden="true" tabindex="-1"></a>new_test <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb215-4"><a href="model-based-class.html#cb215-4" aria-hidden="true" tabindex="-1"></a>new_test<span class="sc">$</span>sample_table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(new_test<span class="sc">$</span>sample_table, env_data_16S[<span class="fu">rownames</span>(new_test<span class="sc">$</span>sample_table), ])</span>
<span id="cb215-5"><a href="model-based-class.html#cb215-5" aria-hidden="true" tabindex="-1"></a><span class="co"># pH response variable</span></span>
<span id="cb215-6"><a href="model-based-class.html#cb215-6" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_classifier<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> new_test, <span class="at">y.response =</span> <span class="st">&quot;pH&quot;</span>, <span class="at">x.predictors =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb215-7"><a href="model-based-class.html#cb215-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_split</span>(<span class="at">prop.train =</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>)</span>
<span id="cb215-8"><a href="model-based-class.html#cb215-8" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">set_trainControl</span>()</span>
<span id="cb215-9"><a href="model-based-class.html#cb215-9" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_train</span>()</span>
<span id="cb215-10"><a href="model-based-class.html#cb215-10" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_train</span>
<span id="cb215-11"><a href="model-based-class.html#cb215-11" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_predict</span>()</span>
<span id="cb215-12"><a href="model-based-class.html#cb215-12" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span>res_predict</span>
<span id="cb215-13"><a href="model-based-class.html#cb215-13" aria-hidden="true" tabindex="-1"></a><span class="co"># feature importance</span></span>
<span id="cb215-14"><a href="model-based-class.html#cb215-14" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_feature_imp</span>()</span>
<span id="cb215-15"><a href="model-based-class.html#cb215-15" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(t1<span class="sc">$</span>res_feature_imp)</span>
<span id="cb215-16"><a href="model-based-class.html#cb215-16" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_feature_imp</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;IncNodePurity&quot;</span>)</span></code></pre></div>
</div>
<div id="key-points-8" class="section level3 hasAnchor" number="6.4.3">
<h3><span class="header-section-number">6.4.3</span> Key points<a href="model-based-class.html#key-points-8" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>cal_feature_sel(): perform feature selection</li>
</ul>
</div>
<div id="other-function-1" class="section level3 hasAnchor" number="6.4.4">
<h3><span class="header-section-number">6.4.4</span> Other function<a href="model-based-class.html#other-function-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>cal_feature_imp(): get feature importance from the training model when method is “rf”</li>
<li>cal_preProcess(): Pre-process (centering, scaling etc.) of the feature data based on the caret::preProcess function.</li>
</ul>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Beck_Machine_2014" class="csl-entry">
Beck, Daniel, and James A Foster. 2014. <span>“Machine Learning Techniques Accurately Classify Microbial Communities by Bacterial Vaginosis Characteristics.”</span> Journal Article. <em>PloS One</em> 9 (2): e87830.
</div>
<div id="ref-Chase_null_2011" class="csl-entry">
Chase, Jonathan M, Nathan JB Kraft, Kevin G Smith, Mark Vellend, and Brian D Inouye. 2011. <span>“Using Null Models to Disentangle Variation in Community Dissimilarity from Variation in α-Diversity.”</span> Journal Article. <em>Ecosphere</em> 2 (2): art24. <a href="https://doi.org/10.1890/ES10-00117.1">https://doi.org/10.1890/ES10-00117.1</a>.
</div>
<div id="ref-Coyte_Theecology_2015" class="csl-entry">
Coyte, Katharine Z, Jonas Schluter, and Kevin R Foster. 2015. <span>“The Ecology of the Microbiome: Networks, Competition, and Stability.”</span> Journal Article. <em>Science</em> 350 (6261): 663–66.
</div>
<div id="ref-Betaregression_2010" class="csl-entry">
Cribari-Neto, Francisco, and Achim Zeileis. 2010. <span>“Beta Regression in <span>R</span>.”</span> <em>Journal of Statistical Software</em> 34 (2): 1–24. <a href="https://doi.org/10.18637/jss.v034.i02">https://doi.org/10.18637/jss.v034.i02</a>.
</div>
<div id="ref-Deng_Molecular_2012" class="csl-entry">
Deng, Ye, Yi-Huei Jiang, Yunfeng Yang, Zhili He, Feng Luo, and Jizhong Zhou. 2012. <span>“Molecular Ecological Network Analyses.”</span> Journal Article. <em>BMC Bioinformatics</em> 13 (1): 113.
</div>
<div id="ref-Fang_CCLasso_2015" class="csl-entry">
Fang, Huaying, Chengcheng Huang, Hongyu Zhao, and Minghua Deng. 2015. <span>“CCLasso: Correlation Inference for Compositional Data Through Lasso.”</span> Journal Article. <em>Bioinformatics</em> 31 (19): 3172–3318. <a href="https://doi.org/10.1093/bioinformatics/btv349">https://doi.org/10.1093/bioinformatics/btv349</a>.
</div>
<div id="ref-Faust_Microbial_2012" class="csl-entry">
Faust, Karoline, and Jeroen Raes. 2012. <span>“Microbial Interactions: From Networks to Models.”</span> Journal Article. <em>Nature Reviews Microbiology</em> 10 (8): 538–50.
</div>
<div id="ref-Fernandes_Unifying_2014" class="csl-entry">
Fernandes, Andrew D, Jennifer Ns Reid, Jean M Macklaim, Thomas A McMurrough, David R Edgell, and Gregory B Gloor. 2014. <span>“Unifying the Analysis of High-Throughput Sequencing Datasets: Characterizing RNA-Seq, 16S rRNA Gene Sequencing and Selective Growth Experiments by Compositional Data Analysis.”</span> Journal Article. <em>Microbiome</em> 2 (1): 1–13.
</div>
<div id="ref-Friedman_Inferring_2012" class="csl-entry">
Friedman, Jonathan, and Eric J Alm. 2012. <span>“Inferring Correlation Networks from Genomic Survey Data.”</span> Journal Article. <em>PLOS Computational Biology</em> 8 (9): e1002687. <a href="https://doi.org/10.1371/journal.pcbi.1002687">https://doi.org/10.1371/journal.pcbi.1002687</a>.
</div>
<div id="ref-Picante_Kembel_2010" class="csl-entry">
Kembel, S. W., P. D. Cowan, M. R. Helmus, W. K. Cornwell, H. Morlon, D. D. Ackerly, S. P. Blomberg, and C. O. Webb. 2010. <span>“Picante: <span>R</span> Tools for Integrating Phylogenies and Ecology.”</span> <em>Bioinformatics</em> 26: 1463–64.
</div>
<div id="ref-Kursa_Feature_2010" class="csl-entry">
Kursa, Miron B, and Witold R Rudnicki. 2010. <span>“Feature Selection with the Boruta Package.”</span> Journal Article. <em>J Stat Softw</em> 36 (11): 1–13.
</div>
<div id="ref-Kurtz_Sparse_2015" class="csl-entry">
Kurtz, Z. D., C. L. Muller, E. R. Miraldi, D. R. Littman, M. J. Blaser, and R. A. Bonneau. 2015. <span>“Sparse and Compositionally Robust Inference of Microbial Ecological Networks.”</span> Journal Article. <em>PLoS Comput Biol</em> 11 (5): e1004226. <a href="https://doi.org/10.1371/journal.pcbi.1004226">https://doi.org/10.1371/journal.pcbi.1004226</a>.
</div>
<div id="ref-Li_BEEMStatic_2021" class="csl-entry">
Li, C., T. V. Av-Shalom, J. W. G. Tan, J. S. Kwah, K. R. Chng, and N. Nagarajan. 2021. <span>“BEEM-Static: Accurate Inference of Ecological Interactions from Cross-Sectional Microbiome Data.”</span> Journal Article. <em>PLoS Comput Biol</em> 17 (9): e1009343. <a href="https://doi.org/10.1371/journal.pcbi.1009343">https://doi.org/10.1371/journal.pcbi.1009343</a>.
</div>
<div id="ref-Lin_Analysis_2020" class="csl-entry">
Lin, Huang, and Shyamal Das Peddada. 2020. <span>“Analysis of Compositions of Microbiomes with Bias Correction.”</span> Journal Article. <em>Nature Communications</em> 11 (1): 1–11.
</div>
<div id="ref-Liu_Long_term_2017" class="csl-entry">
Liu, Chi, Minjie Yao, James C Stegen, Junpeng Rui, Jiabao Li, and Xiangzhen Li. 2017. <span>“Long-Term Nitrogen Addition Affects the Phylogenetic Turnover of Soil Microbial Community Responding to Moisture Pulse.”</span> Journal Article. <em>Scientific Reports</em> 7 (1): 17492.
</div>
<div id="ref-Michael_Moderated_2014" class="csl-entry">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-Nearing_Microbiome_2022" class="csl-entry">
Nearing, Jacob T, Gavin M Douglas, Molly G Hayes, Jocelyn MacDonald, Dhwani K Desai, Nicole Allward, Casey Jones, Robyn J Wright, Akhilesh S Dhanani, and André M Comeau. 2022. <span>“Microbiome Differential Abundance Methods Produce Different Results Across 38 Datasets.”</span> Journal Article. <em>Nature Communications</em> 13 (1): 1–16.
</div>
<div id="ref-Ning_general_2019" class="csl-entry">
Ning, D., Y. Deng, J. M. Tiedje, and J. Zhou. 2019. <span>“A General Framework for Quantitatively Assessing Ecological Stochasticity.”</span> Journal Article. <em>Proc Natl Acad Sci U S A</em> 116 (34): 16892–98. <a href="https://doi.org/10.1073/pnas.1904623116">https://doi.org/10.1073/pnas.1904623116</a>.
</div>
<div id="ref-Paulson_Differential_2013" class="csl-entry">
Paulson, Joseph N, O Colin Stine, Héctor Corrada Bravo, and Mihai Pop. 2013. <span>“Differential Abundance Analysis for Microbial Marker-Gene Surveys.”</span> Journal Article. <em>Nature Methods</em> 10 (12): 1200–1202.
</div>
<div id="ref-Peschel_NetCoMi_2021" class="csl-entry">
Peschel, Stefanie, Christian L Müller, Erika von Mutius, Anne-Laure Boulesteix, and Martin Depner. 2021. <span>“NetCoMi: Network Construction and Comparison for Microbiome Data in r.”</span> Journal Article. <em>Briefings in Bioinformatics</em> 22 (4): bbaa290. <a href="https://doi.org/10.1093/bib/bbaa290">https://doi.org/10.1093/bib/bbaa290</a>.
</div>
<div id="ref-Poore_Microbiome_2020" class="csl-entry">
Poore, Gregory D, Evguenia Kopylova, Qiyun Zhu, Carolina Carpenter, Serena Fraraccio, Stephen Wandro, Tomasz Kosciolek, Stefan Janssen, Jessica Metcalf, and Se Jin Song. 2020. <span>“Microbiome Analyses of Blood and Tissues Suggest Cancer Diagnostic Approach.”</span> Journal Article. <em>Nature</em> 579 (7800): 567–74.
</div>
<div id="ref-Segata_Metagenomic_2011" class="csl-entry">
Segata, Nicola, Jacques Izard, Levi Waldron, Dirk Gevers, Larisa Miropolsky, Wendy S Garrett, and Curtis Huttenhower. 2011. <span>“Metagenomic Biomarker Discovery and Explanation.”</span> Journal Article. <em>Genome Biology</em> 12 (6): R60.
</div>
<div id="ref-Stegen_Quantifying_2013" class="csl-entry">
Stegen, James C, Xueju Lin, Jim K Fredrickson, Xingyuan Chen, David W Kennedy, Christopher J Murray, Mark L Rockhold, and Allan Konopka. 2013. <span>“Quantifying Community Assembly Processes and Identifying Features That Impose Them.”</span> Journal Article. <em>The ISME Journal</em> 7 (11): 2069–79.
</div>
<div id="ref-Tackmann_Rapid_2019" class="csl-entry">
Tackmann, J., J. F. Matias Rodrigues, and C. von Mering. 2019. <span>“Rapid Inference of Direct Interactions in Large-Scale Ecological Networks from Heterogeneous Microbial Sequencing Data.”</span> Journal Article. <em>Cell Systems</em> 9 (3): 286–296 e8. <a href="https://doi.org/10.1016/j.cels.2019.08.002">https://doi.org/10.1016/j.cels.2019.08.002</a>.
</div>
<div id="ref-Webb_Phylogenies_2002" class="csl-entry">
Webb, Campbell O, David D Ackerly, Mark A McPeek, and Michael J Donoghue. 2002. <span>“Phylogenies and Community Ecology.”</span> Journal Article. <em>Annu. Rev. Ecol. Syst.</em> 33: 475–505. <a href="https://doi.org/10.1146/annurev.ecolsys.33.010802.150448">https://doi.org/10.1146/annurev.ecolsys.33.010802.150448</a>.
</div>
<div id="ref-White_Statistical_2009" class="csl-entry">
White, JR, N Nagarajan, and M Pop. 2009. <span>“Statistical Methods for Detecting Differentially Abundant Features in Clinical Metagenomic Samples.”</span> Journal Article. <em>PLoS Computational Biology</em> 5 (4): e1000352.
</div>
<div id="ref-Wilhelm_Predicting_2021" class="csl-entry">
Wilhelm, Roland C, Harold M van Es, and Daniel H Buckley. 2021. <span>“Predicting Measures of Soil Health Using the Microbiome and Supervised Machine Learning.”</span> Journal Article. <em>Soil Biology and Biochemistry</em>, 108472.
</div>
<div id="ref-Yatsunenko_Human_2012" class="csl-entry">
Yatsunenko, Tanya, Federico E Rey, Mark J Manary, Indi Trehan, Maria Gloria Dominguez-Bello, Monica Contreras, Magda Magris, Glida Hidalgo, Robert N Baldassano, and Andrey P Anokhin. 2012. <span>“Human Gut Microbiome Viewed Across Age and Geography.”</span> Journal Article. <em>Nature</em> 486 (7402): 222–27.
</div>
<div id="ref-LinDA_Zhou_2022" class="csl-entry">
Zhou, H., K. He, J. Chen, and X. Zhang. 2022. <span>“LinDA: Linear Models for Differential Abundance Analysis of Microbiome Compositional Data.”</span> Journal Article. <em>Genome Biol</em> 23 (1): 95. <a href="https://doi.org/10.1186/s13059-022-02655-5">https://doi.org/10.1186/s13059-022-02655-5</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="diversity-based-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="explainable-class.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/05-Model-based_class.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
