<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Notes | Tutorial for R microeco package (v0.13.0)</title>
  <meta name="description" content="The tutorial for R microeco, file2meco, meconetcomp and mecodev packages" />
  <meta name="generator" content="bookdown 0.30 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Notes | Tutorial for R microeco package (v0.13.0)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecodev packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Notes | Tutorial for R microeco package (v0.13.0)" />
  
  <meta name="twitter:description" content="The tutorial for R microeco, file2meco, meconetcomp and mecodev packages" />
  

<meta name="author" content="Chi Liu, Felipe R. P. Mansoldo, Umer Zeeshan Ijaz, Chenhao Li, Yang Cao, Jarrod J. Scott, Yaoming Cui, Alane B. Vermelho, Minjie Yao, Xiangzhen Li" />


<meta name="date" content="2022-12-15" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="other-dataset.html"/>
<link rel="next" href="mecodev-package.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#rtools"><i class="fa fa-check"></i><b>2.4</b> RTools</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.5</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#important-packages"><i class="fa fa-check"></i><b>2.5.1</b> Important packages</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#cran-packages"><i class="fa fa-check"></i><b>2.5.2</b> CRAN packages</a></li>
<li class="chapter" data-level="2.5.3" data-path="intro.html"><a href="intro.html#ggtree"><i class="fa fa-check"></i><b>2.5.3</b> ggtree</a></li>
<li class="chapter" data-level="2.5.4" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.5.4</b> Gephi</a></li>
<li class="chapter" data-level="2.5.5" data-path="intro.html"><a href="intro.html#tax4fun"><i class="fa fa-check"></i><b>2.5.5</b> Tax4Fun</a></li>
<li class="chapter" data-level="2.5.6" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.5.6</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#plotting"><i class="fa fa-check"></i><b>2.6</b> Plotting</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#prepare-the-example-data"><i class="fa fa-check"></i><b>3.1.1</b> Prepare the example data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#how-to-read-your-files-to-microtable-object"><i class="fa fa-check"></i><b>3.1.2</b> How to read your files to microtable object?</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#functions-in-microtable-class"><i class="fa fa-check"></i><b>3.1.3</b> Functions in microtable class</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#subset-of-samples"><i class="fa fa-check"></i><b>3.1.4</b> subset of samples</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.5</b> Other examples</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.6</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-2"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-4"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#network-comparison"><i class="fa fa-check"></i><b>6.2.2</b> Network comparison</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.3</b> Key points</a></li>
<li class="chapter" data-level="6.2.4" data-path="model-based-class.html"><a href="model-based-class.html#other-functions"><i class="fa fa-check"></i><b>6.2.4</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies</a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples</a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points</a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-7"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME</a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2</a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann"><i class="fa fa-check"></i><b>8.3</b> HUMAnN</a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn</a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2"><i class="fa fa-check"></i><b>8.5</b> Kraken2</a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB</a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq</a></li>
<li class="chapter" data-level="8.8" data-path="file2meco-package.html"><a href="file2meco-package.html#picrust2"><i class="fa fa-check"></i><b>8.8</b> PICRUSt2</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset</a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data</a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> <code>clone</code> function</a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> <code>save</code> function</a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#github-issues"><i class="fa fa-check"></i><b>10.3</b> Github Issues</a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.4</b> change object</a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.5</b> group order</a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.6</b> add layers to plot</a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#colors-for-many-groups"><i class="fa fa-check"></i><b>10.7</b> colors for many groups</a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>11</b> mecodev package</a>
<ul>
<li class="chapter" data-level="11.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>11.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="11.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_convert"><i class="fa fa-check"></i><b>11.2</b> trans_convert</a></li>
<li class="chapter" data-level="11.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>11.3</b> trans_ts</a></li>
<li class="chapter" data-level="11.4" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>11.4</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>12</b> Other examples</a>
<ul>
<li class="chapter" data-level="12.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>12.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="12.2" data-path="other-examples-1.html"><a href="other-examples-1.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>12.2</b> The importance of tidy_taxonomy function</a></li>
<li class="chapter" data-level="12.3" data-path="other-examples-1.html"><a href="other-examples-1.html#show-the-abundance-of-unknown-taxa"><i class="fa fa-check"></i><b>12.3</b> Show the abundance of unknown taxa</a></li>
<li class="chapter" data-level="12.4" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>12.4</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="12.5" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>12.5</b> The use of phylogenetic tree</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v0.13.0)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="notes" class="section level1 hasAnchor" number="10">
<h1><span class="header-section-number">Chapter 10</span> Notes<a href="notes.html#notes" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>We show some other important points here.</p>
<div id="clone-function" class="section level2 hasAnchor" number="10.1">
<h2><span class="header-section-number">10.1</span> <code>clone</code> function<a href="notes.html#clone-function" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>R6 class has a special copy mechanism which is different from S3 and S4.
<strong>If you want to copy an object completely, you should use the function <code>clone</code> instead of direct assignment.</strong></p>
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="notes.html#cb261-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use clone to copy completely</span></span>
<span id="cb261-2"><a href="notes.html#cb261-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb261-3"><a href="notes.html#cb261-3" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">clone</span>(t1)</span>
<span id="cb261-4"><a href="notes.html#cb261-4" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>sample_table <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb261-5"><a href="notes.html#cb261-5" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(t2, t1)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="notes.html#cb263-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this operation is usually unuseful, because changing t2 will also affect t1</span></span>
<span id="cb263-2"><a href="notes.html#cb263-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> t1</span>
<span id="cb263-3"><a href="notes.html#cb263-3" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>sample_table <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb263-4"><a href="notes.html#cb263-4" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(t2, t1)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="save-function" class="section level2 hasAnchor" number="10.2">
<h2><span class="header-section-number">10.2</span> <code>save</code> function<a href="notes.html#save-function" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>As an R6 object is an encapsulated environment, there is also another useful advantage, i.e. the data repeatability.
By applying R6 classes, the analyzed data in ‘R Environment’ can have minimum numbers of intermediate files.
It is also very convenient to save the analyzed data (i.e. object) to local computer with <code>save</code> function for back-up and communication with others.
This is also the recommended way to send data for testing in ‘Github Issues’.</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="notes.html#cb265-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save &#39;dataset&#39; to local computer</span></span>
<span id="cb265-2"><a href="notes.html#cb265-2" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(dataset, <span class="at">file =</span> <span class="st">&quot;dataset.RData&quot;</span>)</span>
<span id="cb265-3"><a href="notes.html#cb265-3" aria-hidden="true" tabindex="-1"></a><span class="co"># dataset.RData can be compressed into zip format for the upload in &#39;Github Issues&#39;</span></span>
<span id="cb265-4"><a href="notes.html#cb265-4" aria-hidden="true" tabindex="-1"></a><span class="co"># load the data by others in their R project</span></span>
<span id="cb265-5"><a href="notes.html#cb265-5" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;dataset.RData&quot;</span>)</span></code></pre></div>
<p>The saved object is only related with the package version when you save it and has no matter with the updated package when you load it again.
So it is repeatable.</p>
</div>
<div id="github-issues" class="section level2 hasAnchor" number="10.3">
<h2><span class="header-section-number">10.3</span> Github Issues<a href="notes.html#github-issues" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Please write the question/bug/suggestion in ‘Github Issues’ (<a href="https://github.com/ChiLiubio/microeco/issues" class="uri">https://github.com/ChiLiubio/microeco/issues</a>).
Several points are recommended.</p>
<ul>
<li><p>Before creating a ‘New issue’, please first skim through the ‘Issues’ including ‘Open’ and ‘Closed’.
Then go ahead if no similar issue is found or the similar issue is not solved.</p></li>
<li><p>Please remember to provide your package version when writing a ‘New issue’.</p></li>
<li><p>For the specific issue related with data, please provide the data and scripts for the reproduction.
If the data has not been published, please provide a toy example or a subset data as far as possible.</p></li>
<li><p>To provide the data, please use save function as previous part shows (<a href="https://chiliubio.github.io/microeco_tutorial/notes.html#save-function" class="uri">https://chiliubio.github.io/microeco_tutorial/notes.html#save-function</a>),
and paste the compressed zip format data.</p></li>
<li><p>When pasting the script directly, please use ``` at the beginning and end of scripts to mark the code block for the readability
as # can cause format chaos without code tag.</p></li>
</ul>
</div>
<div id="change-object" class="section level2 hasAnchor" number="10.4">
<h2><span class="header-section-number">10.4</span> change object<a href="notes.html#change-object" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>All the classes are set public, meaning that you can <strong>change, add or remove the objects stored in them as you want</strong>.</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="notes.html#cb266-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add a matrix you think useful</span></span>
<span id="cb266-2"><a href="notes.html#cb266-2" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>my_matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span>, <span class="at">nrow =</span> <span class="dv">4</span>, <span class="at">ncol =</span> <span class="dv">4</span>)</span>
<span id="cb266-3"><a href="notes.html#cb266-3" aria-hidden="true" tabindex="-1"></a><span class="co"># change the information</span></span>
<span id="cb266-4"><a href="notes.html#cb266-4" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>sample_table <span class="sc">%&lt;&gt;%</span> .[, <span class="sc">-</span><span class="dv">2</span>]</span></code></pre></div>
</div>
<div id="group-order" class="section level2 hasAnchor" number="10.5">
<h2><span class="header-section-number">10.5</span> group order<a href="notes.html#group-order" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>If you want to reorder the groups, <strong>assign the factors may be the most simplest way.</strong></p>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="notes.html#cb267-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataset)</span>
<span id="cb267-2"><a href="notes.html#cb267-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb267-3"><a href="notes.html#cb267-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">ordination =</span> <span class="st">&quot;PCoA&quot;</span>)</span></code></pre></div>
<pre><code>## The ordination result is stored in object$res_ordination ...</code></pre>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="notes.html#cb269-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="microeco-tutorial_files/figure-html/unnamed-chunk-286-1.png" width="672" /></p>
<p>Then we assign factors to the ‘Group’ in sample_table of dataset.
We can find the changed group order in the legend and colors in the plot.</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="notes.html#cb270-1" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>sample_table<span class="sc">$</span>Group <span class="sc">%&lt;&gt;%</span> <span class="fu">factor</span>(., <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;IW&quot;</span>, <span class="st">&quot;TW&quot;</span>, <span class="st">&quot;CW&quot;</span>))</span>
<span id="cb270-2"><a href="notes.html#cb270-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(dataset<span class="sc">$</span>sample_table)</span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    90 obs. of  4 variables:
##  $ SampleID: chr  &quot;S1&quot; &quot;S2&quot; &quot;S3&quot; &quot;S4&quot; ...
##  $ Group   : Factor w/ 3 levels &quot;IW&quot;,&quot;TW&quot;,&quot;CW&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##  $ Type    : chr  &quot;NE&quot; &quot;NE&quot; &quot;NE&quot; &quot;NE&quot; ...
##  $ Saline  : chr  &quot;Non-saline soil&quot; &quot;Non-saline soil&quot; &quot;Non-saline soil&quot; &quot;Non-saline soil&quot; ...</code></pre>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="notes.html#cb272-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb272-2"><a href="notes.html#cb272-2" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">ordination =</span> <span class="st">&quot;PCoA&quot;</span>)</span></code></pre></div>
<pre><code>## The ordination result is stored in object$res_ordination ...</code></pre>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="notes.html#cb274-1" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>)</span></code></pre></div>
<p><img src="microeco-tutorial_files/figure-html/unnamed-chunk-287-1.png" width="672" /></p>
</div>
<div id="add-layers-to-plot" class="section level2 hasAnchor" number="10.6">
<h2><span class="header-section-number">10.6</span> add layers to plot<a href="notes.html#add-layers-to-plot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Most of the plots are generated by applying the ggplot2 package.
The important parameters in the plotting functions are configured according to our experience.
If the inner parameters can not enough, the user can add the layers to the plot like the following operation or
make the plot using the data (generally data.frame class) stored in the object.</p>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="notes.html#cb275-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The groupmean parameter can be used to obtain the group-mean barplot.</span></span>
<span id="cb275-2"><a href="notes.html#cb275-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">groupmean =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb275-3"><a href="notes.html#cb275-3" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">others_color =</span> <span class="st">&quot;grey70&quot;</span>, <span class="at">legend_text_italic =</span> <span class="cn">FALSE</span>)</span>
<span id="cb275-4"><a href="notes.html#cb275-4" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">theme_classic</span>() <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>))</span></code></pre></div>
<p><img src="Images/plot_bar_mean_classic.png" width="400px" style="display: block; margin: auto;" /></p>
</div>
<div id="colors-for-many-groups" class="section level2 hasAnchor" number="10.7">
<h2><span class="header-section-number">10.7</span> colors for many groups<a href="notes.html#colors-for-many-groups" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In general, the color palette in the ploting function is enough for the settings of groups or samples,
such as RColorBrewer::brewer.pal(12, “Paired”).
If the user has too many groups (e.g. 20), the color_palette_20 can be used. It has 20 colors with paired combinations.
If this is still not enough, the randomcoloR package can be used for generating more colors.</p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="notes.html#cb276-1" aria-hidden="true" tabindex="-1"></a>color_palette_20</span></code></pre></div>
<pre><code>##  [1] &quot;#1f77b4&quot; &quot;#aec7e8&quot; &quot;#ff7f0e&quot; &quot;#ffbb78&quot; &quot;#2ca02c&quot; &quot;#98df8a&quot; &quot;#d62728&quot;
##  [8] &quot;#ff9896&quot; &quot;#9467bd&quot; &quot;#c5b0d5&quot; &quot;#8c564b&quot; &quot;#c49c94&quot; &quot;#e377c2&quot; &quot;#f7b6d2&quot;
## [15] &quot;#7f7f7f&quot; &quot;#c7c7c7&quot; &quot;#bcbd22&quot; &quot;#dbdb8d&quot; &quot;#17becf&quot; &quot;#9edae5&quot;</code></pre>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="notes.html#cb278-1" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">col =</span> color_palette_20)</span></code></pre></div>
</div>
<div id="customized-data-input" class="section level2 hasAnchor" number="10.8">
<h2><span class="header-section-number">10.8</span> customized data input<a href="notes.html#customized-data-input" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>From v0.7.0, trans_alpha, trans_env and trans_venn classes can accept NULL dataset input for some customized usage of some functions in those classes.</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="notes.html#cb279-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>()</span>
<span id="cb279-2"><a href="notes.html#cb279-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_alpha<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> <span class="cn">NULL</span>, <span class="at">group =</span> <span class="cn">NULL</span>)</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="other-dataset.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="mecodev-package.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/09-Notes.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
