<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 14 Other examples | Tutorial for R microeco package (v1.8.0)</title>
  <meta name="description" content="Examples for R microeco (v1.8.0), file2meco (v0.8.0), meconetcomp (v0.5.1) and mecoturn packages" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 14 Other examples | Tutorial for R microeco package (v1.8.0)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Examples for R microeco (v1.8.0), file2meco (v0.8.0), meconetcomp (v0.5.1) and mecoturn packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 14 Other examples | Tutorial for R microeco package (v1.8.0)" />
  
  <meta name="twitter:description" content="Examples for R microeco (v1.8.0), file2meco (v0.8.0), meconetcomp (v0.5.1) and mecoturn packages" />
  



<meta name="date" content="2024-06-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="mecoturn-package.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#rtools"><i class="fa fa-check"></i><b>2.4</b> RTools</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.5</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#description"><i class="fa fa-check"></i><b>2.5.1</b> Description</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#cran-packages"><i class="fa fa-check"></i><b>2.5.2</b> CRAN packages</a></li>
<li class="chapter" data-level="2.5.3" data-path="intro.html"><a href="intro.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor packages</a></li>
<li class="chapter" data-level="2.5.4" data-path="intro.html"><a href="intro.html#github-packages"><i class="fa fa-check"></i><b>2.5.4</b> Github packages</a></li>
<li class="chapter" data-level="2.5.5" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.5.5</b> Gephi</a></li>
<li class="chapter" data-level="2.5.6" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.5.6</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#plot"><i class="fa fa-check"></i><b>2.6</b> Plot</a></li>
<li class="chapter" data-level="2.7" data-path="intro.html"><a href="intro.html#rstudio"><i class="fa fa-check"></i><b>2.7</b> Rstudio</a></li>
<li class="chapter" data-level="2.8" data-path="intro.html"><a href="intro.html#citation"><i class="fa fa-check"></i><b>2.8</b> Citation</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#prepare-the-example-data"><i class="fa fa-check"></i><b>3.1.1</b> Prepare the example data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#how-to-read-your-files-to-microtable-object"><i class="fa fa-check"></i><b>3.1.2</b> How to read your files to microtable object?</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#functions-in-microtable-class"><i class="fa fa-check"></i><b>3.1.3</b> Functions in microtable class</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#merge-taxa-or-samples"><i class="fa fa-check"></i><b>3.1.4</b> merge taxa or samples</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-class.html"><a href="basic-class.html#subset-of-samples"><i class="fa fa-check"></i><b>3.1.5</b> subset of samples</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-class.html"><a href="basic-class.html#subset-of-taxa"><i class="fa fa-check"></i><b>3.1.6</b> subset of taxa</a></li>
<li class="chapter" data-level="3.1.7" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.7</b> Other examples</a></li>
<li class="chapter" data-level="3.1.8" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.8</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-2"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-4"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#more-options"><i class="fa fa-check"></i><b>6.1.2</b> More options</a></li>
<li class="chapter" data-level="6.1.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.3</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#network-comparison"><i class="fa fa-check"></i><b>6.2.2</b> Network comparison</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.3</b> Key points</a></li>
<li class="chapter" data-level="6.2.4" data-path="model-based-class.html"><a href="model-based-class.html#other-functions"><i class="fa fa-check"></i><b>6.2.4</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies</a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples</a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points</a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-7"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME</a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2</a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann"><i class="fa fa-check"></i><b>8.3</b> HUMAnN</a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn</a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2bracken"><i class="fa fa-check"></i><b>8.5</b> Kraken2/Bracken</a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdbpcycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB/PCycDB</a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq</a></li>
<li class="chapter" data-level="8.8" data-path="file2meco-package.html"><a href="file2meco-package.html#picrust2"><i class="fa fa-check"></i><b>8.8</b> PICRUSt2</a></li>
<li class="chapter" data-level="8.9" data-path="file2meco-package.html"><a href="file2meco-package.html#viromescan"><i class="fa fa-check"></i><b>8.9</b> ViromeScan</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset</a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data</a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> <code>clone</code> function</a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> <code>save</code> function</a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#github-issues"><i class="fa fa-check"></i><b>10.3</b> Github Issues</a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.4</b> group order</a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.5</b> add layers to plot</a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#colors-for-more-groups"><i class="fa fa-check"></i><b>10.6</b> colors for more groups</a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.7</b> change object</a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html"><i class="fa fa-check"></i><b>11</b> meconetcomp package</a>
<ul>
<li class="chapter" data-level="11.1" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-modularity-for-all-networks"><i class="fa fa-check"></i><b>11.1</b> Network modularity for all networks</a></li>
<li class="chapter" data-level="11.2" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-topological-attributes-for-all-networks"><i class="fa fa-check"></i><b>11.2</b> Network topological attributes for all networks</a></li>
<li class="chapter" data-level="11.3" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#node-and-edge-properties-extraction-for-all-networks"><i class="fa fa-check"></i><b>11.3</b> Node and edge properties extraction for all networks</a></li>
<li class="chapter" data-level="11.4" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-nodes-across-networks"><i class="fa fa-check"></i><b>11.4</b> Compare nodes across networks</a></li>
<li class="chapter" data-level="11.5" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-edges-across-networks"><i class="fa fa-check"></i><b>11.5</b> Compare edges across networks</a></li>
<li class="chapter" data-level="11.6" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#extract-overlapped-edges-of-networks-to-a-new-network"><i class="fa fa-check"></i><b>11.6</b> Extract overlapped edges of networks to a new network</a></li>
<li class="chapter" data-level="11.7" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-phylogenetic-distances-of-paired-nodes-in-edges"><i class="fa fa-check"></i><b>11.7</b> Compare phylogenetic distances of paired nodes in edges</a></li>
<li class="chapter" data-level="11.8" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-node-sources-of-edges-across-networks"><i class="fa fa-check"></i><b>11.8</b> Compare node sources of edges across networks</a></li>
<li class="chapter" data-level="11.9" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-topological-properties-of-sub-networks"><i class="fa fa-check"></i><b>11.9</b> Compare topological properties of sub-networks</a></li>
<li class="chapter" data-level="11.10" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#robustness-of-network"><i class="fa fa-check"></i><b>11.10</b> Robustness of network</a></li>
<li class="chapter" data-level="11.11" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#vulnerability-of-nodes"><i class="fa fa-check"></i><b>11.11</b> Vulnerability of nodes</a></li>
<li class="chapter" data-level="11.12" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#cohesion"><i class="fa fa-check"></i><b>11.12</b> Cohesion</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>12</b> mecodev package</a>
<ul>
<li class="chapter" data-level="12.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>12.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="12.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>12.2</b> trans_ts</a></li>
<li class="chapter" data-level="12.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>12.3</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mecoturn-package.html"><a href="mecoturn-package.html"><i class="fa fa-check"></i><b>13</b> mecoturn package</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mecoturn-package.html"><a href="mecoturn-package.html#betaturn-class"><i class="fa fa-check"></i><b>13.1</b> betaturn class</a></li>
<li class="chapter" data-level="13.2" data-path="mecoturn-package.html"><a href="mecoturn-package.html#taxaturn-class"><i class="fa fa-check"></i><b>13.2</b> taxaturn class</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>14</b> Other examples</a>
<ul>
<li class="chapter" data-level="14.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>14.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="14.2" data-path="other-examples-1.html"><a href="other-examples-1.html#tidy_taxonomy-function"><i class="fa fa-check"></i><b>14.2</b> tidy_taxonomy function</a></li>
<li class="chapter" data-level="14.3" data-path="other-examples-1.html"><a href="other-examples-1.html#show-the-abundance-of-unknown-taxa"><i class="fa fa-check"></i><b>14.3</b> Show the abundance of unknown taxa</a></li>
<li class="chapter" data-level="14.4" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>14.4</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="14.5" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>14.5</b> The use of phylogenetic tree</a></li>
<li class="chapter" data-level="14.6" data-path="other-examples-1.html"><a href="other-examples-1.html#trans_norm"><i class="fa fa-check"></i><b>14.6</b> trans_norm</a></li>
<li class="chapter" data-level="14.7" data-path="other-examples-1.html"><a href="other-examples-1.html#mantel-test-correlation-heatmap"><i class="fa fa-check"></i><b>14.7</b> Mantel test + Correlation heatmap</a></li>
<li class="chapter" data-level="14.8" data-path="other-examples-1.html"><a href="other-examples-1.html#pcoa-boxplot"><i class="fa fa-check"></i><b>14.8</b> PCoA + boxplot</a></li>
<li class="chapter" data-level="14.9" data-path="other-examples-1.html"><a href="other-examples-1.html#faprotax-differential-test"><i class="fa fa-check"></i><b>14.9</b> FAPROTAX + differential test</a></li>
<li class="chapter" data-level="14.10" data-path="other-examples-1.html"><a href="other-examples-1.html#rda-results-output"><i class="fa fa-check"></i><b>14.10</b> RDA results output</a></li>
<li class="chapter" data-level="14.11" data-path="other-examples-1.html"><a href="other-examples-1.html#betanribetanti-for-each-group"><i class="fa fa-check"></i><b>14.11</b> betaNRI/betaNTI for each group</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v1.8.0)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="other-examples-1" class="section level1 hasAnchor" number="14">
<h1><span class="header-section-number">Chapter 14</span> Other examples<a href="other-examples-1.html#other-examples-1" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>We’ve gathered a few tricks to make several issues easier to solve.
We encourage users to contribute some unique, special or helpful examples inspired by the packages.</p>
<div id="custom-taxa-order-in-bar-plot" class="section level2 hasAnchor" number="14.1">
<h2><span class="header-section-number">14.1</span> Custom taxa order in bar plot<a href="other-examples-1.html#custom-taxa-order-in-bar-plot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The flexibility of the package design can be reflected on many aspects.
In some cases, it is possible to tactfully handle some particular challenges.
For example, Dr.Jarrod contribute a cool answer to the question that how to use custom taxa and the order in bar plot by modifying the data inside the object.
This is a discussion topic in microeco Discussions (<a href="https://github.com/ChiLiubio/microeco/discussions/45" class="uri">https://github.com/ChiLiubio/microeco/discussions/45</a>).</p>
</div>
<div id="tidy_taxonomy-function" class="section level2 hasAnchor" number="14.2">
<h2><span class="header-section-number">14.2</span> tidy_taxonomy function<a href="other-examples-1.html#tidy_taxonomy-function" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The taxonomic classification with standard prefix is very important for some analyses,
e.g. taxonomic abundance plotting and biomarker finding.
The <code>tidy_taxonomy</code> function in microeco package is designed to make the taxa having standard prefix.
See those Issues with the detailed examples: (<a href="https://github.com/ChiLiubio/microeco/issues/32" class="uri">https://github.com/ChiLiubio/microeco/issues/32</a>), (<a href="https://github.com/ChiLiubio/microeco/issues/22" class="uri">https://github.com/ChiLiubio/microeco/issues/22</a>),
(<a href="https://github.com/ChiLiubio/microeco/issues/231" class="uri">https://github.com/ChiLiubio/microeco/issues/231</a>), (<a href="https://github.com/ChiLiubio/microeco/issues/254" class="uri">https://github.com/ChiLiubio/microeco/issues/254</a>) and (<a href="https://github.com/ChiLiubio/microeco/issues/338" class="uri">https://github.com/ChiLiubio/microeco/issues/338</a>).</p>
</div>
<div id="show-the-abundance-of-unknown-taxa" class="section level2 hasAnchor" number="14.3">
<h2><span class="header-section-number">14.3</span> Show the abundance of unknown taxa<a href="other-examples-1.html#show-the-abundance-of-unknown-taxa" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the <code>trans_abund</code> class, the default operation can filter the unclassified taxa, such as those with ’g__’.
Sometimes, these unknown taxa may be useful to represent uncultured species.
Under the circumstance, the abundance of unknown taxa can be used to show the ratio of unknown taxa in the system.
Please see the solution in this topic (<a href="https://github.com/ChiLiubio/microeco/issues/165" class="uri">https://github.com/ChiLiubio/microeco/issues/165</a>).</p>
</div>
<div id="question-of-prefix-in-the-taxa" class="section level2 hasAnchor" number="14.4">
<h2><span class="header-section-number">14.4</span> Question of prefix in the taxa<a href="other-examples-1.html#question-of-prefix-in-the-taxa" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The prefix of taxa in taxonomic table may affect the following performance of plotting, e.g. text in legend.
Please see those Issues (<a href="https://github.com/ChiLiubio/microeco/issues/32" class="uri">https://github.com/ChiLiubio/microeco/issues/32</a>), (<a href="https://github.com/ChiLiubio/microeco/issues/7" class="uri">https://github.com/ChiLiubio/microeco/issues/7</a>)
and (<a href="https://github.com/ChiLiubio/microeco/issues/15" class="uri">https://github.com/ChiLiubio/microeco/issues/15</a>).</p>
</div>
<div id="the-use-of-phylogenetic-tree" class="section level2 hasAnchor" number="14.5">
<h2><span class="header-section-number">14.5</span> The use of phylogenetic tree<a href="other-examples-1.html#the-use-of-phylogenetic-tree" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One of Issues referred to the basic use of phylogenetic tree in the microeco package (<a href="https://github.com/ChiLiubio/microeco/issues/33" class="uri">https://github.com/ChiLiubio/microeco/issues/33</a>).</p>
</div>
<div id="trans_norm" class="section level2 hasAnchor" number="14.6">
<h2><span class="header-section-number">14.6</span> trans_norm<a href="other-examples-1.html#trans_norm" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The class trans_norm in microeco package (&gt;= 1.6.0) provides several data normalization or transformation approaches for the microtable object or data.frame object.
The format of output is same with the input.
For the details and references of the approaches, please refer to the help document of the class with the command <code>?trans_norm</code>.</p>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="other-examples-1.html#cb385-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb385-2"><a href="other-examples-1.html#cb385-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataset)</span>
<span id="cb385-3"><a href="other-examples-1.html#cb385-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_norm<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset)</span>
<span id="cb385-4"><a href="other-examples-1.html#cb385-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Geometric mean of pairwise ratios</span></span>
<span id="cb385-5"><a href="other-examples-1.html#cb385-5" aria-hidden="true" tabindex="-1"></a>d1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">norm</span>(<span class="at">method =</span> <span class="st">&quot;GMPR&quot;</span>)</span>
<span id="cb385-6"><a href="other-examples-1.html#cb385-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Centered log-ratio normalization</span></span>
<span id="cb385-7"><a href="other-examples-1.html#cb385-7" aria-hidden="true" tabindex="-1"></a>d2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">norm</span>(<span class="at">method =</span> <span class="st">&quot;clr&quot;</span>)</span>
<span id="cb385-8"><a href="other-examples-1.html#cb385-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Robust centered log-ratio normalization</span></span>
<span id="cb385-9"><a href="other-examples-1.html#cb385-9" aria-hidden="true" tabindex="-1"></a>d2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">norm</span>(<span class="at">method =</span> <span class="st">&quot;rclr&quot;</span>)</span>
<span id="cb385-10"><a href="other-examples-1.html#cb385-10" aria-hidden="true" tabindex="-1"></a><span class="co"># d2 is another microtable object</span></span>
<span id="cb385-11"><a href="other-examples-1.html#cb385-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Cumulative sum scaling normalization</span></span>
<span id="cb385-12"><a href="other-examples-1.html#cb385-12" aria-hidden="true" tabindex="-1"></a>d2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">norm</span>(<span class="at">method =</span> <span class="st">&quot;CCS&quot;</span>)</span>
<span id="cb385-13"><a href="other-examples-1.html#cb385-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Total sum scaling, dividing counts by the sequencing depth</span></span>
<span id="cb385-14"><a href="other-examples-1.html#cb385-14" aria-hidden="true" tabindex="-1"></a>d2 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">norm</span>(<span class="at">method =</span> <span class="st">&quot;TSS&quot;</span>)</span>
<span id="cb385-15"><a href="other-examples-1.html#cb385-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-16"><a href="other-examples-1.html#cb385-16" aria-hidden="true" tabindex="-1"></a><span class="co"># log transformation</span></span>
<span id="cb385-17"><a href="other-examples-1.html#cb385-17" aria-hidden="true" tabindex="-1"></a>d3 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">norm</span>(<span class="at">method =</span> <span class="st">&quot;log&quot;</span>)</span></code></pre></div>
</div>
<div id="mantel-test-correlation-heatmap" class="section level2 hasAnchor" number="14.7">
<h2><span class="header-section-number">14.7</span> Mantel test + Correlation heatmap<a href="other-examples-1.html#mantel-test-correlation-heatmap" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Sometimes, mantel test results and correlations of environmental factors are visualized together to deliver more information.
In this example, we show how to extend the mantel test results from <code>trans_env</code> class with the package ggcor.
Please install <code>ggcor</code> package according to the document (<a href="https://chiliubio.github.io/microeco_tutorial/intro.html#github-packages" class="uri">https://chiliubio.github.io/microeco_tutorial/intro.html#github-packages</a>).</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb386-1"><a href="other-examples-1.html#cb386-1" aria-hidden="true" tabindex="-1"></a><span class="co"># prepare data</span></span>
<span id="cb386-2"><a href="other-examples-1.html#cb386-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb386-3"><a href="other-examples-1.html#cb386-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb386-4"><a href="other-examples-1.html#cb386-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataset)</span>
<span id="cb386-5"><a href="other-examples-1.html#cb386-5" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(env_data_16S)</span>
<span id="cb386-6"><a href="other-examples-1.html#cb386-6" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>sample_table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(dataset<span class="sc">$</span>sample_table, env_data_16S[<span class="fu">rownames</span>(dataset<span class="sc">$</span>sample_table), ])</span>
<span id="cb386-7"><a href="other-examples-1.html#cb386-7" aria-hidden="true" tabindex="-1"></a><span class="co"># extract two phyla to show the steps</span></span>
<span id="cb386-8"><a href="other-examples-1.html#cb386-8" aria-hidden="true" tabindex="-1"></a>d1 <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb386-9"><a href="other-examples-1.html#cb386-9" aria-hidden="true" tabindex="-1"></a>d1<span class="sc">$</span>tax_table <span class="ot">&lt;-</span> d1<span class="sc">$</span>tax_table[d1<span class="sc">$</span>tax_table<span class="sc">$</span>Phylum <span class="sc">==</span> <span class="st">&quot;p__Proteobacteria&quot;</span>, ]</span>
<span id="cb386-10"><a href="other-examples-1.html#cb386-10" aria-hidden="true" tabindex="-1"></a>d1<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb386-11"><a href="other-examples-1.html#cb386-11" aria-hidden="true" tabindex="-1"></a>d1<span class="sc">$</span><span class="fu">cal_betadiv</span>()</span>
<span id="cb386-12"><a href="other-examples-1.html#cb386-12" aria-hidden="true" tabindex="-1"></a>d2 <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb386-13"><a href="other-examples-1.html#cb386-13" aria-hidden="true" tabindex="-1"></a>d2<span class="sc">$</span>tax_table <span class="ot">&lt;-</span> d2<span class="sc">$</span>tax_table[d2<span class="sc">$</span>tax_table<span class="sc">$</span>Phylum <span class="sc">==</span> <span class="st">&quot;p__Bacteroidetes&quot;</span>, ]</span>
<span id="cb386-14"><a href="other-examples-1.html#cb386-14" aria-hidden="true" tabindex="-1"></a>d2<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb386-15"><a href="other-examples-1.html#cb386-15" aria-hidden="true" tabindex="-1"></a>d2<span class="sc">$</span><span class="fu">cal_betadiv</span>()</span>
<span id="cb386-16"><a href="other-examples-1.html#cb386-16" aria-hidden="true" tabindex="-1"></a><span class="co"># first perform mantel test</span></span>
<span id="cb386-17"><a href="other-examples-1.html#cb386-17" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> d1, <span class="at">env_cols =</span> <span class="dv">8</span><span class="sc">:</span><span class="dv">15</span>)</span>
<span id="cb386-18"><a href="other-examples-1.html#cb386-18" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_mantel</span>(<span class="at">use_measure =</span> <span class="st">&quot;bray&quot;</span>, <span class="at">partial_mantel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb386-19"><a href="other-examples-1.html#cb386-19" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> d2, <span class="at">env_cols =</span> <span class="dv">8</span><span class="sc">:</span><span class="dv">15</span>)</span>
<span id="cb386-20"><a href="other-examples-1.html#cb386-20" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_mantel</span>(<span class="at">use_measure =</span> <span class="st">&quot;bray&quot;</span>, <span class="at">partial_mantel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb386-21"><a href="other-examples-1.html#cb386-21" aria-hidden="true" tabindex="-1"></a><span class="co"># extract a part of the results </span></span>
<span id="cb386-22"><a href="other-examples-1.html#cb386-22" aria-hidden="true" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">spec =</span> <span class="st">&quot;Proteobacteria&quot;</span>, t1<span class="sc">$</span>res_mantel) <span class="sc">%&gt;%</span> .[, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">6</span>, <span class="dv">8</span>)]</span>
<span id="cb386-23"><a href="other-examples-1.html#cb386-23" aria-hidden="true" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">spec =</span> <span class="st">&quot;Bacteroidetes&quot;</span>, t2<span class="sc">$</span>res_mantel) <span class="sc">%&gt;%</span> .[, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">6</span>, <span class="dv">8</span>)]</span>
<span id="cb386-24"><a href="other-examples-1.html#cb386-24" aria-hidden="true" tabindex="-1"></a><span class="co"># rename columns</span></span>
<span id="cb386-25"><a href="other-examples-1.html#cb386-25" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(x1) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x2) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;spec&quot;</span>, <span class="st">&quot;env&quot;</span>, <span class="st">&quot;r&quot;</span>, <span class="st">&quot;p.value&quot;</span>)</span>
<span id="cb386-26"><a href="other-examples-1.html#cb386-26" aria-hidden="true" tabindex="-1"></a><span class="co"># generate interval data</span></span>
<span id="cb386-27"><a href="other-examples-1.html#cb386-27" aria-hidden="true" tabindex="-1"></a>x1 <span class="sc">%&lt;&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">rd =</span> <span class="fu">cut</span>(r, <span class="at">breaks =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">0.3</span>, <span class="fl">0.6</span>, <span class="cn">Inf</span>), <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;&lt; 0.3&quot;</span>, <span class="st">&quot;0.3 - 0.6&quot;</span>, <span class="st">&quot;&gt;= 0.6&quot;</span>)),</span>
<span id="cb386-28"><a href="other-examples-1.html#cb386-28" aria-hidden="true" tabindex="-1"></a>    <span class="at">pd =</span> <span class="fu">cut</span>(p.value, <span class="at">breaks =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="cn">Inf</span>), <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;&lt; 0.01&quot;</span>, <span class="st">&quot;0.01 - 0.05&quot;</span>, <span class="st">&quot;&gt;= 0.05&quot;</span>)))</span>
<span id="cb386-29"><a href="other-examples-1.html#cb386-29" aria-hidden="true" tabindex="-1"></a>x2 <span class="sc">%&lt;&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">rd =</span> <span class="fu">cut</span>(r, <span class="at">breaks =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">0.3</span>, <span class="fl">0.6</span>, <span class="cn">Inf</span>), <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;&lt; 0.3&quot;</span>, <span class="st">&quot;0.3 - 0.6&quot;</span>, <span class="st">&quot;&gt;= 0.6&quot;</span>)),</span>
<span id="cb386-30"><a href="other-examples-1.html#cb386-30" aria-hidden="true" tabindex="-1"></a>    <span class="at">pd =</span> <span class="fu">cut</span>(p.value, <span class="at">breaks =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">0.01</span>, <span class="fl">0.05</span>, <span class="cn">Inf</span>), <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;&lt; 0.01&quot;</span>, <span class="st">&quot;0.01 - 0.05&quot;</span>, <span class="st">&quot;&gt;= 0.05&quot;</span>)))</span>
<span id="cb386-31"><a href="other-examples-1.html#cb386-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-32"><a href="other-examples-1.html#cb386-32" aria-hidden="true" tabindex="-1"></a><span class="co"># cobine two tables</span></span>
<span id="cb386-33"><a href="other-examples-1.html#cb386-33" aria-hidden="true" tabindex="-1"></a>plot_table <span class="ot">&lt;-</span> <span class="fu">rbind</span>(x1, x2)</span>
<span id="cb386-34"><a href="other-examples-1.html#cb386-34" aria-hidden="true" tabindex="-1"></a><span class="co"># install ggcor following the steps (https://chiliubio.github.io/microeco_tutorial/intro.html#github-packages)</span></span>
<span id="cb386-35"><a href="other-examples-1.html#cb386-35" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb386-36"><a href="other-examples-1.html#cb386-36" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggcor)</span>
<span id="cb386-37"><a href="other-examples-1.html#cb386-37" aria-hidden="true" tabindex="-1"></a><span class="fu">set_scale</span>()</span>
<span id="cb386-38"><a href="other-examples-1.html#cb386-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-39"><a href="other-examples-1.html#cb386-39" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> <span class="fu">quickcor</span>(t1<span class="sc">$</span>data_env, <span class="at">type =</span> <span class="st">&quot;upper&quot;</span>, <span class="at">cor.test =</span> <span class="cn">TRUE</span>, <span class="at">show.diag =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb386-40"><a href="other-examples-1.html#cb386-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_square</span>() <span class="sc">+</span></span>
<span id="cb386-41"><a href="other-examples-1.html#cb386-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_mark</span>(<span class="at">sig.thres =</span> <span class="fl">0.05</span>, <span class="at">markonly =</span> <span class="cn">TRUE</span>, <span class="at">color =</span> <span class="st">&quot;black&quot;</span>, <span class="at">size =</span> <span class="dv">6</span>) <span class="sc">+</span></span>
<span id="cb386-42"><a href="other-examples-1.html#cb386-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">anno_link</span>(<span class="fu">aes</span>(<span class="at">colour =</span> pd, <span class="at">size =</span> rd), <span class="at">data =</span> plot_table) <span class="sc">+</span></span>
<span id="cb386-43"><a href="other-examples-1.html#cb386-43" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_size_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">1.5</span>, <span class="dv">3</span>)) <span class="sc">+</span></span>
<span id="cb386-44"><a href="other-examples-1.html#cb386-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">&quot;#D95F02&quot;</span>, <span class="st">&quot;#1B9E77&quot;</span>, <span class="st">&quot;#A2A2A288&quot;</span>)) <span class="sc">+</span></span>
<span id="cb386-45"><a href="other-examples-1.html#cb386-45" aria-hidden="true" tabindex="-1"></a>    <span class="fu">guides</span>(<span class="at">size =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">&quot;Mantel&#39;s r&quot;</span>, <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">colour =</span> <span class="st">&quot;grey35&quot;</span>), <span class="at">order =</span> <span class="dv">2</span>),</span>
<span id="cb386-46"><a href="other-examples-1.html#cb386-46" aria-hidden="true" tabindex="-1"></a>           <span class="at">colour =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">&quot;Mantel&#39;s p&quot;</span>, <span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>), <span class="at">order =</span> <span class="dv">1</span>),</span>
<span id="cb386-47"><a href="other-examples-1.html#cb386-47" aria-hidden="true" tabindex="-1"></a>           <span class="at">fill =</span> <span class="fu">guide_colorbar</span>(<span class="at">title =</span> <span class="st">&quot;Pearson&#39;s r&quot;</span>, <span class="at">order =</span> <span class="dv">3</span>))</span>
<span id="cb386-48"><a href="other-examples-1.html#cb386-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb386-49"><a href="other-examples-1.html#cb386-49" aria-hidden="true" tabindex="-1"></a>g1</span></code></pre></div>
<p><img src="Images/plot_env_mantel_ggcor.png" width="650px" style="display: block; margin: auto;" /></p>
</div>
<div id="pcoa-boxplot" class="section level2 hasAnchor" number="14.8">
<h2><span class="header-section-number">14.8</span> PCoA + boxplot<a href="other-examples-1.html#pcoa-boxplot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the sections involving trans_venn and trans_env classes,
we mentioned some combination graphs generated based on the <code>aplot</code> package.
Here, we continue to use the <code>aplot</code> and <code>microeco</code> packages to demonstrate how to add boxplots of coordinate scores to PCoA graphs.
Similar combination graphs can be easily manipulated, so we no longer include this type of visualization in the plotting functions of ordination methods.
Firstly, we use <code>trans_beta</code> class to compute PCoA, then perform differential test based on the <code>cal_diff</code> function in <code>trans_env</code> class.</p>
<div class="sourceCode" id="cb387"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb387-1"><a href="other-examples-1.html#cb387-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb387-2"><a href="other-examples-1.html#cb387-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb387-3"><a href="other-examples-1.html#cb387-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb387-4"><a href="other-examples-1.html#cb387-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(aplot)</span>
<span id="cb387-5"><a href="other-examples-1.html#cb387-5" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>())</span>
<span id="cb387-6"><a href="other-examples-1.html#cb387-6" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataset)</span>
<span id="cb387-7"><a href="other-examples-1.html#cb387-7" aria-hidden="true" tabindex="-1"></a><span class="co"># PCoA</span></span>
<span id="cb387-8"><a href="other-examples-1.html#cb387-8" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;bray&quot;</span>)</span>
<span id="cb387-9"><a href="other-examples-1.html#cb387-9" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">method =</span> <span class="st">&quot;PCoA&quot;</span>)</span>
<span id="cb387-10"><a href="other-examples-1.html#cb387-10" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the axis scores</span></span>
<span id="cb387-11"><a href="other-examples-1.html#cb387-11" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> t1<span class="sc">$</span>res_ordination<span class="sc">$</span>scores</span>
<span id="cb387-12"><a href="other-examples-1.html#cb387-12" aria-hidden="true" tabindex="-1"></a><span class="co"># differential test with trans_env class</span></span>
<span id="cb387-13"><a href="other-examples-1.html#cb387-13" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> tmp[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])</span>
<span id="cb387-14"><a href="other-examples-1.html#cb387-14" aria-hidden="true" tabindex="-1"></a><span class="co"># &#39;KW_dunn&#39; for non-parametric test</span></span>
<span id="cb387-15"><a href="other-examples-1.html#cb387-15" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_diff</span>(<span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">method =</span> <span class="st">&quot;anova&quot;</span>)</span></code></pre></div>
<p>Then, employ the plotting functions in <code>t1</code> and <code>t2</code> objects for visualization.</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb388-1"><a href="other-examples-1.html#cb388-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_type =</span> <span class="fu">c</span>(<span class="st">&quot;point&quot;</span>, <span class="st">&quot;ellipse&quot;</span>))</span>
<span id="cb388-2"><a href="other-examples-1.html#cb388-2" aria-hidden="true" tabindex="-1"></a><span class="co"># groups order in p2 is same with p1; use legend.position = &quot;none&quot; to remove redundant legend</span></span>
<span id="cb388-3"><a href="other-examples-1.html#cb388-3" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> t2<span class="sc">$</span><span class="fu">plot_diff</span>(<span class="at">measure =</span> <span class="st">&quot;PCo1&quot;</span>, <span class="at">add_sig =</span> T) <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">coord_flip</span>() <span class="sc">+</span> </span>
<span id="cb388-4"><a href="other-examples-1.html#cb388-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>, <span class="at">axis.title.x =</span> <span class="fu">element_blank</span>(), <span class="at">axis.text.y =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>())</span>
<span id="cb388-5"><a href="other-examples-1.html#cb388-5" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> t2<span class="sc">$</span><span class="fu">plot_diff</span>(<span class="at">measure =</span> <span class="st">&quot;PCo2&quot;</span>, <span class="at">add_sig =</span> T) <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb388-6"><a href="other-examples-1.html#cb388-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;none&quot;</span>, <span class="at">axis.title.y =</span> <span class="fu">element_blank</span>(), <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>())</span>
<span id="cb388-7"><a href="other-examples-1.html#cb388-7" aria-hidden="true" tabindex="-1"></a><span class="co"># height of the upper figure and width of the right-hand figure are both 0.2-fold of the main figure</span></span>
<span id="cb388-8"><a href="other-examples-1.html#cb388-8" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> p1 <span class="sc">%&gt;%</span> <span class="fu">insert_top</span>(p2, <span class="at">height =</span> <span class="fl">0.2</span>) <span class="sc">%&gt;%</span> <span class="fu">insert_right</span>(p3, <span class="at">width =</span> <span class="fl">0.2</span>)</span>
<span id="cb388-9"><a href="other-examples-1.html#cb388-9" aria-hidden="true" tabindex="-1"></a>g</span></code></pre></div>
<p><img src="Images/Other_PCoA_boxplot1.png" width="650px" style="display: block; margin: auto;" /></p>
<p>At this point,
we notice that the horizontal axis of the upper figure and the vertical axis of the right-hand figure do not correspond precisely with those of the main figure.
Therefore, if we proceed with these graphs,
the axis ticks of the upper and right-hand figures should be kept.
If the user requires complete correspondence of the ticks,
further control over the coordinate axes should be implemented.
Here, we manually control the coordinate range.</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb389-1"><a href="other-examples-1.html#cb389-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use 1.4-fold of the scores as axis ranges</span></span>
<span id="cb389-2"><a href="other-examples-1.html#cb389-2" aria-hidden="true" tabindex="-1"></a>x_lim <span class="ot">&lt;-</span> <span class="fu">range</span>(tmp[, <span class="dv">1</span>]) <span class="sc">*</span> <span class="fl">1.4</span></span>
<span id="cb389-3"><a href="other-examples-1.html#cb389-3" aria-hidden="true" tabindex="-1"></a>y_lim <span class="ot">&lt;-</span> <span class="fu">range</span>(tmp[, <span class="dv">2</span>]) <span class="sc">*</span> <span class="fl">1.4</span></span>
<span id="cb389-4"><a href="other-examples-1.html#cb389-4" aria-hidden="true" tabindex="-1"></a><span class="co"># limit x and y axis without any extension</span></span>
<span id="cb389-5"><a href="other-examples-1.html#cb389-5" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> p1 <span class="sc">+</span> <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> y_lim, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span> </span>
<span id="cb389-6"><a href="other-examples-1.html#cb389-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> x_lim, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>))</span>
<span id="cb389-7"><a href="other-examples-1.html#cb389-7" aria-hidden="true" tabindex="-1"></a><span class="co"># limit x axis of upper figure (it&#39;s y axis when flipped)</span></span>
<span id="cb389-8"><a href="other-examples-1.html#cb389-8" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> p2 <span class="sc">+</span> <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> x_lim, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>))</span>
<span id="cb389-9"><a href="other-examples-1.html#cb389-9" aria-hidden="true" tabindex="-1"></a><span class="co"># limit y axis of right-hand figure</span></span>
<span id="cb389-10"><a href="other-examples-1.html#cb389-10" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> p3 <span class="sc">+</span> <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> y_lim, <span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>))</span>
<span id="cb389-11"><a href="other-examples-1.html#cb389-11" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> p1 <span class="sc">%&gt;%</span> <span class="fu">insert_top</span>(p2, <span class="at">height =</span> <span class="fl">0.2</span>) <span class="sc">%&gt;%</span> <span class="fu">insert_right</span>(p3, <span class="at">width =</span> <span class="fl">0.2</span>)</span>
<span id="cb389-12"><a href="other-examples-1.html#cb389-12" aria-hidden="true" tabindex="-1"></a>g</span></code></pre></div>
<p>Now, the ticks of the coordinate axes fully correspond, and we may selectively remove the ticks on the upper and right-hand figures.</p>
<div class="sourceCode" id="cb390"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb390-1"><a href="other-examples-1.html#cb390-1" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> p2 <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.x =</span> <span class="fu">element_blank</span>())</span>
<span id="cb390-2"><a href="other-examples-1.html#cb390-2" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> p3 <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.y =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.y =</span> <span class="fu">element_blank</span>())</span>
<span id="cb390-3"><a href="other-examples-1.html#cb390-3" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> p1 <span class="sc">%&gt;%</span> <span class="fu">insert_top</span>(p2, <span class="at">height =</span> <span class="fl">0.2</span>) <span class="sc">%&gt;%</span> <span class="fu">insert_right</span>(p3, <span class="at">width =</span> <span class="fl">0.2</span>)</span>
<span id="cb390-4"><a href="other-examples-1.html#cb390-4" aria-hidden="true" tabindex="-1"></a>g</span>
<span id="cb390-5"><a href="other-examples-1.html#cb390-5" aria-hidden="true" tabindex="-1"></a><span class="co"># save g to computer</span></span>
<span id="cb390-6"><a href="other-examples-1.html#cb390-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">&quot;test1.pdf&quot;</span>, g, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">height=</span> <span class="dv">6</span>)</span></code></pre></div>
<p><img src="Images/Other_PCoA_boxplot2.png" width="650px" style="display: block; margin: auto;" /></p>
</div>
<div id="faprotax-differential-test" class="section level2 hasAnchor" number="14.9">
<h2><span class="header-section-number">14.9</span> FAPROTAX + differential test<a href="other-examples-1.html#faprotax-differential-test" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the <code>trans_func</code> section, we showed an example of correlations between the abundances of OTUs with the traits from FAPROTAX database and environmental factors.
Now, we go ahead and use <code>trans_diff</code> class to demonstrate the differential test of the abundances of OTUs with the traits across groups.</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb391-1"><a href="other-examples-1.html#cb391-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb391-2"><a href="other-examples-1.html#cb391-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataset)</span>
<span id="cb391-3"><a href="other-examples-1.html#cb391-3" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(dataset)</span>
<span id="cb391-4"><a href="other-examples-1.html#cb391-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_spe_func</span>(<span class="at">prok_database =</span> <span class="st">&quot;FAPROTAX&quot;</span>)</span>
<span id="cb391-5"><a href="other-examples-1.html#cb391-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">abundance_weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb391-6"><a href="other-examples-1.html#cb391-6" aria-hidden="true" tabindex="-1"></a><span class="co"># it is better to clone a dataset</span></span>
<span id="cb391-7"><a href="other-examples-1.html#cb391-7" aria-hidden="true" tabindex="-1"></a>tmp_mt <span class="ot">&lt;-</span> <span class="fu">clone</span>(dataset)</span>
<span id="cb391-8"><a href="other-examples-1.html#cb391-8" aria-hidden="true" tabindex="-1"></a><span class="co"># transpose res_spe_func_perc to be a data.frame like taxonomic abundance</span></span>
<span id="cb391-9"><a href="other-examples-1.html#cb391-9" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(t1<span class="sc">$</span>res_spe_func_perc), <span class="at">check.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb391-10"><a href="other-examples-1.html#cb391-10" aria-hidden="true" tabindex="-1"></a><span class="co"># assign the table back to taxa_abund list for further analysis</span></span>
<span id="cb391-11"><a href="other-examples-1.html#cb391-11" aria-hidden="true" tabindex="-1"></a>tmp_mt<span class="sc">$</span>taxa_abund<span class="sc">$</span>func <span class="ot">&lt;-</span> tmp</span>
<span id="cb391-12"><a href="other-examples-1.html#cb391-12" aria-hidden="true" tabindex="-1"></a><span class="co"># select the &quot;func&quot; in taxa_abund list in trans_diff</span></span>
<span id="cb391-13"><a href="other-examples-1.html#cb391-13" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> tmp_mt, <span class="at">method =</span> <span class="st">&quot;anova&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;func&quot;</span>)</span>
<span id="cb391-14"><a href="other-examples-1.html#cb391-14" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">add_sig =</span> T) <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">ylab</span>(<span class="st">&quot;Relative abundance (%)&quot;</span>)</span></code></pre></div>
</div>
<div id="rda-results-output" class="section level2 hasAnchor" number="14.10">
<h2><span class="header-section-number">14.10</span> RDA results output<a href="other-examples-1.html#rda-results-output" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In some analyses, it is often necessary to preserve intermediate files of various analysis results,
and not all files are in the data.frame format.
For example, in RDA analysis, the <code>res_ordination_R2</code> generated by <code>cal_ordination</code> function is a numeric vector,
and the <code>res_ordination_envfit</code> generated by the <code>cal_ordination_envfit</code> function is of envfit class.
For those outputs, we can utilize the <code>capture.output</code> function to redirect the contents displayed in the dialogue box into a file.</p>
<div class="sourceCode" id="cb392"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb392-1"><a href="other-examples-1.html#cb392-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb392-2"><a href="other-examples-1.html#cb392-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(dataset)</span>
<span id="cb392-3"><a href="other-examples-1.html#cb392-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(env_data_16S)</span>
<span id="cb392-4"><a href="other-examples-1.html#cb392-4" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_env<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">add_data =</span> env_data_16S[, <span class="dv">4</span><span class="sc">:</span><span class="dv">11</span>])</span>
<span id="cb392-5"><a href="other-examples-1.html#cb392-5" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination</span>(<span class="at">method =</span> <span class="st">&quot;RDA&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Genus&quot;</span>)</span>
<span id="cb392-6"><a href="other-examples-1.html#cb392-6" aria-hidden="true" tabindex="-1"></a><span class="co"># get the significance of the terms</span></span>
<span id="cb392-7"><a href="other-examples-1.html#cb392-7" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination_anova</span>()</span>
<span id="cb392-8"><a href="other-examples-1.html#cb392-8" aria-hidden="true" tabindex="-1"></a><span class="co"># fit factors onto the ordination to get R2 for each factor</span></span>
<span id="cb392-9"><a href="other-examples-1.html#cb392-9" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_ordination_envfit</span>()</span>
<span id="cb392-10"><a href="other-examples-1.html#cb392-10" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">trans_ordination</span>(<span class="at">adjust_arrow_length =</span> <span class="cn">TRUE</span>)</span>
<span id="cb392-11"><a href="other-examples-1.html#cb392-11" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">plot_ordination</span>(<span class="at">plot_color =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">plot_shape =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb392-12"><a href="other-examples-1.html#cb392-12" aria-hidden="true" tabindex="-1"></a>ggplot2<span class="sc">::</span><span class="fu">ggsave</span>(<span class="st">&quot;RDA.pdf&quot;</span>, g1, <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="fl">6.5</span>)</span>
<span id="cb392-13"><a href="other-examples-1.html#cb392-13" aria-hidden="true" tabindex="-1"></a><span class="co"># use capture.output to save output</span></span>
<span id="cb392-14"><a href="other-examples-1.html#cb392-14" aria-hidden="true" tabindex="-1"></a><span class="fu">capture.output</span>(t1<span class="sc">$</span>res_ordination_R2, <span class="at">file =</span> <span class="st">&quot;RDA_R2.txt&quot;</span>)</span>
<span id="cb392-15"><a href="other-examples-1.html#cb392-15" aria-hidden="true" tabindex="-1"></a><span class="fu">capture.output</span>(t1<span class="sc">$</span>res_ordination_envfit, <span class="at">file =</span> <span class="st">&quot;RDA_envfit.txt&quot;</span>)</span>
<span id="cb392-16"><a href="other-examples-1.html#cb392-16" aria-hidden="true" tabindex="-1"></a><span class="co"># save data.frame objects</span></span>
<span id="cb392-17"><a href="other-examples-1.html#cb392-17" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(t1<span class="sc">$</span>res_ordination_terms, <span class="st">&quot;RDA_anova_termsig.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb392-18"><a href="other-examples-1.html#cb392-18" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(t1<span class="sc">$</span>res_ordination_axis, <span class="st">&quot;RDA_anova_axissig.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb392-19"><a href="other-examples-1.html#cb392-19" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(t1<span class="sc">$</span>res_ordination_trans<span class="sc">$</span>df_sites, <span class="st">&quot;RDA_axis_sample.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb392-20"><a href="other-examples-1.html#cb392-20" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(t1<span class="sc">$</span>res_ordination_trans<span class="sc">$</span>df_arrows, <span class="st">&quot;RDA_axis_term.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span>
<span id="cb392-21"><a href="other-examples-1.html#cb392-21" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(t1<span class="sc">$</span>res_ordination_trans<span class="sc">$</span>df_arrows_spe, <span class="st">&quot;RDA_axis_taxa.txt&quot;</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>)</span></code></pre></div>
</div>
<div id="betanribetanti-for-each-group" class="section level2 hasAnchor" number="14.11">
<h2><span class="header-section-number">14.11</span> betaNRI/betaNTI for each group<a href="other-examples-1.html#betanribetanti-for-each-group" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Sometimes, if you want to perform betaNRI/betaNTI for each group individually (one group is considered one species pool),
you should calculate the index for each group, respectively.</p>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb393-1"><a href="other-examples-1.html#cb393-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb393-2"><a href="other-examples-1.html#cb393-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb393-3"><a href="other-examples-1.html#cb393-3" aria-hidden="true" tabindex="-1"></a><span class="co"># we create a list to store the trans_nullmodel results.</span></span>
<span id="cb393-4"><a href="other-examples-1.html#cb393-4" aria-hidden="true" tabindex="-1"></a>sesbeta_each <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb393-5"><a href="other-examples-1.html#cb393-5" aria-hidden="true" tabindex="-1"></a>group_col <span class="ot">&lt;-</span> <span class="st">&quot;Group&quot;</span></span>
<span id="cb393-6"><a href="other-examples-1.html#cb393-6" aria-hidden="true" tabindex="-1"></a>all_groups <span class="ot">&lt;-</span> <span class="fu">unique</span>(dataset<span class="sc">$</span>sample_table[, group_col])</span>
<span id="cb393-7"><a href="other-examples-1.html#cb393-7" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate for each group, respectively</span></span>
<span id="cb393-8"><a href="other-examples-1.html#cb393-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> all_groups){</span>
<span id="cb393-9"><a href="other-examples-1.html#cb393-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># like the above operation, but need provide &#39;group&#39; and &#39;select_group&#39;</span></span>
<span id="cb393-10"><a href="other-examples-1.html#cb393-10" aria-hidden="true" tabindex="-1"></a>    test <span class="ot">&lt;-</span> trans_nullmodel<span class="sc">$</span><span class="fu">new</span>(dataset, <span class="at">group =</span> group_col, <span class="at">select_group =</span> i, <span class="at">filter_thres =</span> <span class="fl">0.0005</span>)</span>
<span id="cb393-11"><a href="other-examples-1.html#cb393-11" aria-hidden="true" tabindex="-1"></a>    test<span class="sc">$</span><span class="fu">cal_ses_betampd</span>(<span class="at">runs =</span> <span class="dv">500</span>, <span class="at">abundance.weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb393-12"><a href="other-examples-1.html#cb393-12" aria-hidden="true" tabindex="-1"></a>    sesbeta_each[[i]] <span class="ot">&lt;-</span> test<span class="sc">$</span>res_ses_betampd</span>
<span id="cb393-13"><a href="other-examples-1.html#cb393-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb393-14"><a href="other-examples-1.html#cb393-14" aria-hidden="true" tabindex="-1"></a><span class="co"># merge and reshape to generate one symmetrical matrix</span></span>
<span id="cb393-15"><a href="other-examples-1.html#cb393-15" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">lapply</span>(sesbeta_each, reshape2<span class="sc">::</span>melt) <span class="sc">%&gt;%</span> </span>
<span id="cb393-16"><a href="other-examples-1.html#cb393-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">do.call</span>(rbind, .) <span class="sc">%&gt;%</span></span>
<span id="cb393-17"><a href="other-examples-1.html#cb393-17" aria-hidden="true" tabindex="-1"></a>    reshape2<span class="sc">::</span><span class="fu">dcast</span>(., Var1<span class="sc">~</span>Var2, <span class="at">value.var =</span> <span class="st">&quot;value&quot;</span>)</span>
<span id="cb393-18"><a href="other-examples-1.html#cb393-18" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(test) <span class="ot">&lt;-</span> test[, <span class="dv">1</span>]</span>
<span id="cb393-19"><a href="other-examples-1.html#cb393-19" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> test[, <span class="sc">-</span><span class="dv">1</span>, drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb393-20"><a href="other-examples-1.html#cb393-20" aria-hidden="true" tabindex="-1"></a><span class="co"># like the above operation</span></span>
<span id="cb393-21"><a href="other-examples-1.html#cb393-21" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>beta_diversity[[<span class="st">&quot;betaNRI&quot;</span>]] <span class="ot">&lt;-</span> test</span>
<span id="cb393-22"><a href="other-examples-1.html#cb393-22" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_beta<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">measure =</span> <span class="st">&quot;betaNRI&quot;</span>)</span>
<span id="cb393-23"><a href="other-examples-1.html#cb393-23" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance</span>()</span>
<span id="cb393-24"><a href="other-examples-1.html#cb393-24" aria-hidden="true" tabindex="-1"></a><span class="co"># statistical analysis</span></span>
<span id="cb393-25"><a href="other-examples-1.html#cb393-25" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_group_distance_diff</span>(<span class="at">method =</span> <span class="st">&quot;wilcox&quot;</span>)</span>
<span id="cb393-26"><a href="other-examples-1.html#cb393-26" aria-hidden="true" tabindex="-1"></a>g1 <span class="ot">&lt;-</span> t2<span class="sc">$</span><span class="fu">plot_group_distance</span>(<span class="at">boxplot_add =</span> <span class="st">&quot;mean&quot;</span>)</span>
<span id="cb393-27"><a href="other-examples-1.html#cb393-27" aria-hidden="true" tabindex="-1"></a>g1 <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">2</span>, <span class="at">linetype =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="Images/trans_nullmodel_betaNRI_each_dataset.png" width="550px" style="display: block; margin: auto;" /></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="mecoturn-package.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/13-Other_examples.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
