<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 9 Other dataset | Tutorial for R microeco package (v1.9.0)</title>
  <meta name="description" content="Examples for R microeco (v1.9.0), file2meco (v0.8.0), meconetcomp (v0.5.1) and mecoturn packages" />
  <meta name="generator" content="bookdown 0.40 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 9 Other dataset | Tutorial for R microeco package (v1.9.0)" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Examples for R microeco (v1.9.0), file2meco (v0.8.0), meconetcomp (v0.5.1) and mecoturn packages" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 9 Other dataset | Tutorial for R microeco package (v1.9.0)" />
  
  <meta name="twitter:description" content="Examples for R microeco (v1.9.0), file2meco (v0.8.0), meconetcomp (v0.5.1) and mecoturn packages" />
  



<meta name="date" content="2024-08-18" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="file2meco-package.html"/>
<link rel="next" href="notes.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#framework"><i class="fa fa-check"></i><b>2.1</b> Framework</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#rtools"><i class="fa fa-check"></i><b>2.4</b> RTools</a></li>
<li class="chapter" data-level="2.5" data-path="intro.html"><a href="intro.html#dependence"><i class="fa fa-check"></i><b>2.5</b> Dependence</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="intro.html"><a href="intro.html#description"><i class="fa fa-check"></i><b>2.5.1</b> Description</a></li>
<li class="chapter" data-level="2.5.2" data-path="intro.html"><a href="intro.html#cran-packages"><i class="fa fa-check"></i><b>2.5.2</b> CRAN packages</a></li>
<li class="chapter" data-level="2.5.3" data-path="intro.html"><a href="intro.html#bioconductor-packages"><i class="fa fa-check"></i><b>2.5.3</b> Bioconductor packages</a></li>
<li class="chapter" data-level="2.5.4" data-path="intro.html"><a href="intro.html#github-packages"><i class="fa fa-check"></i><b>2.5.4</b> Github packages</a></li>
<li class="chapter" data-level="2.5.5" data-path="intro.html"><a href="intro.html#gephi"><i class="fa fa-check"></i><b>2.5.5</b> Gephi</a></li>
<li class="chapter" data-level="2.5.6" data-path="intro.html"><a href="intro.html#tax4fun2"><i class="fa fa-check"></i><b>2.5.6</b> Tax4Fun2</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="intro.html"><a href="intro.html#plot"><i class="fa fa-check"></i><b>2.6</b> Plot</a></li>
<li class="chapter" data-level="2.7" data-path="intro.html"><a href="intro.html#rstudio"><i class="fa fa-check"></i><b>2.7</b> Rstudio</a></li>
<li class="chapter" data-level="2.8" data-path="intro.html"><a href="intro.html#citation"><i class="fa fa-check"></i><b>2.8</b> Citation</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> Basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a>
<ul>
<li class="chapter" data-level="3.1.1" data-path="basic-class.html"><a href="basic-class.html#prepare-the-example-data"><i class="fa fa-check"></i><b>3.1.1</b> Prepare the example data</a></li>
<li class="chapter" data-level="3.1.2" data-path="basic-class.html"><a href="basic-class.html#how-to-read-your-files-to-microtable-object"><i class="fa fa-check"></i><b>3.1.2</b> How to read your files to microtable object?</a></li>
<li class="chapter" data-level="3.1.3" data-path="basic-class.html"><a href="basic-class.html#functions-in-microtable-class"><i class="fa fa-check"></i><b>3.1.3</b> Functions in microtable class</a></li>
<li class="chapter" data-level="3.1.4" data-path="basic-class.html"><a href="basic-class.html#merge-taxa-or-samples"><i class="fa fa-check"></i><b>3.1.4</b> merge taxa or samples</a></li>
<li class="chapter" data-level="3.1.5" data-path="basic-class.html"><a href="basic-class.html#subset-of-samples"><i class="fa fa-check"></i><b>3.1.5</b> subset of samples</a></li>
<li class="chapter" data-level="3.1.6" data-path="basic-class.html"><a href="basic-class.html#subset-of-taxa"><i class="fa fa-check"></i><b>3.1.6</b> subset of taxa</a></li>
<li class="chapter" data-level="3.1.7" data-path="basic-class.html"><a href="basic-class.html#other-examples"><i class="fa fa-check"></i><b>3.1.7</b> Other examples</a></li>
<li class="chapter" data-level="3.1.8" data-path="basic-class.html"><a href="basic-class.html#key-points"><i class="fa fa-check"></i><b>3.1.8</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="composition-based-class.html"><a href="composition-based-class.html"><i class="fa fa-check"></i><b>4</b> Composition-based class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example"><i class="fa fa-check"></i><b>4.1.1</b> Example</a></li>
<li class="chapter" data-level="4.1.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-1"><i class="fa fa-check"></i><b>4.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="composition-based-class.html"><a href="composition-based-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="composition-based-class.html"><a href="composition-based-class.html#example-1"><i class="fa fa-check"></i><b>4.2.1</b> Example</a></li>
<li class="chapter" data-level="4.2.2" data-path="composition-based-class.html"><a href="composition-based-class.html#key-points-2"><i class="fa fa-check"></i><b>4.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="diversity-based-class.html"><a href="diversity-based-class.html"><i class="fa fa-check"></i><b>5</b> Diversity-based class</a>
<ul>
<li class="chapter" data-level="5.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>5.1</b> trans_alpha class</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-2"><i class="fa fa-check"></i><b>5.1.1</b> Example</a></li>
<li class="chapter" data-level="5.1.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-3"><i class="fa fa-check"></i><b>5.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#trans_beta-class"><i class="fa fa-check"></i><b>5.2</b> trans_beta class</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="diversity-based-class.html"><a href="diversity-based-class.html#example-3"><i class="fa fa-check"></i><b>5.2.1</b> Example</a></li>
<li class="chapter" data-level="5.2.2" data-path="diversity-based-class.html"><a href="diversity-based-class.html#key-points-4"><i class="fa fa-check"></i><b>5.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="model-based-class.html"><a href="model-based-class.html"><i class="fa fa-check"></i><b>6</b> Model-based class</a>
<ul>
<li class="chapter" data-level="6.1" data-path="model-based-class.html"><a href="model-based-class.html#trans_diff-class"><i class="fa fa-check"></i><b>6.1</b> trans_diff class</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="model-based-class.html"><a href="model-based-class.html#example-4"><i class="fa fa-check"></i><b>6.1.1</b> Example</a></li>
<li class="chapter" data-level="6.1.2" data-path="model-based-class.html"><a href="model-based-class.html#more-options"><i class="fa fa-check"></i><b>6.1.2</b> More options</a></li>
<li class="chapter" data-level="6.1.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-5"><i class="fa fa-check"></i><b>6.1.3</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="model-based-class.html"><a href="model-based-class.html#trans_network-class"><i class="fa fa-check"></i><b>6.2</b> trans_network class</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="model-based-class.html"><a href="model-based-class.html#example-5"><i class="fa fa-check"></i><b>6.2.1</b> Example</a></li>
<li class="chapter" data-level="6.2.2" data-path="model-based-class.html"><a href="model-based-class.html#network-comparison"><i class="fa fa-check"></i><b>6.2.2</b> Network comparison</a></li>
<li class="chapter" data-level="6.2.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-6"><i class="fa fa-check"></i><b>6.2.3</b> Key points</a></li>
<li class="chapter" data-level="6.2.4" data-path="model-based-class.html"><a href="model-based-class.html#other-functions"><i class="fa fa-check"></i><b>6.2.4</b> Other functions</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="model-based-class.html"><a href="model-based-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>6.3</b> trans_nullmodel class</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="model-based-class.html"><a href="model-based-class.html#example-6"><i class="fa fa-check"></i><b>6.3.1</b> Example</a></li>
<li class="chapter" data-level="6.3.2" data-path="model-based-class.html"><a href="model-based-class.html#key-points-7"><i class="fa fa-check"></i><b>6.3.2</b> Key points</a></li>
<li class="chapter" data-level="6.3.3" data-path="model-based-class.html"><a href="model-based-class.html#other-function"><i class="fa fa-check"></i><b>6.3.3</b> Other function</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="model-based-class.html"><a href="model-based-class.html#trans_classifier-class"><i class="fa fa-check"></i><b>6.4</b> trans_classifier class</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="model-based-class.html"><a href="model-based-class.html#dependencies"><i class="fa fa-check"></i><b>6.4.1</b> Dependencies</a></li>
<li class="chapter" data-level="6.4.2" data-path="model-based-class.html"><a href="model-based-class.html#examples"><i class="fa fa-check"></i><b>6.4.2</b> Examples</a></li>
<li class="chapter" data-level="6.4.3" data-path="model-based-class.html"><a href="model-based-class.html#key-points-8"><i class="fa fa-check"></i><b>6.4.3</b> Key points</a></li>
<li class="chapter" data-level="6.4.4" data-path="model-based-class.html"><a href="model-based-class.html#other-function-1"><i class="fa fa-check"></i><b>6.4.4</b> Other function</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="explainable-class.html"><a href="explainable-class.html"><i class="fa fa-check"></i><b>7</b> Explainable class</a>
<ul>
<li class="chapter" data-level="7.1" data-path="explainable-class.html"><a href="explainable-class.html#trans_env-class"><i class="fa fa-check"></i><b>7.1</b> trans_env class</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="explainable-class.html"><a href="explainable-class.html#example-7"><i class="fa fa-check"></i><b>7.1.1</b> Example</a></li>
<li class="chapter" data-level="7.1.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-9"><i class="fa fa-check"></i><b>7.1.2</b> Key points</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="explainable-class.html"><a href="explainable-class.html#trans_func-class"><i class="fa fa-check"></i><b>7.2</b> trans_func class</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="explainable-class.html"><a href="explainable-class.html#example-8"><i class="fa fa-check"></i><b>7.2.1</b> Example</a></li>
<li class="chapter" data-level="7.2.2" data-path="explainable-class.html"><a href="explainable-class.html#key-points-10"><i class="fa fa-check"></i><b>7.2.2</b> Key points</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="file2meco-package.html"><a href="file2meco-package.html"><i class="fa fa-check"></i><b>8</b> file2meco package</a>
<ul>
<li class="chapter" data-level="8.1" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime"><i class="fa fa-check"></i><b>8.1</b> QIIME</a></li>
<li class="chapter" data-level="8.2" data-path="file2meco-package.html"><a href="file2meco-package.html#qiime2"><i class="fa fa-check"></i><b>8.2</b> QIIME2</a></li>
<li class="chapter" data-level="8.3" data-path="file2meco-package.html"><a href="file2meco-package.html#humann"><i class="fa fa-check"></i><b>8.3</b> HUMAnN</a></li>
<li class="chapter" data-level="8.4" data-path="file2meco-package.html"><a href="file2meco-package.html#metaphlan"><i class="fa fa-check"></i><b>8.4</b> MetaPhlAn</a></li>
<li class="chapter" data-level="8.5" data-path="file2meco-package.html"><a href="file2meco-package.html#kraken2bracken"><i class="fa fa-check"></i><b>8.5</b> Kraken2/Bracken</a></li>
<li class="chapter" data-level="8.6" data-path="file2meco-package.html"><a href="file2meco-package.html#ncycdbpcycdb"><i class="fa fa-check"></i><b>8.6</b> NCycDB/PCycDB</a></li>
<li class="chapter" data-level="8.7" data-path="file2meco-package.html"><a href="file2meco-package.html#phyloseq"><i class="fa fa-check"></i><b>8.7</b> phyloseq</a></li>
<li class="chapter" data-level="8.8" data-path="file2meco-package.html"><a href="file2meco-package.html#picrust2"><i class="fa fa-check"></i><b>8.8</b> PICRUSt2</a></li>
<li class="chapter" data-level="8.9" data-path="file2meco-package.html"><a href="file2meco-package.html#viromescan"><i class="fa fa-check"></i><b>8.9</b> ViromeScan</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="other-dataset.html"><a href="other-dataset.html"><i class="fa fa-check"></i><b>9</b> Other dataset</a>
<ul>
<li class="chapter" data-level="9.1" data-path="other-dataset.html"><a href="other-dataset.html#fungi-data"><i class="fa fa-check"></i><b>9.1</b> Fungi data</a></li>
<li class="chapter" data-level="9.2" data-path="other-dataset.html"><a href="other-dataset.html#metagenomic-data"><i class="fa fa-check"></i><b>9.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="9.3" data-path="other-dataset.html"><a href="other-dataset.html#gut-microbiome"><i class="fa fa-check"></i><b>9.3</b> Gut microbiome</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>10</b> Notes</a>
<ul>
<li class="chapter" data-level="10.1" data-path="notes.html"><a href="notes.html#clone-function"><i class="fa fa-check"></i><b>10.1</b> <code>clone</code> function</a></li>
<li class="chapter" data-level="10.2" data-path="notes.html"><a href="notes.html#save-function"><i class="fa fa-check"></i><b>10.2</b> <code>save</code> function</a></li>
<li class="chapter" data-level="10.3" data-path="notes.html"><a href="notes.html#github-issues"><i class="fa fa-check"></i><b>10.3</b> Github Issues</a></li>
<li class="chapter" data-level="10.4" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>10.4</b> group order</a></li>
<li class="chapter" data-level="10.5" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>10.5</b> add layers to plot</a></li>
<li class="chapter" data-level="10.6" data-path="notes.html"><a href="notes.html#colors-for-more-groups"><i class="fa fa-check"></i><b>10.6</b> colors for more groups</a></li>
<li class="chapter" data-level="10.7" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>10.7</b> change object</a></li>
<li class="chapter" data-level="10.8" data-path="notes.html"><a href="notes.html#customized-data-input"><i class="fa fa-check"></i><b>10.8</b> customized data input</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html"><i class="fa fa-check"></i><b>11</b> meconetcomp package</a>
<ul>
<li class="chapter" data-level="11.1" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-modularity-for-all-networks"><i class="fa fa-check"></i><b>11.1</b> Network modularity for all networks</a></li>
<li class="chapter" data-level="11.2" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#network-topological-attributes-for-all-networks"><i class="fa fa-check"></i><b>11.2</b> Network topological attributes for all networks</a></li>
<li class="chapter" data-level="11.3" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#node-and-edge-properties-extraction-for-all-networks"><i class="fa fa-check"></i><b>11.3</b> Node and edge properties extraction for all networks</a></li>
<li class="chapter" data-level="11.4" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-nodes-across-networks"><i class="fa fa-check"></i><b>11.4</b> Compare nodes across networks</a></li>
<li class="chapter" data-level="11.5" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-edges-across-networks"><i class="fa fa-check"></i><b>11.5</b> Compare edges across networks</a></li>
<li class="chapter" data-level="11.6" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#extract-overlapped-edges-of-networks-to-a-new-network"><i class="fa fa-check"></i><b>11.6</b> Extract overlapped edges of networks to a new network</a></li>
<li class="chapter" data-level="11.7" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-phylogenetic-distances-of-paired-nodes-in-edges"><i class="fa fa-check"></i><b>11.7</b> Compare phylogenetic distances of paired nodes in edges</a></li>
<li class="chapter" data-level="11.8" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-node-sources-of-edges-across-networks"><i class="fa fa-check"></i><b>11.8</b> Compare node sources of edges across networks</a></li>
<li class="chapter" data-level="11.9" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#compare-topological-properties-of-sub-networks"><i class="fa fa-check"></i><b>11.9</b> Compare topological properties of sub-networks</a></li>
<li class="chapter" data-level="11.10" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#robustness-of-network"><i class="fa fa-check"></i><b>11.10</b> Robustness of network</a></li>
<li class="chapter" data-level="11.11" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#vulnerability-of-nodes"><i class="fa fa-check"></i><b>11.11</b> Vulnerability of nodes</a></li>
<li class="chapter" data-level="11.12" data-path="meconetcomp-package.html"><a href="meconetcomp-package.html#cohesion"><i class="fa fa-check"></i><b>11.12</b> Cohesion</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>12</b> mecodev package</a>
<ul>
<li class="chapter" data-level="12.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>12.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="12.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_ts"><i class="fa fa-check"></i><b>12.2</b> trans_ts</a></li>
<li class="chapter" data-level="12.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_gamma"><i class="fa fa-check"></i><b>12.3</b> trans_gamma</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="mecoturn-package.html"><a href="mecoturn-package.html"><i class="fa fa-check"></i><b>13</b> mecoturn package</a>
<ul>
<li class="chapter" data-level="13.1" data-path="mecoturn-package.html"><a href="mecoturn-package.html#betaturn-class"><i class="fa fa-check"></i><b>13.1</b> betaturn class</a></li>
<li class="chapter" data-level="13.2" data-path="mecoturn-package.html"><a href="mecoturn-package.html#taxaturn-class"><i class="fa fa-check"></i><b>13.2</b> taxaturn class</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="other-examples-1.html"><a href="other-examples-1.html"><i class="fa fa-check"></i><b>14</b> Other examples</a>
<ul>
<li class="chapter" data-level="14.1" data-path="other-examples-1.html"><a href="other-examples-1.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>14.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="14.2" data-path="other-examples-1.html"><a href="other-examples-1.html#tidy_taxonomy-function"><i class="fa fa-check"></i><b>14.2</b> tidy_taxonomy function</a></li>
<li class="chapter" data-level="14.3" data-path="other-examples-1.html"><a href="other-examples-1.html#show-the-abundance-of-unknown-taxa"><i class="fa fa-check"></i><b>14.3</b> Show the abundance of unknown taxa</a></li>
<li class="chapter" data-level="14.4" data-path="other-examples-1.html"><a href="other-examples-1.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>14.4</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="14.5" data-path="other-examples-1.html"><a href="other-examples-1.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>14.5</b> The use of phylogenetic tree</a></li>
<li class="chapter" data-level="14.6" data-path="other-examples-1.html"><a href="other-examples-1.html#trans_norm"><i class="fa fa-check"></i><b>14.6</b> trans_norm</a></li>
<li class="chapter" data-level="14.7" data-path="other-examples-1.html"><a href="other-examples-1.html#mantel-test-correlation-heatmap"><i class="fa fa-check"></i><b>14.7</b> Mantel test + Correlation heatmap</a></li>
<li class="chapter" data-level="14.8" data-path="other-examples-1.html"><a href="other-examples-1.html#pcoa-boxplot"><i class="fa fa-check"></i><b>14.8</b> PCoA + boxplot</a></li>
<li class="chapter" data-level="14.9" data-path="other-examples-1.html"><a href="other-examples-1.html#faprotax-differential-test"><i class="fa fa-check"></i><b>14.9</b> FAPROTAX + differential test</a></li>
<li class="chapter" data-level="14.10" data-path="other-examples-1.html"><a href="other-examples-1.html#rda-results-output"><i class="fa fa-check"></i><b>14.10</b> RDA results output</a></li>
<li class="chapter" data-level="14.11" data-path="other-examples-1.html"><a href="other-examples-1.html#betanribetanti-for-each-group"><i class="fa fa-check"></i><b>14.11</b> betaNRI/betaNTI for each group</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Tutorial for R microeco package (v1.9.0)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="other-dataset" class="section level1 hasAnchor" number="9">
<h1><span class="header-section-number">Chapter 9</span> Other dataset<a href="other-dataset.html#other-dataset" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="fungi-data" class="section level2 hasAnchor" number="9.1">
<h2><span class="header-section-number">9.1</span> Fungi data<a href="other-dataset.html#fungi-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Another ITS sequencing dataset <span class="citation">(<a href="#ref-Gao_Strong_2019" role="doc-biblioref">Gao et al. 2019</a>)</span> is also stored in the example data of the package.
Here, we use it as an example to show the use of FUNGuild database<span class="citation">(<a href="#ref-Nguyen_FUNGuild_2016" role="doc-biblioref">Nguyen et al. 2016</a>)</span>.
FungalTraits <span class="citation">(<a href="#ref-Polme_FungalTraits_2020" role="doc-biblioref">Põlme et al. 2020</a>)</span> database is also available for identifying fungal traits.</p>
<div class="sourceCode" id="cb301"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb301-1"><a href="other-dataset.html#cb301-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb301-2"><a href="other-dataset.html#cb301-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb301-3"><a href="other-dataset.html#cb301-3" aria-hidden="true" tabindex="-1"></a><span class="co"># load ITS data</span></span>
<span id="cb301-4"><a href="other-dataset.html#cb301-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(sample_info_ITS)</span>
<span id="cb301-5"><a href="other-dataset.html#cb301-5" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(otu_table_ITS)</span>
<span id="cb301-6"><a href="other-dataset.html#cb301-6" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(taxonomy_table_ITS)</span>
<span id="cb301-7"><a href="other-dataset.html#cb301-7" aria-hidden="true" tabindex="-1"></a><span class="co"># create microtable object</span></span>
<span id="cb301-8"><a href="other-dataset.html#cb301-8" aria-hidden="true" tabindex="-1"></a>meco_fungi <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">sample_table =</span> sample_info_ITS, <span class="at">otu_table =</span> otu_table_ITS, <span class="at">tax_table =</span> taxonomy_table_ITS)</span>
<span id="cb301-9"><a href="other-dataset.html#cb301-9" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the taxa not assigned in the Kingdom &quot;k__Fungi&quot;</span></span>
<span id="cb301-10"><a href="other-dataset.html#cb301-10" aria-hidden="true" tabindex="-1"></a>meco_fungi<span class="sc">$</span>tax_table <span class="sc">%&lt;&gt;%</span> base<span class="sc">::</span><span class="fu">subset</span>(Kingdom <span class="sc">==</span> <span class="st">&quot;k__Fungi&quot;</span>)</span>
<span id="cb301-11"><a href="other-dataset.html#cb301-11" aria-hidden="true" tabindex="-1"></a><span class="co"># use tidy_dataset() to make OTUs and samples information consistent across files</span></span>
<span id="cb301-12"><a href="other-dataset.html#cb301-12" aria-hidden="true" tabindex="-1"></a>meco_fungi<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span></code></pre></div>
<p>Then, let’s perform the fungal trait prediction.</p>
<div class="sourceCode" id="cb302"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb302-1"><a href="other-dataset.html#cb302-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(meco_fungi)</span>
<span id="cb302-2"><a href="other-dataset.html#cb302-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Both FungalTraits and FUNGuild are available</span></span>
<span id="cb302-3"><a href="other-dataset.html#cb302-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_spe_func</span>(<span class="at">fungi_database =</span> <span class="st">&quot;FungalTraits&quot;</span>)</span>
<span id="cb302-4"><a href="other-dataset.html#cb302-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">abundance_weighted =</span> <span class="cn">TRUE</span>)</span>
<span id="cb302-5"><a href="other-dataset.html#cb302-5" aria-hidden="true" tabindex="-1"></a><span class="co"># return: t1$res_spe_func_perc</span></span></code></pre></div>
<p>Then we show an example on the differential test of traits across groups.</p>
<div class="sourceCode" id="cb303"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb303-1"><a href="other-dataset.html#cb303-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clone meco_fungi for differential test</span></span>
<span id="cb303-2"><a href="other-dataset.html#cb303-2" aria-hidden="true" tabindex="-1"></a>tmp_mt <span class="ot">&lt;-</span> <span class="fu">clone</span>(meco_fungi)</span>
<span id="cb303-3"><a href="other-dataset.html#cb303-3" aria-hidden="true" tabindex="-1"></a><span class="co"># we need assign the table to taxa_abund manually</span></span>
<span id="cb303-4"><a href="other-dataset.html#cb303-4" aria-hidden="true" tabindex="-1"></a>tmp_mt<span class="sc">$</span>taxa_abund<span class="sc">$</span>func <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(t1<span class="sc">$</span>res_spe_func_perc), <span class="at">check.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb303-5"><a href="other-dataset.html#cb303-5" aria-hidden="true" tabindex="-1"></a><span class="co"># use trans_diff class to perform differential test</span></span>
<span id="cb303-6"><a href="other-dataset.html#cb303-6" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> tmp_mt, <span class="at">method =</span> <span class="st">&quot;anova&quot;</span>, <span class="at">group =</span> <span class="st">&quot;sample_type&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;func&quot;</span>)</span>
<span id="cb303-7"><a href="other-dataset.html#cb303-7" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_diff_abund</span>(<span class="at">add_sig =</span> T, <span class="at">simplify_names =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">ylab</span>(<span class="st">&quot;Relative abundance (%)&quot;</span>)</span></code></pre></div>
<p>Then we show the distribution of traits across modules in network.</p>
<div class="sourceCode" id="cb304"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb304-1"><a href="other-dataset.html#cb304-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_network object</span></span>
<span id="cb304-2"><a href="other-dataset.html#cb304-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> meco_fungi, <span class="at">cal_cor =</span> <span class="st">&quot;WGCNA&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.000001</span>, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb304-3"><a href="other-dataset.html#cb304-3" aria-hidden="true" tabindex="-1"></a><span class="co"># create correlation network </span></span>
<span id="cb304-4"><a href="other-dataset.html#cb304-4" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">COR_p_thres =</span> <span class="fl">0.05</span>, <span class="at">COR_cut =</span> <span class="fl">0.6</span>)</span>
<span id="cb304-5"><a href="other-dataset.html#cb304-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add modules</span></span>
<span id="cb304-6"><a href="other-dataset.html#cb304-6" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_module</span>()</span>
<span id="cb304-7"><a href="other-dataset.html#cb304-7" aria-hidden="true" tabindex="-1"></a><span class="co"># convert module info to microtable object</span></span>
<span id="cb304-8"><a href="other-dataset.html#cb304-8" aria-hidden="true" tabindex="-1"></a>meco_module <span class="ot">&lt;-</span> t1<span class="sc">$</span><span class="fu">trans_comm</span>(<span class="at">use_col =</span> <span class="st">&quot;module&quot;</span>)</span>
<span id="cb304-9"><a href="other-dataset.html#cb304-9" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_func object</span></span>
<span id="cb304-10"><a href="other-dataset.html#cb304-10" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(meco_module)</span>
<span id="cb304-11"><a href="other-dataset.html#cb304-11" aria-hidden="true" tabindex="-1"></a><span class="co"># identify species traits, automatically select database for prokaryotes or fungi</span></span>
<span id="cb304-12"><a href="other-dataset.html#cb304-12" aria-hidden="true" tabindex="-1"></a><span class="co"># fungi_database = &quot;FungalTraits&quot; for the FungalTraits database</span></span>
<span id="cb304-13"><a href="other-dataset.html#cb304-13" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func</span>(<span class="at">fungi_database =</span> <span class="st">&quot;FUNGuild&quot;</span>)</span>
<span id="cb304-14"><a href="other-dataset.html#cb304-14" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate abundance-unweighted functional redundancy of each trait for each network module</span></span>
<span id="cb304-15"><a href="other-dataset.html#cb304-15" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">abundance_weighted =</span> <span class="cn">FALSE</span>)</span>
<span id="cb304-16"><a href="other-dataset.html#cb304-16" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the functional redundancy of network modules</span></span>
<span id="cb304-17"><a href="other-dataset.html#cb304-17" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_spe_func_perc</span>(<span class="at">order_x =</span> <span class="fu">paste0</span>(<span class="st">&quot;M&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>))</span></code></pre></div>
</div>
<div id="metagenomic-data" class="section level2 hasAnchor" number="9.2">
<h2><span class="header-section-number">9.2</span> Metagenomic data<a href="other-dataset.html#metagenomic-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In the file2meco package part,
we provide several examples to transform the output files of some famous metagenomic tools (e.g. HUMAnN and kraken2) to
the microtable object directly.
Here, we show some detailed examples using KEGG pathway results.</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="other-dataset.html#cb305-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb305-2"><a href="other-dataset.html#cb305-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb305-3"><a href="other-dataset.html#cb305-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb305-4"><a href="other-dataset.html#cb305-4" aria-hidden="true" tabindex="-1"></a>?humann2meco</span>
<span id="cb305-5"><a href="other-dataset.html#cb305-5" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_metagenome_sample_info.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb305-6"><a href="other-dataset.html#cb305-6" aria-hidden="true" tabindex="-1"></a>match_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_metagenome_match_table.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb305-7"><a href="other-dataset.html#cb305-7" aria-hidden="true" tabindex="-1"></a><span class="co"># use KEGG pathway based HUMAnN result</span></span>
<span id="cb305-8"><a href="other-dataset.html#cb305-8" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example_HUMAnN_KEGG_abund.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb305-9"><a href="other-dataset.html#cb305-9" aria-hidden="true" tabindex="-1"></a><span class="co"># match_table parameter can be used to replace sample names</span></span>
<span id="cb305-10"><a href="other-dataset.html#cb305-10" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">humann2meco</span>(abund_file_path, <span class="at">db =</span> <span class="st">&quot;KEGG&quot;</span>, <span class="at">sample_table =</span> sample_file_path, <span class="at">match_table =</span> match_file_path)</span>
<span id="cb305-11"><a href="other-dataset.html#cb305-11" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the unclassified pathway in Level.1</span></span>
<span id="cb305-12"><a href="other-dataset.html#cb305-12" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>tax_table <span class="sc">%&lt;&gt;%</span> <span class="fu">subset</span>(Level<span class="fl">.1</span> <span class="sc">!=</span> <span class="st">&quot;unclassified&quot;</span>)</span>
<span id="cb305-13"><a href="other-dataset.html#cb305-13" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb305-14"><a href="other-dataset.html#cb305-14" aria-hidden="true" tabindex="-1"></a><span class="co"># rel = FALSE donot use relative abundance, use the raw RPK</span></span>
<span id="cb305-15"><a href="other-dataset.html#cb305-15" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">select_cols =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">rel =</span> <span class="cn">FALSE</span>)</span>
<span id="cb305-16"><a href="other-dataset.html#cb305-16" aria-hidden="true" tabindex="-1"></a><span class="co"># use_percentage = FALSE disable percentage for relative abundance</span></span>
<span id="cb305-17"><a href="other-dataset.html#cb305-17" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(test, <span class="at">taxrank =</span> <span class="st">&quot;Level.2&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">use_percentage =</span> <span class="cn">FALSE</span>)</span>
<span id="cb305-18"><a href="other-dataset.html#cb305-18" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">bar_full =</span> <span class="cn">FALSE</span>, <span class="at">xtext_angle =</span> <span class="dv">30</span>) <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">ylab</span>(<span class="st">&quot;Abundance (RPK)&quot;</span>)</span></code></pre></div>
<p><img src="Images/file2meco_HUMANN_KEGG_bar.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Then, we select both function and taxa to see which taxa those high abundant pathways come from.</p>
<div class="sourceCode" id="cb306"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb306-1"><a href="other-dataset.html#cb306-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This operation is more flexible</span></span>
<span id="cb306-2"><a href="other-dataset.html#cb306-2" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">select_cols =</span> <span class="fu">c</span>(<span class="st">&quot;Level.1&quot;</span>, <span class="st">&quot;Phylum&quot;</span>, <span class="st">&quot;Genus&quot;</span>), <span class="at">rel =</span> <span class="cn">FALSE</span>)</span>
<span id="cb306-3"><a href="other-dataset.html#cb306-3" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>taxa_abund<span class="sc">$</span>Level<span class="fl">.1</span> <span class="sc">%&lt;&gt;%</span> .[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;unclass&quot;</span>, <span class="fu">rownames</span>(.)), ]</span>
<span id="cb306-4"><a href="other-dataset.html#cb306-4" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span>taxa_abund<span class="sc">$</span>Phylum <span class="sc">%&lt;&gt;%</span> .[<span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;unclass&quot;</span>, <span class="fu">rownames</span>(.)), ]</span>
<span id="cb306-5"><a href="other-dataset.html#cb306-5" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(test, <span class="at">taxrank =</span> <span class="st">&quot;Phylum&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">use_percentage =</span> <span class="cn">FALSE</span>)</span>
<span id="cb306-6"><a href="other-dataset.html#cb306-6" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_bar</span>(<span class="at">facet =</span> <span class="st">&quot;Group&quot;</span>, <span class="at">bar_full =</span> <span class="cn">FALSE</span>, <span class="at">xtext_angle =</span> <span class="dv">30</span>) <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">ylab</span>(<span class="st">&quot;Abundance (RPK)&quot;</span>)</span></code></pre></div>
<p><img src="Images/file2meco_HUMANN_KEGG_bar_taxafunc.png" width="800px" style="display: block; margin: auto;" /></p>
<p>Let’s run LEfSe to find some functional biomarkers to differentiate two groups.</p>
<div class="sourceCode" id="cb307"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb307-1"><a href="other-dataset.html#cb307-1" aria-hidden="true" tabindex="-1"></a><span class="co"># functional biomarker</span></span>
<span id="cb307-2"><a href="other-dataset.html#cb307-2" aria-hidden="true" tabindex="-1"></a>test<span class="sc">$</span><span class="fu">cal_abund</span>(<span class="at">select_cols =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">rel =</span> <span class="cn">TRUE</span>)</span>
<span id="cb307-3"><a href="other-dataset.html#cb307-3" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(test, <span class="at">method =</span> <span class="st">&quot;lefse&quot;</span>, <span class="at">group =</span> <span class="st">&quot;Group&quot;</span>)</span>
<span id="cb307-4"><a href="other-dataset.html#cb307-4" aria-hidden="true" tabindex="-1"></a>test1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">threshold =</span> <span class="dv">3</span>)</span></code></pre></div>
<p><img src="Images/file2meco_HUMANN_KEGG_lefse_bar.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<div id="gut-microbiome" class="section level2 hasAnchor" number="9.3">
<h2><span class="header-section-number">9.3</span> Gut microbiome<a href="other-dataset.html#gut-microbiome" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We use mouse gut data stored in file2meco package to show the input of QIIME2 file and the use of metabolic trait database NJC19 database<span class="citation">(<a href="#ref-Lim_Large_2020" role="doc-biblioref">Lim et al. 2020</a>)</span>.</p>
<div class="sourceCode" id="cb308"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb308-1"><a href="other-dataset.html#cb308-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb308-2"><a href="other-dataset.html#cb308-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb308-3"><a href="other-dataset.html#cb308-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb308-4"><a href="other-dataset.html#cb308-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb308-5"><a href="other-dataset.html#cb308-5" aria-hidden="true" tabindex="-1"></a><span class="co"># use data files inside the file2meco package.</span></span>
<span id="cb308-6"><a href="other-dataset.html#cb308-6" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;dada2_table.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb308-7"><a href="other-dataset.html#cb308-7" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sample-metadata.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb308-8"><a href="other-dataset.html#cb308-8" aria-hidden="true" tabindex="-1"></a>taxonomy_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;taxonomy.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb308-9"><a href="other-dataset.html#cb308-9" aria-hidden="true" tabindex="-1"></a><span class="co"># construct microtable object</span></span>
<span id="cb308-10"><a href="other-dataset.html#cb308-10" aria-hidden="true" tabindex="-1"></a>data1 <span class="ot">&lt;-</span> <span class="fu">qiime2meco</span>(abund_file_path, <span class="at">sample_table =</span> sample_file_path, <span class="at">taxonomy_table =</span> taxonomy_file_path)</span>
<span id="cb308-11"><a href="other-dataset.html#cb308-11" aria-hidden="true" tabindex="-1"></a>data1<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb308-12"><a href="other-dataset.html#cb308-12" aria-hidden="true" tabindex="-1"></a><span class="co"># revise the species names in tax_table as the information in the example file is not standard</span></span>
<span id="cb308-13"><a href="other-dataset.html#cb308-13" aria-hidden="true" tabindex="-1"></a>select_rows <span class="ot">&lt;-</span> data1<span class="sc">$</span>tax_table<span class="sc">$</span>Species <span class="sc">!=</span> <span class="st">&quot;s__&quot;</span></span>
<span id="cb308-14"><a href="other-dataset.html#cb308-14" aria-hidden="true" tabindex="-1"></a>data1<span class="sc">$</span>tax_table<span class="sc">$</span>Species[select_rows] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;s__&quot;</span>, <span class="fu">gsub</span>(<span class="st">&quot;g__&quot;</span>, <span class="st">&quot;&quot;</span>, data1<span class="sc">$</span>tax_table<span class="sc">$</span>Genus[select_rows]), <span class="st">&quot; &quot;</span>, <span class="fu">gsub</span>(<span class="st">&quot;s__&quot;</span>, <span class="st">&quot;&quot;</span>, data1<span class="sc">$</span>tax_table<span class="sc">$</span>Species[select_rows]))</span>
<span id="cb308-15"><a href="other-dataset.html#cb308-15" aria-hidden="true" tabindex="-1"></a><span class="co"># taxonomic abundance</span></span>
<span id="cb308-16"><a href="other-dataset.html#cb308-16" aria-hidden="true" tabindex="-1"></a>data1<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb308-17"><a href="other-dataset.html#cb308-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb308-18"><a href="other-dataset.html#cb308-18" aria-hidden="true" tabindex="-1"></a><span class="co"># create object of trans_func</span></span>
<span id="cb308-19"><a href="other-dataset.html#cb308-19" aria-hidden="true" tabindex="-1"></a>data2 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(data1)</span>
<span id="cb308-20"><a href="other-dataset.html#cb308-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Select NJC19 database</span></span>
<span id="cb308-21"><a href="other-dataset.html#cb308-21" aria-hidden="true" tabindex="-1"></a>data2<span class="sc">$</span><span class="fu">cal_spe_func</span>(<span class="at">prok_database =</span> <span class="st">&quot;NJC19&quot;</span>)</span>
<span id="cb308-22"><a href="other-dataset.html#cb308-22" aria-hidden="true" tabindex="-1"></a><span class="co"># get the trait percentage data</span></span>
<span id="cb308-23"><a href="other-dataset.html#cb308-23" aria-hidden="true" tabindex="-1"></a>data2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">abundance_weighted =</span> <span class="cn">FALSE</span>)</span>
<span id="cb308-24"><a href="other-dataset.html#cb308-24" aria-hidden="true" tabindex="-1"></a><span class="co"># inset the trait percentage result into taxa_abund of microtable object</span></span>
<span id="cb308-25"><a href="other-dataset.html#cb308-25" aria-hidden="true" tabindex="-1"></a>data1<span class="sc">$</span>taxa_abund<span class="sc">$</span>Trait <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(data2<span class="sc">$</span>res_spe_func_perc))</span>
<span id="cb308-26"><a href="other-dataset.html#cb308-26" aria-hidden="true" tabindex="-1"></a><span class="co"># use trans_abund to plot</span></span>
<span id="cb308-27"><a href="other-dataset.html#cb308-27" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> data1, <span class="at">taxrank =</span> <span class="st">&quot;Trait&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">use_percentage =</span> <span class="cn">FALSE</span>)</span>
<span id="cb308-28"><a href="other-dataset.html#cb308-28" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_box</span>(<span class="at">group =</span> <span class="st">&quot;donor_status&quot;</span>, <span class="at">xtext_angle =</span> <span class="dv">30</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Relative population abundance (%)&quot;</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">13</span>))</span></code></pre></div>
<p><img src="Images/plot_mouse_NJC19.png" width="800px" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb309"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb309-1"><a href="other-dataset.html#cb309-1" aria-hidden="true" tabindex="-1"></a><span class="co"># differential abundance test of the traits percentage and use random forest to find biomarkers</span></span>
<span id="cb309-2"><a href="other-dataset.html#cb309-2" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_diff<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> data1, <span class="at">method =</span> <span class="st">&quot;rf&quot;</span>, <span class="at">group =</span> <span class="st">&quot;donor_status&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;Trait&quot;</span>)</span>
<span id="cb309-3"><a href="other-dataset.html#cb309-3" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_diff_bar</span>(<span class="at">use_number =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code></pre></div>
<p><img src="Images/plot_mouse_NJC19_rf.png" width="800px" style="display: block; margin: auto;" /></p>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Gao_Strong_2019" class="csl-entry">
Gao, Cheng, Liliam Montoya, Ling Xu, Mary Madera, Joy Hollingsworth, Elizabeth Purdom, Robert B Hutmacher, Jeffery A Dahlberg, Devin Coleman-Derr, and Peggy G Lemaux. 2019. <span>“Strong Succession in Arbuscular Mycorrhizal Fungal Communities.”</span> Journal Article. <em>The ISME Journal</em> 13 (1): 214.
</div>
<div id="ref-Lim_Large_2020" class="csl-entry">
Lim, Roktaek, Josephine Jill T. Cabatbat, Thomas L. P. Martin, Haneul Kim, Seunghyeon Kim, Jaeyun Sung, Cheol-Min Ghim, and Pan-Jun Kim. 2020. <span>“Large-Scale Metabolic Interaction Network of the Mouse and Human Gut Microbiota.”</span> Journal Article. <em>Scientific Data</em> 7 (1). <a href="https://doi.org/10.1038/s41597-020-0516-5">https://doi.org/10.1038/s41597-020-0516-5</a>.
</div>
<div id="ref-Nguyen_FUNGuild_2016" class="csl-entry">
Nguyen, Nhu H., Zewei Song, Scott T. Bates, Sara Branco, Leho Tedersoo, Jon Menke, Jonathan S. Schilling, and Peter G. Kennedy. 2016. <span>“FUNGuild: An Open Annotation Tool for Parsing Fungal Community Datasets by Ecological Guild.”</span> Journal Article. <em>Fungal Ecology</em> 20 (1): 241–48.
</div>
<div id="ref-Polme_FungalTraits_2020" class="csl-entry">
Põlme, Sergei, Kessy Abarenkov, R. Henrik Nilsson, Björn D. Lindahl, Karina Engelbrecht Clemmensen, Havard Kauserud, Nhu Nguyen, et al. 2020. <span>“FungalTraits: A User-Friendly Traits Database of Fungi and Fungus-Like Stramenopiles.”</span> Journal Article. <em>Fungal Diversity</em> 105 (1): 1–16. <a href="https://doi.org/10.1007/s13225-020-00466-2">https://doi.org/10.1007/s13225-020-00466-2</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="file2meco-package.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="notes.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/08-Otherdata.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
