<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Other data | microeco: an R package for data mining in microbial community ecology</title>
  <meta name="description" content="The tutorial for microeco package" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Other data | microeco: an R package for data mining in microbial community ecology" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="The tutorial for microeco package" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Other data | microeco: an R package for data mining in microbial community ecology" />
  
  <meta name="twitter:description" content="The tutorial for microeco package" />
  

<meta name="author" content="Chi Liu, Felipe Mansoldo, Jarrod J. Scott, Alane B. Vermelho, Minjie Yao" />


<meta name="date" content="2021-08-20" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="extended-class.html"/>
<link rel="next" href="notes.html"/>
<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">microeco tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#features"><i class="fa fa-check"></i><b>2.1</b> Features</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#r6-class"><i class="fa fa-check"></i><b>2.2</b> R6 Class</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#help"><i class="fa fa-check"></i><b>2.3</b> Help</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="basic-class.html"><a href="basic-class.html"><i class="fa fa-check"></i><b>3</b> basic class</a>
<ul>
<li class="chapter" data-level="3.1" data-path="basic-class.html"><a href="basic-class.html#microtable-class"><i class="fa fa-check"></i><b>3.1</b> microtable class</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="extended-class.html"><a href="extended-class.html"><i class="fa fa-check"></i><b>4</b> Extended class</a>
<ul>
<li class="chapter" data-level="4.1" data-path="extended-class.html"><a href="extended-class.html#trans_abund-class"><i class="fa fa-check"></i><b>4.1</b> trans_abund class</a></li>
<li class="chapter" data-level="4.2" data-path="extended-class.html"><a href="extended-class.html#trans_venn-class"><i class="fa fa-check"></i><b>4.2</b> trans_venn class</a></li>
<li class="chapter" data-level="4.3" data-path="extended-class.html"><a href="extended-class.html#trans_alpha-class"><i class="fa fa-check"></i><b>4.3</b> trans_alpha class</a></li>
<li class="chapter" data-level="4.4" data-path="extended-class.html"><a href="extended-class.html#trans_beta-class"><i class="fa fa-check"></i><b>4.4</b> trans_beta class</a></li>
<li class="chapter" data-level="4.5" data-path="extended-class.html"><a href="extended-class.html#trans_diff-class"><i class="fa fa-check"></i><b>4.5</b> trans_diff class</a></li>
<li class="chapter" data-level="4.6" data-path="extended-class.html"><a href="extended-class.html#trans_env-class"><i class="fa fa-check"></i><b>4.6</b> trans_env class</a></li>
<li class="chapter" data-level="4.7" data-path="extended-class.html"><a href="extended-class.html#trans_nullmodel-class"><i class="fa fa-check"></i><b>4.7</b> trans_nullmodel class</a></li>
<li class="chapter" data-level="4.8" data-path="extended-class.html"><a href="extended-class.html#trans_network-class"><i class="fa fa-check"></i><b>4.8</b> trans_network class</a></li>
<li class="chapter" data-level="4.9" data-path="extended-class.html"><a href="extended-class.html#trans_func-class"><i class="fa fa-check"></i><b>4.9</b> trans_func class</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="other-data.html"><a href="other-data.html"><i class="fa fa-check"></i><b>5</b> Other data</a>
<ul>
<li class="chapter" data-level="5.1" data-path="other-data.html"><a href="other-data.html#fungi-data"><i class="fa fa-check"></i><b>5.1</b> Fungi data</a></li>
<li class="chapter" data-level="5.2" data-path="other-data.html"><a href="other-data.html#metagenomic-data"><i class="fa fa-check"></i><b>5.2</b> Metagenomic data</a></li>
<li class="chapter" data-level="5.3" data-path="other-data.html"><a href="other-data.html#animal-data"><i class="fa fa-check"></i><b>5.3</b> Animal data</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="notes.html"><a href="notes.html"><i class="fa fa-check"></i><b>6</b> Notes</a>
<ul>
<li class="chapter" data-level="6.1" data-path="notes.html"><a href="notes.html#clone"><i class="fa fa-check"></i><b>6.1</b> clone</a></li>
<li class="chapter" data-level="6.2" data-path="notes.html"><a href="notes.html#add-layers-to-plot"><i class="fa fa-check"></i><b>6.2</b> add layers to plot</a></li>
<li class="chapter" data-level="6.3" data-path="notes.html"><a href="notes.html#subset-of-samples"><i class="fa fa-check"></i><b>6.3</b> subset of samples</a></li>
<li class="chapter" data-level="6.4" data-path="notes.html"><a href="notes.html#change-object"><i class="fa fa-check"></i><b>6.4</b> change object</a></li>
<li class="chapter" data-level="6.5" data-path="notes.html"><a href="notes.html#customized-taxonomic-table"><i class="fa fa-check"></i><b>6.5</b> customized taxonomic table</a></li>
<li class="chapter" data-level="6.6" data-path="notes.html"><a href="notes.html#group-order"><i class="fa fa-check"></i><b>6.6</b> group order</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="mecodev-package.html"><a href="mecodev-package.html"><i class="fa fa-check"></i><b>7</b> mecodev package</a>
<ul>
<li class="chapter" data-level="7.1" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_rarefy"><i class="fa fa-check"></i><b>7.1</b> trans_rarefy</a></li>
<li class="chapter" data-level="7.2" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_convert"><i class="fa fa-check"></i><b>7.2</b> trans_convert</a></li>
<li class="chapter" data-level="7.3" data-path="mecodev-package.html"><a href="mecodev-package.html#trans_netchord"><i class="fa fa-check"></i><b>7.3</b> trans_netchord</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="other-examples.html"><a href="other-examples.html"><i class="fa fa-check"></i><b>8</b> Other examples</a>
<ul>
<li class="chapter" data-level="8.1" data-path="other-examples.html"><a href="other-examples.html#custom-taxa-order-in-bar-plot"><i class="fa fa-check"></i><b>8.1</b> Custom taxa order in bar plot</a></li>
<li class="chapter" data-level="8.2" data-path="other-examples.html"><a href="other-examples.html#question-of-prefix-in-the-taxa"><i class="fa fa-check"></i><b>8.2</b> Question of prefix in the taxa</a></li>
<li class="chapter" data-level="8.3" data-path="other-examples.html"><a href="other-examples.html#the-use-of-phylogenetic-tree"><i class="fa fa-check"></i><b>8.3</b> The use of phylogenetic tree</a></li>
<li class="chapter" data-level="8.4" data-path="other-examples.html"><a href="other-examples.html#the-importance-of-tidy_taxonomy-function"><i class="fa fa-check"></i><b>8.4</b> The importance of tidy_taxonomy function</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">microeco: an R package for data mining in microbial community ecology</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="other-data" class="section level1" number="5">
<h1><span class="header-section-number">Chapter 5</span> Other data</h1>
<div id="fungi-data" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Fungi data</h2>
<p>The above examples are shown with the prokaryotic data.
Now, we use the ITS amplicon sequencing dataset as an example to show the use of FUNGuild database<span class="citation">(<a href="#ref-Nguyen_FUNGuild_2016" role="doc-biblioref">Nguyen et al. 2016</a>)</span>.
FungalTraits <span class="citation">(<a href="#ref-Polme_FungalTraits_2020" role="doc-biblioref">Põlme et al. 2020</a>)</span> database is also available for identifying fungal traits.</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="other-data.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load ITS data</span></span>
<span id="cb172-2"><a href="other-data.html#cb172-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(sample_info_ITS)</span>
<span id="cb172-3"><a href="other-data.html#cb172-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(otu_table_ITS)</span>
<span id="cb172-4"><a href="other-data.html#cb172-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(taxonomy_table_ITS)</span>
<span id="cb172-5"><a href="other-data.html#cb172-5" aria-hidden="true" tabindex="-1"></a><span class="co"># create microtable object</span></span>
<span id="cb172-6"><a href="other-data.html#cb172-6" aria-hidden="true" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> microtable<span class="sc">$</span><span class="fu">new</span>(<span class="at">sample_table =</span> sample_info_ITS, <span class="at">otu_table =</span> otu_table_ITS, <span class="at">tax_table =</span> taxonomy_table_ITS)</span>
<span id="cb172-7"><a href="other-data.html#cb172-7" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the taxa not assigned in the Kingdom &quot;k__Fungi&quot;</span></span>
<span id="cb172-8"><a href="other-data.html#cb172-8" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span>tax_table <span class="sc">%&lt;&gt;%</span> base<span class="sc">::</span><span class="fu">subset</span>(Kingdom <span class="sc">==</span> <span class="st">&quot;k__Fungi&quot;</span>)</span>
<span id="cb172-9"><a href="other-data.html#cb172-9" aria-hidden="true" tabindex="-1"></a><span class="co"># use tidy_dataset() to make OTUs and samples information consistent across files</span></span>
<span id="cb172-10"><a href="other-data.html#cb172-10" aria-hidden="true" tabindex="-1"></a>dataset<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb172-11"><a href="other-data.html#cb172-11" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_network object</span></span>
<span id="cb172-12"><a href="other-data.html#cb172-12" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_network<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> dataset, <span class="at">cal_cor =</span> <span class="st">&quot;WGCNA&quot;</span>, <span class="at">taxa_level =</span> <span class="st">&quot;OTU&quot;</span>, <span class="at">filter_thres =</span> <span class="fl">0.000001</span>, <span class="at">cor_method =</span> <span class="st">&quot;spearman&quot;</span>)</span>
<span id="cb172-13"><a href="other-data.html#cb172-13" aria-hidden="true" tabindex="-1"></a><span class="co"># create correlation network </span></span>
<span id="cb172-14"><a href="other-data.html#cb172-14" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_network</span>(<span class="at">p_thres =</span> <span class="fl">0.05</span>, <span class="at">COR_cut =</span> <span class="fl">0.6</span>)</span>
<span id="cb172-15"><a href="other-data.html#cb172-15" aria-hidden="true" tabindex="-1"></a><span class="co"># add modules</span></span>
<span id="cb172-16"><a href="other-data.html#cb172-16" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_module</span>()</span>
<span id="cb172-17"><a href="other-data.html#cb172-17" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate node topological properties</span></span>
<span id="cb172-18"><a href="other-data.html#cb172-18" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">cal_node_type</span>()</span>
<span id="cb172-19"><a href="other-data.html#cb172-19" aria-hidden="true" tabindex="-1"></a>node_type_table <span class="ot">&lt;-</span> t1<span class="sc">$</span>res_node_type</span>
<span id="cb172-20"><a href="other-data.html#cb172-20" aria-hidden="true" tabindex="-1"></a><span class="co"># create trans_func object</span></span>
<span id="cb172-21"><a href="other-data.html#cb172-21" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(dataset)</span>
<span id="cb172-22"><a href="other-data.html#cb172-22" aria-hidden="true" tabindex="-1"></a><span class="co"># identify species traits, automatically select database for prokaryotes or fungi</span></span>
<span id="cb172-23"><a href="other-data.html#cb172-23" aria-hidden="true" tabindex="-1"></a><span class="co"># fungi_database = &quot;FungalTraits&quot; for the FungalTraits database</span></span>
<span id="cb172-24"><a href="other-data.html#cb172-24" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func</span>(<span class="at">fungi_database =</span> <span class="st">&quot;FUNGuild&quot;</span>)</span>
<span id="cb172-25"><a href="other-data.html#cb172-25" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate abundance-unweighted functional redundancy of each trait for each network module</span></span>
<span id="cb172-26"><a href="other-data.html#cb172-26" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">use_community =</span> <span class="cn">FALSE</span>, <span class="at">node_type_table =</span> node_type_table)</span>
<span id="cb172-27"><a href="other-data.html#cb172-27" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the functional redundancy of network modules</span></span>
<span id="cb172-28"><a href="other-data.html#cb172-28" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span><span class="fu">plot_spe_func_perc</span>(<span class="at">select_samples =</span> <span class="fu">paste0</span>(<span class="st">&quot;M&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>))</span></code></pre></div>
<p><img src="Images/plot_func_perc_module_fungi.png" width="700px" style="display: block; margin: auto;" /></p>
</div>
<div id="metagenomic-data" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Metagenomic data</h2>
<p>Many methods in microeco package can be used not only for the traditional community abundance table, i.e. species-sample table,
but also for other data, such as metagenomic and metabolomic data.
In the README of file2meco package (<a href="https://github.com/ChiLiubio/file2meco" class="uri">https://github.com/ChiLiubio/file2meco</a>),
we show several examples of metagenomic data from HUMAnN and Ncyc.</p>
</div>
<div id="animal-data" class="section level2" number="5.3">
<h2><span class="header-section-number">5.3</span> Animal data</h2>
<p>We use mouse gut data stored in file2meco package to show the input of QIIME2 file and the use of metabolic trait database NJC19 database<span class="citation">(<a href="#ref-Lim_Large_2020" role="doc-biblioref">Lim et al. 2020</a>)</span>.</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="other-data.html#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(microeco)</span>
<span id="cb173-2"><a href="other-data.html#cb173-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(file2meco)</span>
<span id="cb173-3"><a href="other-data.html#cb173-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use data files inside the file2meco package.</span></span>
<span id="cb173-4"><a href="other-data.html#cb173-4" aria-hidden="true" tabindex="-1"></a>abund_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;dada2_table.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb173-5"><a href="other-data.html#cb173-5" aria-hidden="true" tabindex="-1"></a>sample_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;sample-metadata.tsv&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb173-6"><a href="other-data.html#cb173-6" aria-hidden="true" tabindex="-1"></a>taxonomy_file_path <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;taxonomy.qza&quot;</span>, <span class="at">package=</span><span class="st">&quot;file2meco&quot;</span>)</span>
<span id="cb173-7"><a href="other-data.html#cb173-7" aria-hidden="true" tabindex="-1"></a><span class="co"># construct microtable object</span></span>
<span id="cb173-8"><a href="other-data.html#cb173-8" aria-hidden="true" tabindex="-1"></a>data1 <span class="ot">&lt;-</span> <span class="fu">qiime2meco</span>(<span class="at">ASV_data =</span> abund_file_path, <span class="at">sample_data =</span> sample_file_path, <span class="at">taxonomy_data =</span> taxonomy_file_path)</span>
<span id="cb173-9"><a href="other-data.html#cb173-9" aria-hidden="true" tabindex="-1"></a>data1<span class="sc">$</span><span class="fu">tidy_dataset</span>()</span>
<span id="cb173-10"><a href="other-data.html#cb173-10" aria-hidden="true" tabindex="-1"></a><span class="co"># We correct the species names in tax_table</span></span>
<span id="cb173-11"><a href="other-data.html#cb173-11" aria-hidden="true" tabindex="-1"></a>select_rows <span class="ot">&lt;-</span> data1<span class="sc">$</span>tax_table<span class="sc">$</span>Species <span class="sc">!=</span> <span class="st">&quot;s__&quot;</span></span>
<span id="cb173-12"><a href="other-data.html#cb173-12" aria-hidden="true" tabindex="-1"></a>data1<span class="sc">$</span>tax_table<span class="sc">$</span>Species[select_rows] <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;s__&quot;</span>, <span class="fu">gsub</span>(<span class="st">&quot;g__&quot;</span>, <span class="st">&quot;&quot;</span>, data1<span class="sc">$</span>tax_table<span class="sc">$</span>Genus[select_rows]), <span class="st">&quot; &quot;</span>, <span class="fu">gsub</span>(<span class="st">&quot;s__&quot;</span>, <span class="st">&quot;&quot;</span>, data1<span class="sc">$</span>tax_table<span class="sc">$</span>Species[select_rows]))</span>
<span id="cb173-13"><a href="other-data.html#cb173-13" aria-hidden="true" tabindex="-1"></a><span class="co"># taxonomic abundance</span></span>
<span id="cb173-14"><a href="other-data.html#cb173-14" aria-hidden="true" tabindex="-1"></a>data1<span class="sc">$</span><span class="fu">cal_abund</span>()</span>
<span id="cb173-15"><a href="other-data.html#cb173-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-16"><a href="other-data.html#cb173-16" aria-hidden="true" tabindex="-1"></a><span class="co"># create object of trans_func</span></span>
<span id="cb173-17"><a href="other-data.html#cb173-17" aria-hidden="true" tabindex="-1"></a>data2 <span class="ot">&lt;-</span> trans_func<span class="sc">$</span><span class="fu">new</span>(data1)</span>
<span id="cb173-18"><a href="other-data.html#cb173-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Select NJC19 database</span></span>
<span id="cb173-19"><a href="other-data.html#cb173-19" aria-hidden="true" tabindex="-1"></a>data2<span class="sc">$</span><span class="fu">cal_spe_func</span>(<span class="at">prok_database =</span> <span class="st">&quot;NJC19&quot;</span>)</span>
<span id="cb173-20"><a href="other-data.html#cb173-20" aria-hidden="true" tabindex="-1"></a><span class="co"># get the trait percentage data</span></span>
<span id="cb173-21"><a href="other-data.html#cb173-21" aria-hidden="true" tabindex="-1"></a>data2<span class="sc">$</span><span class="fu">cal_spe_func_perc</span>(<span class="at">use_community =</span> <span class="cn">TRUE</span>)</span>
<span id="cb173-22"><a href="other-data.html#cb173-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb173-23"><a href="other-data.html#cb173-23" aria-hidden="true" tabindex="-1"></a><span class="co"># inset the trait percentage result into taxa_abund of microtable object</span></span>
<span id="cb173-24"><a href="other-data.html#cb173-24" aria-hidden="true" tabindex="-1"></a>data1<span class="sc">$</span>taxa_abund<span class="sc">$</span>Trait <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(data2<span class="sc">$</span>res_spe_func_perc))</span>
<span id="cb173-25"><a href="other-data.html#cb173-25" aria-hidden="true" tabindex="-1"></a><span class="co"># use trans_abund to plot</span></span>
<span id="cb173-26"><a href="other-data.html#cb173-26" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> trans_abund<span class="sc">$</span><span class="fu">new</span>(<span class="at">dataset =</span> data1, <span class="at">taxrank =</span> <span class="st">&quot;Trait&quot;</span>, <span class="at">ntaxa =</span> <span class="dv">10</span>, <span class="at">use_percentage =</span> <span class="cn">FALSE</span>)</span>
<span id="cb173-27"><a href="other-data.html#cb173-27" aria-hidden="true" tabindex="-1"></a>t1<span class="sc">$</span><span class="fu">plot_box</span>(<span class="at">group =</span> <span class="st">&quot;donor_status&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Relative population abundance (%)&quot;</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">13</span>))</span></code></pre></div>
<p><img src="Images/plot_mouse_NJC19.png" width="800px" style="display: block; margin: auto;" /></p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Lim_Large_2020" class="csl-entry">
Lim, Roktaek, Josephine Jill T. Cabatbat, Thomas L. P. Martin, Haneul Kim, Seunghyeon Kim, Jaeyun Sung, Cheol-Min Ghim, and Pan-Jun Kim. 2020. <span>“Large-Scale Metabolic Interaction Network of the Mouse and Human Gut Microbiota.”</span> Journal Article. <em>Scientific Data</em> 7 (1). <a href="https://doi.org/10.1038/s41597-020-0516-5">https://doi.org/10.1038/s41597-020-0516-5</a>.
</div>
<div id="ref-Nguyen_FUNGuild_2016" class="csl-entry">
Nguyen, Nhu H., Zewei Song, Scott T. Bates, Sara Branco, Leho Tedersoo, Jon Menke, Jonathan S. Schilling, and Peter G. Kennedy. 2016. <span>“FUNGuild: An Open Annotation Tool for Parsing Fungal Community Datasets by Ecological Guild.”</span> Journal Article. <em>Fungal Ecology</em> 20 (1): 241–48.
</div>
<div id="ref-Polme_FungalTraits_2020" class="csl-entry">
Põlme, Sergei, Kessy Abarenkov, R. Henrik Nilsson, Björn D. Lindahl, Karina Engelbrecht Clemmensen, Havard Kauserud, Nhu Nguyen, et al. 2020. <span>“FungalTraits: A User-Friendly Traits Database of Fungi and Fungus-Like Stramenopiles.”</span> Journal Article. <em>Fungal Diversity</em> 105 (1): 1–16. <a href="https://doi.org/10.1007/s13225-020-00466-2">https://doi.org/10.1007/s13225-020-00466-2</a>.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="extended-class.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="notes.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/04-Otherdata.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["microeco-tutorial.pdf", "microeco-tutorial.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
