\name{PathwayInformationKEGGBacArch}
\alias{PathwayInformationKEGGBacArch}
\docType{data}
\title{
%%   ~~ data name/kind ... ~~
}
\description{
%%  ~~ A concise (1-5 lines) description of the dataset. ~~
}
\usage{data(PathwayInformationKEGGBacArch)}
\format{
  A data frame with 340 observations on the following 4 variables.
  \describe{
    \item{\code{V3}}{a factor with levels \code{ko00010} \code{ko00020} \code{ko00030} \code{ko00040} \code{ko00051} \code{ko00052} \code{ko00053} \code{ko00061} \code{ko00062} \code{ko00071} \code{ko00072} \code{ko00073} \code{ko00100} \code{ko00120} \code{ko00121} \code{ko00130} \code{ko00140} \code{ko00190} \code{ko00195} \code{ko00196} \code{ko00230} \code{ko00231} \code{ko00232} \code{ko00240} \code{ko00250} \code{ko00253} \code{ko00260} \code{ko00270} \code{ko00280} \code{ko00281} \code{ko00290} \code{ko00300} \code{ko00310} \code{ko00311} \code{ko00312} \code{ko00330} \code{ko00331} \code{ko00340} \code{ko00350} \code{ko00351} \code{ko00360} \code{ko00361} \code{ko00362} \code{ko00363} \code{ko00364} \code{ko00380} \code{ko00400} \code{ko00401} \code{ko00402} \code{ko00410} \code{ko00430} \code{ko00440} \code{ko00450} \code{ko00460} \code{ko00471} \code{ko00472} \code{ko00473} \code{ko00480} \code{ko00500} \code{ko00510} \code{ko00511} \code{ko00512} \code{ko00513} \code{ko00514} \code{ko00520} \code{ko00521} \code{ko00522} \code{ko00523} \code{ko00524} \code{ko00531} \code{ko00532} \code{ko00533} \code{ko00534} \code{ko00540} \code{ko00550} \code{ko00561} \code{ko00562} \code{ko00563} \code{ko00564} \code{ko00565} \code{ko00590} \code{ko00591} \code{ko00592} \code{ko00600} \code{ko00601} \code{ko00603} \code{ko00604} \code{ko00620} \code{ko00621} \code{ko00622} \code{ko00623} \code{ko00624} \code{ko00625} \code{ko00626} \code{ko00627} \code{ko00630} \code{ko00633} \code{ko00640} \code{ko00642} \code{ko00643} \code{ko00650} \code{ko00660} \code{ko00670} \code{ko00680} \code{ko00710} \code{ko00720} \code{ko00730} \code{ko00740} \code{ko00750} \code{ko00760} \code{ko00770} \code{ko00780} \code{ko00785} \code{ko00790} \code{ko00791} \code{ko00830} \code{ko00860} \code{ko00900} \code{ko00901} \code{ko00902} \code{ko00903} \code{ko00904} \code{ko00905} \code{ko00906} \code{ko00908} \code{ko00909} \code{ko00910} \code{ko00920} \code{ko00930} \code{ko00940} \code{ko00941} \code{ko00942} \code{ko00943} \code{ko00944} \code{ko00945} \code{ko00950} \code{ko00960} \code{ko00965} \code{ko00966} \code{ko00970} \code{ko00980} \code{ko00981} \code{ko00982} \code{ko00983} \code{ko00984} \code{ko01040} \code{ko01051} \code{ko01053} \code{ko01054} \code{ko01055} \code{ko01056} \code{ko01057} \code{ko01058} \code{ko02010} \code{ko02020} \code{ko02030} \code{ko02040} \code{ko02060} \code{ko03008} \code{ko03010} \code{ko03013} \code{ko03015} \code{ko03018} \code{ko03020} \code{ko03022} \code{ko03030} \code{ko03040} \code{ko03050} \code{ko03060} \code{ko03070} \code{ko03320} \code{ko03410} \code{ko03420} \code{ko03430} \code{ko03440} \code{ko03450} \code{ko03460} \code{ko04010} \code{ko04011} \code{ko04012} \code{ko04013} \code{ko04020} \code{ko04060} \code{ko04062} \code{ko04064} \code{ko04070} \code{ko04075} \code{ko04080} \code{ko04110} \code{ko04111} \code{ko04112} \code{ko04113} \code{ko04114} \code{ko04115} \code{ko04120} \code{ko04122} \code{ko04130} \code{ko04140} \code{ko04141} \code{ko04142} \code{ko04144} \code{ko04145} \code{ko04146} \code{ko04150} \code{ko04210} \code{ko04260} \code{ko04270} \code{ko04310} \code{ko04320} \code{ko04330} \code{ko04340} \code{ko04350} \code{ko04360} \code{ko04370} \code{ko04380} \code{ko04510} \code{ko04512} \code{ko04514} \code{ko04520} \code{ko04530} \code{ko04540} \code{ko04610} \code{ko04612} \code{ko04614} \code{ko04620} \code{ko04621} \code{ko04622} \code{ko04623} \code{ko04626} \code{ko04630} \code{ko04640} \code{ko04650} \code{ko04660} \code{ko04662} \code{ko04664} \code{ko04666} \code{ko04670} \code{ko04672} \code{ko04710} \code{ko04711} \code{ko04712} \code{ko04720} \code{ko04721} \code{ko04722} \code{ko04723} \code{ko04724} \code{ko04725} \code{ko04726} \code{ko04727} \code{ko04728} \code{ko04730} \code{ko04740} \code{ko04742} \code{ko04744} \code{ko04745} \code{ko04810} \code{ko04910} \code{ko04912} \code{ko04914} \code{ko04916} \code{ko04920} \code{ko04930} \code{ko04940} \code{ko04950} \code{ko04960} \code{ko04961} \code{ko04962} \code{ko04964} \code{ko04966} \code{ko04970} \code{ko04971} \code{ko04972} \code{ko04973} \code{ko04974} \code{ko04975} \code{ko04976} \code{ko04977} \code{ko04978} \code{ko05010} \code{ko05012} \code{ko05014} \code{ko05016} \code{ko05020} \code{ko05030} \code{ko05031} \code{ko05032} \code{ko05033} \code{ko05034} \code{ko05100} \code{ko05110} \code{ko05111} \code{ko05120} \code{ko05130} \code{ko05131} \code{ko05132} \code{ko05133} \code{ko05134} \code{ko05140} \code{ko05142} \code{ko05143} \code{ko05144} \code{ko05145} \code{ko05146} \code{ko05150} \code{ko05152} \code{ko05160} \code{ko05162} \code{ko05164} \code{ko05166} \code{ko05168} \code{ko05169} \code{ko05200} \code{ko05202} \code{ko05203} \code{ko05204} \code{ko05210} \code{ko05211} \code{ko05212} \code{ko05213} \code{ko05214} \code{ko05215} \code{ko05216} \code{ko05217} \code{ko05218} \code{ko05219} \code{ko05220} \code{ko05221} \code{ko05222} \code{ko05223} \code{ko05310} \code{ko05320} \code{ko05322} \code{ko05323} \code{ko05330} \code{ko05332} \code{ko05340} \code{ko05410} \code{ko05412} \code{ko05414} \code{ko05416}}
    \item{\code{V4}}{a factor with levels \code{ABC transporters} \code{Acridone alkaloid biosynthesis} \code{Acute myeloid leukemia} \code{Adherens junction} \code{Adipocytokine signaling pathway} \code{African trypanosomiasis} \code{Alanine, aspartate and glutamate metabolism} \code{Alcoholism} \code{Aldosterone-regulated sodium reabsorption} \code{Allograft rejection} \code{alpha-Linolenic acid metabolism} \code{Alzheimers disease} \code{Aminoacyl-tRNA biosynthesis} \code{Aminobenzoate degradation} \code{Amino sugar and nucleotide sugar metabolism} \code{Amoebiasis} \code{Amphetamine addiction} \code{Amyotrophic lateral sclerosis (ALS)} \code{Anthocyanin biosynthesis} \code{Antigen processing and presentation} \code{Apoptosis} \code{Arachidonic acid metabolism} \code{Arginine and proline metabolism} \code{Arrhythmogenic right ventricular cardiomyopathy (ARVC)} \code{Ascorbate and aldarate metabolism} \code{Asthma} \code{Atrazine degradation} \code{Autoimmune thyroid disease} \code{Axon guidance} \code{Bacterial chemotaxis} \code{Bacterial invasion of epithelial cells} \code{Bacterial secretion system} \code{Basal cell carcinoma} \code{Basal transcription factors} \code{Base excision repair} \code{B cell receptor signaling pathway} \code{Benzoate degradation} \code{Benzoxazinoid biosynthesis} \code{beta-Alanine metabolism} \code{beta-Lactam resistance} \code{Betalain biosynthesis} \code{Bile secretion} \code{Biosynthesis of 12-, 14- and 16-membered macrolides} \code{Biosynthesis of ansamycins} \code{Biosynthesis of siderophore group nonribosomal peptides} \code{Biosynthesis of type II polyketide backbone} \code{Biosynthesis of type II polyketide products} \code{Biosynthesis of unsaturated fatty acids} \code{Biosynthesis of vancomycin group antibiotics} \code{Biotin metabolism} \code{Bisphenol degradation} \code{Bladder cancer} \code{Brassinosteroid biosynthesis} \code{Butanoate metabolism} \code{Butirosin and neomycin biosynthesis} \code{C5-Branched dibasic acid metabolism} \code{Caffeine metabolism} \code{Calcium signaling pathway} \code{Caprolactam degradation} \code{Carbohydrate digestion and absorption} \code{Carbon fixation in photosynthetic organisms} \code{Carbon fixation pathways in prokaryotes} \code{Cardiac muscle contraction} \code{Carotenoid biosynthesis} \code{Cell adhesion molecules (CAMs)} \code{Cell cycle} \code{Cell cycle - Caulobacter} \code{Cell cycle - yeast} \code{Chagas disease (American trypanosomiasis)} \code{Chemical carcinogenesis} \code{Chemokine signaling pathway} \code{Chloroalkane and chloroalkene degradation} \code{Chlorocyclohexane and chlorobenzene degradation} \code{Cholinergic synapse} \code{Chronic myeloid leukemia} \code{Circadian rhythm - fly} \code{Circadian rhythm - mammal} \code{Circadian rhythm - plant} \code{Citrate cycle (TCA cycle)} \code{Clavulanic acid biosynthesis} \code{Cocaine addiction} \code{Collecting duct acid secretion} \code{Colorectal cancer} \code{Complement and coagulation cascades} \code{Cutin, suberine and wax biosynthesis} \code{Cyanoamino acid metabolism} \code{Cysteine and methionine metabolism} \code{Cytokine-cytokine receptor interaction} \code{Cytosolic DNA-sensing pathway} \code{D-Alanine metabolism} \code{D-Arginine and D-ornithine metabolism} \code{DDT degradation} \code{D-Glutamine and D-glutamate metabolism} \code{Dilated cardiomyopathy} \code{Dioxin degradation} \code{Diterpenoid biosynthesis} \code{DNA replication} \code{Dopaminergic synapse} \code{Dorso-ventral axis formation} \code{Drug metabolism - cytochrome P450} \code{Drug metabolism - other enzymes} \code{ECM-receptor interaction} \code{Endocrine and other factor-regulated calcium reabsorption} \code{Endocytosis} \code{Endometrial cancer} \code{Epithelial cell signaling in Helicobacter pylori infection} \code{Epstein-Barr virus infection} \code{ErbB signaling pathway} \code{Ether lipid metabolism} \code{Ethylbenzene degradation} \code{Fanconi anemia pathway} \code{Fat digestion and absorption} \code{Fatty acid biosynthesis} \code{Fatty acid elongation} \code{Fatty acid metabolism} \code{Fc epsilon RI signaling pathway} \code{Fc gamma R-mediated phagocytosis} \code{Flagellar assembly} \code{Flavone and flavonol biosynthesis} \code{Flavonoid biosynthesis} \code{Fluorobenzoate degradation} \code{Focal adhesion} \code{Folate biosynthesis} \code{Fructose and mannose metabolism} \code{GABAergic synapse} \code{Galactose metabolism} \code{Gap junction} \code{Gastric acid secretion} \code{Geraniol degradation} \code{Glioma} \code{Glucosinolate biosynthesis} \code{Glutamatergic synapse} \code{Glutathione metabolism} \code{Glycerolipid metabolism} \code{Glycerophospholipid metabolism} \code{Glycine, serine and threonine metabolism} \code{Glycolysis / Gluconeogenesis} \code{Glycosaminoglycan biosynthesis - chondroitin sulfate} \code{Glycosaminoglycan biosynthesis - heparan sulfate} \code{Glycosaminoglycan biosynthesis - keratan sulfate} \code{Glycosaminoglycan degradation} \code{Glycosphingolipid biosynthesis - ganglio series} \code{Glycosphingolipid biosynthesis - globo series} \code{Glycosphingolipid biosynthesis - lacto and neolacto series} \code{Glycosylphosphatidylinositol(GPI)-anchor biosynthesis} \code{Glyoxylate and dicarboxylate metabolism} \code{GnRH signaling pathway} \code{Graft-versus-host disease} \code{Hedgehog signaling pathway} \code{Hematopoietic cell lineage} \code{Hepatitis C} \code{Herpes simplex infection} \code{Histidine metabolism} \code{Homologous recombination} \code{HTLV-I infection} \code{Huntingtons disease} \code{Hypertrophic cardiomyopathy (HCM)} \code{Indole alkaloid biosynthesis} \code{Influenza A} \code{Inositol phosphate metabolism} \code{Insect hormone biosynthesis} \code{Insulin signaling pathway} \code{Intestinal immune network for IgA production} \code{Isoflavonoid biosynthesis} \code{Isoquinoline alkaloid biosynthesis} \code{Jak-STAT signaling pathway} \code{Legionellosis} \code{Leishmaniasis} \code{Leukocyte transendothelial migration} \code{Limonene and pinene degradation} \code{Linoleic acid metabolism} \code{Lipoic acid metabolism} \code{Lipopolysaccharide biosynthesis} \code{Long-term depression} \code{Long-term potentiation} \code{Lysine biosynthesis} \code{Lysine degradation} \code{Lysosome} \code{Malaria} \code{MAPK signaling pathway} \code{MAPK signaling pathway - fly} \code{MAPK signaling pathway - yeast} \code{Maturity onset diabetes of the young} \code{Measles} \code{Meiosis - yeast} \code{Melanogenesis} \code{Melanoma} \code{Metabolism of xenobiotics by cytochrome P450} \code{Methane metabolism} \code{Mineral absorption} \code{Mismatch repair} \code{Monoterpenoid biosynthesis} \code{Morphine addiction} \code{mRNA surveillance pathway} \code{mTOR signaling pathway} \code{Mucin type O-Glycan biosynthesis} \code{Naphthalene degradation} \code{Natural killer cell mediated cytotoxicity} \code{Neuroactive ligand-receptor interaction} \code{Neurotrophin signaling pathway} \code{NF-kappa B signaling pathway} \code{N-Glycan biosynthesis} \code{Nicotinate and nicotinamide metabolism} \code{Nicotine addiction} \code{Nitrogen metabolism} \code{Nitrotoluene degradation} \code{NOD-like receptor signaling pathway} \code{Non-homologous end-joining} \code{Nonribosomal peptide structures} \code{Non-small cell lung cancer} \code{Notch signaling pathway} \code{Novobiocin biosynthesis} \code{Nucleotide excision repair} \code{Olfactory transduction} \code{One carbon pool by folate} \code{Oocyte meiosis} \code{Osteoclast differentiation} \code{Other glycan degradation} \code{Other types of O-glycan biosynthesis} \code{Oxidative phosphorylation} \code{p53 signaling pathway} \code{Pancreatic cancer} \code{Pancreatic secretion} \code{Pantothenate and CoA biosynthesis} \code{Parkinsons disease} \code{Pathogenic Escherichia coli infection} \code{Pathways in cancer} \code{Penicillin and cephalosporin biosynthesis} \code{Pentose and glucuronate interconversions} \code{Pentose phosphate pathway} \code{Peptidoglycan biosynthesis} \code{Peroxisome} \code{Pertussis} \code{Phagosome} \code{Phenylalanine metabolism} \code{Phenylalanine, tyrosine and tryptophan biosynthesis} \code{Phenylpropanoid biosynthesis} \code{Phosphatidylinositol signaling system} \code{Phosphonate and phosphinate metabolism} \code{Phosphotransferase system (PTS)} \code{Photosynthesis} \code{Photosynthesis - antenna proteins} \code{Phototransduction} \code{Phototransduction - fly} \code{Plant hormone signal transduction} \code{Plant-pathogen interaction} \code{Polycyclic aromatic hydrocarbon degradation} \code{Polyketide sugar unit biosynthesis} \code{Porphyrin and chlorophyll metabolism} \code{PPAR signaling pathway} \code{Primary bile acid biosynthesis} \code{Primary immunodeficiency} \code{Prion diseases} \code{Progesterone-mediated oocyte maturation} \code{Propanoate metabolism} \code{Prostate cancer} \code{Proteasome} \code{Protein digestion and absorption} \code{Protein export} \code{Protein processing in endoplasmic reticulum} \code{Proximal tubule bicarbonate reclamation} \code{Purine metabolism} \code{Puromycin biosynthesis} \code{Pyrimidine metabolism} \code{Pyruvate metabolism} \code{Regulation of actin cytoskeleton} \code{Regulation of autophagy} \code{Renal cell carcinoma} \code{Renin-angiotensin system} \code{Retinol metabolism} \code{Retrograde endocannabinoid signaling} \code{Rheumatoid arthritis} \code{Riboflavin metabolism} \code{Ribosome} \code{Ribosome biogenesis in eukaryotes} \code{RIG-I-like receptor signaling pathway} \code{RNA degradation} \code{RNA polymerase} \code{RNA transport} \code{Salivary secretion} \code{Salmonella infection} \code{Secondary bile acid biosynthesis} \code{Selenocompound metabolism} \code{Serotonergic synapse} \code{Sesquiterpenoid and triterpenoid biosynthesis} \code{Shigellosis} \code{Small cell lung cancer} \code{SNARE interactions in vesicular transport} \code{Sphingolipid metabolism} \code{Spliceosome} \code{Staphylococcus aureus infection} \code{Starch and sucrose metabolism} \code{Steroid biosynthesis} \code{Steroid degradation} \code{Steroid hormone biosynthesis} \code{Stilbenoid, diarylheptanoid and gingerol biosynthesis} \code{Streptomycin biosynthesis} \code{Styrene degradation} \code{Sulfur metabolism} \code{Sulfur relay system} \code{Synaptic vesicle cycle} \code{Synthesis and degradation of ketone bodies} \code{Systemic lupus erythematosus} \code{Taste transduction} \code{Taurine and hypotaurine metabolism} \code{T cell receptor signaling pathway} \code{Terpenoid backbone biosynthesis} \code{Tetracycline biosynthesis} \code{TGF-beta signaling pathway} \code{Thiamine metabolism} \code{Thyroid cancer} \code{Tight junction} \code{Toll-like receptor signaling pathway} \code{Toluene degradation} \code{Toxoplasmosis} \code{Transcriptional misregulation in cancer} \code{Tropane, piperidine and pyridine alkaloid biosynthesis} \code{Tryptophan metabolism} \code{Tuberculosis} \code{Two-component system} \code{Type I diabetes mellitus} \code{Type II diabetes mellitus} \code{Tyrosine metabolism} \code{Ubiquinone and other terpenoid-quinone biosynthesis} \code{Ubiquitin mediated proteolysis} \code{Valine, leucine and isoleucine biosynthesis} \code{Valine, leucine and isoleucine degradation} \code{Various types of N-glycan biosynthesis} \code{Vascular smooth muscle contraction} \code{Vasopressin-regulated water reabsorption} \code{VEGF signaling pathway} \code{Vibrio cholerae infection} \code{Vibrio cholerae pathogenic cycle} \code{Viral carcinogenesis} \code{Viral myocarditis} \code{Vitamin B6 metabolism} \code{Vitamin digestion and absorption} \code{Wnt signaling pathway} \code{Xylene degradation} \code{Zeatin biosynthesis}}
    \item{\code{V6}}{a factor with levels \code{Amino Acid Metabolism} \code{Biosynthesis of Other Secondary Metabolites} \code{Cancers} \code{Carbohydrate Metabolism} \code{Cardiovascular Diseases} \code{Cell Communication} \code{Cell Growth and Death} \code{Cell Motility} \code{Circulatory System} \code{Development} \code{Digestive System} \code{Endocrine and Metabolic Diseases} \code{Endocrine System} \code{Energy Metabolism} \code{Environmental Adaptation} \code{Excretory System} \code{Folding, Sorting and Degradation} \code{Glycan Biosynthesis and Metabolism} \code{Immune Diseases} \code{Immune System} \code{Infectious Diseases} \code{Lipid Metabolism} \code{Membrane Transport} \code{Metabolism of Cofactors and Vitamins} \code{Metabolism of Other Amino Acids} \code{Metabolism of Terpenoids and Polyketides} \code{Nervous System} \code{Neurodegenerative Diseases} \code{Nucleotide Metabolism} \code{Replication and Repair} \code{Sensory System} \code{Signaling Molecules and Interaction} \code{Signal Transduction} \code{Substance Dependence} \code{Transcription} \code{Translation} \code{Transport and Catabolism} \code{Xenobiotics Biodegradation and Metabolism}}
    \item{\code{V8}}{a factor with levels \code{Cellular Processes} \code{Environmental Information Processing} \code{Genetic Information Processing} \code{Human Diseases} \code{Metabolism} \code{Organismal Systems}}
  }
}
\details{
%%  ~~ If necessary, more details than the __description__ above ~~
}
\source{
%%  ~~ reference to a publication or URL from which the data were obtained ~~
}
\references{
%%  ~~ possibly secondary sources and usages ~~
}
\examples{
data(PathwayInformationKEGGBacArch)
## maybe str(PathwayInformationKEGGBacArch) ; plot(PathwayInformationKEGGBacArch) ...
}
\keyword{datasets}
